{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "47e7307b",
   "metadata": {},
   "source": [
    "# Auto-generate BTE annotations for BioThings SEMMEDDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ab5663",
   "metadata": {},
   "source": [
    "This notebook walks a developer through the process of [taking the SEMMEDDB database](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html) and using this data to auto-generate the x-bte operations for [BTE](https://github.com/biothings/BioThings_Explorer_TRAPI). This is needed for BTE to query + process the responses from the [BioThings SEMMEDDB API](https://pending.biothings.io/semmeddb). \n",
    "\n",
    "---\n",
    "\n",
    "When one sees this:  \n",
    "**PAUSE**\n",
    "\n",
    "read the accompanying text that will explain what the developer needs to do before running the code chunks below it\n",
    "\n",
    "---\n",
    "\n",
    "The [yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_operations.yaml) [segments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_list.yaml) generated by this notebook are added to [this file](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml) to make [yaml used for the smartapi registration](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) for BioThings SEMMEDDB API."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43745613",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712186ee",
   "metadata": {},
   "source": [
    "Requirements:\n",
    "* get bmt from pip (see instructions here https://github.com/biolink/biolink-model-toolkit/)\n",
    "* get ruamel.yaml from pip (see instructions here https://yaml.readthedocs.io/en/latest/install.html)\n",
    "\n",
    "Files:\n",
    "* Get SEMMEDDB PREDICATION CSV here: https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or a direct link [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SRDEF.txt)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5ee5aac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:06:26.245383Z",
     "start_time": "2021-11-10T07:06:24.591112Z"
    }
   },
   "outputs": [],
   "source": [
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "## get using pip https://github.com/biolink/biolink-model-toolkit/\n",
    "from bmt import Toolkit\n",
    "\n",
    "## get from pip (using instead of pyyaml's import yaml)\n",
    "import ruamel.yaml as ryml\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "## used in trying things out\n",
    "# import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af532d5",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Check and correct the path for `raw_data_location`\n",
    "* Check that the columns specified in `usecols` and `names` match the columns of the PREDICATION [file](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html). \n",
    "* Check that the `na_values`, `sep` are correct. One can use a command in Terminal like `head`\n",
    "* If there are encoding issues, try different encodings. latin1 was used and worked, ref: https://stackoverflow.com/questions/61163367/how-to-resolve-unicodedecodeerror-in-pandas-read-csv-while-loading-dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73408978",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:41.314785Z",
     "start_time": "2021-11-10T07:06:26.247323Z"
    }
   },
   "outputs": [],
   "source": [
    "raw_data_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"semmedVER43_2021_R_PREDICATION.csv\")\n",
    "\n",
    "raw_data = pd.read_csv(raw_data_location, header=None, sep=\",\", encoding=\"latin1\",\n",
    "                          usecols=[3, 6, 7, 10, 11],\n",
    "                          names=[\"PREDICATE\",\"SUBJECT_SEMTYPE\",\"SUBJECT_NOVELTY\",\n",
    "                                 \"OBJECT_SEMTYPE\", \"OBJECT_NOVELTY\"],\n",
    "                          na_values=r\"\\N\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8150ed6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:41.326573Z",
     "start_time": "2021-11-10T07:07:41.316999Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(113863366, 5)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57d2dc4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:41.341193Z",
     "start_time": "2021-11-10T07:07:41.329953Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>SUBJECT_NOVELTY</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>OBJECT_NOVELTY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>0</td>\n",
       "      <td>humn</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PREDICATE SUBJECT_SEMTYPE  SUBJECT_NOVELTY OBJECT_SEMTYPE  OBJECT_NOVELTY\n",
       "0  PROCESS_OF            virs                1           mamm             1.0\n",
       "1         ISA            virs                1           virs             1.0\n",
       "2         ISA            virs                1           virs             1.0\n",
       "3         ISA            virs                1           virs             1.0\n",
       "4  PROCESS_OF            dsyn                0           humn             0.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b68180",
   "metadata": {},
   "source": [
    "## Basic Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc722c4",
   "metadata": {},
   "source": [
    "### keep only novelty = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b8cf0b",
   "metadata": {},
   "source": [
    "now filter it to only keep novelty == 1 for both subject / object...\n",
    "\n",
    "since those with novelty == 0 probably aren't very helpful / interesting to Translator. The entities with novelty == 0 are [listed](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html) in the SEMMEDDB GENERIC_CONCEPT table files, which can be downloaded [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c362bfe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:47.978374Z",
     "start_time": "2021-11-10T07:07:41.343318Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = raw_data[(raw_data[\"SUBJECT_NOVELTY\"] == 1) &\n",
    "                         (raw_data[\"OBJECT_NOVELTY\"] == 1)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68a11e89",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:47.985386Z",
     "start_time": "2021-11-10T07:07:47.980849Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(78688677, 5)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9abae6d5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:49.748820Z",
     "start_time": "2021-11-10T07:07:47.987549Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove the novelty stuff since that will make computations faster and it's always 1 now\n",
    "filtered_data = filtered_data[['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66309970",
   "metadata": {},
   "source": [
    "what semantic types are even there? we have to prune down to what we actually want BTE operations on...\n",
    "\n",
    "interesting since the stats from the official website say 127 semantic types and 54 predicates\n",
    "https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67a2db8e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:59.808693Z",
     "start_time": "2021-11-10T07:07:49.750725Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "68"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes)  ## got 133\n",
    "len(object_semtypes)   ## got 135\n",
    "len(predicates)        ## got 68"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b788a7d",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Review the 3 sets above to see if there are things I want to remove. The normal entity semantic types have 4-letter codes..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88f7eeff",
   "metadata": {},
   "source": [
    "### Remove semantic types that we don't want to make operations from"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6085739",
   "metadata": {},
   "source": [
    "so there's some object stuff that I'm going to remove..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "69d37a01",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:07:59.815783Z",
     "start_time": "2021-11-10T07:07:59.810807Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C0030193',\n",
       " 'C0030705',\n",
       " 'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gngm,aapp',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'podg,humn',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "object_semtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26561f10",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:05.500187Z",
     "start_time": "2021-11-10T07:07:59.819775Z"
    }
   },
   "outputs": [],
   "source": [
    "## make the set of stuff we want to remove\n",
    "removal1 = set([\"C0030193\", \"C0030705\", \"gngm,aapp\", \"podg,humn\"])\n",
    "\n",
    "## remove it from the data \n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(removal1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f18986e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:14.978680Z",
     "start_time": "2021-11-10T07:08:05.501819Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 133: didn't change\n",
    "len(object_semtypes)  ## was 135, now 131: decreased by 4 (expected)\n",
    "len(predicates)       ## was 68, now 65: decreased by 3 (cool)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff20381b",
   "metadata": {},
   "source": [
    "### Remove Predicates we don't want to make operations from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09163f4e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:14.985000Z",
     "start_time": "2021-11-10T07:08:14.981122Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADMINISTERED_TO',\n",
       " 'AFFECTS',\n",
       " 'ASSOCIATED_WITH',\n",
       " 'AUGMENTS',\n",
       " 'CAUSES',\n",
       " 'COEXISTS_WITH',\n",
       " 'COMPLICATES',\n",
       " 'CONVERTS_TO',\n",
       " 'DIAGNOSES',\n",
       " 'DISRUPTS',\n",
       " 'INHIBITS',\n",
       " 'INTERACTS_WITH',\n",
       " 'ISA',\n",
       " 'LOCATION_OF',\n",
       " 'MANIFESTATION_OF',\n",
       " 'MEASUREMENT_OF',\n",
       " 'MEASURES',\n",
       " 'METHOD_OF',\n",
       " 'NEG_ADMINISTERED_TO',\n",
       " 'NEG_AFFECTS',\n",
       " 'NEG_ASSOCIATED_WITH',\n",
       " 'NEG_AUGMENTS',\n",
       " 'NEG_CAUSES',\n",
       " 'NEG_COEXISTS_WITH',\n",
       " 'NEG_COMPLICATES',\n",
       " 'NEG_CONVERTS_TO',\n",
       " 'NEG_DIAGNOSES',\n",
       " 'NEG_DISRUPTS',\n",
       " 'NEG_INHIBITS',\n",
       " 'NEG_INTERACTS_WITH',\n",
       " 'NEG_ISA',\n",
       " 'NEG_LOCATION_OF',\n",
       " 'NEG_MANIFESTATION_OF',\n",
       " 'NEG_MEASUREMENT_OF',\n",
       " 'NEG_MEASURES',\n",
       " 'NEG_METHOD_OF',\n",
       " 'NEG_OCCURS_IN',\n",
       " 'NEG_PART_OF',\n",
       " 'NEG_PRECEDES',\n",
       " 'NEG_PREDISPOSES',\n",
       " 'NEG_PREVENTS',\n",
       " 'NEG_PROCESS_OF',\n",
       " 'NEG_PRODUCES',\n",
       " 'NEG_STIMULATES',\n",
       " 'NEG_TREATS',\n",
       " 'NEG_USES',\n",
       " 'NEG_higher_than',\n",
       " 'NEG_lower_than',\n",
       " 'NEG_same_as',\n",
       " 'NOM',\n",
       " 'OCCURS_IN',\n",
       " 'PART_OF',\n",
       " 'PRECEDES',\n",
       " 'PREDISPOSES',\n",
       " 'PREP',\n",
       " 'PREVENTS',\n",
       " 'PROCESS_OF',\n",
       " 'PRODUCES',\n",
       " 'STIMULATES',\n",
       " 'TREATS',\n",
       " 'USES',\n",
       " 'compared_with',\n",
       " 'higher_than',\n",
       " 'lower_than',\n",
       " 'same_as'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "106d9b99",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Decide what predicates you want to remove\n",
    "\n",
    "**Current logic**\n",
    "\n",
    "Translator currently isn't great with negation, so removing those\n",
    "\n",
    "Also some predicates seem to actually be practice phrases? see https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-486#Sec26\n",
    "\n",
    "And I didn't find the following relationships useful: ISA (subclass relationship), same_as (equivalent / done just as well), lower_than / higher_than (these two entities were compared and 1 was better (higher) or worse (lower) than the other), and compared_with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dec1bcc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:14.990501Z",
     "start_time": "2021-11-10T07:08:14.987162Z"
    }
   },
   "outputs": [],
   "source": [
    "negative_preds = [\"NEG_ADMINISTERED_TO\", \"NEG_AFFECTS\", \"NEG_ASSOCIATED_WITH\",\n",
    "                  \"NEG_AUGMENTS\", \"NEG_CAUSES\", \"NEG_COEXISTS_WITH\", \"NEG_COMPLICATES\",\n",
    "                  \"NEG_CONVERTS_TO\", \"NEG_DIAGNOSES\", \"NEG_DISRUPTS\", \"NEG_INHIBITS\",\n",
    "                  \"NEG_INTERACTS_WITH\", \"NEG_ISA\", \"NEG_LOCATION_OF\", \"NEG_MANIFESTATION_OF\",\n",
    "                  \"NEG_MEASUREMENT_OF\", \"NEG_MEASURES\", \"NEG_METHOD_OF\", \"NEG_OCCURS_IN\",\n",
    "                  \"NEG_PART_OF\", \"NEG_PRECEDES\", \"NEG_PREDISPOSES\", \"NEG_PREVENTS\",\n",
    "                  \"NEG_PROCESS_OF\", \"NEG_PRODUCES\", \"NEG_STIMULATES\", \"NEG_TREATS\", \"NEG_USES\",\n",
    "                  \"NEG_higher_than\", \"NEG_lower_than\", \"NEG_same_as\"\n",
    "                 ]\n",
    "\n",
    "practice_phrase = [\"VERB\", \"NOM\", \"PREP\"]\n",
    "\n",
    "dont_like = [\"ISA\", \"same_as\", \"lower_than\", \"higher_than\", \"compared_with\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ab643b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:31.243921Z",
     "start_time": "2021-11-10T07:08:14.991905Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(negative_preds)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(practice_phrase)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(dont_like)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f3d145",
   "metadata": {},
   "source": [
    "now look at the stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b6244458",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:39.918650Z",
     "start_time": "2021-11-10T07:08:31.245313Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 127: decreased by 6\n",
    "len(object_semtypes)  ## was 131, now 127: decreased by 4\n",
    "len(predicates)       ## was 65, now 27: decreased by 38 (was expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e31d94",
   "metadata": {},
   "source": [
    "so....that's how many kinds of subjects, objects, and predicates we have to go forward with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4184a17f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:53.294709Z",
     "start_time": "2021-11-10T07:08:39.920870Z"
    }
   },
   "outputs": [],
   "source": [
    "combos = filtered_data.value_counts().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "24036b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:53.299727Z",
     "start_time": "2021-11-10T07:08:53.295931Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14796, 4)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos.shape\n",
    "## so that's still a lot...."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e998d9e8",
   "metadata": {},
   "source": [
    "## Specific Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330bba7a",
   "metadata": {},
   "source": [
    "so we actually only need to use certain semantic types...it depends on how they map to the biolink model\n",
    "\n",
    "As a reminder:\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or a direct link [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SRDEF.txt)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91866eb6",
   "metadata": {},
   "source": [
    "### Ingest SEMMED semantic info and biolink model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8d1bbf26",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:53.304643Z",
     "start_time": "2021-11-10T07:08:53.301828Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"2020AA\", \"SRDEF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8f859145",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:53.313704Z",
     "start_time": "2021-11-10T07:08:53.306662Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef = pd.read_csv(srdef_location, sep=\"|\", header=0, index_col=False,\n",
    "                    names=[\"Record Type (RT)\",\n",
    "                           \"Unique Identifier (UI)\",\n",
    "                           \"Full Name (STY/RL)\", \n",
    "                           \"Tree Number (STN/RTN)\", \n",
    "                           \"Definition (DEF)\",\n",
    "                           \"Examples (EX)\",\n",
    "                           \"Usage Note (UN)\",\n",
    "                           \"Non-Human Flag (NH)\",\n",
    "                           \"Abbreviation (ABR)\",\n",
    "                           \"Inverse Relation (RIN)\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d470fa70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:53.325906Z",
     "start_time": "2021-11-10T07:08:53.315191Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Record Type (RT)</th>\n",
       "      <th>Unique Identifier (UI)</th>\n",
       "      <th>Full Name (STY/RL)</th>\n",
       "      <th>Tree Number (STN/RTN)</th>\n",
       "      <th>Definition (DEF)</th>\n",
       "      <th>Examples (EX)</th>\n",
       "      <th>Usage Note (UN)</th>\n",
       "      <th>Non-Human Flag (NH)</th>\n",
       "      <th>Abbreviation (ABR)</th>\n",
       "      <th>Inverse Relation (RIN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>STY</td>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>A1.4.1.2.1.7</td>\n",
       "      <td>Amino acids and chains of amino acids connecte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>When the concept is both an enzyme and a prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aapp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Record Type (RT) Unique Identifier (UI)               Full Name (STY/RL)  \\\n",
       "97              STY                   T116  Amino Acid, Peptide, or Protein   \n",
       "\n",
       "   Tree Number (STN/RTN)                                   Definition (DEF)  \\\n",
       "97          A1.4.1.2.1.7  Amino acids and chains of amino acids connecte...   \n",
       "\n",
       "    Examples (EX)                                    Usage Note (UN)  \\\n",
       "97            NaN  When the concept is both an enzyme and a prote...   \n",
       "\n",
       "   Non-Human Flag (NH) Abbreviation (ABR) Inverse Relation (RIN)  \n",
       "97                 NaN               aapp                    NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srdef[srdef['Abbreviation (ABR)'] == 'aapp']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f01880",
   "metadata": {},
   "source": [
    "### Get the biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "62059226",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:56.601579Z",
     "start_time": "2021-11-10T07:08:53.328415Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:ToolkitGenerator:Range of slot 'in taxon' (organism taxon) does not line with the domain of its inverse (taxon of)\n",
      "WARNING:ToolkitGenerator:Range of slot 'taxon of' (named thing) does not line with the domain of its inverse (in taxon)\n"
     ]
    }
   ],
   "source": [
    "bmt_tool = Toolkit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "89c14528",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:56.606375Z",
     "start_time": "2021-11-10T07:08:56.603297Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_entity_types = srdef[srdef['Record Type (RT)'] == 'STY'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c2c14436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:08:56.625605Z",
     "start_time": "2021-11-10T07:08:56.608125Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting biolink-mapping, in the format needed to create operations\n",
    "SEMMED_entity_types['BiolinkMapping'] = [bmt_tool.get_element_by_mapping('STY:'+i)\n",
    "                                         for i in SEMMED_entity_types['Unique Identifier (UI)']]\n",
    "SEMMED_entity_types['BiolinkMapping'] = [i.title().replace(\" \", \"\") if isinstance(i, str)\n",
    "                                         else i for i in SEMMED_entity_types['BiolinkMapping']]\n",
    "\n",
    "## clean up the df\n",
    "SEMMED_entity_types = SEMMED_entity_types[['Full Name (STY/RL)', \n",
    "                                           'Abbreviation (ABR)', \n",
    "                                           'BiolinkMapping']].copy()\n",
    "SEMMED_entity_types.columns = ['FullName', \"Abbrev\", \"BiolinkMapping\"]\n",
    "SEMMED_entity_types.sort_values(by='Abbrev', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e245634",
   "metadata": {},
   "source": [
    "### Remove semantic types in data but not in SRDEF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ad0dbc8",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "One can choose whether to do this removal step or not...currently we remove them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8c8c04ce",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:07.814625Z",
     "start_time": "2021-11-10T07:08:56.627325Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "removing these that are in data, but not in SRDEF:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'carb',\n",
       " 'eico',\n",
       " 'invt',\n",
       " 'lipd',\n",
       " 'nsba',\n",
       " 'opco',\n",
       " 'orgm',\n",
       " 'rich',\n",
       " 'strd'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "only_in_data = data_types - set(SEMMED_entity_types['Abbrev'])\n",
    "print(\"removing these that are in data, but not in SRDEF:\")\n",
    "only_in_data\n",
    "\n",
    "filtered_data = filtered_data[ ~ filtered_data['SUBJECT_SEMTYPE'].isin(only_in_data)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(only_in_data)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2a5e2806",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:16.151068Z",
     "start_time": "2021-11-10T07:09:07.816458Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "117"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 127, now 117: decreased by 10 (expected)\n",
    "len(object_semtypes)  ## was 127, now 117: decreased by 10 (expected)\n",
    "len(predicates)       ## was 27, still 27 (expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51111407",
   "metadata": {},
   "source": [
    "### Clean up the mapping file to only have terms in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "22daf133",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:16.156948Z",
     "start_time": "2021-11-10T07:09:16.153085Z"
    }
   },
   "outputs": [],
   "source": [
    "## refresh set of all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "\n",
    "## get the subset of semantic network terms that are actually in the data\n",
    "SEMMED_entity_types = SEMMED_entity_types[SEMMED_entity_types['Abbrev'].isin(data_types)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d30960",
   "metadata": {},
   "source": [
    "### review mappings and deciding what to change, remove"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd9e849",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "This is a place to **STOP** and review all the SEMMED semantic types and their mapping to biolink semantic types...to decide what we are interested in keeping. This involves some knowledge of what biolink semantic types are prioritized in Translator. \n",
    "\n",
    "One can use the definitions of the SEMMED semantic types (from the SRDEF file or [browsing the UMLS vocab online](https://uts.nlm.nih.gov/uts/umls/semantic-network/root)) and the definitions of biolink semantic types (look under [things](https://github.com/biolink/biolink-model/blob/b831767e02f25c7869f760e80567ae05ceefe06c/biolink-model.yaml#L6653))\n",
    "\n",
    "See the last section (section 6) for notes on decisions that were made...\n",
    "\n",
    "---\n",
    "\n",
    "Sections 3.5 + 3.6 below involves this review, changing mappings, and removing some SEMMED semantic types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "40f60ee4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:16.168029Z",
     "start_time": "2021-11-10T07:09:16.159259Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Polypeptide', 'Disease', 'Activity', 'Cohort', None,\n",
       "       'AnatomicalEntity', 'Drug', 'Behavior', 'Phenomenon', 'Device',\n",
       "       'GrossAnatomicalStructure', 'CellularComponent',\n",
       "       'PhysiologicalProcess', 'Cell', 'ChemicalEntity',\n",
       "       'InformationContentEntity', 'ClinicalAttribute', 'Procedure',\n",
       "       'SmallMolecule', 'BiologicalEntity', 'Protein', 'Event',\n",
       "       'DiseaseOrPhenotypicFeature', 'Food', 'GeographicLocation',\n",
       "       'GenomicEntity', 'Agent', 'Publication', 'PhysicalEntity',\n",
       "       'MolecularActivity', 'MolecularEntity', 'NucleicAcidEntity',\n",
       "       'OrganismAttribute', 'PopulationOfIndividualOrganisms',\n",
       "       'PhenotypicFeature', 'Vitamin'], dtype=object)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>Vitamin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    FullName Abbrev BiolinkMapping\n",
       "104  Vitamin   vita        Vitamin"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## code used to review \n",
    "SEMMED_entity_types['BiolinkMapping'].unique()\n",
    "\n",
    "SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'] == 'Vitamin']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e641ae5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:16.186786Z",
     "start_time": "2021-11-10T07:09:16.175626Z"
    }
   },
   "outputs": [],
   "source": [
    "## re-mapping based on putting IDs into normalization service / our operation system...\n",
    "##   we have UMLS for Disease (mydisease, mychem), SmallMolecule (idisk)\n",
    "\n",
    "## Previously were mapped to ChemicalEntity, currently not mapped, so doing the mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'bacs'),'BiolinkMapping'] = 'SmallMolecule'  ## Biologically Active Substance\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'hops'),'BiolinkMapping'] = 'SmallMolecule'  ## Hazardous or Poisonous Substance\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'horm'),'BiolinkMapping'] = 'SmallMolecule'  ## Hormone\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'imft'),'BiolinkMapping'] = 'ChemicalEntity'  ## Immunologic Factor: I didn't know where to put antigens and vaccines...\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'inch'),'BiolinkMapping'] = 'SmallMolecule'  ## Inorganic Chemical\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'orch'),'BiolinkMapping'] = 'SmallMolecule'  ## Organic Chemical\n",
    "\n",
    "## currently mapped to Drug, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'antb'),'BiolinkMapping'] = 'SmallMolecule'  ## Antibiotic\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'phsu'),'BiolinkMapping'] = 'SmallMolecule'  ## Pharmacologic Substance\n",
    "\n",
    "## currently mapped to GenomicEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'gngm'),'BiolinkMapping'] = 'Gene'\n",
    "\n",
    "## currently mapped to NucleicAcidEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'nnon'),'BiolinkMapping'] = 'SmallMolecule'  ## Nucleic Acid, Nucleoside, or Nucleotide\n",
    "\n",
    "## currently mapped to Vitamin, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'vita'),'BiolinkMapping'] = 'SmallMolecule'  ## Vitamin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "a3a9da81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:16.194328Z",
     "start_time": "2021-11-10T07:09:16.188575Z"
    }
   },
   "outputs": [],
   "source": [
    "currently_unused_entities = [\n",
    "    ## mapped to Polypeptide\n",
    "    \"amas\",     \n",
    "    ## all that are mapped to Activity\n",
    "    \"acty\", \"dora\", \"edac\", \"gora\", \"hlca\", \"mcha\", \"ocac\", \"resa\",  \n",
    "    ## all that are mapped to Cohort\n",
    "    \"aggp\", \"famg\", \"podg\", \"prog\",  \n",
    "    ## missing mappings\n",
    "    \"amph\", \"anim\", \"arch\", \"bact\", \"bdsu\", \"bdsy\", \"bird\", \"blor\",\n",
    "    \"bmod\", \"bsoj\", \"chvf\", \"chvs\", \"euka\", \"ffas\", \"fish\", \"fngs\", \n",
    "    \"humn\", \"irda\", \"mamm\", \"ocdi\", \"plnt\", \"rept\", \"sbst\", \"virs\", \"vtbt\",\n",
    "    ## mapped to Anatomical Entity\n",
    "    \"anst\",\n",
    "    ## mapped to Behavior\n",
    "    \"bhvr\", \"inbe\", \"menp\", \"socb\",\n",
    "    ## mapped to Phenomenon\n",
    "    \"biof\", \"eehu\", \"hcpp\", \"lbtr\", \"npop\", \"phpr\",\n",
    "    ## mapped to Device\n",
    "    \"bodm\", \"drdd\", \"medd\", \"resd\",\n",
    "    ## mapped to GrossAnatomicalStructure\n",
    "    \"emst\", \"tisu\", \n",
    "    ## mapped to PhysiologicalProcess\n",
    "    \"genf\",\n",
    "    ## mapped to ChemicalEntity\n",
    "    \"chem\",\n",
    "    ## all that are mapped to InformationContentEntity\n",
    "    \"clas\", \"cnce\", \"ftcn\", \"grpa\", \"idcn\", \"lang\", \"qlco\", \"qnco\", \n",
    "    \"rnlw\", \"spco\", \"tmco\",\n",
    "    ## all that are mapped to ClinicalAttribute\n",
    "    \"clna\",\n",
    "    ## mapped to Procedure\n",
    "    \"lbpr\", \"mbrt\", \n",
    "    ## all that are mapped to SmallMolecule\n",
    "    \"elii\",\n",
    "    ## all that are mapped to BiologicalEntity\n",
    "    \"emod\",\n",
    "    ## mapped to Protein\n",
    "    \"rcpt\",\n",
    "    ## all that are mapped to Event\n",
    "    \"evnt\",\n",
    "    ## all that are mapped to DiseaseOrPhenotypicFeature\n",
    "    \"fndg\",\n",
    "    ## all that are mapped to GeographicLocation\n",
    "    \"geoa\",\n",
    "    ## all that are mapped to Agent\n",
    "    \"grup\", \"hcro\", \"orgt\", \"pros\", \"shro\",\n",
    "    ## all that are mapped to Publication\n",
    "    \"inpr\",\n",
    "    ## all that are mapped to PhysicalEntity\n",
    "    \"mnob\",\n",
    "    ## all that are mapped to MolecularEntity\n",
    "    \"mosq\",\n",
    "    ## mapped to NucleicAcidEntity\n",
    "    \"nusq\",\n",
    "    ## all that are mapped to OrganismAttribute\n",
    "    \"orga\",\n",
    "    ## all that are mapped to PopulationOfIndividualOrganisms\n",
    "    \"popg\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da7aa1e",
   "metadata": {},
   "source": [
    "### prune data based on what node entities to remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "308358e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:26.491044Z",
     "start_time": "2021-11-10T07:09:16.196550Z"
    }
   },
   "outputs": [],
   "source": [
    "## prune this doc\n",
    "SEMMED_entity_types = SEMMED_entity_types[ ~ SEMMED_entity_types['Abbrev'].isin(currently_unused_entities)]\n",
    "\n",
    "## prune the data doc\n",
    "filtered_data = filtered_data[ ~ filtered_data['SUBJECT_SEMTYPE'].isin(currently_unused_entities)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(currently_unused_entities)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "67861c2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.713289Z",
     "start_time": "2021-11-10T07:09:26.492454Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 117, now 35: decreased by 82\n",
    "len(object_semtypes)  ## was 117, now 35: decreased by 82\n",
    "len(predicates)       ## was 27, now 26: decreased by 1 (cool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "647e4065",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.723413Z",
     "start_time": "2021-11-10T07:09:30.714806Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Biologically Active Substance</td>\n",
       "      <td>bacs</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Body Part, Organ, or Organ Component</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Cell Component</td>\n",
       "      <td>celc</td>\n",
       "      <td>CellularComponent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Cell Function</td>\n",
       "      <td>celf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Cell</td>\n",
       "      <td>cell</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Congenital Abnormality</td>\n",
       "      <td>cgab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Clinical Drug</td>\n",
       "      <td>clnd</td>\n",
       "      <td>Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Cell or Molecular Dysfunction</td>\n",
       "      <td>comd</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Diagnostic Procedure</td>\n",
       "      <td>diap</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Enzyme</td>\n",
       "      <td>enzy</td>\n",
       "      <td>Protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Food</td>\n",
       "      <td>food</td>\n",
       "      <td>Food</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Gene or Genome</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>Hazardous or Poisonous Substance</td>\n",
       "      <td>hops</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Hormone</td>\n",
       "      <td>horm</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>Immunologic Factor</td>\n",
       "      <td>imft</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>Inorganic Chemical</td>\n",
       "      <td>inch</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Injury or Poisoning</td>\n",
       "      <td>inpo</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>mobd</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>moft</td>\n",
       "      <td>MolecularActivity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>neop</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Nucleic Acid, Nucleoside, or Nucleotide</td>\n",
       "      <td>nnon</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Organic Chemical</td>\n",
       "      <td>orch</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Organism Function</td>\n",
       "      <td>orgf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Organ or Tissue Function</td>\n",
       "      <td>ortf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Pathologic Function</td>\n",
       "      <td>patf</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Physiologic Function</td>\n",
       "      <td>phsf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Pharmacologic Substance</td>\n",
       "      <td>phsu</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Sign or Symptom</td>\n",
       "      <td>sosy</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    FullName Abbrev            BiolinkMapping\n",
       "97           Amino Acid, Peptide, or Protein   aapp               Polypeptide\n",
       "15                      Acquired Abnormality   acab                   Disease\n",
       "115                   Anatomical Abnormality   anab                   Disease\n",
       "119                               Antibiotic   antb             SmallMolecule\n",
       "101            Biologically Active Substance   bacs             SmallMolecule\n",
       "18      Body Part, Organ, or Organ Component   bpoc  GrossAnatomicalStructure\n",
       "21                            Cell Component   celc         CellularComponent\n",
       "35                             Cell Function   celf      PhysiologicalProcess\n",
       "20                                      Cell   cell                      Cell\n",
       "14                    Congenital Abnormality   cgab                   Disease\n",
       "122                            Clinical Drug   clnd                      Drug\n",
       "41             Cell or Molecular Dysfunction   comd                   Disease\n",
       "52                      Diagnostic Procedure   diap                 Procedure\n",
       "39                       Disease or Syndrome   dsyn                   Disease\n",
       "103                                   Enzyme   enzy                   Protein\n",
       "109                                     Food   food                      Food\n",
       "22                            Gene or Genome   gngm                      Gene\n",
       "107         Hazardous or Poisonous Substance   hops             SmallMolecule\n",
       "102                                  Hormone   horm             SmallMolecule\n",
       "105                       Immunologic Factor   imft            ChemicalEntity\n",
       "121                       Inorganic Chemical   inch             SmallMolecule\n",
       "29                       Injury or Poisoning   inpo                   Disease\n",
       "40          Mental or Behavioral Dysfunction   mobd                   Disease\n",
       "36                        Molecular Function   moft         MolecularActivity\n",
       "116                       Neoplastic Process   neop                   Disease\n",
       "96   Nucleic Acid, Nucleoside, or Nucleotide   nnon             SmallMolecule\n",
       "95                          Organic Chemical   orch             SmallMolecule\n",
       "32                         Organism Function   orgf      PhysiologicalProcess\n",
       "34                  Organ or Tissue Function   ortf      PhysiologicalProcess\n",
       "38                       Pathologic Function   patf                   Disease\n",
       "31                      Physiologic Function   phsf      PhysiologicalProcess\n",
       "99                   Pharmacologic Substance   phsu             SmallMolecule\n",
       "113                          Sign or Symptom   sosy         PhenotypicFeature\n",
       "53       Therapeutic or Preventive Procedure   topp                 Procedure\n",
       "104                                  Vitamin   vita             SmallMolecule"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf6f43",
   "metadata": {},
   "source": [
    "### getting mappings for predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "52496738",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.729377Z",
     "start_time": "2021-11-10T07:09:30.724706Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates = srdef[srdef['Record Type (RT)'] == 'RL'].copy()\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates[['Full Name (STY/RL)', 'Inverse Relation (RIN)'\n",
    "                                      ]].copy()\n",
    "SEMMED_predicates.columns = ['SemmedName', 'SemmedInverse']\n",
    "SEMMED_predicates.sort_values(by='SemmedName', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "635827c9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "The SRDEF file might be missing predicates that are in the data, so check that and see if you would like to add those back in..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6a3d217d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.749743Z",
     "start_time": "2021-11-10T07:09:30.731136Z"
    }
   },
   "outputs": [],
   "source": [
    "## manually add some predicates that are in the data, but not in SRDEF...\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['administered_to', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['augments', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['co-exists_with', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['converts_to', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['inhibits', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['predisposes', '']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['stimulates', '']\n",
    "\n",
    "SEMMED_predicates.sort_values(by='SemmedName', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "979e0da8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.764039Z",
     "start_time": "2021-11-10T07:09:30.751381Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['SemmedCurie']= ['SEMMEDDB:'+i \n",
    "                                   for i in SEMMED_predicates['SemmedName'].str.replace(\"-\", \"\")]\n",
    "\n",
    "SEMMED_predicates['BiolinkMapping'] = [bmt_tool.get_element_by_mapping(i)\n",
    "                                       for i in SEMMED_predicates['SemmedCurie']]\n",
    "\n",
    "SEMMED_predicates['BiolinkMapping'] = SEMMED_predicates['BiolinkMapping'].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5d7f7944",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.775789Z",
     "start_time": "2021-11-10T07:09:30.766234Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedName</th>\n",
       "      <th>SemmedInverse</th>\n",
       "      <th>SemmedCurie</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:administered_to</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMEDDB:affects</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:associated_with</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:augments</td>\n",
       "      <td>positively_regulates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMEDDB:causes</td>\n",
       "      <td>causes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>co-exists_with</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:coexists_with</td>\n",
       "      <td>coexists_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>complicated_by</td>\n",
       "      <td>SEMMEDDB:complicates</td>\n",
       "      <td>interacts_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:converts_to</td>\n",
       "      <td>derives_into</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>diagnosed_by</td>\n",
       "      <td>SEMMEDDB:diagnoses</td>\n",
       "      <td>biomarker_for</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>SEMMEDDB:disrupts</td>\n",
       "      <td>disrupts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:inhibits</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:interacts_with</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>has_location</td>\n",
       "      <td>SEMMEDDB:location_of</td>\n",
       "      <td>location_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>SEMMEDDB:manifestation_of</td>\n",
       "      <td>manifestation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>has_occurrence</td>\n",
       "      <td>SEMMEDDB:occurs_in</td>\n",
       "      <td>occurs_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>has_part</td>\n",
       "      <td>SEMMEDDB:part_of</td>\n",
       "      <td>part_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>follows</td>\n",
       "      <td>SEMMEDDB:precedes</td>\n",
       "      <td>precedes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:predisposes</td>\n",
       "      <td>predisposes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>SEMMEDDB:prevents</td>\n",
       "      <td>prevents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>has_process</td>\n",
       "      <td>SEMMEDDB:process_of</td>\n",
       "      <td>occurs_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>SEMMEDDB:produces</td>\n",
       "      <td>produces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:stimulates</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMEDDB:treats</td>\n",
       "      <td>treats</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedName      SemmedInverse                SemmedCurie  \\\n",
       "54    administered_to                      SEMMEDDB:administered_to   \n",
       "145           affects        affected_by           SEMMEDDB:affects   \n",
       "160   associated_with    associated_with   SEMMEDDB:associated_with   \n",
       "55           augments                             SEMMEDDB:augments   \n",
       "141            causes          caused_by            SEMMEDDB:causes   \n",
       "56     co-exists_with                        SEMMEDDB:coexists_with   \n",
       "143       complicates     complicated_by       SEMMEDDB:complicates   \n",
       "57        converts_to                          SEMMEDDB:converts_to   \n",
       "157         diagnoses       diagnosed_by         SEMMEDDB:diagnoses   \n",
       "140          disrupts       disrupted_by          SEMMEDDB:disrupts   \n",
       "58           inhibits                             SEMMEDDB:inhibits   \n",
       "136    interacts_with     interacts_with    SEMMEDDB:interacts_with   \n",
       "129       location_of       has_location       SEMMEDDB:location_of   \n",
       "144  manifestation_of  has_manifestation  SEMMEDDB:manifestation_of   \n",
       "146         occurs_in     has_occurrence         SEMMEDDB:occurs_in   \n",
       "127           part_of           has_part           SEMMEDDB:part_of   \n",
       "132          precedes            follows          SEMMEDDB:precedes   \n",
       "59        predisposes                          SEMMEDDB:predisposes   \n",
       "142          prevents       prevented_by          SEMMEDDB:prevents   \n",
       "134        process_of        has_process        SEMMEDDB:process_of   \n",
       "138          produces        produced_by          SEMMEDDB:produces   \n",
       "60         stimulates                           SEMMEDDB:stimulates   \n",
       "148            treats         treated_by            SEMMEDDB:treats   \n",
       "\n",
       "                         BiolinkMapping  \n",
       "54                              affects  \n",
       "145                             affects  \n",
       "160                          related_to  \n",
       "55                 positively_regulates  \n",
       "141                              causes  \n",
       "56                        coexists_with  \n",
       "143                      interacts_with  \n",
       "57                         derives_into  \n",
       "157                       biomarker_for  \n",
       "140                            disrupts  \n",
       "58   entity_negatively_regulates_entity  \n",
       "136           physically_interacts_with  \n",
       "129                         location_of  \n",
       "144                    manifestation_of  \n",
       "146                           occurs_in  \n",
       "127                             part_of  \n",
       "132                            precedes  \n",
       "59                          predisposes  \n",
       "142                            prevents  \n",
       "134                           occurs_in  \n",
       "138                            produces  \n",
       "60   entity_positively_regulates_entity  \n",
       "148                              treats  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates[SEMMED_predicates['BiolinkMapping'].notna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c658e0a2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Below some mappings for predicates are changed. Check and decide whether you want to do this and how...\n",
    "\n",
    "---\n",
    "Going to re-map for these entries..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "d7650efb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.781781Z",
     "start_time": "2021-11-10T07:09:30.777340Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## since it's currently mapped to a mixin....move it to a real predicate\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedName'] == 'augments'),'BiolinkMapping'] = 'entity_positively_regulates_entity' \n",
    "\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedName'] == 'complicates'),'BiolinkMapping'] = 'exacerbates' \n",
    "\n",
    "## the \"diagnoses\" predicate is more of a \"this thing distinguishes / identifies this other thing\"...\n",
    "##   so I didn't want to leave it mapped to \"biomarker_for\"\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedName'] == 'diagnoses'),'BiolinkMapping'] = 'related_to' \n",
    "\n",
    "## treats and administered_to (as affects) are\n",
    "##   also a bit strong (it was used but unclear whether it actually helped or had effect)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72039ac9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Check and decide whether you want to do this\n",
    "\n",
    "---\n",
    "\n",
    "Missing mappings for:\n",
    "- measures \n",
    "- method of\n",
    "- uses\n",
    "\n",
    "Going to remove those combos from the data in section 3.8 later..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80e84505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.786607Z",
     "start_time": "2021-11-10T07:09:30.783500Z"
    }
   },
   "outputs": [],
   "source": [
    "## only keep the stuff we found biolink mappings for\n",
    "SEMMED_predicates = SEMMED_predicates[SEMMED_predicates['BiolinkMapping'].notna()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "52cfbbc4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.792883Z",
     "start_time": "2021-11-10T07:09:30.788546Z"
    }
   },
   "outputs": [],
   "source": [
    "## get inverse predicates so generating reverse operations is easier\n",
    "\n",
    "SEMMED_predicates['BiolinkInverse'] = [bmt_tool.get_element(i).inverse \n",
    "                                       if isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "                                       else i \n",
    "                                       for i in SEMMED_predicates['BiolinkMapping']]\n",
    "\n",
    "SEMMED_predicates['BiolinkInverse'] = SEMMED_predicates['BiolinkInverse'].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9718fe3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.804148Z",
     "start_time": "2021-11-10T07:09:30.794864Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedName</th>\n",
       "      <th>SemmedInverse</th>\n",
       "      <th>SemmedCurie</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "      <th>BiolinkInverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:administered_to</td>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMEDDB:affects</td>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:associated_with</td>\n",
       "      <td>related_to</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:augments</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>entity_positively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMEDDB:causes</td>\n",
       "      <td>causes</td>\n",
       "      <td>caused_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>co-exists_with</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:coexists_with</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>coexists_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>complicated_by</td>\n",
       "      <td>SEMMEDDB:complicates</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>exacerbates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:converts_to</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>derives_from</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>diagnosed_by</td>\n",
       "      <td>SEMMEDDB:diagnoses</td>\n",
       "      <td>related_to</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>SEMMEDDB:disrupts</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>disrupted_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:inhibits</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>entity_negatively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:interacts_with</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>has_location</td>\n",
       "      <td>SEMMEDDB:location_of</td>\n",
       "      <td>location_of</td>\n",
       "      <td>located_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>SEMMEDDB:manifestation_of</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>has_manifestation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>has_occurrence</td>\n",
       "      <td>SEMMEDDB:occurs_in</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>contains_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>has_part</td>\n",
       "      <td>SEMMEDDB:part_of</td>\n",
       "      <td>part_of</td>\n",
       "      <td>has_part</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>follows</td>\n",
       "      <td>SEMMEDDB:precedes</td>\n",
       "      <td>precedes</td>\n",
       "      <td>preceded_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:predisposes</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>predisposes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>SEMMEDDB:prevents</td>\n",
       "      <td>prevents</td>\n",
       "      <td>prevented_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>has_process</td>\n",
       "      <td>SEMMEDDB:process_of</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>contains_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>SEMMEDDB:produces</td>\n",
       "      <td>produces</td>\n",
       "      <td>produced_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:stimulates</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>entity_positively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMEDDB:treats</td>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedName      SemmedInverse                SemmedCurie  \\\n",
       "54    administered_to                      SEMMEDDB:administered_to   \n",
       "145           affects        affected_by           SEMMEDDB:affects   \n",
       "160   associated_with    associated_with   SEMMEDDB:associated_with   \n",
       "55           augments                             SEMMEDDB:augments   \n",
       "141            causes          caused_by            SEMMEDDB:causes   \n",
       "56     co-exists_with                        SEMMEDDB:coexists_with   \n",
       "143       complicates     complicated_by       SEMMEDDB:complicates   \n",
       "57        converts_to                          SEMMEDDB:converts_to   \n",
       "157         diagnoses       diagnosed_by         SEMMEDDB:diagnoses   \n",
       "140          disrupts       disrupted_by          SEMMEDDB:disrupts   \n",
       "58           inhibits                             SEMMEDDB:inhibits   \n",
       "136    interacts_with     interacts_with    SEMMEDDB:interacts_with   \n",
       "129       location_of       has_location       SEMMEDDB:location_of   \n",
       "144  manifestation_of  has_manifestation  SEMMEDDB:manifestation_of   \n",
       "146         occurs_in     has_occurrence         SEMMEDDB:occurs_in   \n",
       "127           part_of           has_part           SEMMEDDB:part_of   \n",
       "132          precedes            follows          SEMMEDDB:precedes   \n",
       "59        predisposes                          SEMMEDDB:predisposes   \n",
       "142          prevents       prevented_by          SEMMEDDB:prevents   \n",
       "134        process_of        has_process        SEMMEDDB:process_of   \n",
       "138          produces        produced_by          SEMMEDDB:produces   \n",
       "60         stimulates                           SEMMEDDB:stimulates   \n",
       "148            treats         treated_by            SEMMEDDB:treats   \n",
       "\n",
       "                         BiolinkMapping                         BiolinkInverse  \n",
       "54                              affects                            affected_by  \n",
       "145                             affects                            affected_by  \n",
       "160                          related_to                             related_to  \n",
       "55   entity_positively_regulates_entity  entity_positively_regulated_by_entity  \n",
       "141                              causes                              caused_by  \n",
       "56                        coexists_with                          coexists_with  \n",
       "143                         exacerbates                            exacerbates  \n",
       "57                         derives_into                           derives_from  \n",
       "157                          related_to                             related_to  \n",
       "140                            disrupts                           disrupted_by  \n",
       "58   entity_negatively_regulates_entity  entity_negatively_regulated_by_entity  \n",
       "136           physically_interacts_with              physically_interacts_with  \n",
       "129                         location_of                             located_in  \n",
       "144                    manifestation_of                      has_manifestation  \n",
       "146                           occurs_in                       contains_process  \n",
       "127                             part_of                               has_part  \n",
       "132                            precedes                            preceded_by  \n",
       "59                          predisposes                            predisposes  \n",
       "142                            prevents                           prevented_by  \n",
       "134                           occurs_in                       contains_process  \n",
       "138                            produces                            produced_by  \n",
       "60   entity_positively_regulates_entity  entity_positively_regulated_by_entity  \n",
       "148                              treats                             treated_by  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933327b9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Check if the predicate inverses look correct. Currently some aren't correct because they don't have an inverse but they are directional...\n",
    "\n",
    "This is why they need to be manually changed. \n",
    "\n",
    "Some predicates aren't directional (`symmetric == true`) so those are identical in either direction..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1b12248c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.810225Z",
     "start_time": "2021-11-10T07:09:30.806043Z"
    }
   },
   "outputs": [],
   "source": [
    "## currently no inverse but it needs one\n",
    "\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedName'] == 'complicates'),'BiolinkInverse'] = 'related_to' \n",
    "\n",
    "\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedName'] == 'predisposes'),'BiolinkInverse'] = 'risk_affected_by' \n",
    "## using inverse of the parent term (affects risk for)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4727f8d5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.821737Z",
     "start_time": "2021-11-10T07:09:30.812008Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedName</th>\n",
       "      <th>SemmedInverse</th>\n",
       "      <th>SemmedCurie</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "      <th>BiolinkInverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:administered_to</td>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>SEMMEDDB:affects</td>\n",
       "      <td>affects</td>\n",
       "      <td>affected_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:associated_with</td>\n",
       "      <td>related_to</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:augments</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>entity_positively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>SEMMEDDB:causes</td>\n",
       "      <td>causes</td>\n",
       "      <td>caused_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>co-exists_with</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:coexists_with</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>coexists_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>complicated_by</td>\n",
       "      <td>SEMMEDDB:complicates</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:converts_to</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>derives_from</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>diagnosed_by</td>\n",
       "      <td>SEMMEDDB:diagnoses</td>\n",
       "      <td>related_to</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>SEMMEDDB:disrupts</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>disrupted_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:inhibits</td>\n",
       "      <td>entity_negatively_regulates_entity</td>\n",
       "      <td>entity_negatively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:interacts_with</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "      <td>physically_interacts_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>has_location</td>\n",
       "      <td>SEMMEDDB:location_of</td>\n",
       "      <td>location_of</td>\n",
       "      <td>located_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>SEMMEDDB:manifestation_of</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>has_manifestation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>has_occurrence</td>\n",
       "      <td>SEMMEDDB:occurs_in</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>contains_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>has_part</td>\n",
       "      <td>SEMMEDDB:part_of</td>\n",
       "      <td>part_of</td>\n",
       "      <td>has_part</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>follows</td>\n",
       "      <td>SEMMEDDB:precedes</td>\n",
       "      <td>precedes</td>\n",
       "      <td>preceded_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:predisposes</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>risk_affected_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>SEMMEDDB:prevents</td>\n",
       "      <td>prevents</td>\n",
       "      <td>prevented_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>has_process</td>\n",
       "      <td>SEMMEDDB:process_of</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>contains_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>SEMMEDDB:produces</td>\n",
       "      <td>produces</td>\n",
       "      <td>produced_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td></td>\n",
       "      <td>SEMMEDDB:stimulates</td>\n",
       "      <td>entity_positively_regulates_entity</td>\n",
       "      <td>entity_positively_regulated_by_entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>SEMMEDDB:treats</td>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedName      SemmedInverse                SemmedCurie  \\\n",
       "54    administered_to                      SEMMEDDB:administered_to   \n",
       "145           affects        affected_by           SEMMEDDB:affects   \n",
       "160   associated_with    associated_with   SEMMEDDB:associated_with   \n",
       "55           augments                             SEMMEDDB:augments   \n",
       "141            causes          caused_by            SEMMEDDB:causes   \n",
       "56     co-exists_with                        SEMMEDDB:coexists_with   \n",
       "143       complicates     complicated_by       SEMMEDDB:complicates   \n",
       "57        converts_to                          SEMMEDDB:converts_to   \n",
       "157         diagnoses       diagnosed_by         SEMMEDDB:diagnoses   \n",
       "140          disrupts       disrupted_by          SEMMEDDB:disrupts   \n",
       "58           inhibits                             SEMMEDDB:inhibits   \n",
       "136    interacts_with     interacts_with    SEMMEDDB:interacts_with   \n",
       "129       location_of       has_location       SEMMEDDB:location_of   \n",
       "144  manifestation_of  has_manifestation  SEMMEDDB:manifestation_of   \n",
       "146         occurs_in     has_occurrence         SEMMEDDB:occurs_in   \n",
       "127           part_of           has_part           SEMMEDDB:part_of   \n",
       "132          precedes            follows          SEMMEDDB:precedes   \n",
       "59        predisposes                          SEMMEDDB:predisposes   \n",
       "142          prevents       prevented_by          SEMMEDDB:prevents   \n",
       "134        process_of        has_process        SEMMEDDB:process_of   \n",
       "138          produces        produced_by          SEMMEDDB:produces   \n",
       "60         stimulates                           SEMMEDDB:stimulates   \n",
       "148            treats         treated_by            SEMMEDDB:treats   \n",
       "\n",
       "                         BiolinkMapping                         BiolinkInverse  \n",
       "54                              affects                            affected_by  \n",
       "145                             affects                            affected_by  \n",
       "160                          related_to                             related_to  \n",
       "55   entity_positively_regulates_entity  entity_positively_regulated_by_entity  \n",
       "141                              causes                              caused_by  \n",
       "56                        coexists_with                          coexists_with  \n",
       "143                         exacerbates                             related_to  \n",
       "57                         derives_into                           derives_from  \n",
       "157                          related_to                             related_to  \n",
       "140                            disrupts                           disrupted_by  \n",
       "58   entity_negatively_regulates_entity  entity_negatively_regulated_by_entity  \n",
       "136           physically_interacts_with              physically_interacts_with  \n",
       "129                         location_of                             located_in  \n",
       "144                    manifestation_of                      has_manifestation  \n",
       "146                           occurs_in                       contains_process  \n",
       "127                             part_of                               has_part  \n",
       "132                            precedes                            preceded_by  \n",
       "59                          predisposes                       risk_affected_by  \n",
       "142                            prevents                           prevented_by  \n",
       "134                           occurs_in                       contains_process  \n",
       "138                            produces                            produced_by  \n",
       "60   entity_positively_regulates_entity  entity_positively_regulated_by_entity  \n",
       "148                              treats                             treated_by  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c616c4ff",
   "metadata": {},
   "source": [
    "### prune the data based on what predicates to keep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8b1e2b",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Note that this step is used to remove the predicates that were dropped earlier...\n",
    "\n",
    "now to filter out the predicates I dropped (couldn't map / didn't find useful):\n",
    "- measures\n",
    "- method of\n",
    "- uses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "50027e5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.827545Z",
     "start_time": "2021-11-10T07:09:30.823684Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['NameInDataFormat'] = SEMMED_predicates['SemmedName'].str.upper()\n",
    "SEMMED_predicates['NameInDataFormat'] = SEMMED_predicates['NameInDataFormat'].str.replace(\"-\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5e25a929",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:30.832767Z",
     "start_time": "2021-11-10T07:09:30.829364Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'MEASURES', 'METHOD_OF', 'USES'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_removals = predicates - set(SEMMED_predicates['NameInDataFormat'])\n",
    "more_removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "64cacd52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:33.144444Z",
     "start_time": "2021-11-10T07:09:30.834150Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(more_removals)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "530e3631",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:37.180141Z",
     "start_time": "2021-11-10T07:09:33.145811Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 35, still 35 (expected)\n",
    "len(object_semtypes)  ## was 35, still 35 (expected)\n",
    "len(predicates)       ## was 26, now 23: decreased by 3 (expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e9660d",
   "metadata": {},
   "source": [
    "## Looking at combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "863543b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:09:43.076672Z",
     "start_time": "2021-11-10T07:09:37.182217Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4211, 4)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>707094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>696732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>phsu</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>660654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cell</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>598884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>542301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "0            bpoc  LOCATION_OF           neop  1023005\n",
       "1            bpoc  LOCATION_OF           aapp   953010\n",
       "2            topp       TREATS           dsyn   849825\n",
       "3            cell  LOCATION_OF           aapp   815374\n",
       "4            bpoc  LOCATION_OF           patf   799234\n",
       "5            topp       TREATS           neop   707094\n",
       "6            bpoc  LOCATION_OF           dsyn   696732\n",
       "7            phsu       TREATS           dsyn   660654\n",
       "8            cell      PART_OF           bpoc   598884\n",
       "9            bpoc      PART_OF           bpoc   542301"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at number of combos after this removal\n",
    "combos = filtered_data.value_counts().reset_index()\n",
    "combos.columns = ['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE', 'COUNT']\n",
    "combos.shape\n",
    "combos.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305f3a3a",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "use the code block below to decide how many combos to keep based on how many predications/records there are per combo...\n",
    "\n",
    "---\n",
    "\n",
    "Current:\n",
    "\n",
    "let's try setting a limit that we're only building operations if there are > 100 objects for the triple..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "86920c33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:20.188351Z",
     "start_time": "2021-11-10T07:12:20.178793Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2180</th>\n",
       "      <td>antb</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>celf</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>food</td>\n",
       "      <td>CAUSES</td>\n",
       "      <td>anab</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2182</th>\n",
       "      <td>clnd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>anab</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2183</th>\n",
       "      <td>antb</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>imft</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2184</th>\n",
       "      <td>imft</td>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>inpo</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2185 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "0               bpoc  LOCATION_OF           neop  1023005\n",
       "1               bpoc  LOCATION_OF           aapp   953010\n",
       "2               topp       TREATS           dsyn   849825\n",
       "3               cell  LOCATION_OF           aapp   815374\n",
       "4               bpoc  LOCATION_OF           patf   799234\n",
       "...              ...          ...            ...      ...\n",
       "2180            antb     AUGMENTS           celf      102\n",
       "2181            food       CAUSES           anab      102\n",
       "2182            clnd       TREATS           anab      101\n",
       "2183            antb     INHIBITS           imft      101\n",
       "2184            imft  PREDISPOSES           inpo      101\n",
       "\n",
       "[2185 rows x 4 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos[(combos['COUNT'] > 100)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "2c859df9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:25.964405Z",
     "start_time": "2021-11-10T07:12:25.958876Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2185, 3)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos = combos[(combos['COUNT'] > 100)].copy()\n",
    "filtered_combos.drop(columns='COUNT', inplace=True)\n",
    "filtered_combos.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "420d44a1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:27.248773Z",
     "start_time": "2021-11-10T07:12:27.244930Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['FullName', 'Abbrev', 'BiolinkMapping'], dtype='object')"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b53030eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:29.904644Z",
     "start_time": "2021-11-10T07:12:29.897563Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all subject to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']], \n",
    "                                        how='left', left_on='SUBJECT_SEMTYPE', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject', \n",
    "                           'BiolinkSubject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "739e9d15",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:30.489328Z",
     "start_time": "2021-11-10T07:12:30.482226Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all predicate to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_predicates[['NameInDataFormat', 'BiolinkMapping', 'BiolinkInverse']],\n",
    "                                        how='left', left_on='OriginalPredicate', right_on='NameInDataFormat')\n",
    "\n",
    "filtered_combos.drop(columns = 'NameInDataFormat', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject', \n",
    "                           'BiolinkSubject', 'BiolinkPredicate', 'BiolinkInversePred']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "9e435399",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:32.595947Z",
     "start_time": "2021-11-10T07:12:32.586023Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all object to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']],\n",
    "                                        how='left', left_on='OriginalObject', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject', \n",
    "                           'BiolinkSubject', 'BiolinkPredicate', 'BiolinkInversePred', 'BiolinkObject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "adddd021",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:33.276014Z",
     "start_time": "2021-11-10T07:12:33.267417Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkPredicate</th>\n",
       "      <th>BiolinkInversePred</th>\n",
       "      <th>BiolinkObject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>location_of</td>\n",
       "      <td>located_in</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>location_of</td>\n",
       "      <td>located_in</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>treats</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  OriginalSubject OriginalPredicate OriginalObject            BiolinkSubject  \\\n",
       "0            bpoc       LOCATION_OF           neop  GrossAnatomicalStructure   \n",
       "1            bpoc       LOCATION_OF           aapp  GrossAnatomicalStructure   \n",
       "2            topp            TREATS           dsyn                 Procedure   \n",
       "\n",
       "  BiolinkPredicate BiolinkInversePred BiolinkObject  \n",
       "0      location_of         located_in       Disease  \n",
       "1      location_of         located_in   Polypeptide  \n",
       "2           treats         treated_by       Disease  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870db906",
   "metadata": {},
   "source": [
    "## Generate operation yaml!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344f9ec6",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* If needed, change the code within the functions below to change the x-bte annotations that are made...\n",
    "* the yaml created below refers to umls-subj and umls-obj...those are specified [here close to the bottom](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1fcf921d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:12:37.919271Z",
     "start_time": "2021-11-10T07:12:37.916144Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml=ryml.YAML()\n",
    "folded = ryml.scalarstring.FoldedScalarString"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4849cde4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:17:33.994459Z",
     "start_time": "2021-11-10T07:17:33.988576Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_forward_op(original_subj, original_pred, original_obj,\n",
    "                        biolink_subj, biolink_pred, biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    normal_op_name = f\"{original_subj}-{original_pred}-{original_obj}\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## original direction: subject -> object\n",
    "    normal_op_body = folded(\n",
    "    '{' \\\n",
    "        '\"q\": [ {{ queryInputs | wrap( \\'[\"\\', \\'\",' + f'\"{original_pred}\",\"{original_obj}\",1,1]\\' )' + ' }} ], ' \\\n",
    "        '\"scopes\": [\"subject.umls\", \"predicate\", \"object.semantic_type_abbreviation\", ' \\\n",
    "                   '\"subject.novelty\", \"object.novelty\"]' \\\n",
    "    '}')\n",
    "    \n",
    "    ## create operation...\n",
    "    temp = {\n",
    "        ## original direction: subject -> object\n",
    "        normal_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': 'UMLS',\n",
    "                        'semantic': biolink_subj  ## input is subject!\n",
    "                    }\n",
    "                ],\n",
    "                'requestBodyType': 'object',\n",
    "                'requestBody': {'body': normal_op_body},\n",
    "                'parameters': {\n",
    "                    'fields': 'object.umls,pmid,subject.umls,subject.name,predicate,object.name',\n",
    "                    'size': POST_size\n",
    "                },\n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': 'UMLS',\n",
    "                        'semantic': biolink_obj  ## output is object\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_pred,\n",
    "                'source': 'infores:semmeddb',\n",
    "                'response_mapping': {\n",
    "                    \"$ref\": '#/components/x-bte-response-mapping/umls-obj'  ## matches output as object\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "03e65b67",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:18:03.175790Z",
     "start_time": "2021-11-10T07:18:03.170546Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_reverse_op(original_subj, original_pred, original_obj,\n",
    "                            biolink_subj, biolink_inverse_pred,  ## NOTICE THE INVERSE PRED USED HERE\n",
    "                            biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    rev_op_name = f\"{original_subj}-{original_pred}-{original_obj}-rev\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## reverse direction: object -> subject\n",
    "    rev_op_body = folded(\n",
    "    '{' \\\n",
    "        '\"q\": [ {{ queryInputs | wrap( \\'[\"\\', \\'\",' + f'\"{original_pred}\",\"{original_subj}\",1,1]\\' )' + '}}], ' \\\n",
    "        '\"scopes\": [\"object.umls\", \"predicate\", \"subject.semantic_type_abbreviation\", ' \\\n",
    "                   '\"subject.novelty\", \"object.novelty\"]' \\\n",
    "    '}')\n",
    "    \n",
    "    ## create the operation...\n",
    "    temp = {\n",
    "        ## reverse direction: object -> subject\n",
    "        rev_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': 'UMLS',\n",
    "                        'semantic': biolink_obj  ## input is object!\n",
    "                    }\n",
    "                ],\n",
    "                'requestBodyType': 'object',\n",
    "                'requestBody': {'body': rev_op_body},\n",
    "                'parameters': {\n",
    "                    'fields': 'subject.umls,pmid,subject.name,predicate,object.umls,object.name',\n",
    "                    'size': POST_size\n",
    "                },\n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': 'UMLS',\n",
    "                        'semantic': biolink_subj  ## output is subject\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                'source': 'infores:semmeddb',\n",
    "                'response_mapping': {\n",
    "                    \"$ref\": '#/components/x-bte-response-mapping/umls-subj'  ## matches output as subj\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "db9f9f81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:18:07.536522Z",
     "start_time": "2021-11-10T07:18:07.530595Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_all_operations(combo_df):\n",
    "    op_tracking = set()\n",
    "    \n",
    "    saved = dict()\n",
    "    ## iterate through rows of combos dataframe\n",
    "    for row in combo_df.itertuples(index = False):\n",
    "        \n",
    "        ## forward: only make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,object used here)\n",
    "        forward_op_record = f\"{row.BiolinkSubject}-{row.OriginalPredicate}-{row.OriginalObject}\"\n",
    "        if forward_op_record not in op_tracking:\n",
    "            op_tracking.add(forward_op_record)\n",
    "            saved.update(generate_forward_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_pred = row.BiolinkPredicate,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                            ))\n",
    "            \n",
    "        ## reverse: make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,subject used here)\n",
    "        reverse_op_record = f\"rev-{row.BiolinkObject}-{row.OriginalPredicate}-{row.OriginalSubject}\"\n",
    "        if reverse_op_record not in op_tracking:\n",
    "            op_tracking.add(reverse_op_record)\n",
    "            saved.update(generate_reverse_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_inverse_pred = row.BiolinkInversePred,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                             ))\n",
    "            \n",
    "    final = {\"x-bte-kgs-operations\": saved}\n",
    "    return final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ec9a4",
   "metadata": {},
   "source": [
    "Get the file made!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "3856781d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:18:49.809786Z",
     "start_time": "2021-11-10T07:18:49.738945Z"
    }
   },
   "outputs": [],
   "source": [
    "kgs_operations = generate_all_operations(filtered_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "fdcd4202",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:18:51.770174Z",
     "start_time": "2021-11-10T07:18:51.766700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1883"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd057af2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* it's cool how condensed the operations become, due to the way querying is done (keeping track of not creating duplicated operations using the sets in generate_all_operations function)\n",
    "* set where to download the yamls in the code chunks below\n",
    "  * operations_path\n",
    "  * list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "239dce9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:19:01.218914Z",
     "start_time": "2021-11-10T07:18:57.796286Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.boolean_representation = ['False', 'True']\n",
    "\n",
    "operations_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_operations.yaml\")\n",
    "\n",
    "yaml.dump(kgs_operations, operations_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c946247a",
   "metadata": {},
   "source": [
    "Wait a sec! Need the operations list too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "27403608",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:19:05.195777Z",
     "start_time": "2021-11-10T07:19:05.192822Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_kgs_operations_list(operations_dict):\n",
    "    kgs_op_list = []\n",
    "    for key in operations_dict.keys():\n",
    "        kgs_op_list.append( {\"$ref\": f\"#/components/x-bte-kgs-operations/{key}\"} )\n",
    "    final2 = {\"x-bte-kgs-operations\": kgs_op_list}\n",
    "    return final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "4bec97ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:19:07.667746Z",
     "start_time": "2021-11-10T07:19:07.664616Z"
    }
   },
   "outputs": [],
   "source": [
    "operations_list = generate_kgs_operations_list(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "e64c34db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T07:19:08.955821Z",
     "start_time": "2021-11-10T07:19:08.663910Z"
    }
   },
   "outputs": [],
   "source": [
    "list_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_list.yaml\")\n",
    "\n",
    "yaml.dump(operations_list, list_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421fbcdc",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* now the yaml segments downloaded have to be indented manually and inserted into the correct sections of the smartapi yaml...\n",
    "  * It's easier to do with an IDE like Visual Code studio where one can select large sections of text\n",
    "  * the amount of indent to do and where to put things is specified in the [yaml that acts as a template](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)\n",
    "  * the finished file is meant to be [here](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) so one could select-paste the sections directly there\n",
    "  \n",
    "---\n",
    "\n",
    "The code below is optional, in case one wants to convert the yaml to json (for BTE's test/query endpoint testing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ebd89060",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2021-11-10T08:25:45.918664Z",
     "start_time": "2021-11-10T08:25:36.630127Z"
    }
   },
   "outputs": [],
   "source": [
    "## extra code in case we want to convert to json...\n",
    "\n",
    "yaml_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"smartapi.yaml\")\n",
    "\n",
    "here = yaml.load(yaml_path)\n",
    "\n",
    "json_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"smartapi.json\")\n",
    "\n",
    "with open(json_path, 'w') as file:\n",
    "    json.dump(here, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa95cb1",
   "metadata": {},
   "source": [
    "## Notes on choices here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6bbe17",
   "metadata": {},
   "source": [
    "### Leaving"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0906866",
   "metadata": {},
   "source": [
    "* leaving aapp as Polypeptide (will be a chemical unless Protein is also in its SRI ID resolution (then it'll be Gene too))\n",
    "* leaving Disease mappings as-is, but some seem like they could be PathologicalAnatomicalStructure (acab, anab, cgab), PathologicalProcess (comd, patf), or PhenotypicFeature (inpo) instead\n",
    "* left clnd (Clinical Drug) as Drug since it really seemed like a drug (dosage)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce33fcc5",
   "metadata": {},
   "source": [
    "### Removing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb04a8e",
   "metadata": {},
   "source": [
    "* was mapped to Polypeptide:\n",
    "    * amas: (Amino Acid Sequence) looks like protein \"domains\". Examples: Nuclear Export Signals, DNA Binding Domain\n",
    "* everything mapped to Activity (8)\n",
    "    * acty (Activity) Examples: War, Retirement, Euthanasia, Lifting\n",
    "    * dora (Daily or Recreational Activity) Examples: Physical activity, Light Exercise, Relaxation\n",
    "    * edac (Educational Activity) Examples: Training, Medical Residencies\n",
    "    * gora (Governmental or Regulatory Activity) Examples: Health Care Reform, Advisory Committees\n",
    "    * hlca (Health Care Activity) Examples: follow-up, Diagnosis\n",
    "    * mcha (Machine Activity) Examples: Refridgeration, Neural Network Simulation\n",
    "    * ocac (Occupational Activity) Examples: Promotion, Work, Mining\n",
    "    * resa (Research Activity) Examples: Clinical Trials, research study\n",
    "* everything mapped to Cohort (4)\n",
    "    * aggp (Age Group) Examples: Infant, Child, Adult, Elderly\n",
    "    * famg (Family Group) Examples: spouse, Sister, Foster Parent\n",
    "    * podg (Patient or Disabled Group) Examples: Patients\n",
    "    * prog (Professional or Occupational Group) Examples: Administrators, Employee, Author\n",
    "* missing mappings to biolink-model (25):\n",
    "    * amph (Amphibian) Examples: Toad, Bufo boreas, Anura\n",
    "    * anim (Animal) Examples: Animals, Laboratory /  Control Animal\n",
    "    * arch (Archaeon) Examples: Archaea, halophilic bacteria, Thermoplasma acidophilum\n",
    "    * bact (Bacterium) Examples: Escherichia coli, Salmonella, Borrelia burgdorferi\n",
    "    * bdsu (Body Substance) too general. Examples: Urine, Milk, Lymph, Urine specimen\n",
    "    * bdsy (Body System) too general. Examples: hypothalamic-pituitary-adrenal axis, Neurosecretory Systems\n",
    "    * bird (Bird) Examples: Geese, Passeriformes, Raptors\n",
    "    * blor (Body Location or Region) too general. Examples: Hepatic, Lysosomal, Cytoplasmic\n",
    "    * bmod (Biomedical Occupation or Discipline) Examples: Medicine, Dentistry, Midwifery\n",
    "    * bsoj (Body Space or Junction) too general. Examples: Compartments, Synapses, Cistern\n",
    "    * chvf (Chemical Viewed Functionally) too general. Examples: inhibitors, antagonists, Agent\n",
    "    * chvs (Chemical Viewed Structurally) too general. Examples: particle, solid state, vapor\n",
    "    * euka (Eukaryote) Examples: Wasps, Protozoan parasite\n",
    "    * ffas (Fully Formed Anatomical Structure) Examples: Carcass\n",
    "    * fish (Fish) Examples: Eels, Fishes, Electric Fish\n",
    "    * fngs (Fungus) Examples: Saccharomyces cerevisiae, Alternaria brassicicola, fungus\n",
    "    * humn (Human) Examples: Family, Patients, Males\n",
    "    * irda (Indicator, Reagent, or Diagnostic Aid) Examples: Fluorescent Probes, Chelating Agents\n",
    "    * mamm (Mammal) Examples: Rattus norvegicus, Felis catus, Mus\n",
    "    * ocdi (Occupation or Discipline) Examples: Science, Politics\n",
    "    * plnt (Plant) Examples: Chrysanthemum x morifolium, Pollen, Oryza sativa\n",
    "    * rept (Reptile) Examples: Snakes, Turtles, Reptiles\n",
    "    * sbst (Substance) too general. Examples: Materials, Plastics, Photons, Substance\n",
    "    * virs (Virus) Examples: Herpesvirus 4, Human / GB virus C / Herpesviridae\n",
    "    * vtbt (Vertebrate) Examples: Vertebrates / Poikilotherm, NOS\n",
    "* was mapped to AnatomicalEntity:\n",
    "    * anst (Anatomical Structure) Examples: Entire fetus, Whole body, Cadaver\n",
    "* was mapped to Behavior (4)\n",
    "    * bhvr (Behavior) too general. Examples: Sexuality, Nest Building, Behavioral phenotype\n",
    "    * inbe (Individual Behavior) too general. Examples: impulsivity, Habits, Performance\n",
    "    * menp (Mental Process) too general. Examples: mind control, Learning, experience\n",
    "    * socb (Social Behavior) too general. Examples: Communication, Gestures, Marriage\n",
    "* was mapped to Phenomenon (6)\n",
    "    * biof (Biologic Function) too general. Examples: dose-response relationship, Pharmacodynamics, Anabolism\n",
    "    * eehu (Environmental Effect of Humans) too general. Examples: Sewage, Pollution, Smoke\n",
    "    * hcpp (Human-caused Phenomenon or Process) too general. Now not in API. Examples: particle beam, Conferences, Victimization\n",
    "    * lbtr (Laboratory or Test Result) too general. Examples: False Positive Reactions, Bone Density, Serum Calcium Level\n",
    "    * npop (Natural Phenomenon or Process) too general. Examples: Floods, Fluorescence, Freezing\n",
    "    * phpr (Phenomenon or Process) too general. Examples: Disasters, Acceleration, Feedback\n",
    "* was mapped to Device (4)\n",
    "    * bodm (Biomedical or Dental Material) too general. Examples: Pill, Gel, Talc, calcium phosphate\n",
    "    * drdd (Drug Delivery Device) too general. Examples: Epipen, Skin Patch, Lilly cyanide antidote kit\n",
    "    * medd (Medical Device) too general. Examples: Implants / Denture, Overlay / Silicone gel implant / Swab\n",
    "    * resd (Research Device) too general. Examples: Study models, Slide\n",
    "* was mapped to GrossAnatomicalStructure\n",
    "    * emst (Embryonic Structure) Examples: Chick Embryo, Blastocyst structure, Placenta\n",
    "    * tisu (Tissue) Examples: Tissue specimen, Blood, Human tissue, Mucous Membrane\n",
    "* was mapped to PhysiologicalProcess\n",
    "    * genf (Genetic Function) too general. Examples: Transcription, Genetic / Transcriptional Activation / Recombination, Genetic\n",
    "* was mapped to ChemicalEntity\n",
    "    * chem (Chemical) too general. Examples: Chemicals, Acids, Ligands, Ozone\n",
    "* everything mapped to InformationContentEntity (11)\n",
    "    * clas (Classification) too general. Examples: Research Diagnostic Criteria, Group C\n",
    "    * cnce (Conceptual Entity) Now not in data. 1 matching subject. Examples: LNA\n",
    "    * ftcn (Functional Concept) too general. Examples: Techniques, Intravenous Route of Drug Administration\n",
    "    * grpa (Group Attribute) didn't see in API...\n",
    "    * idcn (Idea or Concept) too general. Examples: Significant, subject, Data, Owner\n",
    "    * lang (Language) too general. Now not in API. Examples: Nuosu Language, Chinook Jargon language\n",
    "    * qlco (Qualitative Concept) too general. Examples: Effect, Associated with, Advanced phase\n",
    "    * qnco (Quantitative Concept) too general. Examples: Calibration, occurrence, degrees Celsius\n",
    "    * rnlw (Regulation or Law) too general. Examples: Medicare, Medicaid, regulatory\n",
    "    * spco (Spatial Concept) too general. Examples: Structure, Longitudinal, Asymmetry\n",
    "    * tmco (Temporal Concept) too general. Examples: New, /period, 24 Hours\n",
    "* everything mapped to ClinicalAttribute\n",
    "    * clna (Clinical Attribute) too general. Examples: response, Renin secretion, BAND PATTERN\n",
    "* was mapped to Procedure\n",
    "    * lbpr (Laboratory Procedure) too general. Examples: Western Blot, Radioimmunoassay, Staining method\n",
    "    * mbrt (Molecular Biology Research Technique) too general. Examples: Polymerase Chain Reaction / Blotting, Northern\n",
    "* everything mapped to SmallMolecule\n",
    "    * elii (Element, Ion, or Isotope) too general. Examples: Atom, Aluminum, Superoxides\n",
    "* everything mapped to BiologicalEntity\n",
    "    * emod (Experimental Model of Disease) too general. Examples: Experimental Autoimmune Encephalomyelitis, Rodent Model\n",
    "* was mapped to Protein\n",
    "    * rcpt (Receptor) too general. Examples: Binding Sites / Receptors, Metabotropic Glutamate\n",
    "* was mapped to Event\n",
    "    * evnt (Event) too general. Now not in API. Examples: Stressful Events\n",
    "* was mapped to DiseaseOrPhenotypicFeature\n",
    "    * fndg (Finding) too general. Examples: spinal cord; lesion, Normal birth weight, Sedentary job\n",
    "* was mapped to GeographicLocation\n",
    "    * geoa (Geographic Area) too general. Examples: Country, Canada\n",
    "* everything mapped to Agent\n",
    "    * grup (Group) too general. Examples: Human, Individual\n",
    "    * hcro (Health Care Related Organization) too general. Examples: Hospitals, Health System\n",
    "    * orgt (Organization) too general. Examples: United Nations, Organization administrative structures\n",
    "    * pros (Professional Society) too general. Examples: Professional Organizations, American Nurses' Association\n",
    "    * shro (Self-help or Relief Organization) too general. Examples: Social Welfare, Support Groups\n",
    "* everything mapped to Publication\n",
    "    * inpr (Intellectual Product) too general. Examples: Methodology, Study models\n",
    "* everything mapped to PhysicalEntity\n",
    "    * mnob (Manufactured Object) too general. Examples: Glass, Manuals\n",
    "* everything mapped to MolecularEntity\n",
    "    * mosq (Molecular Sequence) too general. Now not in API. Examples: Genetic Code\n",
    "* was mapped to NucleicAcidEntity\n",
    "    * nusq (Nucleotide Sequence) too general. Examples: Base Sequence, DNA Sequence, 22q11\n",
    "* was mapped to OrganismAttribute\n",
    "    * orga (Organism Attribute) too general. Examples: Ability, Body Composition\n",
    "* was mapped to PopulationOfIndividualOrganisms\n",
    "    * popg (Population Group) too general. Examples: Male population group, Woman"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": false,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
