{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "47e7307b",
   "metadata": {},
   "source": [
    "# Auto-generate x-bte annotations for BioThings SEMMEDDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ab5663",
   "metadata": {},
   "source": [
    "This notebook walks a developer through the process of [taking the SEMMEDDB database](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html) and using this data to auto-generate the x-bte operations for [BTE](https://github.com/biothings/BioThings_Explorer_TRAPI). This is needed for BTE to query + process the responses from the [BioThings SEMMEDDB API](https://biothings.ncats.io/semmeddb). \n",
    "\n",
    "---\n",
    "\n",
    "When one sees this:  \n",
    "**PAUSE**\n",
    "\n",
    "read the accompanying text that will explain what the developer needs to do before running the code chunks below that text block.\n",
    "\n",
    "---\n",
    "\n",
    "The [yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_operations.yaml) [segments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_list.yaml) generated by this notebook are added to [this file](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml) to make [yaml used for the smartapi registration](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) for BioThings SEMMEDDB API."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43745613",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712186ee",
   "metadata": {},
   "source": [
    "Requirements:\n",
    "* install [Biolink-model Toolkit (bmt)](https://github.com/biolink/biolink-model-toolkit/). I'm [installing as a user](https://biolink.github.io/biolink-model-toolkit/intro/intro.html#for-users) with pip (`pip install bmt`). Using release 1.1.1 at the moment.\n",
    "* install ruamel.yaml with [pip](https://pypi.org/project/ruamel.yaml/). Using 0.17.32 at the moment. It's more recent than the version from [conda-forge](https://anaconda.org/conda-forge/ruamel_yaml). The documentation is [here](https://yaml.readthedocs.io/en/latest/)\n",
    "\n",
    "Files:\n",
    "* Get SEMMEDDB PREDICATION CSV [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html). This notebook was originally made using the version `semmedVER43_2021_R_PREDICATION`\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or pick the latest version [here](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5ee5aac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T20:58:39.620283Z",
     "start_time": "2023-08-07T20:58:39.128441Z"
    }
   },
   "outputs": [],
   "source": [
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "## see above for install instructions\n",
    "from bmt import Toolkit\n",
    "import ruamel.yaml as ryml    ## use ruamel_yaml if using the conda-forge version\n",
    "\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "## used in trying things out\n",
    "# import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af532d5",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Review the code chunk below before running it:\n",
    "* Check and correct the path for `raw_data_location`\n",
    "* Check that the columns specified in `usecols` and `names` match the columns of the PREDICATION [file](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html). \n",
    "* Check that the `na_values`, `sep` are correct. One can use a command in Terminal like `head`\n",
    "* If there are encoding issues, try different encodings. latin1 was used and worked, [ref](https://stackoverflow.com/questions/61163367/how-to-resolve-unicodedecodeerror-in-pandas-read-csv-while-loading-dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73408978",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T20:59:54.024692Z",
     "start_time": "2023-08-07T20:58:39.622649Z"
    }
   },
   "outputs": [],
   "source": [
    "raw_data_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"semmedVER43_2021_R_PREDICATION.csv\")\n",
    "\n",
    "raw_data = pd.read_csv(raw_data_location, header=None, sep=\",\", encoding=\"latin1\",\n",
    "                          usecols=[3, 6, 7, 10, 11],\n",
    "                          names=[\"PREDICATE\",\"SUBJECT_SEMTYPE\",\"SUBJECT_NOVELTY\",\n",
    "                                 \"OBJECT_SEMTYPE\", \"OBJECT_NOVELTY\"],\n",
    "                          na_values=r\"\\N\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8150ed6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T20:59:54.032191Z",
     "start_time": "2023-08-07T20:59:54.026718Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(113863366, 5)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57d2dc4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T20:59:54.044208Z",
     "start_time": "2023-08-07T20:59:54.035217Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>SUBJECT_NOVELTY</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>OBJECT_NOVELTY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>0</td>\n",
       "      <td>humn</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PREDICATE SUBJECT_SEMTYPE  SUBJECT_NOVELTY OBJECT_SEMTYPE  OBJECT_NOVELTY\n",
       "0  PROCESS_OF            virs                1           mamm             1.0\n",
       "1         ISA            virs                1           virs             1.0\n",
       "2         ISA            virs                1           virs             1.0\n",
       "3         ISA            virs                1           virs             1.0\n",
       "4  PROCESS_OF            dsyn                0           humn             0.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b68180",
   "metadata": {},
   "source": [
    "## Basic Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc722c4",
   "metadata": {},
   "source": [
    "### keep only novelty = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b8cf0b",
   "metadata": {},
   "source": [
    "now filter it to only keep novelty == 1 for both subject / object...\n",
    "\n",
    "since those with novelty == 0 probably aren't very helpful / interesting to Translator. The entities with novelty == 0 are [listed](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html) in the SEMMEDDB GENERIC_CONCEPT table files, which can be downloaded [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c362bfe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:00.101815Z",
     "start_time": "2023-08-07T20:59:54.045667Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = raw_data[(raw_data[\"SUBJECT_NOVELTY\"] == 1) &\n",
    "                         (raw_data[\"OBJECT_NOVELTY\"] == 1)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68a11e89",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:00.106479Z",
     "start_time": "2023-08-07T21:00:00.103236Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(78688677, 5)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9abae6d5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:01.671061Z",
     "start_time": "2023-08-07T21:00:00.108221Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove the novelty stuff since that will make computations faster and it's always 1 now\n",
    "filtered_data = filtered_data[['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66309970",
   "metadata": {},
   "source": [
    "what semantic types are even there? we have to prune down to what we actually want BTE operations on...\n",
    "\n",
    "interesting since the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67a2db8e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:11.250711Z",
     "start_time": "2023-08-07T21:00:01.673056Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "68"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes)  ## got 133\n",
    "len(object_semtypes)   ## got 135\n",
    "len(predicates)        ## got 68"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b788a7d",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Review the 3 sets above to see if there are things I want to remove. The normal entity semantic types have 4-letter codes..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88f7eeff",
   "metadata": {},
   "source": [
    "### Remove semantic types that we don't want to make operations from"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6085739",
   "metadata": {},
   "source": [
    "so there's some object stuff that I'm going to remove..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "69d37a01",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:11.256727Z",
     "start_time": "2023-08-07T21:00:11.252563Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C0030193',\n",
       " 'C0030705',\n",
       " 'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gngm,aapp',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'podg,humn',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "object_semtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26561f10",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:15.277701Z",
     "start_time": "2023-08-07T21:00:11.260817Z"
    }
   },
   "outputs": [],
   "source": [
    "## make the set of stuff we want to remove\n",
    "removal1 = set([\"C0030193\", \"C0030705\", \"gngm,aapp\", \"podg,humn\"])\n",
    "\n",
    "## remove it from the data \n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(removal1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f18986e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:24.963669Z",
     "start_time": "2023-08-07T21:00:15.279445Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 133: didn't change\n",
    "len(object_semtypes)  ## was 135, now 131: decreased by 4 (expected)\n",
    "len(predicates)       ## was 68, now 65: decreased by 3 (cool)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff20381b",
   "metadata": {},
   "source": [
    "### Remove Predicates we don't want to make operations from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09163f4e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:24.969068Z",
     "start_time": "2023-08-07T21:00:24.965107Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADMINISTERED_TO',\n",
       " 'AFFECTS',\n",
       " 'ASSOCIATED_WITH',\n",
       " 'AUGMENTS',\n",
       " 'CAUSES',\n",
       " 'COEXISTS_WITH',\n",
       " 'COMPLICATES',\n",
       " 'CONVERTS_TO',\n",
       " 'DIAGNOSES',\n",
       " 'DISRUPTS',\n",
       " 'INHIBITS',\n",
       " 'INTERACTS_WITH',\n",
       " 'ISA',\n",
       " 'LOCATION_OF',\n",
       " 'MANIFESTATION_OF',\n",
       " 'MEASUREMENT_OF',\n",
       " 'MEASURES',\n",
       " 'METHOD_OF',\n",
       " 'NEG_ADMINISTERED_TO',\n",
       " 'NEG_AFFECTS',\n",
       " 'NEG_ASSOCIATED_WITH',\n",
       " 'NEG_AUGMENTS',\n",
       " 'NEG_CAUSES',\n",
       " 'NEG_COEXISTS_WITH',\n",
       " 'NEG_COMPLICATES',\n",
       " 'NEG_CONVERTS_TO',\n",
       " 'NEG_DIAGNOSES',\n",
       " 'NEG_DISRUPTS',\n",
       " 'NEG_INHIBITS',\n",
       " 'NEG_INTERACTS_WITH',\n",
       " 'NEG_ISA',\n",
       " 'NEG_LOCATION_OF',\n",
       " 'NEG_MANIFESTATION_OF',\n",
       " 'NEG_MEASUREMENT_OF',\n",
       " 'NEG_MEASURES',\n",
       " 'NEG_METHOD_OF',\n",
       " 'NEG_OCCURS_IN',\n",
       " 'NEG_PART_OF',\n",
       " 'NEG_PRECEDES',\n",
       " 'NEG_PREDISPOSES',\n",
       " 'NEG_PREVENTS',\n",
       " 'NEG_PROCESS_OF',\n",
       " 'NEG_PRODUCES',\n",
       " 'NEG_STIMULATES',\n",
       " 'NEG_TREATS',\n",
       " 'NEG_USES',\n",
       " 'NEG_higher_than',\n",
       " 'NEG_lower_than',\n",
       " 'NEG_same_as',\n",
       " 'NOM',\n",
       " 'OCCURS_IN',\n",
       " 'PART_OF',\n",
       " 'PRECEDES',\n",
       " 'PREDISPOSES',\n",
       " 'PREP',\n",
       " 'PREVENTS',\n",
       " 'PROCESS_OF',\n",
       " 'PRODUCES',\n",
       " 'STIMULATES',\n",
       " 'TREATS',\n",
       " 'USES',\n",
       " 'compared_with',\n",
       " 'higher_than',\n",
       " 'lower_than',\n",
       " 'same_as'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "106d9b99",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Decide what predicates you want to remove\n",
    "\n",
    "**Current logic**\n",
    "\n",
    "I'm using biolink-model 3.5.3 right now. \n",
    "\n",
    "With biolink-model >= 3.5.1, Translator has reviewed some predicates for domain-predicate and predicate-range combo exclusions - ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "We'll want to keep the predicates that have been reviewed and only do the custom removals below for predicates that:\n",
    "* haven't been reviewed yet\n",
    "* are problematic to use right now: \n",
    "    * Translator currently isn't great with negation, so removing those\n",
    "    * Also some predicates seem to actually be practice phrases? see the [article](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-486#Sec26)\n",
    "    * And I didn't find the following relationships useful: ISA (subclass relationship), lower_than / higher_than (these two entities were compared and 1 was better (higher) or worse (lower) than the other), and compared_with...\n",
    "    \n",
    "And we'll want to adjust the lists below if Translator reviews more predicates in the future. \n",
    "\n",
    "Old note on same_as (previously excluded): it's meaning seems to be \"equivalent / it did just as well as some other thing\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dec1bcc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:24.973686Z",
     "start_time": "2023-08-07T21:00:24.970402Z"
    }
   },
   "outputs": [],
   "source": [
    "negative_preds = [\"NEG_ADMINISTERED_TO\", \"NEG_AFFECTS\", \"NEG_ASSOCIATED_WITH\",\n",
    "                  \"NEG_AUGMENTS\", \"NEG_CAUSES\", \"NEG_COEXISTS_WITH\", \"NEG_COMPLICATES\",\n",
    "                  \"NEG_CONVERTS_TO\", \"NEG_DIAGNOSES\", \"NEG_DISRUPTS\", \"NEG_INHIBITS\",\n",
    "                  \"NEG_INTERACTS_WITH\", \"NEG_ISA\", \"NEG_LOCATION_OF\", \"NEG_MANIFESTATION_OF\",\n",
    "                  \"NEG_MEASUREMENT_OF\", \"NEG_MEASURES\", \"NEG_METHOD_OF\", \"NEG_OCCURS_IN\",\n",
    "                  \"NEG_PART_OF\", \"NEG_PRECEDES\", \"NEG_PREDISPOSES\", \"NEG_PREVENTS\",\n",
    "                  \"NEG_PROCESS_OF\", \"NEG_PRODUCES\", \"NEG_STIMULATES\", \"NEG_TREATS\", \"NEG_USES\",\n",
    "                  \"NEG_higher_than\", \"NEG_lower_than\", \"NEG_same_as\"\n",
    "                 ]\n",
    "\n",
    "practice_phrase = [\"VERB\", \"NOM\", \"PREP\"]\n",
    "\n",
    "dont_like = [\"ISA\", \"lower_than\", \"higher_than\", \"compared_with\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ab643b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:39.744104Z",
     "start_time": "2023-08-07T21:00:24.975478Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(negative_preds)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(practice_phrase)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(dont_like)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f3d145",
   "metadata": {},
   "source": [
    "now look at the stats. \n",
    "\n",
    "(interesting since the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates. 54 / 2 = 27 predicates without negation + 1 for same_as inclusion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b6244458",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:00:48.491523Z",
     "start_time": "2023-08-07T21:00:39.745858Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 127: decreased by 6\n",
    "len(object_semtypes)  ## was 131, now 127: decreased by 4\n",
    "len(predicates)       ## was 65, now 28: decreased by 37 (was expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e31d94",
   "metadata": {},
   "source": [
    "so....that's how many kinds of subjects, objects, and predicates we have to go forward with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4184a17f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.024391Z",
     "start_time": "2023-08-07T21:00:48.493196Z"
    }
   },
   "outputs": [],
   "source": [
    "combos = filtered_data.value_counts().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "24036b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.030116Z",
     "start_time": "2023-08-07T21:01:02.026530Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15469, 4)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos.shape\n",
    "## so that's still a lot....\n",
    "## without same_as, it was 14796"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "949ccb58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.038414Z",
     "start_time": "2023-08-07T21:01:02.032047Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>humn</td>\n",
       "      <td>2143816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1159577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>humn</td>\n",
       "      <td>1000549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    count\n",
       "0            dsyn   PROCESS_OF           humn  2143816\n",
       "1            bpoc      PART_OF           mamm  1159577\n",
       "2            bpoc  LOCATION_OF           neop  1023005\n",
       "3            fndg   PROCESS_OF           humn  1000549\n",
       "4            bpoc  LOCATION_OF           aapp   953010"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e998d9e8",
   "metadata": {},
   "source": [
    "## Mapping and Specific Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330bba7a",
   "metadata": {},
   "source": [
    "First, I'm going to map SEMMED semantic info to biolink-model - this is a legacy from when I used the semantic info + mapping to decide what semantic-types + predicates to write x-bte annotation with. Now, I'll do minimal adjustments.  \n",
    "\n",
    "Then I'll use Translator-curated exclusions to remove some semantic-type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n",
    "\n",
    "As a reminder:\n",
    "\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or pick the latest version [here](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91866eb6",
   "metadata": {},
   "source": [
    "### Ingest SEMMED semantic info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8d1bbf26",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.043068Z",
     "start_time": "2023-08-07T21:01:02.040499Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"2020AA\", \"SRDEF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "8f859145",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.051087Z",
     "start_time": "2023-08-07T21:01:02.045021Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef = pd.read_csv(srdef_location, sep=\"|\", header=0, index_col=False,\n",
    "                    names=[\"Record Type (RT)\",\n",
    "                           \"Unique Identifier (UI)\",\n",
    "                           \"Full Name (STY/RL)\", \n",
    "                           \"Tree Number (STN/RTN)\", \n",
    "                           \"Definition (DEF)\",\n",
    "                           \"Examples (EX)\",\n",
    "                           \"Usage Note (UN)\",\n",
    "                           \"Non-Human Flag (NH)\",\n",
    "                           \"Abbreviation (ABR)\",\n",
    "                           \"Inverse Relation (RIN)\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d470fa70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.062663Z",
     "start_time": "2023-08-07T21:01:02.052529Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Record Type (RT)</th>\n",
       "      <th>Unique Identifier (UI)</th>\n",
       "      <th>Full Name (STY/RL)</th>\n",
       "      <th>Tree Number (STN/RTN)</th>\n",
       "      <th>Definition (DEF)</th>\n",
       "      <th>Examples (EX)</th>\n",
       "      <th>Usage Note (UN)</th>\n",
       "      <th>Non-Human Flag (NH)</th>\n",
       "      <th>Abbreviation (ABR)</th>\n",
       "      <th>Inverse Relation (RIN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>STY</td>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>A1.4.1.2.1.7</td>\n",
       "      <td>Amino acids and chains of amino acids connecte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>When the concept is both an enzyme and a prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aapp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Record Type (RT) Unique Identifier (UI)               Full Name (STY/RL)  \\\n",
       "97              STY                   T116  Amino Acid, Peptide, or Protein   \n",
       "\n",
       "   Tree Number (STN/RTN)                                   Definition (DEF)  \\\n",
       "97          A1.4.1.2.1.7  Amino acids and chains of amino acids connecte...   \n",
       "\n",
       "    Examples (EX)                                    Usage Note (UN)  \\\n",
       "97            NaN  When the concept is both an enzyme and a prote...   \n",
       "\n",
       "   Non-Human Flag (NH) Abbreviation (ABR) Inverse Relation (RIN)  \n",
       "97                 NaN               aapp                    NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srdef[srdef['Abbreviation (ABR)'] == 'aapp']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e245634",
   "metadata": {},
   "source": [
    "### Semantic type processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c2fc216",
   "metadata": {},
   "source": [
    "#### Add semantic types that are in the data but not in SRDEF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49397246",
   "metadata": {},
   "source": [
    "Yao [noticed](https://github.com/biothings/pending.api/issues/30#issuecomment-903609946) that the data file uses \"old SEMMED semantic types\" from 2013AA, and these semantic types didn't exist in the latest SRDEF file (2020AA).\n",
    "\n",
    "In response, Andrew said to [keep](https://github.com/biothings/pending.api/issues/30#issuecomment-903879782) this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "89c14528",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.077442Z",
     "start_time": "2023-08-07T21:01:02.065382Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(126, 3)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>T195</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst\n",
       "119  T195                       Antibiotic   antb"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## simplify the df\n",
    "SEMMED_entity_types = srdef[srdef['Record Type (RT)'] == 'STY'].copy()\n",
    "SEMMED_entity_types = SEMMED_entity_types[['Unique Identifier (UI)',\n",
    "                                           'Full Name (STY/RL)', \n",
    "                                           'Abbreviation (ABR)']].copy()\n",
    "SEMMED_entity_types.columns = ['UI', 'FullName', \"Abbrev\"]\n",
    "SEMMED_entity_types.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_entity_types.shape\n",
    "SEMMED_entity_types[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "8c8c04ce",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.085709Z",
     "start_time": "2023-08-07T21:01:02.079469Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "these semantic types are in the data, but not in SRDEF:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'carb',\n",
       " 'eico',\n",
       " 'invt',\n",
       " 'lipd',\n",
       " 'nsba',\n",
       " 'opco',\n",
       " 'orgm',\n",
       " 'rich',\n",
       " 'strd'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "only_in_data = data_types - set(SEMMED_entity_types['Abbrev'])\n",
    "\n",
    "print(\"these semantic types are in the data, but not in SRDEF:\")\n",
    "only_in_data\n",
    "\n",
    "len(only_in_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533341a2",
   "metadata": {},
   "source": [
    "How I found info on these semantic types from [previous semantic versions](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html):\n",
    "\n",
    "was in 2009AB file (sets with SG_2?)\n",
    "* alga\n",
    "* invt\n",
    "* rich\n",
    "\n",
    "was in 2014AB file (sets with SG_3?)\n",
    "* carb \n",
    "* eico \n",
    "* lipd \n",
    "* nsba \n",
    "* opco\n",
    "* orgm\n",
    "* strd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0400e8e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.092628Z",
     "start_time": "2023-08-07T21:01:02.087529Z"
    }
   },
   "outputs": [],
   "source": [
    "## build the missing info and add to the table\n",
    "missing_semantic_types = [{\"UI\": \"T003\", \"FullName\": \"Alga\", \"Abbrev\": \"alga\"},\n",
    "                          {\"UI\": \"T118\", \"FullName\": \"Carbohydrate\", \"Abbrev\": \"carb\"},\n",
    "                          {\"UI\": \"T111\", \"FullName\": \"Eicosanoid\", \"Abbrev\": \"eico\"},\n",
    "                          {\"UI\": \"T009\", \"FullName\": \"Invertebrate\", \"Abbrev\": \"invt\"},\n",
    "                          {\"UI\": \"T119\", \"FullName\": \"Lipid\", \"Abbrev\": \"lipd\"},\n",
    "                          {\"UI\": \"T124\", \"FullName\": \"Neuroreactive Substance or Biogenic Amine\", \"Abbrev\": \"nsba\"},\n",
    "                          {\"UI\": \"T115\", \"FullName\": \"Organophosphorus Compound\", \"Abbrev\": \"opco\"},\n",
    "                          {\"UI\": \"T001\", \"FullName\": \"Organism\", \"Abbrev\": \"orgm\"},\n",
    "                          {\"UI\": \"T006\", \"FullName\": \"Rickettsia or Chlamydia\", \"Abbrev\": \"rich\"},\n",
    "                          {\"UI\": \"T110\", \"FullName\": \"Steroid\", \"Abbrev\": \"strd\"}\n",
    "                         ]\n",
    "\n",
    "missing_semantic_types = pd.DataFrame.from_records(missing_semantic_types)\n",
    "\n",
    "## add it to the table\n",
    "## SEMMED_entity_types = SEMMED_entity_types.append(missing_semantic_types)\n",
    "\n",
    "SEMMED_entity_types = pd.concat([SEMMED_entity_types, missing_semantic_types])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "683a1f47",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:02.103230Z",
     "start_time": "2023-08-07T21:01:02.094631Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(136, 3)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "0    T003                             Alga   alga\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_entity_types.shape\n",
    "SEMMED_entity_types[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f01880",
   "metadata": {},
   "source": [
    "#### Get the biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a7a170a6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:03.732815Z",
     "start_time": "2023-08-07T21:01:02.105394Z"
    }
   },
   "outputs": [],
   "source": [
    "## as of 2023-08-2: using default which is 3.5.3 \n",
    "bmt_tool = Toolkit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "4cc1fad9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:03.756625Z",
     "start_time": "2023-08-07T21:01:03.734169Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Small Molecule'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bmt_tool.get_element_by_mapping('STY:T123').title()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "c2c14436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:04.235902Z",
     "start_time": "2023-08-07T21:01:03.763744Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting biolink-mapping, in the format needed to create operations\n",
    "SEMMED_entity_types['BiolinkMapping'] = [bmt_tool.get_element_by_mapping('STY:'+i)\n",
    "                                         for i in SEMMED_entity_types['UI']]\n",
    "## put these node categories/semantic-types in the correct format: PascalCase\n",
    "SEMMED_entity_types['BiolinkMapping'] = [i.title().replace(\" \", \"\") if isinstance(i, str)\n",
    "                                         else i for i in SEMMED_entity_types['BiolinkMapping']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "044843d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:04.245284Z",
     "start_time": "2023-08-07T21:01:04.238141Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "0    T003                                 Alga   alga   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "0                        None  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  \n",
       "\n",
       "[136 rows x 4 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51111407",
   "metadata": {},
   "source": [
    "#### Clean up to only have terms in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7fb08392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:04.250847Z",
     "start_time": "2023-08-07T21:01:04.246836Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "note that these semantic types are in SRDEF but not in the data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cnce', 'crbs', 'enty', 'grpa', 'lang', 'phob'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"note that these semantic types are in SRDEF but not in the data\")\n",
    "set(SEMMED_entity_types['Abbrev']) - data_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "22daf133",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:04.255700Z",
     "start_time": "2023-08-07T21:01:04.252452Z"
    }
   },
   "outputs": [],
   "source": [
    "## get the subset of semantic network terms that are actually in the data\n",
    "SEMMED_entity_types = SEMMED_entity_types[SEMMED_entity_types['Abbrev'].isin(data_types)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "86caa9b9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:01:04.265759Z",
     "start_time": "2023-08-07T21:01:04.257735Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "0    T003                                 Alga   alga   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "0                        None  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  \n",
       "\n",
       "[130 rows x 4 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d30960",
   "metadata": {},
   "source": [
    "#### review mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd9e849",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "We are still reviewing / changing mappings, since BTE's responses would probably differ a lot if this behavior was changed. \n",
    "\n",
    "However, we are now removing ONLY semantic-types that lack biolink-model mappings. Other curated semantic-type removals will be done in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "I'm going to keep the old notes + code (commented out), just in case. \n",
    "\n",
    "---\n",
    "\n",
    "Old notes:\n",
    "\n",
    "This is a place to **STOP** and review all the SEMMED semantic types and their mapping to biolink semantic types...to decide what we are interested in keeping. This involves some knowledge of what biolink semantic types are prioritized in Translator. \n",
    "\n",
    "One can use the definitions of the SEMMED semantic types (from the SRDEF file or [browsing the UMLS vocab online](https://uts.nlm.nih.gov/uts/umls/semantic-network/root)) and the definitions of biolink semantic types (look for a comment with 'THINGS' in the biolink-model yaml file)\n",
    "\n",
    "See the last section (section 6) for notes on decisions that were made...\n",
    "\n",
    "Sections 3.5 + 3.6 below involves this review, changing mappings, and removing some SEMMED semantic types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "eb7ea639",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:02:31.364723Z",
     "start_time": "2023-08-07T21:02:31.360736Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Polypeptide', 'Disease', 'Activity', 'Cohort', None,\n",
       "       'Invertebrate', 'OrganismalEntity', 'AnatomicalEntity', 'Drug',\n",
       "       'SmallMolecule', 'Bacterium', 'Behavior', 'Phenomenon', 'Device',\n",
       "       'GrossAnatomicalStructure', 'CellularComponent',\n",
       "       'PhysiologicalProcess', 'Cell', 'ChemicalEntity',\n",
       "       'InformationContentEntity', 'ClinicalAttribute', 'Procedure',\n",
       "       'BiologicalEntity', 'Protein', 'Event',\n",
       "       'DiseaseOrPhenotypicFeature', 'Fungus', 'Food',\n",
       "       'GeographicLocation', 'GenomicEntity', 'Agent', 'Human',\n",
       "       'PathologicalProcess', 'Publication', 'DiagnosticAid', 'Mammal',\n",
       "       'PhysicalEntity', 'MolecularActivity', 'MolecularEntity',\n",
       "       'NucleicAcidEntity', 'OrganismAttribute', 'Plant',\n",
       "       'PopulationOfIndividualOrganisms', 'PhenotypicFeature', 'Virus',\n",
       "       'Vertebrate'], dtype=object)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## code used to review \n",
    "SEMMED_entity_types['BiolinkMapping'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fc902a02",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:02:34.478788Z",
     "start_time": "2023-08-07T21:02:34.470905Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>T194</td>\n",
       "      <td>Archaeon</td>\n",
       "      <td>arch</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>T022</td>\n",
       "      <td>Body System</td>\n",
       "      <td>bdsy</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T012</td>\n",
       "      <td>Bird</td>\n",
       "      <td>bird</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>T091</td>\n",
       "      <td>Biomedical Occupation or Discipline</td>\n",
       "      <td>bmod</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>T120</td>\n",
       "      <td>Chemical Viewed Functionally</td>\n",
       "      <td>chvf</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>T104</td>\n",
       "      <td>Chemical Viewed Structurally</td>\n",
       "      <td>chvs</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>T204</td>\n",
       "      <td>Eukaryote</td>\n",
       "      <td>euka</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>T021</td>\n",
       "      <td>Fully Formed Anatomical Structure</td>\n",
       "      <td>ffas</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T013</td>\n",
       "      <td>Fish</td>\n",
       "      <td>fish</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>T045</td>\n",
       "      <td>Genetic Function</td>\n",
       "      <td>genf</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T009</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>invt</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>T090</td>\n",
       "      <td>Occupation or Discipline</td>\n",
       "      <td>ocdi</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T001</td>\n",
       "      <td>Organism</td>\n",
       "      <td>orgm</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T014</td>\n",
       "      <td>Reptile</td>\n",
       "      <td>rept</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T006</td>\n",
       "      <td>Rickettsia or Chlamydia</td>\n",
       "      <td>rich</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>T167</td>\n",
       "      <td>Substance</td>\n",
       "      <td>sbst</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev BiolinkMapping\n",
       "0    T003                                 Alga   alga           None\n",
       "118  T194                             Archaeon   arch           None\n",
       "17   T022                          Body System   bdsy           None\n",
       "7    T012                                 Bird   bird           None\n",
       "81   T091  Biomedical Occupation or Discipline   bmod           None\n",
       "98   T120         Chemical Viewed Functionally   chvf           None\n",
       "94   T104         Chemical Viewed Structurally   chvs           None\n",
       "125  T204                            Eukaryote   euka           None\n",
       "16   T021    Fully Formed Anatomical Structure   ffas           None\n",
       "8    T013                                 Fish   fish           None\n",
       "37   T045                     Genetic Function   genf           None\n",
       "3    T009                         Invertebrate   invt           None\n",
       "80   T090             Occupation or Discipline   ocdi           None\n",
       "7    T001                             Organism   orgm           None\n",
       "9    T014                              Reptile   rept           None\n",
       "8    T006              Rickettsia or Chlamydia   rich           None\n",
       "108  T167                            Substance   sbst           None"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'] == 'Vertebrate']\n",
    "SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e4f5275",
   "metadata": {},
   "source": [
    "Still re-mapping (not following a strict biolink-model mapping). May revisit later, but concerned about affecting responses to creative-mode queries / requiring template adjustments..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e641ae5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:02:39.405833Z",
     "start_time": "2023-08-07T21:02:39.400813Z"
    }
   },
   "outputs": [],
   "source": [
    "## re-mapping based on putting IDs into normalization service / our operation system...\n",
    "##   we have UMLS for Disease (mydisease, mychem), SmallMolecule (idisk)\n",
    "\n",
    "## leaving aapp as Polypeptide, enzy mapped to Protein\n",
    "## leaving Disease mappings as-is, but some seem like they could be PathologicalAnatomicalStructure \n",
    "##   (acab, anab, cgab) or PathologicalProcess (comd) instead\n",
    "## left clnd (Clinical Drug) as Drug since it really seemed like a drug (dosage)\n",
    "\n",
    "## currently mapped to Drug, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'antb'),'BiolinkMapping'] = 'SmallMolecule'  ## Antibiotic\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'phsu'),'BiolinkMapping'] = 'SmallMolecule'  ## Pharmacologic Substance\n",
    "\n",
    "## currently mapped to GenomicEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'gngm'),'BiolinkMapping'] = 'Gene'\n",
    "\n",
    "## currently mapped to NucleicAcidEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'nnon'),'BiolinkMapping'] = 'SmallMolecule'  ## Nucleic Acid, Nucleoside, or Nucleotide"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cca13d9",
   "metadata": {},
   "source": [
    "We are removing semantic-types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "a0bb0548",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:06:31.804363Z",
     "start_time": "2023-08-07T21:06:31.801550Z"
    }
   },
   "outputs": [],
   "source": [
    "currently_unmapped_types = set(SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'].isna()].Abbrev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bbda9f75",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:06:36.733748Z",
     "start_time": "2023-08-07T21:06:36.730254Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich',\n",
       " 'sbst'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03c050f9",
   "metadata": {},
   "source": [
    "Commenting out: we aren't removing semantic-types here anymore, because we will do so in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a3a9da81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T20:19:41.174425Z",
     "start_time": "2023-08-07T20:19:41.171096Z"
    }
   },
   "outputs": [],
   "source": [
    "# currently_unused_entities = [\n",
    "#     ## missing mappings\n",
    "#     \"arch\", \"bdsy\", \"bird\", \"bmod\", \"chvf\", \"chvs\", \"euka\", \"ffas\", \n",
    "#     \"fish\", \"genf\", \"ocdi\", \"rept\", \"sbst\", \n",
    "#     ## old terms here\n",
    "#     \"alga\", \"invt\", \"orgm\", \"rich\",\n",
    "#     ## mapped to Polypeptide\n",
    "#     \"amas\",     \n",
    "#     ## all that are mapped to Activity\n",
    "#     \"acty\", \"dora\", \"edac\", \"gora\", \"hlca\", \"mcha\", \"ocac\", \"resa\",  \n",
    "#     ## all that are mapped to Cohort\n",
    "#     \"aggp\", \"famg\", \"podg\", \"prog\",  \n",
    "#     ## all that are mapped to Invertebrate\n",
    "#     \"amph\",\n",
    "#     ## all that are mapped to OrganismalEntity\n",
    "#     \"anim\", \n",
    "#     ## all that are mapped to Anatomical Entity\n",
    "#     \"anst\", \"bdsu\", \"blor\", \"bsoj\", \n",
    "#     ## mapped to SmallMolecule\n",
    "#     \"elii\",\n",
    "#     ## mapped to Bacterium\n",
    "#     \"bact\",\n",
    "#     ## all that are mapped to Behavior\n",
    "#     \"bhvr\", \"inbe\", \"menp\", \"socb\",\n",
    "#     ## all that are mapped to Phenomenon\n",
    "#     \"biof\", \"eehu\", \"hcpp\", \"lbtr\", \"npop\", \"phpr\",\n",
    "#     ## all that are mapped to Device\n",
    "#     \"bodm\", \"drdd\", \"medd\", \"resd\",\n",
    "#     ## mapped to GrossAnatomicalStructure\n",
    "#     \"emst\", \"tisu\", \n",
    "#     ## mapped to ChemicalEntity\n",
    "#     \"chem\",\n",
    "#     ## all that are mapped to InformationContentEntity\n",
    "#     \"clas\", \"ftcn\", \"idcn\", \"qlco\", \"qnco\", \"rnlw\", \"spco\", \"tmco\",\n",
    "#     ## all that are mapped to ClinicalAttribute\n",
    "#     \"clna\",\n",
    "#     ## mapped to Procedure\n",
    "#     \"lbpr\", \"mbrt\", \n",
    "#     ## mapped to BiologicalEntity\n",
    "#     \"emod\",\n",
    "#     ## mapped to Protein\n",
    "#     \"rcpt\",\n",
    "#     ## all that are mapped to Event\n",
    "#     \"evnt\",\n",
    "#     ## all that are mapped to DiseaseOrPhenotypicFeature\n",
    "#     \"fndg\",\n",
    "#     ## all that are mapped to Fungus\n",
    "#     \"fngs\",\n",
    "#     ## all that are mapped to GeographicLocation\n",
    "#     \"geoa\",\n",
    "#     ## all that are mapped to Agent\n",
    "#     \"grup\", \"hcro\", \"orgt\", \"pros\", \"shro\",\n",
    "#     ## all that are mapped to Human\n",
    "#     \"humn\", \n",
    "#     ## all that are mapped to Publication\n",
    "#     \"inpr\",\n",
    "#     ## all that are mapped to DiagnosticAid\n",
    "#     \"irda\", \n",
    "#     ## all that are mapped to Mammal\n",
    "#     \"mamm\",  \n",
    "#     ## all that are mapped to PhysicalEntity\n",
    "#     \"mnob\",\n",
    "#     ## all that are mapped to MolecularEntity\n",
    "#     \"mosq\",\n",
    "#     ## mapped to NucleicAcidEntity\n",
    "#     \"nusq\",\n",
    "#     ## all that are mapped to OrganismAttribute\n",
    "#     \"orga\",\n",
    "#     ## all that are mapped to Plant\n",
    "#     \"plnt\", \n",
    "#     ## all that are mapped to PopulationOfIndividualOrganisms\n",
    "#     \"popg\",\n",
    "#     ## all that are mapped to Virus\n",
    "#     \"virs\",\n",
    "#     ## all that are mapped to Vertebrate\n",
    "#     \"vtbt\"\n",
    "# ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da7aa1e",
   "metadata": {},
   "source": [
    "#### prune data based on what node entities to remove"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84e31ea",
   "metadata": {},
   "source": [
    "We are removing semantic-types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4619ca95",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:08:37.164519Z",
     "start_time": "2023-08-07T21:08:37.159771Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich',\n",
       " 'sbst'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types\n",
    "len(currently_unmapped_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "308358e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:08:06.698046Z",
     "start_time": "2023-08-07T21:07:54.972210Z"
    }
   },
   "outputs": [],
   "source": [
    "## prune this doc\n",
    "SEMMED_entity_types = SEMMED_entity_types[ ~ SEMMED_entity_types['Abbrev'].isin(currently_unmapped_types)]\n",
    "\n",
    "## prune the data doc\n",
    "filtered_data = filtered_data[ ~ filtered_data['SUBJECT_SEMTYPE'].isin(currently_unmapped_types)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(currently_unmapped_types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "67861c2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:08:16.682912Z",
     "start_time": "2023-08-07T21:08:08.544814Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "110"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 127, now 112: decreased by 15\n",
    "len(object_semtypes)  ## was 127, now 110: decreased by 17 (number of unmapped entities)\n",
    "len(predicates)       ## was 28, still 28 (expected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "647e4065",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:10.601700Z",
     "start_time": "2023-08-07T21:10:10.594063Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "77   T087                  Amino Acid Sequence   amas   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "77                Polypeptide  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  \n",
       "\n",
       "[113 rows x 4 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03da6943",
   "metadata": {},
   "source": [
    "### Predicate processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf6f43",
   "metadata": {},
   "source": [
    "#### getting mappings for predicates and qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "52496738",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:19.687463Z",
     "start_time": "2023-08-07T21:10:19.682966Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates = srdef[srdef['Record Type (RT)'] == 'RL'].copy()\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates[['Full Name (STY/RL)'\n",
    "                                      ]].copy()\n",
    "SEMMED_predicates.columns = ['SemmedPred']\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "635827c9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "The SRDEF file might be missing predicates that are in the data, so check that using the code block below and see if you would like to add those back in..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b05e90f6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:22.733683Z",
     "start_time": "2023-08-07T21:10:22.728878Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['administered_to',\n",
       " 'augments',\n",
       " 'coexists_with',\n",
       " 'converts_to',\n",
       " 'inhibits',\n",
       " 'predisposes',\n",
       " 'same_as',\n",
       " 'stimulates']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## predicates currently in the filtered data\n",
    "\n",
    "missing_preds = list()\n",
    "\n",
    "for i in predicates:\n",
    "    if i.lower() not in set(SEMMED_predicates['SemmedPred']):\n",
    "        missing_preds.append(i.lower())\n",
    "\n",
    "missing_preds.sort()\n",
    "missing_preds\n",
    "\n",
    "# [i for i in predicates if i.lower() in SEMMED_predicates['SemmedName']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e40796c",
   "metadata": {},
   "source": [
    "Currently we add them back in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "6a3d217d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:25.494642Z",
     "start_time": "2023-08-07T21:10:25.483706Z"
    }
   },
   "outputs": [],
   "source": [
    "## manually add some predicates that are in the data, but not in SRDEF...\n",
    "\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['administered_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['augments']\n",
    "## notice that the data doesn't have a hyphen for coexists...even though SRDEF has hyphen for co-occurs_with\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['coexists_with']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['converts_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['inhibits']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['predisposes']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['same_as']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['stimulates']\n",
    "\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "ec402da4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:27.270284Z",
     "start_time": "2023-08-07T21:10:27.261025Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>adjacent_to</td>\n",
       "      <td>SEMMEDDB:ADJACENT_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>analyzes</td>\n",
       "      <td>SEMMEDDB:ANALYZES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>assesses_effect_of</td>\n",
       "      <td>SEMMEDDB:ASSESSES_EFFECT_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>temporally_related_to</td>\n",
       "      <td>SEMMEDDB:TEMPORALLY_RELATED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>traverses</td>\n",
       "      <td>SEMMEDDB:TRAVERSES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>tributary_of</td>\n",
       "      <td>SEMMEDDB:TRIBUTARY_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SemmedPred                    Semmed_in_BM\n",
       "162            adjacent_to            SEMMEDDB:ADJACENT_TO\n",
       "54         administered_to        SEMMEDDB:ADMINISTERED_TO\n",
       "145                affects                SEMMEDDB:AFFECTS\n",
       "176               analyzes               SEMMEDDB:ANALYZES\n",
       "158     assesses_effect_of     SEMMEDDB:ASSESSES_EFFECT_OF\n",
       "..                     ...                             ...\n",
       "130  temporally_related_to  SEMMEDDB:TEMPORALLY_RELATED_TO\n",
       "166              traverses              SEMMEDDB:TRAVERSES\n",
       "148                 treats                 SEMMEDDB:TREATS\n",
       "178           tributary_of           SEMMEDDB:TRIBUTARY_OF\n",
       "149                   uses                   SEMMEDDB:USES\n",
       "\n",
       "[62 rows x 2 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## since biolink-model mappings doesn't use hyphens\n",
    "SEMMED_predicates['Semmed_in_BM']= ['SEMMEDDB:'+i \n",
    "                                   for i in SEMMED_predicates['SemmedPred'].str.replace(\"-\", \"\")]\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.upper()\n",
    "\n",
    "## Specifically change same_as to be lower-case for the mapping process\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedPred'] == 'same_as'),'Semmed_in_BM'] = 'SEMMEDDB:same_as'\n",
    "\n",
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "979e0da8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:30.151092Z",
     "start_time": "2023-08-07T21:10:29.930870Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting the mapping\n",
    "SEMMED_predicates['BMPred'] = [bmt_tool.get_element_by_mapping(i)\n",
    "                                       for i in SEMMED_predicates['Semmed_in_BM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5d7f7944",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:30.945605Z",
     "start_time": "2023-08-07T21:10:30.935970Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>measurement_of</td>\n",
       "      <td>SEMMEDDB:MEASUREMENT_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO            related to\n",
       "145           affects           SEMMEDDB:AFFECTS               affects\n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related to\n",
       "55           augments          SEMMEDDB:AUGMENTS                  None\n",
       "141            causes            SEMMEDDB:CAUSES                causes\n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists with\n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates\n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives into\n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses\n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts\n",
       "58           inhibits          SEMMEDDB:INHIBITS                  None\n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts with\n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location of\n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation of\n",
       "170    measurement_of    SEMMEDDB:MEASUREMENT_OF                  None\n",
       "156          measures          SEMMEDDB:MEASURES  quantifier qualifier\n",
       "171         method_of         SEMMEDDB:METHOD_OF                  None\n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs in\n",
       "127           part_of           SEMMEDDB:PART_OF               part of\n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes\n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes\n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents\n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs in\n",
       "138          produces          SEMMEDDB:PRODUCES              produces\n",
       "60            same_as           SEMMEDDB:same_as           close match\n",
       "61         stimulates        SEMMEDDB:STIMULATES                  None\n",
       "148            treats            SEMMEDDB:TREATS                treats\n",
       "149              uses              SEMMEDDB:USES             has input"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remember there are 26 predicates in the data\n",
    "len(predicates)\n",
    "pred_compare = [i.lower() for i in predicates]\n",
    "\n",
    "SEMMED_predicates[SEMMED_predicates['SemmedPred'].isin(pred_compare)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c658e0a2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Review the mappings in the table printed by the previous code chunk. \n",
    "\n",
    "The mapping for several SEMMEDDB predicates are missing. \n",
    "\n",
    "I'll add the mappings for some predicates below: and will be added below\n",
    "* augments\n",
    "* inhibits\n",
    "* stimulates\n",
    "\n",
    "I'll keep some as missing (mapping doesn't exist in biolink-model):\n",
    "* measurement_of \n",
    "* method_of\n",
    "\n",
    "Notes:\n",
    "* semmeddb `administered_to` is \"trying something out with a goal\", not that it actually does something...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)\n",
    "* semmeddb `diagnoses` is more of a \"this thing distinguishes / identifies this other thing\"...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T163)\n",
    "* semmeddb `process_of` does seem to map well to biolink-model's `occurs in` so...okay. As an example, see the use of `process_of` in the [API's data](https://biothings.ncats.io/semmeddb/query?q=predicate:PROCESS_OF%20AND%20object.semantic_type_abbreviation:cell)\n",
    "* keeping semmeddb `treats` is more general than biolink-model's `treats`. It's more of \"trying this out as a treatment\". see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5040417f",
   "metadata": {},
   "source": [
    "Starting with the biolink 3.1.1 update, we:\n",
    "* map semmeddb `augments` and `stimulates` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: increased\n",
    "* map semmeddb `inhibits` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: decreased\n",
    "\n",
    "And for the reverse of these operations with qualifiers, we'll want to get the inverted qualified_predicates and we'll use \"subject_\" rather than \"object_\" qualifiers (but the values will remain the same). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "77adf097",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:37.848590Z",
     "start_time": "2023-08-07T21:10:37.838261Z"
    }
   },
   "outputs": [],
   "source": [
    "## complex change: modify table to use qualifiers for semmeddb augments, stimulates, and inhibits\n",
    "\n",
    "## At the moment, only need the following qualifiers added, to map these predicates\n",
    "##   QualPred = qualified_predicate\n",
    "##   ObjAsp = object_aspect_qualifier\n",
    "##   ObjDirect = object_direction_qualifier\n",
    "##   Inv_QualPred = for reverse operations, inverse of qualified_predicate\n",
    "##   Inv_SubjAsp = for reverse operations, subject_aspect_qualifier\n",
    "##   Inv_SubjDirect = for reverse operations, subject_direction_qualifier\n",
    "\n",
    "\n",
    "## First, add the columns. Most values will stay blank (None)\n",
    "SEMMED_predicates['BMQualPred'] = None  \n",
    "SEMMED_predicates['BMObjAsp'] = None \n",
    "SEMMED_predicates['BMObjDirect'] = None\n",
    "SEMMED_predicates['BM_Inv_QualPred'] = None  \n",
    "SEMMED_predicates['BM_Inv_SubjAsp'] = None \n",
    "SEMMED_predicates['BM_Inv_SubjDirect'] = None\n",
    "\n",
    "## Next, mutate the rows for biolink predicates being remapping to predicates + qualifiers\n",
    "for index in SEMMED_predicates.index:\n",
    "    ## semmeddb augments and stimulates should change\n",
    "    if (SEMMED_predicates.loc[index,'SemmedPred']=='augments') or \\\n",
    "    (SEMMED_predicates.loc[index,'SemmedPred']=='stimulates'):\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'increased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'increased'\n",
    "    ## semmeddb inhibits should change\n",
    "    elif SEMMED_predicates.loc[index,'SemmedPred']=='inhibits':\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'decreased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'decreased'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933327b9",
   "metadata": {},
   "source": [
    "Clean up table to only include what we have biolink-model mappings for: this removes measurement_of and method_of predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "80e84505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:40.350062Z",
     "start_time": "2023-08-07T21:10:40.346786Z"
    }
   },
   "outputs": [],
   "source": [
    "## only keep the stuff we found biolink mappings for\n",
    "SEMMED_predicates = SEMMED_predicates[SEMMED_predicates['BMPred'].notna()].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd31635",
   "metadata": {},
   "source": [
    "Now to get the inverse predicates. With biolink-model-toolkit > 0.9.0 and biolink-model > 3.1.1, I've encountered issues retrieving inverse predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f9acdf7f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:43.243625Z",
     "start_time": "2023-08-07T21:10:43.241126Z"
    }
   },
   "outputs": [],
   "source": [
    "## Testing methods of getting inverse\n",
    "# print(bmt_tool.get_element('causes').inverse)  ## get None\n",
    "# print(bmt_tool.get_inverse('subclass of'))   ## get None\n",
    "# print(bmt_tool.get_inverse('subclass_of'))   ## get AttributeError: 'NoneType' object has no attribute 'name'\n",
    "# print(bmt_tool.get_inverse('quantifier_qualifier'))    ## also get AttributeError\n",
    "\n",
    "# for i in SEMMED_predicates['BMPred']:\n",
    "#     print(i)\n",
    "#     if bmt_tool.get_element(i).symmetric:\n",
    "#         print('is symmetrical')\n",
    "#     elif bmt_tool.get_inverse(i):\n",
    "#         print('inverse: ' + bmt_tool.get_inverse(i))\n",
    "#     else:\n",
    "#         print('missing inverse')\n",
    "#     print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "52cfbbc4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:43.747797Z",
     "start_time": "2023-08-07T21:10:43.741669Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['BM_Inv_Pred'] = [i if bmt_tool.get_element(i).symmetric\n",
    "                                      else bmt_tool.get_inverse(i) if bmt_tool.get_inverse(i)\n",
    "                                      else 'MISSING' \n",
    "                                    for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates.reindex(columns=[\n",
    "    'SemmedPred', 'Semmed_in_BM', 'BMPred', 'BMQualPred', 'BMObjAsp', 'BMObjDirect', \n",
    "    'BM_Inv_Pred', 'BM_Inv_QualPred', 'BM_Inv_SubjAsp', 'BM_Inv_SubjDirect'])\n",
    "\n",
    "## old code that doesn't work anymore, was previously used to get inverses for predicates and qualified predicates\n",
    "## get inverses for basic predicates + qualified_predicates to generate reverse operations\n",
    "# SEMMED_predicates['BM_Inv_Pred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "# SEMMED_predicates['BM_Inv_QualPred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if i and isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMQualPred']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "9718fe3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:45.617039Z",
     "start_time": "2023-08-07T21:10:45.605239Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is exacerbated by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is diagnosed by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has predisposing factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is input of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO            related to   \n",
       "145           affects           SEMMEDDB:AFFECTS               affects   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related to   \n",
       "55           augments          SEMMEDDB:AUGMENTS               affects   \n",
       "141            causes            SEMMEDDB:CAUSES                causes   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists with   \n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives into   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts   \n",
       "58           inhibits          SEMMEDDB:INHIBITS               affects   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts with   \n",
       "172               isa               SEMMEDDB:ISA           subclass of   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location of   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation of   \n",
       "156          measures          SEMMEDDB:MEASURES  quantifier qualifier   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs in   \n",
       "127           part_of           SEMMEDDB:PART_OF               part of   \n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes   \n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs in   \n",
       "138          produces          SEMMEDDB:PRODUCES              produces   \n",
       "60            same_as           SEMMEDDB:same_as           close match   \n",
       "61         stimulates        SEMMEDDB:STIMULATES               affects   \n",
       "148            treats            SEMMEDDB:TREATS                treats   \n",
       "149              uses              SEMMEDDB:USES             has input   \n",
       "\n",
       "    BMQualPred               BMObjAsp BMObjDirect              BM_Inv_Pred  \\\n",
       "54        None                   None        None               related to   \n",
       "145       None                   None        None              affected by   \n",
       "160       None                   None        None               related to   \n",
       "55      causes  activity_or_abundance   increased              affected by   \n",
       "141       None                   None        None                caused by   \n",
       "56        None                   None        None            coexists with   \n",
       "143       None                   None        None        is exacerbated by   \n",
       "57        None                   None        None             derives from   \n",
       "157       None                   None        None          is diagnosed by   \n",
       "140       None                   None        None             disrupted by   \n",
       "58      causes  activity_or_abundance   decreased              affected by   \n",
       "136       None                   None        None           interacts with   \n",
       "172       None                   None        None                  MISSING   \n",
       "129       None                   None        None               located in   \n",
       "144       None                   None        None        has manifestation   \n",
       "156       None                   None        None                  MISSING   \n",
       "146       None                   None        None         contains process   \n",
       "127       None                   None        None                 has part   \n",
       "132       None                   None        None              preceded by   \n",
       "59        None                   None        None  has predisposing factor   \n",
       "142       None                   None        None             prevented by   \n",
       "134       None                   None        None         contains process   \n",
       "138       None                   None        None              produced by   \n",
       "60        None                   None        None              close match   \n",
       "61      causes  activity_or_abundance   increased              affected by   \n",
       "148       None                   None        None               treated by   \n",
       "149       None                   None        None              is input of   \n",
       "\n",
       "    BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54             None                   None              None  \n",
       "145            None                   None              None  \n",
       "160            None                   None              None  \n",
       "55        caused_by  activity_or_abundance         increased  \n",
       "141            None                   None              None  \n",
       "56             None                   None              None  \n",
       "143            None                   None              None  \n",
       "57             None                   None              None  \n",
       "157            None                   None              None  \n",
       "140            None                   None              None  \n",
       "58        caused_by  activity_or_abundance         decreased  \n",
       "136            None                   None              None  \n",
       "172            None                   None              None  \n",
       "129            None                   None              None  \n",
       "144            None                   None              None  \n",
       "156            None                   None              None  \n",
       "146            None                   None              None  \n",
       "127            None                   None              None  \n",
       "132            None                   None              None  \n",
       "59             None                   None              None  \n",
       "142            None                   None              None  \n",
       "134            None                   None              None  \n",
       "138            None                   None              None  \n",
       "60             None                   None              None  \n",
       "61        caused_by  activity_or_abundance         increased  \n",
       "148            None                   None              None  \n",
       "149            None                   None              None  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd4046d",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Check if the predicate inverses look correct. \n",
    "\n",
    "For bmt >=1.1.0 and biolink-model >=3.5.2, we want to:\n",
    "* leave the inverse of 'subclass of' as 'MISSING' because no inverse exists, so we'll want to remove it in the next step\n",
    "* leave the inverse of 'quanitifer qualifier' as 'MISSING' since it's actually not a predicate, so we'll want to remove it in the next step\n",
    "\n",
    "Old notes: \n",
    "* Maybe some aren't correct because they don't have an inverse but they are directional...If so, I suggest going back and changing the predicates to ones with inverses. \n",
    "* Some predicates aren't directional (their entry in the biolink-model yaml will have the property `symmetric == true`) so those are identical in either direction..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b121018d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:52.277230Z",
     "start_time": "2023-08-07T21:10:52.268237Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SemmedPred       Semmed_in_BM                BMPred BMQualPred BMObjAsp  \\\n",
       "172        isa       SEMMEDDB:ISA           subclass of       None     None   \n",
       "156   measures  SEMMEDDB:MEASURES  quantifier qualifier       None     None   \n",
       "\n",
       "    BMObjDirect BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "172        None     MISSING            None           None              None  \n",
       "156        None     MISSING            None           None              None  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46765d17",
   "metadata": {},
   "source": [
    "Now we'll make sure all biolink-model-mapped columns are in snake_case, and clean up the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "85224c41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:57.049301Z",
     "start_time": "2023-08-07T21:10:57.044818Z"
    }
   },
   "outputs": [],
   "source": [
    "## FOR RESOURCES, these predicates must all be in snake_case\n",
    "for col in ['BMPred', 'BMQualPred', 'BM_Inv_Pred', 'BM_Inv_QualPred']:\n",
    "    SEMMED_predicates[col] = SEMMED_predicates[col].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "50027e5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:57.573529Z",
     "start_time": "2023-08-07T21:10:57.568013Z"
    }
   },
   "outputs": [],
   "source": [
    "## can mutate and rename the Semmed_in_BM column (not needed anymore) with the format needed\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['SemmedPred'].str.upper()\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.replace(\"-\", \"\")\n",
    "SEMMED_predicates.rename(columns = {'Semmed_in_BM':'Semmed_in_Data'}, inplace = True)\n",
    "\n",
    "## special handling keep lowercase for same_as\n",
    "SEMMED_predicates['Semmed_in_Data'] = [i.lower() if i == 'SAME_AS' else i\n",
    "                                       for i in SEMMED_predicates['Semmed_in_Data']]\n",
    "\n",
    "## Can remove the SemmedPred column: not needed anymore\n",
    "SEMMED_predicates.drop(columns='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "946dd392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:10:58.063359Z",
     "start_time": "2023-08-07T21:10:58.052403Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data                BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO            related_to       None                   None   \n",
       "145           AFFECTS               affects       None                   None   \n",
       "160   ASSOCIATED_WITH            related_to       None                   None   \n",
       "55           AUGMENTS               affects     causes  activity_or_abundance   \n",
       "141            CAUSES                causes       None                   None   \n",
       "56      COEXISTS_WITH         coexists_with       None                   None   \n",
       "143       COMPLICATES           exacerbates       None                   None   \n",
       "57        CONVERTS_TO          derives_into       None                   None   \n",
       "157         DIAGNOSES             diagnoses       None                   None   \n",
       "140          DISRUPTS              disrupts       None                   None   \n",
       "58           INHIBITS               affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH        interacts_with       None                   None   \n",
       "172               ISA           subclass_of       None                   None   \n",
       "129       LOCATION_OF           location_of       None                   None   \n",
       "144  MANIFESTATION_OF      manifestation_of       None                   None   \n",
       "156          MEASURES  quantifier_qualifier       None                   None   \n",
       "146         OCCURS_IN             occurs_in       None                   None   \n",
       "127           PART_OF               part_of       None                   None   \n",
       "132          PRECEDES              precedes       None                   None   \n",
       "59        PREDISPOSES           predisposes       None                   None   \n",
       "142          PREVENTS              prevents       None                   None   \n",
       "134        PROCESS_OF             occurs_in       None                   None   \n",
       "138          PRODUCES              produces       None                   None   \n",
       "60            same_as           close_match       None                   None   \n",
       "61         STIMULATES               affects     causes  activity_or_abundance   \n",
       "148            TREATS                treats       None                   None   \n",
       "149              USES             has_input       None                   None   \n",
       "\n",
       "    BMObjDirect              BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54         None               related_to            None   \n",
       "145        None              affected_by            None   \n",
       "160        None               related_to            None   \n",
       "55    increased              affected_by       caused_by   \n",
       "141        None                caused_by            None   \n",
       "56         None            coexists_with            None   \n",
       "143        None        is_exacerbated_by            None   \n",
       "57         None             derives_from            None   \n",
       "157        None          is_diagnosed_by            None   \n",
       "140        None             disrupted_by            None   \n",
       "58    decreased              affected_by       caused_by   \n",
       "136        None           interacts_with            None   \n",
       "172        None                  MISSING            None   \n",
       "129        None               located_in            None   \n",
       "144        None        has_manifestation            None   \n",
       "156        None                  MISSING            None   \n",
       "146        None         contains_process            None   \n",
       "127        None                 has_part            None   \n",
       "132        None              preceded_by            None   \n",
       "59         None  has_predisposing_factor            None   \n",
       "142        None             prevented_by            None   \n",
       "134        None         contains_process            None   \n",
       "138        None              produced_by            None   \n",
       "60         None              close_match            None   \n",
       "61    increased              affected_by       caused_by   \n",
       "148        None               treated_by            None   \n",
       "149        None              is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "172                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "156                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c616c4ff",
   "metadata": {},
   "source": [
    "#### custom-prune the data of some predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f0387a",
   "metadata": {},
   "source": [
    "We are still doing this because some semmeddb predicates or their mappings are problematic. \n",
    "\n",
    "As of 2023-08-07: the Translator exclusions weren't on a predicate level. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9079161f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:11:02.257998Z",
     "start_time": "2023-08-07T21:11:02.252253Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in data but not in mapping table\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'MEASUREMENT_OF', 'METHOD_OF'}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in mapping file but not in data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ISA'}"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in data but not in mapping table\")\n",
    "predicates - set(SEMMED_predicates['Semmed_in_Data'])\n",
    "\n",
    "print(\"in mapping file but not in data\")\n",
    "set(SEMMED_predicates['Semmed_in_Data']) - predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b8dc8d8c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:11:04.949257Z",
     "start_time": "2023-08-07T21:11:04.940949Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Semmed_in_Data                BMPred BMQualPred BMObjAsp BMObjDirect  \\\n",
       "172            ISA           subclass_of       None     None        None   \n",
       "156       MEASURES  quantifier_qualifier       None     None        None   \n",
       "\n",
       "    BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "172     MISSING            None           None              None  \n",
       "156     MISSING            None           None              None  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8b1e2b",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Remove predicates in the mapping table, but we don't want to use for associations:\n",
    "- no biolink-model mapping: METHOD_OF, MEASUREMENT_OF\n",
    "- no biolink-model inverse (`MISSING`) and other issues:\n",
    "  - ISA: already removed from data during Basic Filtering section 2.3 (predicates)\n",
    "  - MEASURES: mapping isn't to a predicate (quantifier_qualifier)\n",
    "\n",
    "Past: used to exclude USES (not helpful?), but now keeping because the predicate has been reviewed in the Translator-curated exclusions effort. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5e25a929",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:11:28.210098Z",
     "start_time": "2023-08-07T21:11:28.206553Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ISA', 'MEASUREMENT_OF', 'MEASURES', 'METHOD_OF'}"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_removals = {'MEASUREMENT_OF', 'METHOD_OF', 'ISA', 'MEASURES'}\n",
    "more_removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "5a46158c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:11:29.768134Z",
     "start_time": "2023-08-07T21:11:29.755199Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data            BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO        related_to       None                   None   \n",
       "145           AFFECTS           affects       None                   None   \n",
       "160   ASSOCIATED_WITH        related_to       None                   None   \n",
       "55           AUGMENTS           affects     causes  activity_or_abundance   \n",
       "141            CAUSES            causes       None                   None   \n",
       "56      COEXISTS_WITH     coexists_with       None                   None   \n",
       "143       COMPLICATES       exacerbates       None                   None   \n",
       "57        CONVERTS_TO      derives_into       None                   None   \n",
       "157         DIAGNOSES         diagnoses       None                   None   \n",
       "140          DISRUPTS          disrupts       None                   None   \n",
       "58           INHIBITS           affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH    interacts_with       None                   None   \n",
       "129       LOCATION_OF       location_of       None                   None   \n",
       "144  MANIFESTATION_OF  manifestation_of       None                   None   \n",
       "146         OCCURS_IN         occurs_in       None                   None   \n",
       "127           PART_OF           part_of       None                   None   \n",
       "132          PRECEDES          precedes       None                   None   \n",
       "59        PREDISPOSES       predisposes       None                   None   \n",
       "142          PREVENTS          prevents       None                   None   \n",
       "134        PROCESS_OF         occurs_in       None                   None   \n",
       "138          PRODUCES          produces       None                   None   \n",
       "60            same_as       close_match       None                   None   \n",
       "61         STIMULATES           affects     causes  activity_or_abundance   \n",
       "148            TREATS            treats       None                   None   \n",
       "149              USES         has_input       None                   None   \n",
       "\n",
       "    BMObjDirect              BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54         None               related_to            None   \n",
       "145        None              affected_by            None   \n",
       "160        None               related_to            None   \n",
       "55    increased              affected_by       caused_by   \n",
       "141        None                caused_by            None   \n",
       "56         None            coexists_with            None   \n",
       "143        None        is_exacerbated_by            None   \n",
       "57         None             derives_from            None   \n",
       "157        None          is_diagnosed_by            None   \n",
       "140        None             disrupted_by            None   \n",
       "58    decreased              affected_by       caused_by   \n",
       "136        None           interacts_with            None   \n",
       "129        None               located_in            None   \n",
       "144        None        has_manifestation            None   \n",
       "146        None         contains_process            None   \n",
       "127        None                 has_part            None   \n",
       "132        None              preceded_by            None   \n",
       "59         None  has_predisposing_factor            None   \n",
       "142        None             prevented_by            None   \n",
       "134        None         contains_process            None   \n",
       "138        None              produced_by            None   \n",
       "60         None              close_match            None   \n",
       "61    increased              affected_by       caused_by   \n",
       "148        None               treated_by            None   \n",
       "149        None              is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(25, 9)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remove this set from the SEMMED_predicates table\n",
    "SEMMED_predicates = SEMMED_predicates[~ SEMMED_predicates['Semmed_in_Data'].isin(more_removals)]\n",
    "SEMMED_predicates\n",
    "SEMMED_predicates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "64cacd52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:13:00.239727Z",
     "start_time": "2023-08-07T21:12:55.727637Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove this set from the data record\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(more_removals)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "530e3631",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:13:10.105817Z",
     "start_time": "2023-08-07T21:13:01.762578Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "109"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 112, still 112 (expected)\n",
    "len(object_semtypes)  ## was 110, still 109 (cool, not intended)\n",
    "len(predicates)       ## was 28, now 25: decreased by 3 (expected: ISA was already removed earlier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "863543b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:15:37.490939Z",
     "start_time": "2023-08-07T21:15:25.186357Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(13150, 4)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>humn</td>\n",
       "      <td>2143816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1159577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>humn</td>\n",
       "      <td>1000549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>884425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>707094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "0            dsyn   PROCESS_OF           humn  2143816\n",
       "1            bpoc      PART_OF           mamm  1159577\n",
       "2            bpoc  LOCATION_OF           neop  1023005\n",
       "3            fndg   PROCESS_OF           humn  1000549\n",
       "4            bpoc  LOCATION_OF           aapp   953010\n",
       "5            bdsu  LOCATION_OF           aapp   884425\n",
       "6            topp       TREATS           dsyn   849825\n",
       "7            cell  LOCATION_OF           aapp   815374\n",
       "8            bpoc  LOCATION_OF           patf   799234\n",
       "9            topp       TREATS           neop   707094"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at number of combos after this removal\n",
    "combos = filtered_data.value_counts().reset_index()\n",
    "combos.columns = ['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE', 'COUNT']\n",
    "combos.shape\n",
    "combos.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426473de",
   "metadata": {},
   "source": [
    "### Use translator-curated exclusions to prune combos "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2870790d",
   "metadata": {},
   "source": [
    "I'll use Translator-curated exclusions to remove some semantic-type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n",
    "\n",
    "Translator-curated exclusions work on the combo-level. \n",
    "\n",
    "Note: the `_t_code` columns sometimes have the value of `not_found`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "dbd6e003",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:16:34.348615Z",
     "start_time": "2023-08-07T21:16:34.127295Z"
    }
   },
   "outputs": [],
   "source": [
    "imported_exclusions = pd.read_csv(\n",
    "    'https://raw.githubusercontent.com/biolink/biolink-model/master/SEMMEDDB_exclude_list.tsv',\n",
    "    sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "5f404ea6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:16:36.259648Z",
     "start_time": "2023-08-07T21:16:36.249523Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "0                chem                  T103              NaN   \n",
       "1                 NaN                   NaN              NaN   \n",
       "2                chvs                  T104              NaN   \n",
       "3                 NaN                   NaN              NaN   \n",
       "4                chvf                  T120              NaN   \n",
       "\n",
       "  semmed_object_code semmed_object_t_code           exclusion_type  \n",
       "0                NaN                  NaN  semantic type exclusion  \n",
       "1               chem                 T103  semantic type exclusion  \n",
       "2                NaN                  NaN  semantic type exclusion  \n",
       "3               chvs                 T104  semantic type exclusion  \n",
       "4                NaN                  NaN  semantic type exclusion  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(1443, 6)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['semantic type exclusion', 'Domain exclusion', 'Range exclusion'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions.head()\n",
    "imported_exclusions.shape\n",
    "imported_exclusions.exclusion_type.unique()  ## number of unique kinds of exclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99185fc",
   "metadata": {},
   "source": [
    "First, I'm checking that the data fits the \"rules\" I expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "efd0d556",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:48:07.759308Z",
     "start_time": "2023-08-07T21:48:07.752193Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    ## all 'semantic type exclusions' don't consider the predicates, \n",
    "    ##   just specific entities/instances of the types\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_predicate, str):\n",
    "            print('semantic-type exclusion problem: predicate there?')\n",
    "            print(row)\n",
    "    ## 'Domain exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        if (not isinstance(row.semmed_subject_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('domain exclusion problem: missing subject or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_object_code, str):\n",
    "            print('domain exclusion problem: object present')\n",
    "            print(row)\n",
    "    ## 'Range exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        if (not isinstance(row.semmed_object_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('range exclusion problem: missing object or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_subject_code, str):\n",
    "            print('range exclusion problem: subject present')\n",
    "            print(row)\n",
    "    else: \n",
    "        print('some other kind of exclusion?')\n",
    "        print(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea81cf44",
   "metadata": {},
   "source": [
    "Then, I noticed the same_as predicate was not in the expected case, so adjusting this..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2016c1d9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:02:35.150668Z",
     "start_time": "2023-08-07T22:02:35.137555Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          SAME_AS   \n",
       "716                 chvs                  T104          SAME_AS   \n",
       "717                 resa                  T062          SAME_AS   \n",
       "718                 chem                  T103          SAME_AS   \n",
       "719                 qnco                  T081          SAME_AS   \n",
       "720                 ftcn                  T169          SAME_AS   \n",
       "721                 inpr                  T170          SAME_AS   \n",
       "722                 tmco                  T079          SAME_AS   \n",
       "1387                 NaN                   NaN          SAME_AS   \n",
       "1388                 NaN                   NaN          SAME_AS   \n",
       "1389                 NaN                   NaN          SAME_AS   \n",
       "1390                 NaN                   NaN          SAME_AS   \n",
       "1391                 NaN                   NaN          SAME_AS   \n",
       "1392                 NaN                   NaN          SAME_AS   \n",
       "1393                 NaN                   NaN          SAME_AS   \n",
       "1394                 NaN                   NaN          SAME_AS   \n",
       "1395                 NaN                   NaN          SAME_AS   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "dc0f279f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:04:54.192868Z",
     "start_time": "2023-08-07T22:04:54.189775Z"
    }
   },
   "outputs": [],
   "source": [
    "imported_exclusions['semmed_predicate'] = ['same_as' if i == 'SAME_AS' else i \\\n",
    "                                           for i in imported_exclusions['semmed_predicate']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "a35676c0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:05:03.525166Z",
     "start_time": "2023-08-07T22:05:03.512817Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          same_as   \n",
       "716                 chvs                  T104          same_as   \n",
       "717                 resa                  T062          same_as   \n",
       "718                 chem                  T103          same_as   \n",
       "719                 qnco                  T081          same_as   \n",
       "720                 ftcn                  T169          same_as   \n",
       "721                 inpr                  T170          same_as   \n",
       "722                 tmco                  T079          same_as   \n",
       "1387                 NaN                   NaN          same_as   \n",
       "1388                 NaN                   NaN          same_as   \n",
       "1389                 NaN                   NaN          same_as   \n",
       "1390                 NaN                   NaN          same_as   \n",
       "1391                 NaN                   NaN          same_as   \n",
       "1392                 NaN                   NaN          same_as   \n",
       "1393                 NaN                   NaN          same_as   \n",
       "1394                 NaN                   NaN          same_as   \n",
       "1395                 NaN                   NaN          same_as   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e6ba83",
   "metadata": {},
   "source": [
    "Now I can iterate through the `imported_exclusions` table and remove rows from combo that match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "0ce1efaa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T21:50:04.213360Z",
     "start_time": "2023-08-07T21:50:04.203915Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>hlca</td>\n",
       "      <td>T058</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>lbpr</td>\n",
       "      <td>T059</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>796</th>\n",
       "      <td>idcn</td>\n",
       "      <td>T078</td>\n",
       "      <td>USES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>797</th>\n",
       "      <td>qlco</td>\n",
       "      <td>T080</td>\n",
       "      <td>USES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>rcpt</td>\n",
       "      <td>T192</td>\n",
       "      <td>USES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>strd</td>\n",
       "      <td>not_found</td>\n",
       "      <td>USES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>vita</td>\n",
       "      <td>T127</td>\n",
       "      <td>USES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>683 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "118                hlca                  T058  ADMINISTERED_TO   \n",
       "119                inpr                  T170  ADMINISTERED_TO   \n",
       "120                lbpr                  T059  ADMINISTERED_TO   \n",
       "121                resa                  T062  ADMINISTERED_TO   \n",
       "122                qnco                  T081  ADMINISTERED_TO   \n",
       "..                  ...                   ...              ...   \n",
       "796                idcn                  T078             USES   \n",
       "797                qlco                  T080             USES   \n",
       "798                rcpt                  T192             USES   \n",
       "799                strd             not_found             USES   \n",
       "800                vita                  T127             USES   \n",
       "\n",
       "    semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "118                NaN                  NaN  Domain exclusion  \n",
       "119                NaN                  NaN  Domain exclusion  \n",
       "120                NaN                  NaN  Domain exclusion  \n",
       "121                NaN                  NaN  Domain exclusion  \n",
       "122                NaN                  NaN  Domain exclusion  \n",
       "..                 ...                  ...               ...  \n",
       "796                NaN                  NaN  Domain exclusion  \n",
       "797                NaN                  NaN  Domain exclusion  \n",
       "798                NaN                  NaN  Domain exclusion  \n",
       "799                NaN                  NaN  Domain exclusion  \n",
       "800                NaN                  NaN  Domain exclusion  \n",
       "\n",
       "[683 rows x 6 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions[imported_exclusions['exclusion_type'] == 'Domain exclusion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "e5bb1ec1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:18:51.096946Z",
     "start_time": "2023-08-07T22:18:48.222995Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_subject_code, str):  ## exclusion is in the subject\n",
    "            combos = combos[ ~ (combos['SUBJECT_SEMTYPE'] == row.semmed_subject_code)]\n",
    "        else:  ## exclusion should be in object\n",
    "            combos = combos[ ~ (combos['OBJECT_SEMTYPE'] == row.semmed_object_code)]\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        combos = combos[~ ( (combos['SUBJECT_SEMTYPE'] == row.semmed_subject_code ) & \n",
    "                            (combos['PREDICATE'] == row.semmed_predicate ))]\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        combos = combos[~ ( (combos['OBJECT_SEMTYPE'] == row.semmed_object_code ) & \n",
    "                            (combos['PREDICATE'] == row.semmed_predicate ))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "bf7a6309",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:18:52.041626Z",
     "start_time": "2023-08-07T22:18:52.033646Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9667, 4)"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>884425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>707094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>696732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>fndg</td>\n",
       "      <td>670960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>phsu</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>660654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "2             bpoc  LOCATION_OF           neop  1023005\n",
       "4             bpoc  LOCATION_OF           aapp   953010\n",
       "5             bdsu  LOCATION_OF           aapp   884425\n",
       "6             topp       TREATS           dsyn   849825\n",
       "7             cell  LOCATION_OF           aapp   815374\n",
       "8             bpoc  LOCATION_OF           patf   799234\n",
       "9             topp       TREATS           neop   707094\n",
       "10            bpoc  LOCATION_OF           dsyn   696732\n",
       "11            bpoc  LOCATION_OF           fndg   670960\n",
       "13            phsu       TREATS           dsyn   660654"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos.shape\n",
    "combos.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e9660d",
   "metadata": {},
   "source": [
    "## Final Filter: row counts per combo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305f3a3a",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "use the code block below to decide how many combos to keep based on how many predications/records there are per combo...\n",
    "\n",
    "---\n",
    "\n",
    "Current:\n",
    "\n",
    "Counts right now are still based on predication. For now, [we want > 3 publications for each triple](https://github.com/NCATSTranslator/Feedback/issues/100#issuecomment-1632806388), so the count must be > 3 at least..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "d874b8f2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:16:36.163437Z",
     "start_time": "2023-08-07T22:16:36.154622Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>884425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10034</th>\n",
       "      <td>medd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>virs</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10035</th>\n",
       "      <td>anab</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>neop</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10036</th>\n",
       "      <td>opco</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>aggp</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10038</th>\n",
       "      <td>moft</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>gngm</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10041</th>\n",
       "      <td>patf</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>nnon</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7685 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      SUBJECT_SEMTYPE        PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "2                bpoc      LOCATION_OF           neop  1023005\n",
       "4                bpoc      LOCATION_OF           aapp   953010\n",
       "5                bdsu      LOCATION_OF           aapp   884425\n",
       "6                topp           TREATS           dsyn   849825\n",
       "7                cell      LOCATION_OF           aapp   815374\n",
       "...               ...              ...            ...      ...\n",
       "10034            medd           TREATS           virs        4\n",
       "10035            anab          AFFECTS           neop        4\n",
       "10036            opco  ADMINISTERED_TO           aggp        4\n",
       "10038            moft       PROCESS_OF           gngm        4\n",
       "10041            patf         PRODUCES           nnon        4\n",
       "\n",
       "[7685 rows x 4 columns]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos[(combos['COUNT'] > 3)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaf76b09",
   "metadata": {},
   "source": [
    "Hmmm...I'm not happy with some odd same_as Metatriples that are still around, but keeping them in for now..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "86920c33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:19:08.150072Z",
     "start_time": "2023-08-07T22:19:08.139694Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1718</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>orch</td>\n",
       "      <td>2849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3832</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>phsu</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4151</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>aapp</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4196</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>medd</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4402</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>topp</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4575</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>hops</td>\n",
       "      <td>184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4608</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>antb</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4627</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>bacs</td>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5519</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>horm</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5643</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>inch</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5692</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>gngm</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6225</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>nnon</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6292</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>nsba</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6443</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>carb</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6630</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>lipd</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6631</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>irda</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6741</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>elii</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6898</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>opco</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6911</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>clnd</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7591</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>bodm</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8817</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>eico</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8962</th>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>strd</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_SEMTYPE PREDICATE OBJECT_SEMTYPE  COUNT\n",
       "1718            orch   same_as           orch   2849\n",
       "3832            orch   same_as           phsu    343\n",
       "4151            orch   same_as           aapp    257\n",
       "4196            orch   same_as           medd    248\n",
       "4402            orch   same_as           topp    209\n",
       "4575            orch   same_as           hops    184\n",
       "4608            orch   same_as           antb    177\n",
       "4627            orch   same_as           bacs    175\n",
       "5519            orch   same_as           horm     82\n",
       "5643            orch   same_as           inch     74\n",
       "5692            orch   same_as           gngm     72\n",
       "6225            orch   same_as           nnon     48\n",
       "6292            orch   same_as           nsba     46\n",
       "6443            orch   same_as           carb     41\n",
       "6630            orch   same_as           lipd     35\n",
       "6631            orch   same_as           irda     35\n",
       "6741            orch   same_as           elii     33\n",
       "6898            orch   same_as           opco     29\n",
       "6911            orch   same_as           clnd     29\n",
       "7591            orch   same_as           bodm     18\n",
       "8817            orch   same_as           eico      8\n",
       "8962            orch   same_as           strd      7"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos[(combos['COUNT'] > 3) &\n",
    "       (combos['PREDICATE'] == 'same_as') & \n",
    "       (combos['SUBJECT_SEMTYPE'] == 'orch')\n",
    "      ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "2c859df9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:19:28.181481Z",
     "start_time": "2023-08-07T22:19:28.168932Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7685, 3)"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE\n",
       "2            bpoc  LOCATION_OF           neop\n",
       "4            bpoc  LOCATION_OF           aapp\n",
       "5            bdsu  LOCATION_OF           aapp"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos = combos[(combos['COUNT'] > 3)].copy()\n",
    "filtered_combos.drop(columns='COUNT', inplace=True)\n",
    "filtered_combos.shape\n",
    "filtered_combos[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "b53030eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:19:46.782439Z",
     "start_time": "2023-08-07T22:19:46.773570Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all subject to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']], \n",
    "                      how='left', left_on='SUBJECT_SEMTYPE', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject',\n",
    "                  'BiolinkSubject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9e435399",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:19:47.734493Z",
     "start_time": "2023-08-07T22:19:47.727246Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all object to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']],\n",
    "                      how='left', left_on='OriginalObject', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject',\n",
    "                  'BiolinkSubject', 'BiolinkObject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "a0fe51f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:19:49.612370Z",
     "start_time": "2023-08-07T22:19:49.603519Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7680</th>\n",
       "      <td>medd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>virs</td>\n",
       "      <td>Device</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7681</th>\n",
       "      <td>anab</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>neop</td>\n",
       "      <td>Disease</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7682</th>\n",
       "      <td>opco</td>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>aggp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7683</th>\n",
       "      <td>moft</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>gngm</td>\n",
       "      <td>MolecularActivity</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7684</th>\n",
       "      <td>patf</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>nnon</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7685 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     OriginalSubject OriginalPredicate OriginalObject  \\\n",
       "0               bpoc       LOCATION_OF           neop   \n",
       "1               bpoc       LOCATION_OF           aapp   \n",
       "2               bdsu       LOCATION_OF           aapp   \n",
       "3               topp            TREATS           dsyn   \n",
       "4               cell       LOCATION_OF           aapp   \n",
       "...              ...               ...            ...   \n",
       "7680            medd            TREATS           virs   \n",
       "7681            anab           AFFECTS           neop   \n",
       "7682            opco   ADMINISTERED_TO           aggp   \n",
       "7683            moft        PROCESS_OF           gngm   \n",
       "7684            patf          PRODUCES           nnon   \n",
       "\n",
       "                BiolinkSubject  BiolinkObject  \n",
       "0     GrossAnatomicalStructure        Disease  \n",
       "1     GrossAnatomicalStructure    Polypeptide  \n",
       "2             AnatomicalEntity    Polypeptide  \n",
       "3                    Procedure        Disease  \n",
       "4                         Cell    Polypeptide  \n",
       "...                        ...            ...  \n",
       "7680                    Device          Virus  \n",
       "7681                   Disease        Disease  \n",
       "7682             SmallMolecule         Cohort  \n",
       "7683         MolecularActivity           Gene  \n",
       "7684       PathologicalProcess  SmallMolecule  \n",
       "\n",
       "[7685 rows x 5 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "3f42bd07",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:20:11.466282Z",
     "start_time": "2023-08-07T22:20:11.455397Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_combos = filtered_combos.merge(\n",
    "                      SEMMED_predicates,\n",
    "                      how='left', left_on='OriginalPredicate', right_on='Semmed_in_Data')\n",
    "\n",
    "filtered_combos.drop(columns = 'Semmed_in_Data', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "6c651d23",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:20:12.256129Z",
     "start_time": "2023-08-07T22:20:12.245836Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  OriginalSubject OriginalPredicate OriginalObject            BiolinkSubject  \\\n",
       "0            bpoc       LOCATION_OF           neop  GrossAnatomicalStructure   \n",
       "1            bpoc       LOCATION_OF           aapp  GrossAnatomicalStructure   \n",
       "2            bdsu       LOCATION_OF           aapp          AnatomicalEntity   \n",
       "\n",
       "  BiolinkObject       BMPred BMQualPred BMObjAsp BMObjDirect BM_Inv_Pred  \\\n",
       "0       Disease  location_of       None     None        None  located_in   \n",
       "1   Polypeptide  location_of       None     None        None  located_in   \n",
       "2   Polypeptide  location_of       None     None        None  located_in   \n",
       "\n",
       "  BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "0            None           None              None  \n",
       "1            None           None              None  \n",
       "2            None           None              None  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "cc7864f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:20:54.835668Z",
     "start_time": "2023-08-07T22:20:54.813351Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## check to make sure everything is mapped to Biolink successfully\n",
    "filtered_combos[filtered_combos['BiolinkSubject'].isna()]\n",
    "filtered_combos[filtered_combos['BiolinkObject'].isna()]\n",
    "\n",
    "filtered_combos[filtered_combos['BMPred'].isna()]\n",
    "filtered_combos[filtered_combos['BM_Inv_Pred'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62b0c6b7",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Optional Analysis: looking at row counts, organized by biolink combos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4fad7760",
   "metadata": {
    "hidden": true
   },
   "source": [
    "PAUSE\n",
    "\n",
    "Deprecated code: will need to include COUNT column after creating the filtered_combos object..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af677da3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-17T05:16:46.369759Z",
     "start_time": "2022-08-17T05:16:46.362883Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "### deep-copy combos so it's not messed up for the next step\n",
    "biolink_counting = filtered_combos.copy()\n",
    "biolink_counting = biolink_counting.groupby(\n",
    "                       ['BiolinkSubject', 'BiolinkPredicate', 'BiolinkObject']\n",
    "                   ).agg(\n",
    "                       { \"COUNT\": \"sum\"}\n",
    ")\n",
    "## other stuff that can go into agg\n",
    "#                          \"OriginalSubject\": lambda x: set(x),\n",
    "#                          \"OriginalPredicate\": lambda x: set(x),\n",
    "#                          \"OriginalObject\": lambda x: set(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "328305ca",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-17T05:16:47.194635Z",
     "start_time": "2022-08-17T05:16:47.181968Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "biolink_counting.reset_index(inplace = True)\n",
    "\n",
    "biolink_counting.sort_values(by='COUNT', ascending=False, inplace = True)\n",
    "\n",
    "biolink_counting[0:50]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870db906",
   "metadata": {},
   "source": [
    "## Generate operation yaml!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344f9ec6",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* If needed, change the code within the functions below to change the x-bte annotations that are made...\n",
    "  * review the qualifier-generating code! Currently it's very simple because all operations with qualifiers have the same set: qualified_predicate, an aspect_qualifier, and a direction_qualifier...\n",
    "* the yaml created below refers to umls-subj and umls-obj...those are specified [here close to the bottom](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "1fcf921d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:01.952564Z",
     "start_time": "2023-08-07T22:22:01.949813Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml=ryml.YAML()\n",
    "folded = ryml.scalarstring.FoldedScalarString\n",
    "doublequote = ryml.scalarstring.DoubleQuotedScalarString"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "222c6bd1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:02.864847Z",
     "start_time": "2023-08-07T22:22:02.858055Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_forward_op(original_subj, original_pred, original_obj,\n",
    "                        biolink_subj, \n",
    "                        biolink_pred, biolink_qualified_pred, biolink_obj_asp, biolink_obj_direct,\n",
    "                        biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    normal_op_name = f\"{original_subj}-{original_pred}-{original_obj}\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## original direction: subject -> object\n",
    "    normal_op_body = folded(\n",
    "    '{' \n",
    "        '\"q\": {{ queryInputs ' \n",
    "        '| replPrefix(\\'predicate:' + f'{original_pred} AND object.semantic_type_abbreviation:{original_obj}' + ' AND pmid_count:>3 AND subject.umls\\')' \n",
    "        '| dump ' \n",
    "        '}}, ' \n",
    "        '\"scopes\": []' \n",
    "    '}')\n",
    "    \n",
    "    if biolink_qualified_pred:\n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has obj aspect and obj direction\n",
    "      temp = {\n",
    "            ## original direction: subject -> object\n",
    "            normal_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## input is subject!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': normal_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## output is object\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_pred,\n",
    "                    'qualifiers': {\n",
    "                        'qualified_predicate': biolink_qualified_pred,\n",
    "                        'object_aspect_qualifier': biolink_obj_asp,\n",
    "                        'object_direction_qualifier': biolink_obj_direct\n",
    "\n",
    "                    },\n",
    "                    'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-obj'  ## matches output as object\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    else:\n",
    "    ## create operation without qualifiers..\n",
    "        temp = {\n",
    "            ## original direction: subject -> object\n",
    "            normal_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## input is subject!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': normal_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## output is object\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_pred,\n",
    "                    'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-obj'  ## matches output as object\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "        \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "912aa75d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:03.733000Z",
     "start_time": "2023-08-07T22:22:03.725938Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_reverse_op(original_subj, original_pred, original_obj,\n",
    "                        biolink_subj, \n",
    "                        ## NOTICE THE INVERSES USED HERE\n",
    "                        biolink_inverse_pred, biolink_inverse_qualified_pred, \n",
    "                        biolink_inverse_subj_asp, biolink_inverse_subj_direct,\n",
    "                        biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    rev_op_name = f\"{original_subj}-{original_pred}-{original_obj}-rev\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## reverse direction: object -> subject\n",
    "    rev_op_body = folded(\n",
    "    '{' \n",
    "        '\"q\": {{ queryInputs ' \n",
    "        '| replPrefix(\\'predicate:' + f'{original_pred} AND subject.semantic_type_abbreviation:{original_subj}' + ' AND pmid_count:>3 AND object.umls\\')' \n",
    "        '| dump ' \n",
    "        '}}, ' \n",
    "        '\"scopes\": []' \n",
    "    '}')\n",
    "    \n",
    "    if biolink_inverse_qualified_pred:\n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has subj aspect and subj direction\n",
    "        temp = {\n",
    "            ## reverse direction: object -> subject\n",
    "            rev_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## input is object!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': rev_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## output is subject\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                    'qualifiers': {\n",
    "                        'qualified_predicate': biolink_inverse_qualified_pred,  ## use inverse pred!\n",
    "                        ## use subject qualifiers\n",
    "                        'subject_aspect_qualifier': biolink_inverse_subj_asp,         \n",
    "                        'subject_direction_qualifier': biolink_inverse_subj_direct                        \n",
    "                    },                    \n",
    "                    'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-subj'  ## matches output as subj\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    else:\n",
    "    ## create operation without qualifiers..\n",
    "        temp = {\n",
    "            ## reverse direction: object -> subject\n",
    "            rev_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## input is object!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': rev_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## output is subject\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                    'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-subj'  ## matches output as subj\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "db9f9f81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:04.824085Z",
     "start_time": "2023-08-07T22:22:04.818464Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_all_operations(combo_df):\n",
    "    op_tracking = set()\n",
    "    \n",
    "    saved = dict()\n",
    "    ## iterate through rows of combos dataframe\n",
    "    for row in combo_df.itertuples(index = False):\n",
    "        \n",
    "        ## forward: only make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,object used here)\n",
    "        forward_op_record = f\"{row.BiolinkSubject}-{row.OriginalPredicate}-{row.OriginalObject}\"\n",
    "        if forward_op_record not in op_tracking:\n",
    "            op_tracking.add(forward_op_record)\n",
    "            saved.update(generate_forward_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_pred = row.BMPred,\n",
    "                                             biolink_qualified_pred = row.BMQualPred,\n",
    "                                             biolink_obj_asp = row.BMObjAsp,\n",
    "                                             biolink_obj_direct = row.BMObjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                            ))        \n",
    "        \n",
    "        ## reverse: make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,subject used here)\n",
    "        reverse_op_record = f\"rev-{row.BiolinkObject}-{row.OriginalPredicate}-{row.OriginalSubject}\"\n",
    "        if reverse_op_record not in op_tracking:\n",
    "            op_tracking.add(reverse_op_record)\n",
    "            saved.update(generate_reverse_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_inverse_pred = row.BM_Inv_Pred,\n",
    "                                             biolink_inverse_qualified_pred = row.BM_Inv_QualPred,\n",
    "                                             biolink_inverse_subj_asp = row.BM_Inv_SubjAsp,\n",
    "                                             biolink_inverse_subj_direct = row.BM_Inv_SubjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                             ))           \n",
    "            \n",
    "    final = {\"x-bte-kgs-operations\": saved}\n",
    "    return final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ec9a4",
   "metadata": {},
   "source": [
    "Get the file made!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "3856781d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:08.180860Z",
     "start_time": "2023-08-07T22:22:08.033749Z"
    }
   },
   "outputs": [],
   "source": [
    "kgs_operations = generate_all_operations(filtered_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fdcd4202",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:09.767391Z",
     "start_time": "2023-08-07T22:22:09.763557Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6851"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "5495d772",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:22:15.163708Z",
     "start_time": "2023-08-07T22:22:15.159809Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'supportBatch': True,\n",
      "  'useTemplating': True,\n",
      "  'inputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
      "  'requestBodyType': 'object',\n",
      "  'requestBody': {'body': '{\"q\": {{ queryInputs | '\n",
      "                          \"replPrefix('predicate:INHIBITS AND \"\n",
      "                          'subject.semantic_type_abbreviation:aapp AND '\n",
      "                          \"pmid_count:>3 AND object.umls')| dump }}, \"\n",
      "                          '\"scopes\": []}'},\n",
      "  'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
      "                 'size': 1000},\n",
      "  'outputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
      "  'predicate': 'affected_by',\n",
      "  'qualifiers': {'qualified_predicate': 'caused_by',\n",
      "                 'subject_aspect_qualifier': 'activity_or_abundance',\n",
      "                 'subject_direction_qualifier': 'decreased'},\n",
      "  'source': 'infores:semmeddb',\n",
      "  'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'}}]\n"
     ]
    }
   ],
   "source": [
    "## double-check qualifier + rev operation is written properly?\n",
    "for i in kgs_operations['x-bte-kgs-operations'].keys():\n",
    "    if ('INHIBITS' in i) and ('rev' in i):\n",
    "        pprint.pprint(kgs_operations['x-bte-kgs-operations'][i], sort_dicts = False)\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd057af2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* it's cool how condensed the operations become, due to the way querying is done (keeping track of not creating duplicated operations using the sets in generate_all_operations function)\n",
    "* set where to download the yamls in the code chunks below\n",
    "  * operations_path\n",
    "  * list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "239dce9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:23:49.831050Z",
     "start_time": "2023-08-07T22:23:38.675566Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.boolean_representation = ['False', 'True']\n",
    "\n",
    "operations_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_operations.yaml\")\n",
    "\n",
    "yaml.dump(kgs_operations, operations_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c946247a",
   "metadata": {},
   "source": [
    "Wait a sec! Need the operations list too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "27403608",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:23:53.997999Z",
     "start_time": "2023-08-07T22:23:53.995063Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_kgs_operations_list(operations_dict):\n",
    "    kgs_op_list = []\n",
    "    for key in operations_dict.keys():\n",
    "        kgs_op_list.append( {\"$ref\": f\"#/components/x-bte-kgs-operations/{key}\"} )\n",
    "    final2 = {\"x-bte-kgs-operations\": kgs_op_list}\n",
    "    return final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "4bec97ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:23:56.010071Z",
     "start_time": "2023-08-07T22:23:56.006353Z"
    }
   },
   "outputs": [],
   "source": [
    "operations_list = generate_kgs_operations_list(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "e64c34db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-07T22:23:57.786044Z",
     "start_time": "2023-08-07T22:23:56.970745Z"
    }
   },
   "outputs": [],
   "source": [
    "list_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_list.yaml\")\n",
    "\n",
    "yaml.dump(operations_list, list_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421fbcdc",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* now the yaml segments downloaded have to be indented manually and inserted into the correct sections of the smartapi yaml...\n",
    "  * It's easier to do with an IDE like Visual Code studio where one can select large sections of text\n",
    "  * the amount of indent to do and where to put things is specified in the [yaml that acts as a template](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)\n",
    "  * the finished file is meant to be [here](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) so one could select-paste the sections directly there"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa95cb1",
   "metadata": {},
   "source": [
    "## defunct: Notes on previously removed semantic types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb04a8e",
   "metadata": {},
   "source": [
    "* amas: (Amino Acid Sequence) looks like protein \"domains\". Examples: Nuclear Export Signals, DNA Binding Domain\n",
    "* acty (Activity) Examples: War, Retirement, Euthanasia, Lifting\n",
    "* dora (Daily or Recreational Activity) Examples: Physical activity, Light Exercise, Relaxation\n",
    "* edac (Educational Activity) Examples: Training, Medical Residencies\n",
    "* gora (Governmental or Regulatory Activity) Examples: Health Care Reform, Advisory Committees\n",
    "* hlca (Health Care Activity) Examples: follow-up, Diagnosis\n",
    "* mcha (Machine Activity) Examples: Refridgeration, Neural Network Simulation\n",
    "* ocac (Occupational Activity) Examples: Promotion, Work, Mining\n",
    "* resa (Research Activity) Examples: Clinical Trials, research study\n",
    "* aggp (Age Group) Examples: Infant, Child, Adult, Elderly\n",
    "* famg (Family Group) Examples: spouse, Sister, Foster Parent\n",
    "* podg (Patient or Disabled Group) Examples: Patients\n",
    "* prog (Professional or Occupational Group) Examples: Administrators, Employee, Author\n",
    "* algae (Algae)\n",
    "* invt (Invertebrate)\n",
    "* orgm (Organism)\n",
    "* rich (Rickettsia or Chlamydia)\n",
    "* amph (Amphibian) Examples: Toad, Bufo boreas, Anura\n",
    "* anim (Animal) Examples: Animals, Laboratory /  Control Animal\n",
    "* arch (Archaeon) Examples: Archaea, halophilic bacteria, Thermoplasma acidophilum\n",
    "* bact (Bacterium) Examples: Escherichia coli, Salmonella, Borrelia burgdorferi\n",
    "* bdsu (Body Substance) too general. Examples: Urine, Milk, Lymph, Urine specimen\n",
    "* bdsy (Body System) too general. Examples: hypothalamic-pituitary-adrenal axis, Neurosecretory Systems\n",
    "* bird (Bird) Examples: Geese, Passeriformes, Raptors\n",
    "* blor (Body Location or Region) too general. Examples: Hepatic, Lysosomal, Cytoplasmic\n",
    "* bmod (Biomedical Occupation or Discipline) Examples: Medicine, Dentistry, Midwifery\n",
    "* bsoj (Body Space or Junction) too general. Examples: Compartments, Synapses, Cistern\n",
    "* chvf (Chemical Viewed Functionally) too general. Examples: inhibitors, antagonists, Agent\n",
    "* chvs (Chemical Viewed Structurally) too general. Examples: particle, solid state, vapor\n",
    "* euka (Eukaryote) Examples: Wasps, Protozoan parasite\n",
    "* ffas (Fully Formed Anatomical Structure) Examples: Carcass\n",
    "* fish (Fish) Examples: Eels, Fishes, Electric Fish\n",
    "* fngs (Fungus) Examples: Saccharomyces cerevisiae, Alternaria brassicicola, fungus\n",
    "* humn (Human) Examples: Family, Patients, Males\n",
    "* irda (Indicator, Reagent, or Diagnostic Aid) Examples: Fluorescent Probes, Chelating Agents\n",
    "* mamm (Mammal) Examples: Rattus norvegicus, Felis catus, Mus\n",
    "* ocdi (Occupation or Discipline) Examples: Science, Politics\n",
    "* plnt (Plant) Examples: Chrysanthemum x morifolium, Pollen, Oryza sativa\n",
    "* rept (Reptile) Examples: Snakes, Turtles, Reptiles\n",
    "* sbst (Substance) too general. Examples: Materials, Plastics, Photons, Substance\n",
    "* virs (Virus) Examples: Herpesvirus 4, Human / GB virus C / Herpesviridae\n",
    "* vtbt (Vertebrate) Examples: Vertebrates / Poikilotherm, NOS\n",
    "* anst (Anatomical Structure) Examples: Entire fetus, Whole body, Cadaver\n",
    "* bhvr (Behavior) too general. Examples: Sexuality, Nest Building, Behavioral phenotype\n",
    "* inbe (Individual Behavior) too general. Examples: impulsivity, Habits, Performance\n",
    "* menp (Mental Process) too general. Examples: mind control, Learning, experience\n",
    "* socb (Social Behavior) too general. Examples: Communication, Gestures, Marriage\n",
    "* biof (Biologic Function) too general. Examples: dose-response relationship, Pharmacodynamics, Anabolism\n",
    "* eehu (Environmental Effect of Humans) too general. Examples: Sewage, Pollution, Smoke\n",
    "* hcpp (Human-caused Phenomenon or Process) too general. Now not in API. Examples: particle beam, Conferences, Victimization\n",
    "* lbtr (Laboratory or Test Result) too general. Examples: False Positive Reactions, Bone Density, Serum Calcium Level\n",
    "* npop (Natural Phenomenon or Process) too general. Examples: Floods, Fluorescence, Freezing\n",
    "* phpr (Phenomenon or Process) too general. Examples: Disasters, Acceleration, Feedback\n",
    "* bodm (Biomedical or Dental Material) too general. Examples: Pill, Gel, Talc, calcium phosphate\n",
    "* drdd (Drug Delivery Device) too general. Examples: Epipen, Skin Patch, Lilly cyanide antidote kit\n",
    "* medd (Medical Device) too general. Examples: Implants / Denture, Overlay / Silicone gel implant / Swab\n",
    "* resd (Research Device) too general. Examples: Study models, Slide\n",
    "* emst (Embryonic Structure) Examples: Chick Embryo, Blastocyst structure, Placenta\n",
    "* tisu (Tissue) Examples: Tissue specimen, Blood, Human tissue, Mucous Membrane\n",
    "* genf (Genetic Function) too general. Examples: Transcription, Genetic / Transcriptional Activation / Recombination, Genetic\n",
    "* chem (Chemical) too general. Examples: Chemicals, Acids, Ligands, Ozone\n",
    "* clas (Classification) too general. Examples: Research Diagnostic Criteria, Group C\n",
    "* ftcn (Functional Concept) too general. Examples: Techniques, Intravenous Route of Drug Administration\n",
    "* idcn (Idea or Concept) too general. Examples: Significant, subject, Data, Owner\n",
    "* qlco (Qualitative Concept) too general. Examples: Effect, Associated with, Advanced phase\n",
    "* qnco (Quantitative Concept) too general. Examples: Calibration, occurrence, degrees Celsius\n",
    "* rnlw (Regulation or Law) too general. Examples: Medicare, Medicaid, regulatory\n",
    "* spco (Spatial Concept) too general. Examples: Structure, Longitudinal, Asymmetry\n",
    "* tmco (Temporal Concept) too general. Examples: New, /period, 24 Hours\n",
    "* clna (Clinical Attribute) too general. Examples: response, Renin secretion, BAND PATTERN\n",
    "* lbpr (Laboratory Procedure) too general. Examples: Western Blot, Radioimmunoassay, Staining method\n",
    "* mbrt (Molecular Biology Research Technique) too general. Examples: Polymerase Chain Reaction / Blotting, Northern\n",
    "* elii (Element, Ion, or Isotope) too general. Examples: Atom, Aluminum, Superoxides\n",
    "* emod (Experimental Model of Disease) too general. Examples: Experimental Autoimmune Encephalomyelitis, Rodent Model\n",
    "* rcpt (Receptor) too general. Examples: Binding Sites / Receptors, Metabotropic Glutamate\n",
    "* evnt (Event) too general. Now not in API. Examples: Stressful Events\n",
    "* fndg (Finding) too general. Examples: spinal cord; lesion, Normal birth weight, Sedentary job\n",
    "* geoa (Geographic Area) too general. Examples: Country, Canada\n",
    "* grup (Group) too general. Examples: Human, Individual\n",
    "* hcro (Health Care Related Organization) too general. Examples: Hospitals, Health System\n",
    "* orgt (Organization) too general. Examples: United Nations, Organization administrative structures\n",
    "* pros (Professional Society) too general. Examples: Professional Organizations, American Nurses' Association\n",
    "* shro (Self-help or Relief Organization) too general. Examples: Social Welfare, Support Groups\n",
    "* inpr (Intellectual Product) too general. Examples: Methodology, Study models\n",
    "* mnob (Manufactured Object) too general. Examples: Glass, Manuals\n",
    "* mosq (Molecular Sequence) too general. Now not in API. Examples: Genetic Code\n",
    "* nusq (Nucleotide Sequence) too general. Examples: Base Sequence, DNA Sequence, 22q11\n",
    "* orga (Organism Attribute) too general. Examples: Ability, Body Composition\n",
    "* popg (Population Group) too general. Examples: Male population group, Woman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf7f82a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "796.5px",
    "left": "39px",
    "top": "109.633px",
    "width": "536.5px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
