{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "47e7307b",
   "metadata": {},
   "source": [
    "# Auto-generate BTE annotations for BioThings SEMMEDDB"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "58ab5663",
   "metadata": {},
   "source": [
    "This notebook walks a developer through the process of [taking the SEMMEDDB database](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html) and using this data to auto-generate the x-bte operations for [BTE](https://github.com/biothings/BioThings_Explorer_TRAPI). This is needed for BTE to query + process the responses from the [BioThings SEMMEDDB API](https://biothings.ncats.io/semmeddb). \n",
    "\n",
    "---\n",
    "\n",
    "When one sees this:  \n",
    "**PAUSE**\n",
    "\n",
    "read the accompanying text that will explain what the developer needs to do before running the code chunks below that text block.\n",
    "\n",
    "---\n",
    "\n",
    "The [yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_operations.yaml) [segments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_list.yaml) generated by this notebook are added to [this file](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml) to make [yaml used for the smartapi registration](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) for BioThings SEMMEDDB API."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "43745613",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "712186ee",
   "metadata": {},
   "source": [
    "Requirements:\n",
    "* get [Biolink-model Toolkit (bmt)](https://github.com/biolink/biolink-model-toolkit/) from github (using pip to [install](https://biolink.github.io/biolink-model-toolkit/intro/intro.html)). Using release 0.9.0 at the moment.\n",
    "* get ruamel_yaml from [conda-forge](https://anaconda.org/conda-forge/ruamel_yaml). However, the [version from pip](https://pypi.org/project/ruamel.yaml/) should work as well with this code, but you'll have to change the import statement below to import `ruamel.yaml`. The documentation is [here](https://yaml.readthedocs.io/en/latest/)\n",
    "\n",
    "Files:\n",
    "* Get SEMMEDDB PREDICATION CSV [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html). This notebook was originally made using the version `semmedVER43_2021_R_PREDICATION`\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or pick the latest version [here](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5ee5aac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:49:51.251051Z",
     "start_time": "2023-06-06T13:49:43.456305Z"
    }
   },
   "outputs": [],
   "source": [
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "## see above for install instructions\n",
    "from bmt import Toolkit\n",
    "import ruamel_yaml as ryml\n",
    "\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "## used in trying things out\n",
    "# import re"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "0af532d5",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Review the code chunk below before running it:\n",
    "* Check and correct the path for `raw_data_location`\n",
    "* Check that the columns specified in `usecols` and `names` match the columns of the PREDICATION [file](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html). \n",
    "* Check that the `na_values`, `sep` are correct. One can use a command in Terminal like `head`\n",
    "* If there are encoding issues, try different encodings. latin1 was used and worked, [ref](https://stackoverflow.com/questions/61163367/how-to-resolve-unicodedecodeerror-in-pandas-read-csv-while-loading-dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73408978",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:08.934798Z",
     "start_time": "2023-06-06T13:49:51.253525Z"
    }
   },
   "outputs": [],
   "source": [
    "raw_data_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"semmedVER43_2021_R_PREDICATION.csv\")\n",
    "\n",
    "raw_data = pd.read_csv(raw_data_location, header=None, sep=\",\", encoding=\"latin1\",\n",
    "                          usecols=[3, 6, 7, 10, 11],\n",
    "                          names=[\"PREDICATE\",\"SUBJECT_SEMTYPE\",\"SUBJECT_NOVELTY\",\n",
    "                                 \"OBJECT_SEMTYPE\", \"OBJECT_NOVELTY\"],\n",
    "                          na_values=r\"\\N\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8150ed6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:08.944453Z",
     "start_time": "2023-06-06T13:51:08.936255Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(113863366, 5)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57d2dc4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:08.957097Z",
     "start_time": "2023-06-06T13:51:08.947775Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>SUBJECT_NOVELTY</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>OBJECT_NOVELTY</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ISA</td>\n",
       "      <td>virs</td>\n",
       "      <td>1</td>\n",
       "      <td>virs</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>0</td>\n",
       "      <td>humn</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    PREDICATE SUBJECT_SEMTYPE  SUBJECT_NOVELTY OBJECT_SEMTYPE  OBJECT_NOVELTY\n",
       "0  PROCESS_OF            virs                1           mamm             1.0\n",
       "1         ISA            virs                1           virs             1.0\n",
       "2         ISA            virs                1           virs             1.0\n",
       "3         ISA            virs                1           virs             1.0\n",
       "4  PROCESS_OF            dsyn                0           humn             0.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "54b68180",
   "metadata": {},
   "source": [
    "## Basic Filtering"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8dc722c4",
   "metadata": {},
   "source": [
    "### keep only novelty = 1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "32b8cf0b",
   "metadata": {},
   "source": [
    "now filter it to only keep novelty == 1 for both subject / object...\n",
    "\n",
    "since those with novelty == 0 probably aren't very helpful / interesting to Translator. The entities with novelty == 0 are [listed](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html) in the SEMMEDDB GENERIC_CONCEPT table files, which can be downloaded [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6c362bfe",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:15.377738Z",
     "start_time": "2023-06-06T13:51:08.958706Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = raw_data[(raw_data[\"SUBJECT_NOVELTY\"] == 1) &\n",
    "                         (raw_data[\"OBJECT_NOVELTY\"] == 1)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "68a11e89",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:15.382588Z",
     "start_time": "2023-06-06T13:51:15.379346Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(78688677, 5)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9abae6d5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:16.902238Z",
     "start_time": "2023-06-06T13:51:15.384182Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove the novelty stuff since that will make computations faster and it's always 1 now\n",
    "filtered_data = filtered_data[['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE']]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "66309970",
   "metadata": {},
   "source": [
    "what semantic types are even there? we have to prune down to what we actually want BTE operations on...\n",
    "\n",
    "interesting since the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67a2db8e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:25.490823Z",
     "start_time": "2023-06-06T13:51:16.903777Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "135"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "68"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes)  ## got 133\n",
    "len(object_semtypes)   ## got 135\n",
    "len(predicates)        ## got 68"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9b788a7d",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Review the 3 sets above to see if there are things I want to remove. The normal entity semantic types have 4-letter codes..."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "88f7eeff",
   "metadata": {},
   "source": [
    "### Remove semantic types that we don't want to make operations from"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f6085739",
   "metadata": {},
   "source": [
    "so there's some object stuff that I'm going to remove..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "69d37a01",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:25.497451Z",
     "start_time": "2023-06-06T13:51:25.493031Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'C0030193',\n",
       " 'C0030705',\n",
       " 'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gngm,aapp',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'podg,humn',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "object_semtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26561f10",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:29.502925Z",
     "start_time": "2023-06-06T13:51:25.500502Z"
    }
   },
   "outputs": [],
   "source": [
    "## make the set of stuff we want to remove\n",
    "removal1 = set([\"C0030193\", \"C0030705\", \"gngm,aapp\", \"podg,humn\"])\n",
    "\n",
    "## remove it from the data \n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(removal1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1f18986e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:37.779708Z",
     "start_time": "2023-06-06T13:51:29.504761Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "65"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 133: didn't change\n",
    "len(object_semtypes)  ## was 135, now 131: decreased by 4 (expected)\n",
    "len(predicates)       ## was 68, now 65: decreased by 3 (cool)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ff20381b",
   "metadata": {},
   "source": [
    "### Remove Predicates we don't want to make operations from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09163f4e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:37.784703Z",
     "start_time": "2023-06-06T13:51:37.780966Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADMINISTERED_TO',\n",
       " 'AFFECTS',\n",
       " 'ASSOCIATED_WITH',\n",
       " 'AUGMENTS',\n",
       " 'CAUSES',\n",
       " 'COEXISTS_WITH',\n",
       " 'COMPLICATES',\n",
       " 'CONVERTS_TO',\n",
       " 'DIAGNOSES',\n",
       " 'DISRUPTS',\n",
       " 'INHIBITS',\n",
       " 'INTERACTS_WITH',\n",
       " 'ISA',\n",
       " 'LOCATION_OF',\n",
       " 'MANIFESTATION_OF',\n",
       " 'MEASUREMENT_OF',\n",
       " 'MEASURES',\n",
       " 'METHOD_OF',\n",
       " 'NEG_ADMINISTERED_TO',\n",
       " 'NEG_AFFECTS',\n",
       " 'NEG_ASSOCIATED_WITH',\n",
       " 'NEG_AUGMENTS',\n",
       " 'NEG_CAUSES',\n",
       " 'NEG_COEXISTS_WITH',\n",
       " 'NEG_COMPLICATES',\n",
       " 'NEG_CONVERTS_TO',\n",
       " 'NEG_DIAGNOSES',\n",
       " 'NEG_DISRUPTS',\n",
       " 'NEG_INHIBITS',\n",
       " 'NEG_INTERACTS_WITH',\n",
       " 'NEG_ISA',\n",
       " 'NEG_LOCATION_OF',\n",
       " 'NEG_MANIFESTATION_OF',\n",
       " 'NEG_MEASUREMENT_OF',\n",
       " 'NEG_MEASURES',\n",
       " 'NEG_METHOD_OF',\n",
       " 'NEG_OCCURS_IN',\n",
       " 'NEG_PART_OF',\n",
       " 'NEG_PRECEDES',\n",
       " 'NEG_PREDISPOSES',\n",
       " 'NEG_PREVENTS',\n",
       " 'NEG_PROCESS_OF',\n",
       " 'NEG_PRODUCES',\n",
       " 'NEG_STIMULATES',\n",
       " 'NEG_TREATS',\n",
       " 'NEG_USES',\n",
       " 'NEG_higher_than',\n",
       " 'NEG_lower_than',\n",
       " 'NEG_same_as',\n",
       " 'NOM',\n",
       " 'OCCURS_IN',\n",
       " 'PART_OF',\n",
       " 'PRECEDES',\n",
       " 'PREDISPOSES',\n",
       " 'PREP',\n",
       " 'PREVENTS',\n",
       " 'PROCESS_OF',\n",
       " 'PRODUCES',\n",
       " 'STIMULATES',\n",
       " 'TREATS',\n",
       " 'USES',\n",
       " 'compared_with',\n",
       " 'higher_than',\n",
       " 'lower_than',\n",
       " 'same_as'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predicates"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "106d9b99",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* Decide what predicates you want to remove\n",
    "\n",
    "**Current logic**\n",
    "\n",
    "Translator currently isn't great with negation, so removing those\n",
    "\n",
    "Also some predicates seem to actually be practice phrases? see the [article](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-486#Sec26)\n",
    "\n",
    "And I didn't find the following relationships useful: ISA (subclass relationship), same_as (equivalent / it did just as well as some other thing), lower_than / higher_than (these two entities were compared and 1 was better (higher) or worse (lower) than the other), and compared_with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dec1bcc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:37.789285Z",
     "start_time": "2023-06-06T13:51:37.786109Z"
    }
   },
   "outputs": [],
   "source": [
    "negative_preds = [\"NEG_ADMINISTERED_TO\", \"NEG_AFFECTS\", \"NEG_ASSOCIATED_WITH\",\n",
    "                  \"NEG_AUGMENTS\", \"NEG_CAUSES\", \"NEG_COEXISTS_WITH\", \"NEG_COMPLICATES\",\n",
    "                  \"NEG_CONVERTS_TO\", \"NEG_DIAGNOSES\", \"NEG_DISRUPTS\", \"NEG_INHIBITS\",\n",
    "                  \"NEG_INTERACTS_WITH\", \"NEG_ISA\", \"NEG_LOCATION_OF\", \"NEG_MANIFESTATION_OF\",\n",
    "                  \"NEG_MEASUREMENT_OF\", \"NEG_MEASURES\", \"NEG_METHOD_OF\", \"NEG_OCCURS_IN\",\n",
    "                  \"NEG_PART_OF\", \"NEG_PRECEDES\", \"NEG_PREDISPOSES\", \"NEG_PREVENTS\",\n",
    "                  \"NEG_PROCESS_OF\", \"NEG_PRODUCES\", \"NEG_STIMULATES\", \"NEG_TREATS\", \"NEG_USES\",\n",
    "                  \"NEG_higher_than\", \"NEG_lower_than\", \"NEG_same_as\"\n",
    "                 ]\n",
    "\n",
    "practice_phrase = [\"VERB\", \"NOM\", \"PREP\"]\n",
    "\n",
    "dont_like = [\"ISA\", \"same_as\", \"lower_than\", \"higher_than\", \"compared_with\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ab643b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:51:53.191762Z",
     "start_time": "2023-06-06T13:51:37.790753Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(negative_preds)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(practice_phrase)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(dont_like)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5f3d145",
   "metadata": {},
   "source": [
    "now look at the stats. interesting since the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates: f we removed half of 54 because we removed negated predicates, then we get 27...so we match the official stats now."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b6244458",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:00.886374Z",
     "start_time": "2023-06-06T13:51:53.193419Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 133, now 127: decreased by 6\n",
    "len(object_semtypes)  ## was 131, now 127: decreased by 4\n",
    "len(predicates)       ## was 65, now 27: decreased by 38 (was expected)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "94e31d94",
   "metadata": {},
   "source": [
    "so....that's how many kinds of subjects, objects, and predicates we have to go forward with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4184a17f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.937444Z",
     "start_time": "2023-06-06T13:52:00.887992Z"
    }
   },
   "outputs": [],
   "source": [
    "combos = filtered_data.value_counts().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "24036b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.941898Z",
     "start_time": "2023-06-06T13:52:14.938797Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(14796, 4)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos.shape\n",
    "## so that's still a lot...."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e998d9e8",
   "metadata": {},
   "source": [
    "## Specific Filtering"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "330bba7a",
   "metadata": {},
   "source": [
    "so we actually only need to use certain semantic types...it depends on how they map to the biolink model\n",
    "\n",
    "As a reminder:\n",
    "\n",
    "* Get SEMMEDDB SRDEF file needed for interpreting and mapping SEMMED semantic types: \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or pick the latest version [here](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "91866eb6",
   "metadata": {},
   "source": [
    "### Ingest SEMMED semantic info and biolink model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8d1bbf26",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.945569Z",
     "start_time": "2023-06-06T13:52:14.943541Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"2020AA\", \"SRDEF\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8f859145",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.956746Z",
     "start_time": "2023-06-06T13:52:14.947333Z"
    }
   },
   "outputs": [],
   "source": [
    "srdef = pd.read_csv(srdef_location, sep=\"|\", header=0, index_col=False,\n",
    "                    names=[\"Record Type (RT)\",\n",
    "                           \"Unique Identifier (UI)\",\n",
    "                           \"Full Name (STY/RL)\", \n",
    "                           \"Tree Number (STN/RTN)\", \n",
    "                           \"Definition (DEF)\",\n",
    "                           \"Examples (EX)\",\n",
    "                           \"Usage Note (UN)\",\n",
    "                           \"Non-Human Flag (NH)\",\n",
    "                           \"Abbreviation (ABR)\",\n",
    "                           \"Inverse Relation (RIN)\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d470fa70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.968213Z",
     "start_time": "2023-06-06T13:52:14.958404Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Record Type (RT)</th>\n",
       "      <th>Unique Identifier (UI)</th>\n",
       "      <th>Full Name (STY/RL)</th>\n",
       "      <th>Tree Number (STN/RTN)</th>\n",
       "      <th>Definition (DEF)</th>\n",
       "      <th>Examples (EX)</th>\n",
       "      <th>Usage Note (UN)</th>\n",
       "      <th>Non-Human Flag (NH)</th>\n",
       "      <th>Abbreviation (ABR)</th>\n",
       "      <th>Inverse Relation (RIN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>STY</td>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>A1.4.1.2.1.7</td>\n",
       "      <td>Amino acids and chains of amino acids connecte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>When the concept is both an enzyme and a prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aapp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Record Type (RT) Unique Identifier (UI)               Full Name (STY/RL)  \\\n",
       "97              STY                   T116  Amino Acid, Peptide, or Protein   \n",
       "\n",
       "   Tree Number (STN/RTN)                                   Definition (DEF)  \\\n",
       "97          A1.4.1.2.1.7  Amino acids and chains of amino acids connecte...   \n",
       "\n",
       "    Examples (EX)                                    Usage Note (UN)  \\\n",
       "97            NaN  When the concept is both an enzyme and a prote...   \n",
       "\n",
       "   Non-Human Flag (NH) Abbreviation (ABR) Inverse Relation (RIN)  \n",
       "97                 NaN               aapp                    NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "srdef[srdef['Abbreviation (ABR)'] == 'aapp']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2e245634",
   "metadata": {},
   "source": [
    "### Add semantic types that are in the data but not in SRDEF"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "49397246",
   "metadata": {},
   "source": [
    "Yao [noticed](https://github.com/biothings/pending.api/issues/30#issuecomment-903609946) that the data file uses \"old SEMMED semantic types\" from 2013AA, and these semantic types didn't exist in the latest SRDEF file (2020AA).\n",
    "\n",
    "In response, Andrew said to [keep](https://github.com/biothings/pending.api/issues/30#issuecomment-903879782) this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "89c14528",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.980383Z",
     "start_time": "2023-06-06T13:52:14.969796Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(126, 3)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>T195</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst\n",
       "119  T195                       Antibiotic   antb"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## simplify the df\n",
    "SEMMED_entity_types = srdef[srdef['Record Type (RT)'] == 'STY'].copy()\n",
    "SEMMED_entity_types = SEMMED_entity_types[['Unique Identifier (UI)',\n",
    "                                           'Full Name (STY/RL)', \n",
    "                                           'Abbreviation (ABR)']].copy()\n",
    "SEMMED_entity_types.columns = ['UI', 'FullName', \"Abbrev\"]\n",
    "SEMMED_entity_types.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_entity_types.shape\n",
    "SEMMED_entity_types[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8c8c04ce",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.988891Z",
     "start_time": "2023-06-06T13:52:14.982558Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "these semantic types are in the data, but not in SRDEF:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'carb',\n",
       " 'eico',\n",
       " 'invt',\n",
       " 'lipd',\n",
       " 'nsba',\n",
       " 'opco',\n",
       " 'orgm',\n",
       " 'rich',\n",
       " 'strd'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "only_in_data = data_types - set(SEMMED_entity_types['Abbrev'])\n",
    "\n",
    "print(\"these semantic types are in the data, but not in SRDEF:\")\n",
    "only_in_data\n",
    "\n",
    "len(only_in_data)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "533341a2",
   "metadata": {},
   "source": [
    "How I found info on these semantic types from [previous semantic versions](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html):\n",
    "\n",
    "was in 2009AB file (sets with SG_2?)\n",
    "* alga\n",
    "* invt\n",
    "* rich\n",
    "\n",
    "was in 2014AB file (sets with SG_3?)\n",
    "* carb \n",
    "* eico \n",
    "* lipd \n",
    "* nsba \n",
    "* opco\n",
    "* orgm\n",
    "* strd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0400e8e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:14.995465Z",
     "start_time": "2023-06-06T13:52:14.990218Z"
    }
   },
   "outputs": [],
   "source": [
    "## build the missing info and add to the table\n",
    "missing_semantic_types = [{\"UI\": \"T003\", \"FullName\": \"Alga\", \"Abbrev\": \"alga\"},\n",
    "                          {\"UI\": \"T118\", \"FullName\": \"Carbohydrate\", \"Abbrev\": \"carb\"},\n",
    "                          {\"UI\": \"T111\", \"FullName\": \"Eicosanoid\", \"Abbrev\": \"eico\"},\n",
    "                          {\"UI\": \"T009\", \"FullName\": \"Invertebrate\", \"Abbrev\": \"invt\"},\n",
    "                          {\"UI\": \"T119\", \"FullName\": \"Lipid\", \"Abbrev\": \"lipd\"},\n",
    "                          {\"UI\": \"T124\", \"FullName\": \"Neuroreactive Substance or Biogenic Amine\", \"Abbrev\": \"nsba\"},\n",
    "                          {\"UI\": \"T115\", \"FullName\": \"Organophosphorus Compound\", \"Abbrev\": \"opco\"},\n",
    "                          {\"UI\": \"T001\", \"FullName\": \"Organism\", \"Abbrev\": \"orgm\"},\n",
    "                          {\"UI\": \"T006\", \"FullName\": \"Rickettsia or Chlamydia\", \"Abbrev\": \"rich\"},\n",
    "                          {\"UI\": \"T110\", \"FullName\": \"Steroid\", \"Abbrev\": \"strd\"}\n",
    "                         ]\n",
    "\n",
    "missing_semantic_types = pd.DataFrame.from_records(missing_semantic_types)\n",
    "\n",
    "## add it to the table\n",
    "## SEMMED_entity_types = SEMMED_entity_types.append(missing_semantic_types)\n",
    "\n",
    "SEMMED_entity_types = pd.concat([SEMMED_entity_types, missing_semantic_types])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "683a1f47",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:15.004923Z",
     "start_time": "2023-06-06T13:52:14.997056Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(136, 3)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "0    T003                             Alga   alga\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_entity_types.shape\n",
    "SEMMED_entity_types[0:10]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c8f01880",
   "metadata": {},
   "source": [
    "### Get the biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a7a170a6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.091427Z",
     "start_time": "2023-06-06T13:52:15.006801Z"
    }
   },
   "outputs": [],
   "source": [
    "## using biolink 3.1.1\n",
    "bmt_tool = Toolkit('https://raw.githubusercontent.com/biolink/biolink-model/v3.1.1/biolink-model.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4cc1fad9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.435286Z",
     "start_time": "2023-06-06T13:52:22.093119Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Small Molecule'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bmt_tool.get_element_by_mapping('STY:T123').title()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c2c14436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.447464Z",
     "start_time": "2023-06-06T13:52:22.437109Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting biolink-mapping, in the format needed to create operations\n",
    "SEMMED_entity_types['BiolinkMapping'] = [bmt_tool.get_element_by_mapping('STY:'+i)\n",
    "                                         for i in SEMMED_entity_types['UI']]\n",
    "## put these node categories/semantic-types in the correct format: PascalCase\n",
    "SEMMED_entity_types['BiolinkMapping'] = [i.title().replace(\" \", \"\") if isinstance(i, str)\n",
    "                                         else i for i in SEMMED_entity_types['BiolinkMapping']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "044843d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.462358Z",
     "start_time": "2023-06-06T13:52:22.454171Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>136 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "0    T003                                 Alga   alga   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "0                        None  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                        None  \n",
       "104             SmallMolecule  \n",
       "5                        None  \n",
       "\n",
       "[136 rows x 4 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "51111407",
   "metadata": {},
   "source": [
    "### Clean up the semantic type to only have terms in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "7fb08392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.469485Z",
     "start_time": "2023-06-06T13:52:22.463979Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "note that these semantic types are in SRDEF but not in the data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cnce', 'crbs', 'enty', 'grpa', 'lang', 'phob'}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"note that these semantic types are in SRDEF but not in the data\")\n",
    "set(SEMMED_entity_types['Abbrev']) - data_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "22daf133",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.474280Z",
     "start_time": "2023-06-06T13:52:22.471172Z"
    }
   },
   "outputs": [],
   "source": [
    "## get the subset of semantic network terms that are actually in the data\n",
    "SEMMED_entity_types = SEMMED_entity_types[SEMMED_entity_types['Abbrev'].isin(data_types)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "86caa9b9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.484164Z",
     "start_time": "2023-06-06T13:52:22.476301Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>130 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "0    T003                                 Alga   alga   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "0                        None  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                        None  \n",
       "104             SmallMolecule  \n",
       "5                        None  \n",
       "\n",
       "[130 rows x 4 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "89d30960",
   "metadata": {},
   "source": [
    "### review mappings and deciding what to change, remove"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "6fd9e849",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "This is a place to **STOP** and review all the SEMMED semantic types and their mapping to biolink semantic types...to decide what we are interested in keeping. This involves some knowledge of what biolink semantic types are prioritized in Translator. \n",
    "\n",
    "One can use the definitions of the SEMMED semantic types (from the SRDEF file or [browsing the UMLS vocab online](https://uts.nlm.nih.gov/uts/umls/semantic-network/root)) and the definitions of biolink semantic types (look for a comment with 'THINGS' in the biolink-model yaml file)\n",
    "\n",
    "See the last section (section 6) for notes on decisions that were made...\n",
    "\n",
    "---\n",
    "\n",
    "Sections 3.5 + 3.6 below involves this review, changing mappings, and removing some SEMMED semantic types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "eb7ea639",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.490160Z",
     "start_time": "2023-06-06T13:52:22.486269Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Polypeptide', 'Disease', 'Activity', 'Cohort', None,\n",
       "       'AnatomicalEntity', 'Drug', 'SmallMolecule', 'Behavior',\n",
       "       'Phenomenon', 'Device', 'GrossAnatomicalStructure',\n",
       "       'CellularComponent', 'PhysiologicalProcess', 'Cell',\n",
       "       'ChemicalEntity', 'InformationContentEntity', 'ClinicalAttribute',\n",
       "       'Procedure', 'BiologicalEntity', 'Protein', 'Event',\n",
       "       'DiseaseOrPhenotypicFeature', 'Food', 'GeographicLocation',\n",
       "       'GenomicEntity', 'Agent', 'PathologicalProcess', 'Publication',\n",
       "       'PhysicalEntity', 'MolecularActivity', 'MolecularEntity',\n",
       "       'NucleicAcidEntity', 'OrganismAttribute',\n",
       "       'PopulationOfIndividualOrganisms', 'PhenotypicFeature'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## code used to review \n",
    "SEMMED_entity_types['BiolinkMapping'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "fc902a02",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.499403Z",
     "start_time": "2023-06-06T13:52:22.492045Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>T114</td>\n",
       "      <td>Nucleic Acid, Nucleoside, or Nucleotide</td>\n",
       "      <td>nnon</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>T086</td>\n",
       "      <td>Nucleotide Sequence</td>\n",
       "      <td>nusq</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      UI                                 FullName Abbrev     BiolinkMapping\n",
       "96  T114  Nucleic Acid, Nucleoside, or Nucleotide   nnon  NucleicAcidEntity\n",
       "76  T086                      Nucleotide Sequence   nusq  NucleicAcidEntity"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'] == 'NucleicAcidEntity']\n",
    "# SEMMED_entity_types[SEMMED_entity_types['BiolinkMapping'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e641ae5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.507528Z",
     "start_time": "2023-06-06T13:52:22.501795Z"
    }
   },
   "outputs": [],
   "source": [
    "## re-mapping based on putting IDs into normalization service / our operation system...\n",
    "##   we have UMLS for Disease (mydisease, mychem), SmallMolecule (idisk)\n",
    "\n",
    "## leaving aapp as Polypeptide, enzy mapped to Protein\n",
    "## leaving Disease mappings as-is, but some seem like they could be PathologicalAnatomicalStructure \n",
    "##   (acab, anab, cgab) or PathologicalProcess (comd) instead\n",
    "## left clnd (Clinical Drug) as Drug since it really seemed like a drug (dosage)\n",
    "\n",
    "## currently mapped to Drug, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'antb'),'BiolinkMapping'] = 'SmallMolecule'  ## Antibiotic\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'phsu'),'BiolinkMapping'] = 'SmallMolecule'  ## Pharmacologic Substance\n",
    "\n",
    "## currently mapped to GenomicEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'gngm'),'BiolinkMapping'] = 'Gene'\n",
    "\n",
    "## currently mapped to NucleicAcidEntity, but re-mapping here\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'nnon'),'BiolinkMapping'] = 'SmallMolecule'  ## Nucleic Acid, Nucleoside, or Nucleotide\n",
    "\n",
    "## currently doesn't have a mapping\n",
    "SEMMED_entity_types.loc[\n",
    "    (SEMMED_entity_types['Abbrev'] == 'vita'),'BiolinkMapping'] = 'SmallMolecule'  ## Vitamin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a3a9da81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:22.515316Z",
     "start_time": "2023-06-06T13:52:22.509735Z"
    }
   },
   "outputs": [],
   "source": [
    "currently_unused_entities = [\n",
    "    ## mapped to Polypeptide\n",
    "    \"amas\",     \n",
    "    ## all that are mapped to Activity\n",
    "    \"acty\", \"dora\", \"edac\", \"gora\", \"hlca\", \"mcha\", \"ocac\", \"resa\",  \n",
    "    ## all that are mapped to Cohort\n",
    "    \"aggp\", \"famg\", \"podg\", \"prog\",  \n",
    "    ## missing mappings\n",
    "    \"amph\", \"anim\", \"arch\", \"bact\", \"bdsu\", \"bdsy\", \"bird\", \"blor\", \n",
    "    \"bmod\", \"bsoj\", \"chvf\", \"chvs\", \"euka\", \"ffas\", \"fish\", \"fngs\", \n",
    "    \"humn\", \"irda\", \"mamm\", \"ocdi\", \"plnt\", \"rept\", \"sbst\", \"virs\", \"vtbt\",\n",
    "    ## old terms here\n",
    "    \"alga\", \"invt\", \"orgm\", \"rich\",\n",
    "    ## all that are mapped to Anatomical Entity\n",
    "    \"anst\",\n",
    "    ## all that are mapped to Behavior\n",
    "    \"bhvr\", \"inbe\", \"menp\", \"socb\",\n",
    "    ## all that are mapped to Phenomenon\n",
    "    \"biof\", \"eehu\", \"hcpp\", \"lbtr\", \"npop\", \"phpr\",\n",
    "    ## all that are mapped to Device\n",
    "    \"bodm\", \"drdd\", \"medd\", \"resd\",\n",
    "    ## mapped to GrossAnatomicalStructure\n",
    "    \"emst\", \"tisu\", \n",
    "    ## mapped to PhysiologicalProcess\n",
    "    \"genf\",\n",
    "    ## mapped to ChemicalEntity\n",
    "    \"chem\",\n",
    "    ## all that are mapped to InformationContentEntity\n",
    "    \"clas\", \"ftcn\", \"idcn\", \"qlco\", \"qnco\", \"rnlw\", \"spco\", \"tmco\",\n",
    "    ## all that are mapped to ClinicalAttribute\n",
    "    \"clna\",\n",
    "    ## mapped to Procedure\n",
    "    \"lbpr\", \"mbrt\", \n",
    "    ## mapped to SmallMolecule\n",
    "    \"elii\",\n",
    "    ## all that are mapped to BiologicalEntity\n",
    "    \"emod\",\n",
    "    ## mapped to Protein\n",
    "    \"rcpt\",\n",
    "    ## all that are mapped to Event\n",
    "    \"evnt\",\n",
    "    ## all that are mapped to DiseaseOrPhenotypicFeature\n",
    "    \"fndg\",\n",
    "    ## all that are mapped to GeographicLocation\n",
    "    \"geoa\",\n",
    "    ## all that are mapped to Agent\n",
    "    \"grup\", \"hcro\", \"orgt\", \"pros\", \"shro\",\n",
    "    ## all that are mapped to Publication\n",
    "    \"inpr\",\n",
    "    ## all that are mapped to PhysicalEntity\n",
    "    \"mnob\",\n",
    "    ## all that are mapped to MolecularEntity\n",
    "    \"mosq\",\n",
    "    ## mapped to NucleicAcidEntity\n",
    "    \"nusq\",\n",
    "    ## all that are mapped to OrganismAttribute\n",
    "    \"orga\",\n",
    "    ## all that are mapped to PopulationOfIndividualOrganisms\n",
    "    \"popg\"\n",
    "]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "8da7aa1e",
   "metadata": {},
   "source": [
    "### prune data based on what node entities to remove"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "308358e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:35.438114Z",
     "start_time": "2023-06-06T13:52:22.517423Z"
    }
   },
   "outputs": [],
   "source": [
    "## prune this doc\n",
    "SEMMED_entity_types = SEMMED_entity_types[ ~ SEMMED_entity_types['Abbrev'].isin(currently_unused_entities)]\n",
    "\n",
    "## prune the data doc\n",
    "filtered_data = filtered_data[ ~ filtered_data['SUBJECT_SEMTYPE'].isin(currently_unused_entities)]\n",
    "filtered_data = filtered_data[ ~ filtered_data['OBJECT_SEMTYPE'].isin(currently_unused_entities)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "67861c2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.884656Z",
     "start_time": "2023-06-06T13:52:35.440644Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 127, now 41: decreased by 86\n",
    "len(object_semtypes)  ## was 127, now 41: decreased by 86\n",
    "len(predicates)       ## was 27, now 26: decreased by 1 (cool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "647e4065",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.896586Z",
     "start_time": "2023-06-06T13:52:39.886350Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>T195</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>T123</td>\n",
       "      <td>Biologically Active Substance</td>\n",
       "      <td>bacs</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>T023</td>\n",
       "      <td>Body Part, Organ, or Organ Component</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T118</td>\n",
       "      <td>Carbohydrate</td>\n",
       "      <td>carb</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>T026</td>\n",
       "      <td>Cell Component</td>\n",
       "      <td>celc</td>\n",
       "      <td>CellularComponent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>T043</td>\n",
       "      <td>Cell Function</td>\n",
       "      <td>celf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>T025</td>\n",
       "      <td>Cell</td>\n",
       "      <td>cell</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>T019</td>\n",
       "      <td>Congenital Abnormality</td>\n",
       "      <td>cgab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>T200</td>\n",
       "      <td>Clinical Drug</td>\n",
       "      <td>clnd</td>\n",
       "      <td>Drug</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>T049</td>\n",
       "      <td>Cell or Molecular Dysfunction</td>\n",
       "      <td>comd</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>T060</td>\n",
       "      <td>Diagnostic Procedure</td>\n",
       "      <td>diap</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>T047</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T111</td>\n",
       "      <td>Eicosanoid</td>\n",
       "      <td>eico</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>T126</td>\n",
       "      <td>Enzyme</td>\n",
       "      <td>enzy</td>\n",
       "      <td>Protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>T168</td>\n",
       "      <td>Food</td>\n",
       "      <td>food</td>\n",
       "      <td>Food</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>T028</td>\n",
       "      <td>Gene or Genome</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>T131</td>\n",
       "      <td>Hazardous or Poisonous Substance</td>\n",
       "      <td>hops</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>T125</td>\n",
       "      <td>Hormone</td>\n",
       "      <td>horm</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>T129</td>\n",
       "      <td>Immunologic Factor</td>\n",
       "      <td>imft</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>T197</td>\n",
       "      <td>Inorganic Chemical</td>\n",
       "      <td>inch</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>T037</td>\n",
       "      <td>Injury or Poisoning</td>\n",
       "      <td>inpo</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T119</td>\n",
       "      <td>Lipid</td>\n",
       "      <td>lipd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>T048</td>\n",
       "      <td>Mental or Behavioral Dysfunction</td>\n",
       "      <td>mobd</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>T044</td>\n",
       "      <td>Molecular Function</td>\n",
       "      <td>moft</td>\n",
       "      <td>MolecularActivity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>T191</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>neop</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>T114</td>\n",
       "      <td>Nucleic Acid, Nucleoside, or Nucleotide</td>\n",
       "      <td>nnon</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T124</td>\n",
       "      <td>Neuroreactive Substance or Biogenic Amine</td>\n",
       "      <td>nsba</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T115</td>\n",
       "      <td>Organophosphorus Compound</td>\n",
       "      <td>opco</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>T109</td>\n",
       "      <td>Organic Chemical</td>\n",
       "      <td>orch</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>T040</td>\n",
       "      <td>Organism Function</td>\n",
       "      <td>orgf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>T042</td>\n",
       "      <td>Organ or Tissue Function</td>\n",
       "      <td>ortf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>T046</td>\n",
       "      <td>Pathologic Function</td>\n",
       "      <td>patf</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>T039</td>\n",
       "      <td>Physiologic Function</td>\n",
       "      <td>phsf</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>T121</td>\n",
       "      <td>Pharmacologic Substance</td>\n",
       "      <td>phsu</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>T184</td>\n",
       "      <td>Sign or Symptom</td>\n",
       "      <td>sosy</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T110</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>strd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                                   FullName Abbrev  \\\n",
       "97   T116            Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                       Acquired Abnormality   acab   \n",
       "115  T190                     Anatomical Abnormality   anab   \n",
       "119  T195                                 Antibiotic   antb   \n",
       "101  T123              Biologically Active Substance   bacs   \n",
       "18   T023       Body Part, Organ, or Organ Component   bpoc   \n",
       "1    T118                               Carbohydrate   carb   \n",
       "21   T026                             Cell Component   celc   \n",
       "35   T043                              Cell Function   celf   \n",
       "20   T025                                       Cell   cell   \n",
       "14   T019                     Congenital Abnormality   cgab   \n",
       "122  T200                              Clinical Drug   clnd   \n",
       "41   T049              Cell or Molecular Dysfunction   comd   \n",
       "52   T060                       Diagnostic Procedure   diap   \n",
       "39   T047                        Disease or Syndrome   dsyn   \n",
       "2    T111                                 Eicosanoid   eico   \n",
       "103  T126                                     Enzyme   enzy   \n",
       "109  T168                                       Food   food   \n",
       "22   T028                             Gene or Genome   gngm   \n",
       "107  T131           Hazardous or Poisonous Substance   hops   \n",
       "102  T125                                    Hormone   horm   \n",
       "105  T129                         Immunologic Factor   imft   \n",
       "121  T197                         Inorganic Chemical   inch   \n",
       "29   T037                        Injury or Poisoning   inpo   \n",
       "4    T119                                      Lipid   lipd   \n",
       "40   T048           Mental or Behavioral Dysfunction   mobd   \n",
       "36   T044                         Molecular Function   moft   \n",
       "116  T191                         Neoplastic Process   neop   \n",
       "96   T114    Nucleic Acid, Nucleoside, or Nucleotide   nnon   \n",
       "5    T124  Neuroreactive Substance or Biogenic Amine   nsba   \n",
       "6    T115                  Organophosphorus Compound   opco   \n",
       "95   T109                           Organic Chemical   orch   \n",
       "32   T040                          Organism Function   orgf   \n",
       "34   T042                   Organ or Tissue Function   ortf   \n",
       "38   T046                        Pathologic Function   patf   \n",
       "31   T039                       Physiologic Function   phsf   \n",
       "99   T121                    Pharmacologic Substance   phsu   \n",
       "113  T184                            Sign or Symptom   sosy   \n",
       "9    T110                                    Steroid   strd   \n",
       "53   T061        Therapeutic or Preventive Procedure   topp   \n",
       "104  T127                                    Vitamin   vita   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "115                   Disease  \n",
       "119             SmallMolecule  \n",
       "101             SmallMolecule  \n",
       "18   GrossAnatomicalStructure  \n",
       "1               SmallMolecule  \n",
       "21          CellularComponent  \n",
       "35       PhysiologicalProcess  \n",
       "20                       Cell  \n",
       "14                    Disease  \n",
       "122                      Drug  \n",
       "41                    Disease  \n",
       "52                  Procedure  \n",
       "39                    Disease  \n",
       "2               SmallMolecule  \n",
       "103                   Protein  \n",
       "109                      Food  \n",
       "22                       Gene  \n",
       "107             SmallMolecule  \n",
       "102             SmallMolecule  \n",
       "105            ChemicalEntity  \n",
       "121             SmallMolecule  \n",
       "29        PathologicalProcess  \n",
       "4               SmallMolecule  \n",
       "40                    Disease  \n",
       "36          MolecularActivity  \n",
       "116                   Disease  \n",
       "96              SmallMolecule  \n",
       "5               SmallMolecule  \n",
       "6               SmallMolecule  \n",
       "95              SmallMolecule  \n",
       "32       PhysiologicalProcess  \n",
       "34       PhysiologicalProcess  \n",
       "38        PathologicalProcess  \n",
       "31       PhysiologicalProcess  \n",
       "99              SmallMolecule  \n",
       "113         PhenotypicFeature  \n",
       "9               SmallMolecule  \n",
       "53                  Procedure  \n",
       "104             SmallMolecule  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_entity_types"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f4bf6f43",
   "metadata": {},
   "source": [
    "### getting mappings for predicates and qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "52496738",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.903481Z",
     "start_time": "2023-06-06T13:52:39.898507Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates = srdef[srdef['Record Type (RT)'] == 'RL'].copy()\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates[['Full Name (STY/RL)'\n",
    "                                      ]].copy()\n",
    "SEMMED_predicates.columns = ['SemmedPred']\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "635827c9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "The SRDEF file might be missing predicates that are in the data, so check that using the code block below and see if you would like to add those back in..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "b05e90f6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.910776Z",
     "start_time": "2023-06-06T13:52:39.905517Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['administered_to',\n",
       " 'augments',\n",
       " 'coexists_with',\n",
       " 'converts_to',\n",
       " 'inhibits',\n",
       " 'predisposes',\n",
       " 'stimulates']"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## predicates currently in the filtered data\n",
    "\n",
    "missing_preds = list()\n",
    "\n",
    "for i in predicates:\n",
    "    if i.lower() not in set(SEMMED_predicates['SemmedPred']):\n",
    "        missing_preds.append(i.lower())\n",
    "\n",
    "missing_preds.sort()\n",
    "missing_preds\n",
    "\n",
    "# [i for i in predicates if i.lower() in SEMMED_predicates['SemmedName']]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4e40796c",
   "metadata": {},
   "source": [
    "Currently we add them back in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6a3d217d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.924954Z",
     "start_time": "2023-06-06T13:52:39.912492Z"
    }
   },
   "outputs": [],
   "source": [
    "## manually add some predicates that are in the data, but not in SRDEF...\n",
    "\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['administered_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['augments']\n",
    "## notice that the data doesn't have a hyphen for coexists...even though SRDEF has hyphen for co-occurs_with\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['coexists_with']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['converts_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['inhibits']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['predisposes']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['stimulates']\n",
    "\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ec402da4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.936109Z",
     "start_time": "2023-06-06T13:52:39.926741Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>adjacent_to</td>\n",
       "      <td>SEMMEDDB:ADJACENT_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>analyzes</td>\n",
       "      <td>SEMMEDDB:ANALYZES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>assesses_effect_of</td>\n",
       "      <td>SEMMEDDB:ASSESSES_EFFECT_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>temporally_related_to</td>\n",
       "      <td>SEMMEDDB:TEMPORALLY_RELATED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>traverses</td>\n",
       "      <td>SEMMEDDB:TRAVERSES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>tributary_of</td>\n",
       "      <td>SEMMEDDB:TRIBUTARY_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>61 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SemmedPred                    Semmed_in_BM\n",
       "162            adjacent_to            SEMMEDDB:ADJACENT_TO\n",
       "54         administered_to        SEMMEDDB:ADMINISTERED_TO\n",
       "145                affects                SEMMEDDB:AFFECTS\n",
       "176               analyzes               SEMMEDDB:ANALYZES\n",
       "158     assesses_effect_of     SEMMEDDB:ASSESSES_EFFECT_OF\n",
       "..                     ...                             ...\n",
       "130  temporally_related_to  SEMMEDDB:TEMPORALLY_RELATED_TO\n",
       "166              traverses              SEMMEDDB:TRAVERSES\n",
       "148                 treats                 SEMMEDDB:TREATS\n",
       "178           tributary_of           SEMMEDDB:TRIBUTARY_OF\n",
       "149                   uses                   SEMMEDDB:USES\n",
       "\n",
       "[61 rows x 2 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## since biolink model doesn't use hyphens\n",
    "SEMMED_predicates['Semmed_in_BM']= ['SEMMEDDB:'+i \n",
    "                                   for i in SEMMED_predicates['SemmedPred'].str.replace(\"-\", \"\")]\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.upper()\n",
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "979e0da8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.946919Z",
     "start_time": "2023-06-06T13:52:39.937889Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting the mapping\n",
    "SEMMED_predicates['BMPred'] = [bmt_tool.get_element_by_mapping(i)\n",
    "                                       for i in SEMMED_predicates['Semmed_in_BM']]\n",
    "\n",
    "## since the biolink-mapping uses spaces but resources need to use snake_case\n",
    "SEMMED_predicates['BMPred'] = SEMMED_predicates['BMPred'].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5d7f7944",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.958414Z",
     "start_time": "2023-06-06T13:52:39.948661Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has_input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO               affects\n",
       "145           affects           SEMMEDDB:AFFECTS               affects\n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related_to\n",
       "55           augments          SEMMEDDB:AUGMENTS                  None\n",
       "141            causes            SEMMEDDB:CAUSES                causes\n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists_with\n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates\n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives_into\n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses\n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts\n",
       "58           inhibits          SEMMEDDB:INHIBITS                  None\n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts_with\n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location_of\n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation_of\n",
       "156          measures          SEMMEDDB:MEASURES  quantifier_qualifier\n",
       "171         method_of         SEMMEDDB:METHOD_OF                  None\n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs_in\n",
       "127           part_of           SEMMEDDB:PART_OF               part_of\n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes\n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes\n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents\n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs_in\n",
       "138          produces          SEMMEDDB:PRODUCES              produces\n",
       "60         stimulates        SEMMEDDB:STIMULATES                  None\n",
       "148            treats            SEMMEDDB:TREATS                treats\n",
       "149              uses              SEMMEDDB:USES             has_input"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remember there are 26 predicates in the data\n",
    "len(predicates)\n",
    "pred_compare = [i.lower() for i in predicates]\n",
    "\n",
    "SEMMED_predicates[SEMMED_predicates['SemmedPred'].isin(pred_compare)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c658e0a2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Review the mappings in the table printed by the previous code chunk. \n",
    "\n",
    "For the biolink 3.1.1 update, we:\n",
    "* map semmeddb `augments` and `stimulates` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: increased\n",
    "* map semmeddb `inhibits` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: decreased\n",
    "\n",
    "Notes:\n",
    "* keeping the mapping for semmeddb `administered_to` to biolink-model's `affects` but feel like it's not quite right since it's \"trying something out with a goal\", not that it actually does something...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)\n",
    "* semmeddb `diagnoses` is more of a \"this thing distinguishes / identifies this other thing\"...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T163)\n",
    "* semmeddb `process_of` does seem to map well to biolink-model's `occurs_in` so...okay. As an example, see the use of `process_of` in the [API's data](https://biothings.ncats.io/semmeddb/query?q=predicate:PROCESS_OF%20AND%20object.semantic_type_abbreviation:cell)\n",
    "* keeping semmeddb `treats` is more general than biolink-model's `treats`. It's more of \"trying this out as a treatment\". see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "77adf097",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.968666Z",
     "start_time": "2023-06-06T13:52:39.960162Z"
    }
   },
   "outputs": [],
   "source": [
    "## complex change: modify table to use qualifiers for semmeddb augments, stimulates, and inhibits\n",
    "\n",
    "## At the moment, only need the following qualifiers added, to map these predicates\n",
    "##   QualPred = qualified_predicate\n",
    "##   ObjAsp = object_aspect_qualifier\n",
    "##   ObjDirect = object_direction_qualifier\n",
    "\n",
    "## First, add the columns. Most values will stay blank (None)\n",
    "SEMMED_predicates['BMQualPred'] = None  \n",
    "SEMMED_predicates['BMObjAsp'] = None \n",
    "SEMMED_predicates['BMObjDirect'] = None\n",
    "\n",
    "## Next, mutate the rows for biolink predicates being remapping to predicates + qualifiers\n",
    "for index in SEMMED_predicates.index:\n",
    "    ## semmeddb augments and stimulates should change\n",
    "    if (SEMMED_predicates.loc[index,'SemmedPred']=='augments') or \\\n",
    "       (SEMMED_predicates.loc[index,'SemmedPred']=='stimulates'):\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'increased'\n",
    "    ## semmeddb inhibits should change\n",
    "    elif SEMMED_predicates.loc[index,'SemmedPred']=='inhibits':\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'decreased'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "80e84505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.973452Z",
     "start_time": "2023-06-06T13:52:39.970602Z"
    }
   },
   "outputs": [],
   "source": [
    "## only keep the stuff we found biolink mappings for\n",
    "SEMMED_predicates = SEMMED_predicates[SEMMED_predicates['BMPred'].notna()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "52cfbbc4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.980828Z",
     "start_time": "2023-06-06T13:52:39.975236Z"
    }
   },
   "outputs": [],
   "source": [
    "## get inverses for basic predicates + qualified_predicates to generate reverse operations\n",
    "\n",
    "SEMMED_predicates['BMPred_Inv'] = [bmt_tool.get_element(i).inverse \n",
    "                                       if isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "                                       else i \n",
    "                                       for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "SEMMED_predicates['BMQualPred_Inv'] = [bmt_tool.get_element(i).inverse \n",
    "                                       if i and isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "                                       else i \n",
    "                                       for i in SEMMED_predicates['BMQualPred']]\n",
    "\n",
    "\n",
    "\n",
    "SEMMED_predicates['BMPred_Inv'] = SEMMED_predicates['BMPred_Inv'].str.replace(\" \", \"_\")\n",
    "SEMMED_predicates['BMQualPred_Inv'] = SEMMED_predicates['BMQualPred_Inv'].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9718fe3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.993027Z",
     "start_time": "2023-06-06T13:52:39.982487Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO               affects   \n",
       "145           affects           SEMMEDDB:AFFECTS               affects   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related_to   \n",
       "55           augments          SEMMEDDB:AUGMENTS               affects   \n",
       "141            causes            SEMMEDDB:CAUSES                causes   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists_with   \n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives_into   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts   \n",
       "58           inhibits          SEMMEDDB:INHIBITS               affects   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts_with   \n",
       "172               isa               SEMMEDDB:ISA           subclass_of   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location_of   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation_of   \n",
       "156          measures          SEMMEDDB:MEASURES  quantifier_qualifier   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs_in   \n",
       "127           part_of           SEMMEDDB:PART_OF               part_of   \n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes   \n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs_in   \n",
       "138          produces          SEMMEDDB:PRODUCES              produces   \n",
       "60         stimulates        SEMMEDDB:STIMULATES               affects   \n",
       "148            treats            SEMMEDDB:TREATS                treats   \n",
       "149              uses              SEMMEDDB:USES             has_input   \n",
       "\n",
       "    BMQualPred               BMObjAsp BMObjDirect               BMPred_Inv  \\\n",
       "54        None                   None        None              affected_by   \n",
       "145       None                   None        None              affected_by   \n",
       "160       None                   None        None               related_to   \n",
       "55      causes  activity_or_abundance   increased              affected_by   \n",
       "141       None                   None        None                caused_by   \n",
       "56        None                   None        None            coexists_with   \n",
       "143       None                   None        None        is_exacerbated_by   \n",
       "57        None                   None        None             derives_from   \n",
       "157       None                   None        None          is_diagnosed_by   \n",
       "140       None                   None        None             disrupted_by   \n",
       "58      causes  activity_or_abundance   decreased              affected_by   \n",
       "136       None                   None        None           interacts_with   \n",
       "172       None                   None        None              subclass_of   \n",
       "129       None                   None        None               located_in   \n",
       "144       None                   None        None        has_manifestation   \n",
       "156       None                   None        None     quantifier_qualifier   \n",
       "146       None                   None        None         contains_process   \n",
       "127       None                   None        None                 has_part   \n",
       "132       None                   None        None              preceded_by   \n",
       "59        None                   None        None  has_predisposing_factor   \n",
       "142       None                   None        None             prevented_by   \n",
       "134       None                   None        None         contains_process   \n",
       "138       None                   None        None              produced_by   \n",
       "60      causes  activity_or_abundance   increased              affected_by   \n",
       "148       None                   None        None               treated_by   \n",
       "149       None                   None        None              is_input_of   \n",
       "\n",
       "    BMQualPred_Inv  \n",
       "54            None  \n",
       "145           None  \n",
       "160           None  \n",
       "55       caused_by  \n",
       "141           None  \n",
       "56            None  \n",
       "143           None  \n",
       "57            None  \n",
       "157           None  \n",
       "140           None  \n",
       "58       caused_by  \n",
       "136           None  \n",
       "172           None  \n",
       "129           None  \n",
       "144           None  \n",
       "156           None  \n",
       "146           None  \n",
       "127           None  \n",
       "132           None  \n",
       "59            None  \n",
       "142           None  \n",
       "134           None  \n",
       "138           None  \n",
       "60       caused_by  \n",
       "148           None  \n",
       "149           None  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "933327b9",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Check if the predicate inverses look correct. For biolink 3.0.3, we don't need to correct any inverses.\n",
    "\n",
    "* Maybe some aren't correct because they don't have an inverse but they are directional...If so, I suggest going back and changing the predicates to ones with inverses. \n",
    "* Some predicates aren't directional (their entry in the biolink-model yaml will have the property `symmetric == true`) so those are identical in either direction...\n",
    "\n",
    "<br>\n",
    "\n",
    "However, for the biolink 3.1.1 update, we need more than the inverted predicates...we need the qualifiers for the reverse operations:\n",
    "* \"object_\" qualifiers will invert to \"subject_\" qualifiers, but will keep their exact value \n",
    "* Already did above: when retrieving the inverted predicates, we also retrieved the inverted qualified_predicates\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6dfd3797",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:39.997794Z",
     "start_time": "2023-06-06T13:52:39.994751Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['BMInv_SubjAsp'] = SEMMED_predicates['BMObjAsp'] \n",
    "SEMMED_predicates['BMInv_SubjDirect'] = SEMMED_predicates['BMObjDirect'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f7239c6f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.012157Z",
     "start_time": "2023-06-06T13:52:39.999332Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO               affects   \n",
       "145           affects           SEMMEDDB:AFFECTS               affects   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related_to   \n",
       "55           augments          SEMMEDDB:AUGMENTS               affects   \n",
       "141            causes            SEMMEDDB:CAUSES                causes   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists_with   \n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives_into   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts   \n",
       "58           inhibits          SEMMEDDB:INHIBITS               affects   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts_with   \n",
       "172               isa               SEMMEDDB:ISA           subclass_of   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location_of   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation_of   \n",
       "156          measures          SEMMEDDB:MEASURES  quantifier_qualifier   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs_in   \n",
       "127           part_of           SEMMEDDB:PART_OF               part_of   \n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes   \n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs_in   \n",
       "138          produces          SEMMEDDB:PRODUCES              produces   \n",
       "60         stimulates        SEMMEDDB:STIMULATES               affects   \n",
       "148            treats            SEMMEDDB:TREATS                treats   \n",
       "149              uses              SEMMEDDB:USES             has_input   \n",
       "\n",
       "    BMQualPred               BMObjAsp BMObjDirect               BMPred_Inv  \\\n",
       "54        None                   None        None              affected_by   \n",
       "145       None                   None        None              affected_by   \n",
       "160       None                   None        None               related_to   \n",
       "55      causes  activity_or_abundance   increased              affected_by   \n",
       "141       None                   None        None                caused_by   \n",
       "56        None                   None        None            coexists_with   \n",
       "143       None                   None        None        is_exacerbated_by   \n",
       "57        None                   None        None             derives_from   \n",
       "157       None                   None        None          is_diagnosed_by   \n",
       "140       None                   None        None             disrupted_by   \n",
       "58      causes  activity_or_abundance   decreased              affected_by   \n",
       "136       None                   None        None           interacts_with   \n",
       "172       None                   None        None              subclass_of   \n",
       "129       None                   None        None               located_in   \n",
       "144       None                   None        None        has_manifestation   \n",
       "156       None                   None        None     quantifier_qualifier   \n",
       "146       None                   None        None         contains_process   \n",
       "127       None                   None        None                 has_part   \n",
       "132       None                   None        None              preceded_by   \n",
       "59        None                   None        None  has_predisposing_factor   \n",
       "142       None                   None        None             prevented_by   \n",
       "134       None                   None        None         contains_process   \n",
       "138       None                   None        None              produced_by   \n",
       "60      causes  activity_or_abundance   increased              affected_by   \n",
       "148       None                   None        None               treated_by   \n",
       "149       None                   None        None              is_input_of   \n",
       "\n",
       "    BMQualPred_Inv          BMInv_SubjAsp BMInv_SubjDirect  \n",
       "54            None                   None             None  \n",
       "145           None                   None             None  \n",
       "160           None                   None             None  \n",
       "55       caused_by  activity_or_abundance        increased  \n",
       "141           None                   None             None  \n",
       "56            None                   None             None  \n",
       "143           None                   None             None  \n",
       "57            None                   None             None  \n",
       "157           None                   None             None  \n",
       "140           None                   None             None  \n",
       "58       caused_by  activity_or_abundance        decreased  \n",
       "136           None                   None             None  \n",
       "172           None                   None             None  \n",
       "129           None                   None             None  \n",
       "144           None                   None             None  \n",
       "156           None                   None             None  \n",
       "146           None                   None             None  \n",
       "127           None                   None             None  \n",
       "132           None                   None             None  \n",
       "59            None                   None             None  \n",
       "142           None                   None             None  \n",
       "134           None                   None             None  \n",
       "138           None                   None             None  \n",
       "60       caused_by  activity_or_abundance        increased  \n",
       "148           None                   None             None  \n",
       "149           None                   None             None  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c616c4ff",
   "metadata": {},
   "source": [
    "### prune the data based on what predicates to keep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "50027e5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.019287Z",
     "start_time": "2023-06-06T13:52:40.013861Z"
    }
   },
   "outputs": [],
   "source": [
    "## can mutate and rename the Semmed_in_BM column (not needed anymore) with the format needed\n",
    "\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['SemmedPred'].str.upper()\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.replace(\"-\", \"\")\n",
    "SEMMED_predicates.rename(columns = {'Semmed_in_BM':'Semmed_in_Data'}, inplace = True)\n",
    "\n",
    "## Can remove the SemmedPred column: not needed anymore\n",
    "SEMMED_predicates.drop(columns='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "946dd392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.033117Z",
     "start_time": "2023-06-06T13:52:40.021084Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data                BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO               affects       None                   None   \n",
       "145           AFFECTS               affects       None                   None   \n",
       "160   ASSOCIATED_WITH            related_to       None                   None   \n",
       "55           AUGMENTS               affects     causes  activity_or_abundance   \n",
       "141            CAUSES                causes       None                   None   \n",
       "56      COEXISTS_WITH         coexists_with       None                   None   \n",
       "143       COMPLICATES           exacerbates       None                   None   \n",
       "57        CONVERTS_TO          derives_into       None                   None   \n",
       "157         DIAGNOSES             diagnoses       None                   None   \n",
       "140          DISRUPTS              disrupts       None                   None   \n",
       "58           INHIBITS               affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH        interacts_with       None                   None   \n",
       "172               ISA           subclass_of       None                   None   \n",
       "129       LOCATION_OF           location_of       None                   None   \n",
       "144  MANIFESTATION_OF      manifestation_of       None                   None   \n",
       "156          MEASURES  quantifier_qualifier       None                   None   \n",
       "146         OCCURS_IN             occurs_in       None                   None   \n",
       "127           PART_OF               part_of       None                   None   \n",
       "132          PRECEDES              precedes       None                   None   \n",
       "59        PREDISPOSES           predisposes       None                   None   \n",
       "142          PREVENTS              prevents       None                   None   \n",
       "134        PROCESS_OF             occurs_in       None                   None   \n",
       "138          PRODUCES              produces       None                   None   \n",
       "60         STIMULATES               affects     causes  activity_or_abundance   \n",
       "148            TREATS                treats       None                   None   \n",
       "149              USES             has_input       None                   None   \n",
       "\n",
       "    BMObjDirect               BMPred_Inv BMQualPred_Inv  \\\n",
       "54         None              affected_by           None   \n",
       "145        None              affected_by           None   \n",
       "160        None               related_to           None   \n",
       "55    increased              affected_by      caused_by   \n",
       "141        None                caused_by           None   \n",
       "56         None            coexists_with           None   \n",
       "143        None        is_exacerbated_by           None   \n",
       "57         None             derives_from           None   \n",
       "157        None          is_diagnosed_by           None   \n",
       "140        None             disrupted_by           None   \n",
       "58    decreased              affected_by      caused_by   \n",
       "136        None           interacts_with           None   \n",
       "172        None              subclass_of           None   \n",
       "129        None               located_in           None   \n",
       "144        None        has_manifestation           None   \n",
       "156        None     quantifier_qualifier           None   \n",
       "146        None         contains_process           None   \n",
       "127        None                 has_part           None   \n",
       "132        None              preceded_by           None   \n",
       "59         None  has_predisposing_factor           None   \n",
       "142        None             prevented_by           None   \n",
       "134        None         contains_process           None   \n",
       "138        None              produced_by           None   \n",
       "60    increased              affected_by      caused_by   \n",
       "148        None               treated_by           None   \n",
       "149        None              is_input_of           None   \n",
       "\n",
       "             BMInv_SubjAsp BMInv_SubjDirect  \n",
       "54                    None             None  \n",
       "145                   None             None  \n",
       "160                   None             None  \n",
       "55   activity_or_abundance        increased  \n",
       "141                   None             None  \n",
       "56                    None             None  \n",
       "143                   None             None  \n",
       "57                    None             None  \n",
       "157                   None             None  \n",
       "140                   None             None  \n",
       "58   activity_or_abundance        decreased  \n",
       "136                   None             None  \n",
       "172                   None             None  \n",
       "129                   None             None  \n",
       "144                   None             None  \n",
       "156                   None             None  \n",
       "146                   None             None  \n",
       "127                   None             None  \n",
       "132                   None             None  \n",
       "59                    None             None  \n",
       "142                   None             None  \n",
       "134                   None             None  \n",
       "138                   None             None  \n",
       "60   activity_or_abundance        increased  \n",
       "148                   None             None  \n",
       "149                   None             None  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9c8b1e2b",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "Remove predicates in the mapping table, but we don't want to use for associations:\n",
    "- no mapping: METHOD_OF\n",
    "- already removed from data: ISA\n",
    "- don't want to use it as an association predicate after seeing the mapping: \n",
    "  - MEASURES\n",
    "  - USES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9079161f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.040184Z",
     "start_time": "2023-06-06T13:52:40.034753Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in data but not in mapping table\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'METHOD_OF'}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in mapping file but not in data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ISA'}"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in data but not in mapping table\")\n",
    "predicates - set(SEMMED_predicates['Semmed_in_Data'])\n",
    "\n",
    "print(\"in mapping file but not in data\")\n",
    "set(SEMMED_predicates['Semmed_in_Data']) - predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5e25a929",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.044856Z",
     "start_time": "2023-06-06T13:52:40.041844Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ISA', 'MEASURES', 'METHOD_OF', 'USES'}"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_removals = {'METHOD_OF', 'ISA', 'MEASURES', 'USES'}\n",
    "more_removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5a46158c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:40.058318Z",
     "start_time": "2023-06-06T13:52:40.046463Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data            BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO           affects       None                   None   \n",
       "145           AFFECTS           affects       None                   None   \n",
       "160   ASSOCIATED_WITH        related_to       None                   None   \n",
       "55           AUGMENTS           affects     causes  activity_or_abundance   \n",
       "141            CAUSES            causes       None                   None   \n",
       "56      COEXISTS_WITH     coexists_with       None                   None   \n",
       "143       COMPLICATES       exacerbates       None                   None   \n",
       "57        CONVERTS_TO      derives_into       None                   None   \n",
       "157         DIAGNOSES         diagnoses       None                   None   \n",
       "140          DISRUPTS          disrupts       None                   None   \n",
       "58           INHIBITS           affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH    interacts_with       None                   None   \n",
       "129       LOCATION_OF       location_of       None                   None   \n",
       "144  MANIFESTATION_OF  manifestation_of       None                   None   \n",
       "146         OCCURS_IN         occurs_in       None                   None   \n",
       "127           PART_OF           part_of       None                   None   \n",
       "132          PRECEDES          precedes       None                   None   \n",
       "59        PREDISPOSES       predisposes       None                   None   \n",
       "142          PREVENTS          prevents       None                   None   \n",
       "134        PROCESS_OF         occurs_in       None                   None   \n",
       "138          PRODUCES          produces       None                   None   \n",
       "60         STIMULATES           affects     causes  activity_or_abundance   \n",
       "148            TREATS            treats       None                   None   \n",
       "\n",
       "    BMObjDirect               BMPred_Inv BMQualPred_Inv  \\\n",
       "54         None              affected_by           None   \n",
       "145        None              affected_by           None   \n",
       "160        None               related_to           None   \n",
       "55    increased              affected_by      caused_by   \n",
       "141        None                caused_by           None   \n",
       "56         None            coexists_with           None   \n",
       "143        None        is_exacerbated_by           None   \n",
       "57         None             derives_from           None   \n",
       "157        None          is_diagnosed_by           None   \n",
       "140        None             disrupted_by           None   \n",
       "58    decreased              affected_by      caused_by   \n",
       "136        None           interacts_with           None   \n",
       "129        None               located_in           None   \n",
       "144        None        has_manifestation           None   \n",
       "146        None         contains_process           None   \n",
       "127        None                 has_part           None   \n",
       "132        None              preceded_by           None   \n",
       "59         None  has_predisposing_factor           None   \n",
       "142        None             prevented_by           None   \n",
       "134        None         contains_process           None   \n",
       "138        None              produced_by           None   \n",
       "60    increased              affected_by      caused_by   \n",
       "148        None               treated_by           None   \n",
       "\n",
       "             BMInv_SubjAsp BMInv_SubjDirect  \n",
       "54                    None             None  \n",
       "145                   None             None  \n",
       "160                   None             None  \n",
       "55   activity_or_abundance        increased  \n",
       "141                   None             None  \n",
       "56                    None             None  \n",
       "143                   None             None  \n",
       "57                    None             None  \n",
       "157                   None             None  \n",
       "140                   None             None  \n",
       "58   activity_or_abundance        decreased  \n",
       "136                   None             None  \n",
       "129                   None             None  \n",
       "144                   None             None  \n",
       "146                   None             None  \n",
       "127                   None             None  \n",
       "132                   None             None  \n",
       "59                    None             None  \n",
       "142                   None             None  \n",
       "134                   None             None  \n",
       "138                   None             None  \n",
       "60   activity_or_abundance        increased  \n",
       "148                   None             None  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(23, 9)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remove this set from the SEMMED_predicates table\n",
    "SEMMED_predicates = SEMMED_predicates[~ SEMMED_predicates['Semmed_in_Data'].isin(more_removals)]\n",
    "SEMMED_predicates\n",
    "SEMMED_predicates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "64cacd52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:42.362664Z",
     "start_time": "2023-06-06T13:52:40.059850Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove this set from the data record\n",
    "filtered_data = filtered_data[ ~ filtered_data['PREDICATE'].isin(more_removals)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "530e3631",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-06-06T13:52:46.365832Z",
     "start_time": "2023-06-06T13:52:42.364647Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(filtered_data[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(filtered_data[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(filtered_data[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 41, still 41 (expected)\n",
    "len(object_semtypes)  ## was 41, still 41 (expected)\n",
    "len(predicates)       ## was 26, now 23: decreased by 3 (expected: ISA was already removed earlier)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e1e9660d",
   "metadata": {},
   "source": [
    "## Final Filter: row counts per combo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "863543b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.822341Z",
     "start_time": "2023-03-23T19:17:44.690657Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5936, 4)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>707094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>696732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>phsu</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>660654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>cell</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>598884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>542301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "0            bpoc  LOCATION_OF           neop  1023005\n",
       "1            bpoc  LOCATION_OF           aapp   953010\n",
       "2            topp       TREATS           dsyn   849825\n",
       "3            cell  LOCATION_OF           aapp   815374\n",
       "4            bpoc  LOCATION_OF           patf   799234\n",
       "5            topp       TREATS           neop   707094\n",
       "6            bpoc  LOCATION_OF           dsyn   696732\n",
       "7            phsu       TREATS           dsyn   660654\n",
       "8            cell      PART_OF           bpoc   598884\n",
       "9            bpoc      PART_OF           bpoc   542301"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at number of combos after this removal\n",
    "combos = filtered_data.value_counts().reset_index()\n",
    "combos.columns = ['SUBJECT_SEMTYPE', 'PREDICATE', 'OBJECT_SEMTYPE', 'COUNT']\n",
    "combos.shape\n",
    "combos.head(10)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "305f3a3a",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "use the code block below to decide how many combos to keep based on how many predications/records there are per combo...\n",
    "\n",
    "---\n",
    "\n",
    "Current:\n",
    "\n",
    "let's try setting a limit that we're only building operations if there are > 100 objects for the triple..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d874b8f2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.832827Z",
     "start_time": "2023-03-23T19:17:49.823700Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>1023005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>953010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>849825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>815374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>799234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3027</th>\n",
       "      <td>lipd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3028</th>\n",
       "      <td>strd</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>neop</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3029</th>\n",
       "      <td>clnd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>anab</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030</th>\n",
       "      <td>imft</td>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>inpo</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3031</th>\n",
       "      <td>antb</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>imft</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3032 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE    COUNT\n",
       "0               bpoc  LOCATION_OF           neop  1023005\n",
       "1               bpoc  LOCATION_OF           aapp   953010\n",
       "2               topp       TREATS           dsyn   849825\n",
       "3               cell  LOCATION_OF           aapp   815374\n",
       "4               bpoc  LOCATION_OF           patf   799234\n",
       "...              ...          ...            ...      ...\n",
       "3027            lipd       TREATS           neop      102\n",
       "3028            strd     AUGMENTS           neop      101\n",
       "3029            clnd       TREATS           anab      101\n",
       "3030            imft  PREDISPOSES           inpo      101\n",
       "3031            antb     INHIBITS           imft      101\n",
       "\n",
       "[3032 rows x 4 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combos[(combos['COUNT'] > 100)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "86920c33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.842396Z",
     "start_time": "2023-03-23T19:17:49.834491Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>COUNT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>carb</td>\n",
       "      <td>17938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>610</th>\n",
       "      <td>aapp</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>carb</td>\n",
       "      <td>7547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794</th>\n",
       "      <td>aapp</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>carb</td>\n",
       "      <td>4555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961</th>\n",
       "      <td>aapp</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>carb</td>\n",
       "      <td>3296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2127</th>\n",
       "      <td>aapp</td>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>carb</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2383</th>\n",
       "      <td>aapp</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>carb</td>\n",
       "      <td>289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_SEMTYPE       PREDICATE OBJECT_SEMTYPE  COUNT\n",
       "320             aapp  INTERACTS_WITH           carb  17938\n",
       "610             aapp   COEXISTS_WITH           carb   7547\n",
       "794             aapp      STIMULATES           carb   4555\n",
       "961             aapp        INHIBITS           carb   3296\n",
       "2127            aapp     CONVERTS_TO           carb    438\n",
       "2383            aapp        PRODUCES           carb    289"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## checking that the aapp - PRODUCES - carb triple appears w/ this limit\n",
    "## https://biothings.ncats.io/semmeddb/query?q=object.umls:C0043369%20AND%20subject.umls:C0002003\n",
    "## from this issue: https://github.com/biothings/BioThings_Explorer_TRAPI/issues/317\n",
    "combos[(combos['COUNT'] > 100) &\n",
    "       (combos['OBJECT_SEMTYPE'] == 'carb') &\n",
    "       (combos['SUBJECT_SEMTYPE'] == 'aapp')]\n",
    "## okay we have the desired triple with this limit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2c859df9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.851816Z",
     "start_time": "2023-03-23T19:17:49.843923Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3032, 3)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_SEMTYPE    PREDICATE OBJECT_SEMTYPE\n",
       "0            bpoc  LOCATION_OF           neop\n",
       "1            bpoc  LOCATION_OF           aapp\n",
       "2            topp       TREATS           dsyn"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos = combos[(combos['COUNT'] > 100)].copy()\n",
    "filtered_combos.drop(columns='COUNT', inplace=True)\n",
    "filtered_combos.shape\n",
    "filtered_combos[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "b53030eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.859086Z",
     "start_time": "2023-03-23T19:17:49.853149Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all subject to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']], \n",
    "                      how='left', left_on='SUBJECT_SEMTYPE', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject',\n",
    "                  'BiolinkSubject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "9e435399",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.869158Z",
     "start_time": "2023-03-23T19:17:49.860443Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all object to biolink\n",
    "\n",
    "filtered_combos = filtered_combos.merge(SEMMED_entity_types[['Abbrev', 'BiolinkMapping']],\n",
    "                      how='left', left_on='OriginalObject', right_on='Abbrev')\n",
    "\n",
    "filtered_combos.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "filtered_combos.columns = ['OriginalSubject', 'OriginalPredicate', 'OriginalObject',\n",
    "                  'BiolinkSubject', 'BiolinkObject']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a0fe51f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.878224Z",
     "start_time": "2023-03-23T19:17:49.870547Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3027</th>\n",
       "      <td>lipd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>neop</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3028</th>\n",
       "      <td>strd</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>neop</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3029</th>\n",
       "      <td>clnd</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>anab</td>\n",
       "      <td>Drug</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030</th>\n",
       "      <td>imft</td>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>inpo</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3031</th>\n",
       "      <td>antb</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>imft</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3032 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     OriginalSubject OriginalPredicate OriginalObject  \\\n",
       "0               bpoc       LOCATION_OF           neop   \n",
       "1               bpoc       LOCATION_OF           aapp   \n",
       "2               topp            TREATS           dsyn   \n",
       "3               cell       LOCATION_OF           aapp   \n",
       "4               bpoc       LOCATION_OF           patf   \n",
       "...              ...               ...            ...   \n",
       "3027            lipd            TREATS           neop   \n",
       "3028            strd          AUGMENTS           neop   \n",
       "3029            clnd            TREATS           anab   \n",
       "3030            imft       PREDISPOSES           inpo   \n",
       "3031            antb          INHIBITS           imft   \n",
       "\n",
       "                BiolinkSubject        BiolinkObject  \n",
       "0     GrossAnatomicalStructure              Disease  \n",
       "1     GrossAnatomicalStructure          Polypeptide  \n",
       "2                    Procedure              Disease  \n",
       "3                         Cell          Polypeptide  \n",
       "4     GrossAnatomicalStructure  PathologicalProcess  \n",
       "...                        ...                  ...  \n",
       "3027             SmallMolecule              Disease  \n",
       "3028             SmallMolecule              Disease  \n",
       "3029                      Drug              Disease  \n",
       "3030            ChemicalEntity  PathologicalProcess  \n",
       "3031             SmallMolecule       ChemicalEntity  \n",
       "\n",
       "[3032 rows x 5 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3f42bd07",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.884236Z",
     "start_time": "2023-03-23T19:17:49.879715Z"
    }
   },
   "outputs": [],
   "source": [
    "filtered_combos = filtered_combos.merge(\n",
    "                      SEMMED_predicates,\n",
    "                      how='left', left_on='OriginalPredicate', right_on='Semmed_in_Data')\n",
    "\n",
    "filtered_combos.drop(columns = 'Semmed_in_Data', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "6c651d23",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.894905Z",
     "start_time": "2023-03-23T19:17:49.885534Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  OriginalSubject OriginalPredicate OriginalObject            BiolinkSubject  \\\n",
       "0            bpoc       LOCATION_OF           neop  GrossAnatomicalStructure   \n",
       "1            bpoc       LOCATION_OF           aapp  GrossAnatomicalStructure   \n",
       "2            topp            TREATS           dsyn                 Procedure   \n",
       "\n",
       "  BiolinkObject       BMPred BMQualPred BMObjAsp BMObjDirect  BMPred_Inv  \\\n",
       "0       Disease  location_of       None     None        None  located_in   \n",
       "1   Polypeptide  location_of       None     None        None  located_in   \n",
       "2       Disease       treats       None     None        None  treated_by   \n",
       "\n",
       "  BMQualPred_Inv BMInv_SubjAsp BMInv_SubjDirect  \n",
       "0           None          None             None  \n",
       "1           None          None             None  \n",
       "2           None          None             None  "
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_combos[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "cc7864f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T19:17:49.915709Z",
     "start_time": "2023-03-23T19:17:49.896527Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BMPred_Inv, BMQualPred_Inv, BMInv_SubjAsp, BMInv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BMPred_Inv, BMQualPred_Inv, BMInv_SubjAsp, BMInv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BMPred_Inv, BMQualPred_Inv, BMInv_SubjAsp, BMInv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BMPred_Inv</th>\n",
       "      <th>BMQualPred_Inv</th>\n",
       "      <th>BMInv_SubjAsp</th>\n",
       "      <th>BMInv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [OriginalSubject, OriginalPredicate, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BMPred_Inv, BMQualPred_Inv, BMInv_SubjAsp, BMInv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## check to make sure everything is mapped to Biolink successfully\n",
    "filtered_combos[filtered_combos['BiolinkSubject'].isna()]\n",
    "filtered_combos[filtered_combos['BiolinkObject'].isna()]\n",
    "\n",
    "filtered_combos[filtered_combos['BMPred'].isna()]\n",
    "filtered_combos[filtered_combos['BMPred_Inv'].isna()]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "62b0c6b7",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Optional Analysis: looking at row counts, organized by biolink combos"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4fad7760",
   "metadata": {
    "hidden": true
   },
   "source": [
    "PAUSE\n",
    "\n",
    "Deprecated code: will need to include COUNT column after creating the filtered_combos object..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af677da3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-17T05:16:46.369759Z",
     "start_time": "2022-08-17T05:16:46.362883Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "### deep-copy combos so it's not messed up for the next step\n",
    "biolink_counting = filtered_combos.copy()\n",
    "biolink_counting = biolink_counting.groupby(\n",
    "                       ['BiolinkSubject', 'BiolinkPredicate', 'BiolinkObject']\n",
    "                   ).agg(\n",
    "                       { \"COUNT\": \"sum\"}\n",
    ")\n",
    "## other stuff that can go into agg\n",
    "#                          \"OriginalSubject\": lambda x: set(x),\n",
    "#                          \"OriginalPredicate\": lambda x: set(x),\n",
    "#                          \"OriginalObject\": lambda x: set(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "328305ca",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-17T05:16:47.194635Z",
     "start_time": "2022-08-17T05:16:47.181968Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "biolink_counting.reset_index(inplace = True)\n",
    "\n",
    "biolink_counting.sort_values(by='COUNT', ascending=False, inplace = True)\n",
    "\n",
    "biolink_counting[0:50]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "870db906",
   "metadata": {},
   "source": [
    "## Generate operation yaml!"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "344f9ec6",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* If needed, change the code within the functions below to change the x-bte annotations that are made...\n",
    "  * review the qualifier-generating code! Currently it's very simple because all operations with qualifiers have the same set: qualified_predicate, an aspect_qualifier, and a direction_qualifier...\n",
    "* the yaml created below refers to umls-subj and umls-obj...those are specified [here close to the bottom](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "1fcf921d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:38:19.820578Z",
     "start_time": "2023-03-23T23:38:19.817435Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml=ryml.YAML()\n",
    "folded = ryml.scalarstring.FoldedScalarString\n",
    "doublequote = ryml.scalarstring.DoubleQuotedScalarString"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "222c6bd1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:42:59.361323Z",
     "start_time": "2023-03-23T23:42:59.350478Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_forward_op(original_subj, original_pred, original_obj,\n",
    "                        biolink_subj, \n",
    "                        biolink_pred, biolink_qualified_pred, biolink_obj_asp, biolink_obj_direct,\n",
    "                        biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    normal_op_name = f\"{original_subj}-{original_pred}-{original_obj}\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## original direction: subject -> object\n",
    "    normal_op_body = folded(\n",
    "    '{' \n",
    "        '\"q\": {{ queryInputs ' \n",
    "        '| replPrefix(\\'predicate:' + f'{original_pred} AND object.semantic_type_abbreviation:{original_obj}' + ' AND pmid_count:>3 AND subject.umls\\')' \n",
    "        '| dump ' \n",
    "        '}}, ' \n",
    "        '\"scopes\": []' \n",
    "    '}')\n",
    "    \n",
    "    if biolink_qualified_pred:\n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has obj aspect and obj direction\n",
    "      temp = {\n",
    "            ## original direction: subject -> object\n",
    "            normal_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## input is subject!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': normal_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## output is object\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_pred,\n",
    "                    'qualifiers': {\n",
    "                        'qualified_predicate': biolink_qualified_pred,\n",
    "                        'object_aspect_qualifier': biolink_obj_asp,\n",
    "                        'object_direction_qualifier': biolink_obj_direct\n",
    "\n",
    "                    },\n",
    "                    'source': 'infores:semmeddb',\n",
    "#                     'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-obj'  ## matches output as object\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    else:\n",
    "    ## create operation without qualifiers..\n",
    "        temp = {\n",
    "            ## original direction: subject -> object\n",
    "            normal_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## input is subject!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': normal_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## output is object\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_pred,\n",
    "                    'source': 'infores:semmeddb',\n",
    "#                     'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-obj'  ## matches output as object\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "        \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "912aa75d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:43:06.821089Z",
     "start_time": "2023-03-23T23:43:06.813695Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_reverse_op(original_subj, original_pred, original_obj,\n",
    "                        biolink_subj, \n",
    "                        ## NOTICE THE INVERSES USED HERE\n",
    "                        biolink_inverse_pred, biolink_inverse_qualified_pred, \n",
    "                        biolink_inverse_subj_asp, biolink_inverse_subj_direct,\n",
    "                        biolink_obj):\n",
    "    ## set size parameter for biothings POST query. This will change depending on what is set for the API\n",
    "    POST_size = 1000  \n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    rev_op_name = f\"{original_subj}-{original_pred}-{original_obj}-rev\"\n",
    "    \n",
    "    ## USE FOLDED in order to have the quotes handled properly (no escape \\) in the dumped document\n",
    "    ## reverse direction: object -> subject\n",
    "    rev_op_body = folded(\n",
    "    '{' \n",
    "        '\"q\": {{ queryInputs ' \n",
    "        '| replPrefix(\\'predicate:' + f'{original_pred} AND subject.semantic_type_abbreviation:{original_subj}' + ' AND pmid_count:>3 AND object.umls\\')' \n",
    "        '| dump ' \n",
    "        '}}, ' \n",
    "        '\"scopes\": []' \n",
    "    '}')\n",
    "    \n",
    "    if biolink_inverse_qualified_pred:\n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has subj aspect and subj direction\n",
    "        temp = {\n",
    "            ## reverse direction: object -> subject\n",
    "            rev_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## input is object!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': rev_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## output is subject\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                    'qualifiers': {\n",
    "                        'qualified_predicate': biolink_inverse_qualified_pred,  ## use inverse pred!\n",
    "                        ## use subject qualifiers\n",
    "                        'subject_aspect_qualifier': biolink_inverse_subj_asp,         \n",
    "                        'subject_direction_qualifier': biolink_inverse_subj_direct                        \n",
    "                    },                    \n",
    "                    'source': 'infores:semmeddb',\n",
    "#                     'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-subj'  ## matches output as subj\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    else:\n",
    "    ## create operation without qualifiers..\n",
    "        temp = {\n",
    "            ## reverse direction: object -> subject\n",
    "            rev_op_name: [\n",
    "                {\n",
    "                    'supportBatch': True,\n",
    "                    'useTemplating': True,\n",
    "                    'inputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_obj  ## input is object!\n",
    "                        }\n",
    "                    ],\n",
    "                    'requestBodyType': 'object',\n",
    "                    'requestBody': {'body': rev_op_body},\n",
    "                    'parameters': {\n",
    "                        'fields': 'object.umls,predication.pmid,predication.sentence,subject.umls,subject.name,object.name',\n",
    "                        'size': POST_size\n",
    "                    },\n",
    "                    'outputs': [\n",
    "                        {\n",
    "                            'id': 'UMLS',\n",
    "                            'semantic': biolink_subj  ## output is subject\n",
    "                        }\n",
    "                    ],\n",
    "                    'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                    'source': 'infores:semmeddb',\n",
    "#                     'source': doublequote('infores:semmeddb'),\n",
    "                    'response_mapping': {\n",
    "                        \"$ref\": '#/components/x-bte-response-mapping/umls-subj'  ## matches output as subj\n",
    "                    }\n",
    "                }\n",
    "            ]\n",
    "        }\n",
    "    \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "db9f9f81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:38:23.126436Z",
     "start_time": "2023-03-23T23:38:23.120208Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_all_operations(combo_df):\n",
    "    op_tracking = set()\n",
    "    \n",
    "    saved = dict()\n",
    "    ## iterate through rows of combos dataframe\n",
    "    for row in combo_df.itertuples(index = False):\n",
    "        \n",
    "        ## forward: only make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,object used here)\n",
    "        forward_op_record = f\"{row.BiolinkSubject}-{row.OriginalPredicate}-{row.OriginalObject}\"\n",
    "        if forward_op_record not in op_tracking:\n",
    "            op_tracking.add(forward_op_record)\n",
    "            saved.update(generate_forward_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_pred = row.BMPred,\n",
    "                                             biolink_qualified_pred = row.BMQualPred,\n",
    "                                             biolink_obj_asp = row.BMObjAsp,\n",
    "                                             biolink_obj_direct = row.BMObjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                            ))        \n",
    "        \n",
    "        ## reverse: make operation if it's not going to be a dupe\n",
    "        ##          dupes happen when query ends up being the same (predicate,subject used here)\n",
    "        reverse_op_record = f\"rev-{row.BiolinkObject}-{row.OriginalPredicate}-{row.OriginalSubject}\"\n",
    "        if reverse_op_record not in op_tracking:\n",
    "            op_tracking.add(reverse_op_record)\n",
    "            saved.update(generate_reverse_op(original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_inverse_pred = row.BMPred_Inv,\n",
    "                                             biolink_inverse_qualified_pred = row.BMQualPred_Inv,\n",
    "                                             biolink_inverse_subj_asp = row.BMInv_SubjAsp,\n",
    "                                             biolink_inverse_subj_direct = row.BMInv_SubjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                             ))           \n",
    "            \n",
    "    final = {\"x-bte-kgs-operations\": saved}\n",
    "    return final"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5e2ec9a4",
   "metadata": {},
   "source": [
    "Get the file made!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "3856781d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:43:11.045549Z",
     "start_time": "2023-03-23T23:43:11.021272Z"
    }
   },
   "outputs": [],
   "source": [
    "kgs_operations = generate_all_operations(filtered_combos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "fdcd4202",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:43:11.867100Z",
     "start_time": "2023-03-23T23:43:11.863585Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2497"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "5495d772",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:43:12.615832Z",
     "start_time": "2023-03-23T23:43:12.611131Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'supportBatch': True,\n",
      "  'useTemplating': True,\n",
      "  'inputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
      "  'requestBodyType': 'object',\n",
      "  'requestBody': {'body': '{\"q\": {{ queryInputs | '\n",
      "                          \"replPrefix('predicate:INHIBITS AND \"\n",
      "                          'subject.semantic_type_abbreviation:aapp AND '\n",
      "                          \"pmid_count:>5 AND object.umls')| dump }}, \"\n",
      "                          '\"scopes\": []}'},\n",
      "  'parameters': {'fields': 'object.umls,predication.pmid,pmid_count,predication_count,subject.umls,subject.name,predicate,object.name',\n",
      "                 'size': 1000},\n",
      "  'outputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
      "  'predicate': 'affected_by',\n",
      "  'qualifiers': {'qualified_predicate': 'caused_by',\n",
      "                 'subject_aspect_qualifier': 'activity_or_abundance',\n",
      "                 'subject_direction_qualifier': 'decreased'},\n",
      "  'source': 'infores:semmeddb',\n",
      "  'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'}}]\n"
     ]
    }
   ],
   "source": [
    "for i in kgs_operations['x-bte-kgs-operations'].keys():\n",
    "    if ('INHIBITS' in i) and ('rev' in i):\n",
    "        pprint.pprint(kgs_operations['x-bte-kgs-operations'][i], sort_dicts = False)\n",
    "        break"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "fd057af2",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* it's cool how condensed the operations become, due to the way querying is done (keeping track of not creating duplicated operations using the sets in generate_all_operations function)\n",
    "* set where to download the yamls in the code chunks below\n",
    "  * operations_path\n",
    "  * list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "239dce9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-03-23T23:43:21.640380Z",
     "start_time": "2023-03-23T23:43:17.459402Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.boolean_representation = ['False', 'True']\n",
    "\n",
    "operations_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"semmeddb2_specific\", \"generated_operations2.yaml\")\n",
    "\n",
    "yaml.dump(kgs_operations, operations_path)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "c946247a",
   "metadata": {},
   "source": [
    "Wait a sec! Need the operations list too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27403608",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-08T22:52:45.928167Z",
     "start_time": "2022-12-08T22:52:45.924961Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_kgs_operations_list(operations_dict):\n",
    "    kgs_op_list = []\n",
    "    for key in operations_dict.keys():\n",
    "        kgs_op_list.append( {\"$ref\": f\"#/components/x-bte-kgs-operations/{key}\"} )\n",
    "    final2 = {\"x-bte-kgs-operations\": kgs_op_list}\n",
    "    return final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bec97ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-08T22:52:47.614059Z",
     "start_time": "2022-12-08T22:52:47.611199Z"
    }
   },
   "outputs": [],
   "source": [
    "operations_list = generate_kgs_operations_list(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e64c34db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-12-08T22:52:48.628066Z",
     "start_time": "2022-12-08T22:52:48.289221Z"
    }
   },
   "outputs": [],
   "source": [
    "list_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_list.yaml\")\n",
    "\n",
    "yaml.dump(operations_list, list_path)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "421fbcdc",
   "metadata": {},
   "source": [
    "**PAUSE**\n",
    "\n",
    "* now the yaml segments downloaded have to be indented manually and inserted into the correct sections of the smartapi yaml...\n",
    "  * It's easier to do with an IDE like Visual Code studio where one can select large sections of text\n",
    "  * the amount of indent to do and where to put things is specified in the [yaml that acts as a template](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)\n",
    "  * the finished file is meant to be [here](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) so one could select-paste the sections directly there\n",
    "  \n",
    "---\n",
    "\n",
    "The code below is optional, in case one wants to convert the yaml to json (for BTE's test/query endpoint testing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebd89060",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-17T05:24:00.017107Z",
     "start_time": "2022-08-17T05:23:49.455429Z"
    }
   },
   "outputs": [],
   "source": [
    "## extra code in case we want to convert to json...\n",
    "\n",
    "yaml_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"smartapi.yaml\")\n",
    "\n",
    "here = yaml.load(yaml_path)\n",
    "\n",
    "json_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"smartapi.json\")\n",
    "\n",
    "with open(json_path, 'w') as file:\n",
    "    json.dump(here, file)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "efa95cb1",
   "metadata": {},
   "source": [
    "## Notes on choices here"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ce33fcc5",
   "metadata": {},
   "source": [
    "### Removing"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "7eb04a8e",
   "metadata": {},
   "source": [
    "* was mapped to Polypeptide:\n",
    "    * amas: (Amino Acid Sequence) looks like protein \"domains\". Examples: Nuclear Export Signals, DNA Binding Domain\n",
    "* everything mapped to Activity (8)\n",
    "    * acty (Activity) Examples: War, Retirement, Euthanasia, Lifting\n",
    "    * dora (Daily or Recreational Activity) Examples: Physical activity, Light Exercise, Relaxation\n",
    "    * edac (Educational Activity) Examples: Training, Medical Residencies\n",
    "    * gora (Governmental or Regulatory Activity) Examples: Health Care Reform, Advisory Committees\n",
    "    * hlca (Health Care Activity) Examples: follow-up, Diagnosis\n",
    "    * mcha (Machine Activity) Examples: Refridgeration, Neural Network Simulation\n",
    "    * ocac (Occupational Activity) Examples: Promotion, Work, Mining\n",
    "    * resa (Research Activity) Examples: Clinical Trials, research study\n",
    "* everything mapped to Cohort (4)\n",
    "    * aggp (Age Group) Examples: Infant, Child, Adult, Elderly\n",
    "    * famg (Family Group) Examples: spouse, Sister, Foster Parent\n",
    "    * podg (Patient or Disabled Group) Examples: Patients\n",
    "    * prog (Professional or Occupational Group) Examples: Administrators, Employee, Author\n",
    "* older semantic types (4)\n",
    "    * algae (Algae)\n",
    "    * invt (Invertebrate)\n",
    "    * orgm (Organism)\n",
    "    * rich (Rickettsia or Chlamydia)\n",
    "* missing mappings to biolink-model (25):\n",
    "    * amph (Amphibian) Examples: Toad, Bufo boreas, Anura\n",
    "    * anim (Animal) Examples: Animals, Laboratory /  Control Animal\n",
    "    * arch (Archaeon) Examples: Archaea, halophilic bacteria, Thermoplasma acidophilum\n",
    "    * bact (Bacterium) Examples: Escherichia coli, Salmonella, Borrelia burgdorferi\n",
    "    * bdsu (Body Substance) too general. Examples: Urine, Milk, Lymph, Urine specimen\n",
    "    * bdsy (Body System) too general. Examples: hypothalamic-pituitary-adrenal axis, Neurosecretory Systems\n",
    "    * bird (Bird) Examples: Geese, Passeriformes, Raptors\n",
    "    * blor (Body Location or Region) too general. Examples: Hepatic, Lysosomal, Cytoplasmic\n",
    "    * bmod (Biomedical Occupation or Discipline) Examples: Medicine, Dentistry, Midwifery\n",
    "    * bsoj (Body Space or Junction) too general. Examples: Compartments, Synapses, Cistern\n",
    "    * chvf (Chemical Viewed Functionally) too general. Examples: inhibitors, antagonists, Agent\n",
    "    * chvs (Chemical Viewed Structurally) too general. Examples: particle, solid state, vapor\n",
    "    * euka (Eukaryote) Examples: Wasps, Protozoan parasite\n",
    "    * ffas (Fully Formed Anatomical Structure) Examples: Carcass\n",
    "    * fish (Fish) Examples: Eels, Fishes, Electric Fish\n",
    "    * fngs (Fungus) Examples: Saccharomyces cerevisiae, Alternaria brassicicola, fungus\n",
    "    * humn (Human) Examples: Family, Patients, Males\n",
    "    * irda (Indicator, Reagent, or Diagnostic Aid) Examples: Fluorescent Probes, Chelating Agents\n",
    "    * mamm (Mammal) Examples: Rattus norvegicus, Felis catus, Mus\n",
    "    * ocdi (Occupation or Discipline) Examples: Science, Politics\n",
    "    * plnt (Plant) Examples: Chrysanthemum x morifolium, Pollen, Oryza sativa\n",
    "    * rept (Reptile) Examples: Snakes, Turtles, Reptiles\n",
    "    * sbst (Substance) too general. Examples: Materials, Plastics, Photons, Substance\n",
    "    * virs (Virus) Examples: Herpesvirus 4, Human / GB virus C / Herpesviridae\n",
    "    * vtbt (Vertebrate) Examples: Vertebrates / Poikilotherm, NOS\n",
    "* was mapped to AnatomicalEntity:\n",
    "    * anst (Anatomical Structure) Examples: Entire fetus, Whole body, Cadaver\n",
    "* was mapped to Behavior (4)\n",
    "    * bhvr (Behavior) too general. Examples: Sexuality, Nest Building, Behavioral phenotype\n",
    "    * inbe (Individual Behavior) too general. Examples: impulsivity, Habits, Performance\n",
    "    * menp (Mental Process) too general. Examples: mind control, Learning, experience\n",
    "    * socb (Social Behavior) too general. Examples: Communication, Gestures, Marriage\n",
    "* was mapped to Phenomenon (6)\n",
    "    * biof (Biologic Function) too general. Examples: dose-response relationship, Pharmacodynamics, Anabolism\n",
    "    * eehu (Environmental Effect of Humans) too general. Examples: Sewage, Pollution, Smoke\n",
    "    * hcpp (Human-caused Phenomenon or Process) too general. Now not in API. Examples: particle beam, Conferences, Victimization\n",
    "    * lbtr (Laboratory or Test Result) too general. Examples: False Positive Reactions, Bone Density, Serum Calcium Level\n",
    "    * npop (Natural Phenomenon or Process) too general. Examples: Floods, Fluorescence, Freezing\n",
    "    * phpr (Phenomenon or Process) too general. Examples: Disasters, Acceleration, Feedback\n",
    "* was mapped to Device (4)\n",
    "    * bodm (Biomedical or Dental Material) too general. Examples: Pill, Gel, Talc, calcium phosphate\n",
    "    * drdd (Drug Delivery Device) too general. Examples: Epipen, Skin Patch, Lilly cyanide antidote kit\n",
    "    * medd (Medical Device) too general. Examples: Implants / Denture, Overlay / Silicone gel implant / Swab\n",
    "    * resd (Research Device) too general. Examples: Study models, Slide\n",
    "* was mapped to GrossAnatomicalStructure\n",
    "    * emst (Embryonic Structure) Examples: Chick Embryo, Blastocyst structure, Placenta\n",
    "    * tisu (Tissue) Examples: Tissue specimen, Blood, Human tissue, Mucous Membrane\n",
    "* was mapped to PhysiologicalProcess\n",
    "    * genf (Genetic Function) too general. Examples: Transcription, Genetic / Transcriptional Activation / Recombination, Genetic\n",
    "* was mapped to ChemicalEntity\n",
    "    * chem (Chemical) too general. Examples: Chemicals, Acids, Ligands, Ozone\n",
    "* everything mapped to InformationContentEntity (8)\n",
    "    * clas (Classification) too general. Examples: Research Diagnostic Criteria, Group C\n",
    "    * ftcn (Functional Concept) too general. Examples: Techniques, Intravenous Route of Drug Administration\n",
    "    * idcn (Idea or Concept) too general. Examples: Significant, subject, Data, Owner\n",
    "    * qlco (Qualitative Concept) too general. Examples: Effect, Associated with, Advanced phase\n",
    "    * qnco (Quantitative Concept) too general. Examples: Calibration, occurrence, degrees Celsius\n",
    "    * rnlw (Regulation or Law) too general. Examples: Medicare, Medicaid, regulatory\n",
    "    * spco (Spatial Concept) too general. Examples: Structure, Longitudinal, Asymmetry\n",
    "    * tmco (Temporal Concept) too general. Examples: New, /period, 24 Hours\n",
    "* everything mapped to ClinicalAttribute\n",
    "    * clna (Clinical Attribute) too general. Examples: response, Renin secretion, BAND PATTERN\n",
    "* was mapped to Procedure\n",
    "    * lbpr (Laboratory Procedure) too general. Examples: Western Blot, Radioimmunoassay, Staining method\n",
    "    * mbrt (Molecular Biology Research Technique) too general. Examples: Polymerase Chain Reaction / Blotting, Northern\n",
    "* everything mapped to SmallMolecule\n",
    "    * elii (Element, Ion, or Isotope) too general. Examples: Atom, Aluminum, Superoxides\n",
    "* everything mapped to BiologicalEntity\n",
    "    * emod (Experimental Model of Disease) too general. Examples: Experimental Autoimmune Encephalomyelitis, Rodent Model\n",
    "* was mapped to Protein\n",
    "    * rcpt (Receptor) too general. Examples: Binding Sites / Receptors, Metabotropic Glutamate\n",
    "* was mapped to Event\n",
    "    * evnt (Event) too general. Now not in API. Examples: Stressful Events\n",
    "* was mapped to DiseaseOrPhenotypicFeature\n",
    "    * fndg (Finding) too general. Examples: spinal cord; lesion, Normal birth weight, Sedentary job\n",
    "* was mapped to GeographicLocation\n",
    "    * geoa (Geographic Area) too general. Examples: Country, Canada\n",
    "* everything mapped to Agent\n",
    "    * grup (Group) too general. Examples: Human, Individual\n",
    "    * hcro (Health Care Related Organization) too general. Examples: Hospitals, Health System\n",
    "    * orgt (Organization) too general. Examples: United Nations, Organization administrative structures\n",
    "    * pros (Professional Society) too general. Examples: Professional Organizations, American Nurses' Association\n",
    "    * shro (Self-help or Relief Organization) too general. Examples: Social Welfare, Support Groups\n",
    "* everything mapped to Publication\n",
    "    * inpr (Intellectual Product) too general. Examples: Methodology, Study models\n",
    "* everything mapped to PhysicalEntity\n",
    "    * mnob (Manufactured Object) too general. Examples: Glass, Manuals\n",
    "* everything mapped to MolecularEntity\n",
    "    * mosq (Molecular Sequence) too general. Now not in API. Examples: Genetic Code\n",
    "* was mapped to NucleicAcidEntity\n",
    "    * nusq (Nucleotide Sequence) too general. Examples: Base Sequence, DNA Sequence, 22q11\n",
    "* was mapped to OrganismAttribute\n",
    "    * orga (Organism Attribute) too general. Examples: Ability, Body Composition\n",
    "* was mapped to PopulationOfIndividualOrganisms\n",
    "    * popg (Population Group) too general. Examples: Male population group, Woman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf7f82a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "796.5px",
    "left": "39px",
    "top": "109.633px",
    "width": "362.5px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
