{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "47e7307b",
   "metadata": {},
   "source": [
    "# Auto-generate x-bte annotations for BioThings SEMMEDDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ab5663",
   "metadata": {},
   "source": [
    "This notebook walks a developer through the process of taking the [BioThings SEMMEDDB](https://biothings.ncats.io/semmeddb) meta-triple + API-doc-count file [generated during the API deployment process](https://github.com/biothings/semmeddb/blob/main/pred_meta_triples.ipynb) and using this data to auto-generate the x-bte operations for [BTE](https://github.com/biothings/BioThings_Explorer_TRAPI). This is needed for BTE to query + process the responses from the BioThings SEMMEDDB API. \n",
    "\n",
    "---\n",
    "\n",
    "When one sees this:  \n",
    "# PAUSE\n",
    "\n",
    "read the accompanying text that will explain what the developer needs to do before running the code chunks below that text block.\n",
    "\n",
    "---\n",
    "\n",
    "The [yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_operations.yaml) [segments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_list.yaml) generated by this notebook are added to [this file](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml) to make the [SmartAPI yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) for BioThings SEMMEDDB API, which is registered in the [SmartAPI Registry](https://smart-api.info/registry)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43745613",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712186ee",
   "metadata": {},
   "source": [
    "Requirements:\n",
    "* install [Biolink-model Toolkit (bmt)](https://github.com/biolink/biolink-model-toolkit/). I'm [installing as a user](https://biolink.github.io/biolink-model-toolkit/intro/intro.html#for-users) with pip (`pip install bmt`). Using release 1.1.1 at the moment.\n",
    "* install ruamel.yaml with [pip](https://pypi.org/project/ruamel.yaml/). Using 0.17.32 at the moment. It's more recent than the version from [conda-forge](https://anaconda.org/conda-forge/ruamel_yaml). The documentation is [here](https://yaml.readthedocs.io/en/latest/)\n",
    "\n",
    "Files to obtain before starting:\n",
    "* Get the meta-triple + API-doc-count file from the dev for BioThings SEMMEDDB. I've saved the file I'm using to the same directory as this notebook. This file was provided 2023-08-10 by Yao, and it includes counts of matching API documents/records for each metatriple. Generally, the total set of meta-triples should be fairly constant between data releases, but the doc counts may differ. \n",
    "* Get SEMMEDDB SRDEF file needed for interpreting SEMMED semantic types: I'm using 2023AA right now. \n",
    "  * download the compressed file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/download.html) or pick the latest Sem Net file [here](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11497815",
   "metadata": {},
   "source": [
    "### Set config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5ee5aac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:39.127720Z",
     "start_time": "2023-11-01T00:02:38.670913Z"
    }
   },
   "outputs": [],
   "source": [
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "## see above for install instructions\n",
    "from bmt import Toolkit\n",
    "import ruamel.yaml as ryml    ## use ruamel_yaml if using the conda-forge version\n",
    "\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "## used in trying things out\n",
    "# import re\n",
    "# import matplotlib.pyplot as plt   ## for plotting doc counts\n",
    "\n",
    "\n",
    "## set up ruamel.yaml\n",
    "yaml=ryml.YAML()\n",
    "folded = ryml.scalarstring.FoldedScalarString\n",
    "doublequote = ryml.scalarstring.DoubleQuotedScalarString"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af532d5",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* Set variables used in x-bte annotation\n",
    "  * pmid_count: discussed in [this Translator issue](https://github.com/NCATSTranslator/Feedback/issues/100#issuecomment-1632806388)\n",
    "* Set the `metatriples_path` to your local metatriples + API-doc-counts file\n",
    "* Set the `srdef_location` to your local SRDEF file\n",
    "* Check the url to the [Translator-curated exclusions (tsv file)](https://github.com/biolink/biolink-model/blob/master/SEMMEDDB_exclude_list.tsv) to remove some SEMMEDDB semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73408978",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:41.606833Z",
     "start_time": "2023-11-01T00:02:41.499166Z"
    }
   },
   "outputs": [],
   "source": [
    "## variables used for x-bte annotation\n",
    "## set size parameter for biothings POST query. This is 1000 unless this API has custom settings for this\n",
    "POST_size = 1000  \n",
    "## pmid_count: only allow numbers greater than this\n",
    "pmid_count = 3\n",
    "\n",
    "## metatriples + API-doc-counts file\n",
    "metatriples_path = pathlib.Path.home().joinpath(\n",
    "    \"Desktop\", \"translator-api-registry\", \"semmeddb\", \n",
    "    \"semmedVER43_2023_R_PREDICATION.116080_clean_meta_triples.tsv\")\n",
    "\n",
    "imported_metatriples = pd.read_csv(metatriples_path, sep=\"\\t\")\n",
    "\n",
    "\n",
    "## SEMMED Semantic info file\n",
    "srdef_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"2023AA\", \"SRDEF\")\n",
    "\n",
    "srdef = pd.read_csv(srdef_location, sep=\"|\", header=0, index_col=False,\n",
    "                    names=[\"Record Type (RT)\",\n",
    "                           \"Unique Identifier (UI)\",\n",
    "                           \"Full Name (STY/RL)\", \n",
    "                           \"Tree Number (STN/RTN)\", \n",
    "                           \"Definition (DEF)\",\n",
    "                           \"Examples (EX)\",\n",
    "                           \"Usage Note (UN)\",\n",
    "                           \"Non-Human Flag (NH)\",\n",
    "                           \"Abbreviation (ABR)\",\n",
    "                           \"Inverse Relation (RIN)\"])\n",
    "\n",
    "\n",
    "## import the Translator-curated exclusions tsv\n",
    "imported_exclusions = pd.read_csv(\n",
    "    'https://raw.githubusercontent.com/biolink/biolink-model/master/SEMMEDDB_exclude_list.tsv',\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "## where the output yamls go\n",
    "operations_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_operations.yaml\")\n",
    "\n",
    "list_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_list.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8150ed6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:44.016288Z",
     "start_time": "2023-11-01T00:02:44.010537Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34582, 7)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57d2dc4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:44.527624Z",
     "start_time": "2023-11-01T00:02:44.518346Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2344345</td>\n",
       "      <td>2147949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1183119</td>\n",
       "      <td>1133560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1159426</td>\n",
       "      <td>1011905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1088609</td>\n",
       "      <td>998529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>876862</td>\n",
       "      <td>804853</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            bpoc      PART_OF          umls           mamm   \n",
       "2           umls            fndg   PROCESS_OF          umls           humn   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2344345      2147949  \n",
       "1  1183119      1133560  \n",
       "2  1159426      1011905  \n",
       "3  1088609       998529  \n",
       "4   876862       804853  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b68180",
   "metadata": {},
   "source": [
    "## Basic Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b8cf0b",
   "metadata": {},
   "source": [
    "Notes:\n",
    "* This BioThings API's parser already filters out all API docs where novelty == 0, since those with novelty == 0 probably aren't very helpful / interesting to Translator. The entities with novelty == 0 are [listed](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html) in the SEMMEDDB GENERIC_CONCEPT table files, which can be downloaded [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403fe7b4",
   "metadata": {},
   "source": [
    "First, I'm doing a quick look at what exists in this meta-triple file:\n",
    "* two ID-namespaces for subject + object: umls + ncbigene\n",
    "  * will want special handling for ncbigene: biolink-model only recognizes NCBIGene as a valid namespace for specific categories, like Gene \n",
    "* all semmeddb semantic types are 4-letter abbreviations / codes\n",
    "  * 132 unique values for subject\n",
    "  * 131 unique values for object\n",
    "* semmeddb predicates are a mix of different cases...\n",
    "  * 63 unique values\n",
    "\n",
    "Side-note: the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say semmeddb's vocab has 127 semantic types and 54 predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "181df62c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:49.390915Z",
     "start_time": "2023-11-01T00:02:49.372438Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['umls', 'ncbigene'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['umls', 'ncbigene'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['aapp', 'gngm', 'bacs', 'horm', 'phsu', 'enzy', 'imft', 'rcpt',\n",
       "       'nsba', 'nusq', 'nnon', 'orch', 'lipd', 'hops', 'irda', 'carb',\n",
       "       'bodm', 'antb', 'vita', 'strd'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['aapp', 'gngm', 'enzy', 'phsu', 'bacs', 'horm', 'irda', 'imft',\n",
       "       'nsba', 'nusq', 'rcpt', 'nnon', 'antb', 'orch', 'hops', 'bodm'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.SUBJECT_PREFIX.unique()\n",
    "imported_metatriples.OBJECT_PREFIX.unique()\n",
    "\n",
    "imported_metatriples[imported_metatriples['SUBJECT_PREFIX'] == 'ncbigene'].SUBJECT_SEMTYPE.unique()\n",
    "imported_metatriples[imported_metatriples['OBJECT_PREFIX'] == 'ncbigene'].OBJECT_SEMTYPE.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a52d0b4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:50.112680Z",
     "start_time": "2023-11-01T00:02:50.100473Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "63"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## 132\n",
    "len(object_semtypes)  ## 131\n",
    "len(predicates)       ## 63"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d0dd2ed7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:52.462759Z",
     "start_time": "2023-11-01T00:02:52.456269Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'evnt',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.SUBJECT_SEMTYPE.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b27b5f43",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:55.256871Z",
     "start_time": "2023-11-01T00:02:55.250502Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.OBJECT_SEMTYPE.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5765d868",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:02:57.534253Z",
     "start_time": "2023-11-01T00:02:57.528199Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADMINISTERED_TO',\n",
       " 'AFFECTS',\n",
       " 'ASSOCIATED_WITH',\n",
       " 'AUGMENTS',\n",
       " 'CAUSES',\n",
       " 'COEXISTS_WITH',\n",
       " 'COMPLICATES',\n",
       " 'CONVERTS_TO',\n",
       " 'DIAGNOSES',\n",
       " 'DISRUPTS',\n",
       " 'INHIBITS',\n",
       " 'INTERACTS_WITH',\n",
       " 'ISA',\n",
       " 'LOCATION_OF',\n",
       " 'MANIFESTATION_OF',\n",
       " 'MEASUREMENT_OF',\n",
       " 'MEASURES',\n",
       " 'METHOD_OF',\n",
       " 'NEG_ADMINISTERED_TO',\n",
       " 'NEG_AFFECTS',\n",
       " 'NEG_ASSOCIATED_WITH',\n",
       " 'NEG_AUGMENTS',\n",
       " 'NEG_CAUSES',\n",
       " 'NEG_COEXISTS_WITH',\n",
       " 'NEG_COMPLICATES',\n",
       " 'NEG_CONVERTS_TO',\n",
       " 'NEG_DIAGNOSES',\n",
       " 'NEG_DISRUPTS',\n",
       " 'NEG_INHIBITS',\n",
       " 'NEG_INTERACTS_WITH',\n",
       " 'NEG_ISA',\n",
       " 'NEG_LOCATION_OF',\n",
       " 'NEG_MANIFESTATION_OF',\n",
       " 'NEG_MEASUREMENT_OF',\n",
       " 'NEG_MEASURES',\n",
       " 'NEG_METHOD_OF',\n",
       " 'NEG_OCCURS_IN',\n",
       " 'NEG_PART_OF',\n",
       " 'NEG_PRECEDES',\n",
       " 'NEG_PREDISPOSES',\n",
       " 'NEG_PREVENTS',\n",
       " 'NEG_PROCESS_OF',\n",
       " 'NEG_PRODUCES',\n",
       " 'NEG_STIMULATES',\n",
       " 'NEG_TREATS',\n",
       " 'NEG_USES',\n",
       " 'NEG_higher_than',\n",
       " 'NEG_lower_than',\n",
       " 'NEG_same_as',\n",
       " 'OCCURS_IN',\n",
       " 'PART_OF',\n",
       " 'PRECEDES',\n",
       " 'PREDISPOSES',\n",
       " 'PREVENTS',\n",
       " 'PROCESS_OF',\n",
       " 'PRODUCES',\n",
       " 'STIMULATES',\n",
       " 'TREATS',\n",
       " 'USES',\n",
       " 'compared_with',\n",
       " 'higher_than',\n",
       " 'lower_than',\n",
       " 'same_as'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.PREDICATE.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff20381b",
   "metadata": {},
   "source": [
    "### Remove Predicates we don't want to make operations from"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "106d9b99",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* Decide what predicates you want to remove\n",
    "\n",
    "**Current logic**\n",
    "\n",
    "I'm using biolink-model 3.5.3 right now. \n",
    "\n",
    "With biolink-model >= 3.5.1, Translator has reviewed some predicates for domain-predicate and predicate-range combo exclusions - ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "We'll want to keep the predicates that have been reviewed and only do the custom removals below for predicates that:\n",
    "* haven't been reviewed yet\n",
    "* are problematic to use right now: \n",
    "    * Translator currently isn't great with negation, so removing those\n",
    "    * I didn't find the following relationships useful: \n",
    "      * ISA (subclass relationship)\n",
    "      * lower_than / higher_than (these two entities were compared and 1 was better (higher) or worse (lower) than the other)\n",
    "      * compared_with...\n",
    "    \n",
    "And we'll want to adjust the lists below if Translator reviews more predicates in the future. \n",
    "\n",
    "Old notes:\n",
    "* same_as (previously excluded): itss meaning seems to be \"equivalent / it did just as well as some other thing\"\n",
    "* If you see `[\"VERB\", \"NOM\", \"PREP\"]`, these seem to be practice phrases that we wouldn't want to use. see the [article](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-486#Sec26)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dec1bcc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:01.876607Z",
     "start_time": "2023-11-01T00:03:01.870466Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negative_preds = []\n",
    "\n",
    "for i in imported_metatriples.PREDICATE.unique():\n",
    "    if i.startswith('NEG_'):\n",
    "        negative_preds.append(i)\n",
    "        \n",
    "len(negative_preds)\n",
    "        \n",
    "        \n",
    "dont_like = [\"ISA\", \"lower_than\", \"higher_than\", \"compared_with\"]\n",
    "\n",
    "## so 31 + 4 = 35 predicates are being removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ab643b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:02.630334Z",
     "start_time": "2023-11-01T00:03:02.622042Z"
    }
   },
   "outputs": [],
   "source": [
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(negative_preds)]\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(dont_like)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f3d145",
   "metadata": {},
   "source": [
    "now look at the stats. \n",
    "\n",
    "(interesting since the stats from the [official website](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates. 54 / 2 = 27 predicates without negation + 1 for same_as inclusion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6244458",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:05.831879Z",
     "start_time": "2023-11-01T00:03:05.822018Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 132, now 127: decreased by 5\n",
    "len(object_semtypes)  ## was 131, now 127: decreased by 4\n",
    "len(predicates)       ## was 63, now 28: decreased by 35 (was expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e31d94",
   "metadata": {},
   "source": [
    "so....that's how many meta-triples we have to go forward with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24036b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:09.121690Z",
     "start_time": "2023-11-01T00:03:09.118038Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18256, 7)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape\n",
    "## so that's still a lot...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "949ccb58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:09.800477Z",
     "start_time": "2023-11-01T00:03:09.792534Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2344345</td>\n",
       "      <td>2147949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1183119</td>\n",
       "      <td>1133560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1159426</td>\n",
       "      <td>1011905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1088609</td>\n",
       "      <td>998529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>876862</td>\n",
       "      <td>804853</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            bpoc      PART_OF          umls           mamm   \n",
       "2           umls            fndg   PROCESS_OF          umls           humn   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2344345      2147949  \n",
       "1  1183119      1133560  \n",
       "2  1159426      1011905  \n",
       "3  1088609       998529  \n",
       "4   876862       804853  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e998d9e8",
   "metadata": {},
   "source": [
    "## Mapping and Specific Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330bba7a",
   "metadata": {},
   "source": [
    "First, I'm going to map SEMMED semantic info to biolink-model - this is a legacy from when I used the full names + mapping to decide what SEMMED semantic types and predicates to keep for x-bte annotation with. Now, I'll do minimal adjustments.  \n",
    "\n",
    "Then I'll use Translator-curated exclusions to remove some SEMMED semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d470fa70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:13.283952Z",
     "start_time": "2023-11-01T00:03:13.274254Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Record Type (RT)</th>\n",
       "      <th>Unique Identifier (UI)</th>\n",
       "      <th>Full Name (STY/RL)</th>\n",
       "      <th>Tree Number (STN/RTN)</th>\n",
       "      <th>Definition (DEF)</th>\n",
       "      <th>Examples (EX)</th>\n",
       "      <th>Usage Note (UN)</th>\n",
       "      <th>Non-Human Flag (NH)</th>\n",
       "      <th>Abbreviation (ABR)</th>\n",
       "      <th>Inverse Relation (RIN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>STY</td>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>A1.4.1.2.1.7</td>\n",
       "      <td>Amino acids and chains of amino acids connecte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>When the concept is both an enzyme and a prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aapp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Record Type (RT) Unique Identifier (UI)               Full Name (STY/RL)  \\\n",
       "97              STY                   T116  Amino Acid, Peptide, or Protein   \n",
       "\n",
       "   Tree Number (STN/RTN)                                   Definition (DEF)  \\\n",
       "97          A1.4.1.2.1.7  Amino acids and chains of amino acids connecte...   \n",
       "\n",
       "    Examples (EX)                                    Usage Note (UN)  \\\n",
       "97            NaN  When the concept is both an enzyme and a prote...   \n",
       "\n",
       "   Non-Human Flag (NH) Abbreviation (ABR) Inverse Relation (RIN)  \n",
       "97                 NaN               aapp                    NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## quick look at srdef info\n",
    "srdef[srdef['Abbreviation (ABR)'] == 'aapp']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e245634",
   "metadata": {},
   "source": [
    "### Custom SEMMED semantic type processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "89c14528",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:16.916662Z",
     "start_time": "2023-11-01T00:03:16.906103Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(126, 3)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>T195</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst\n",
       "119  T195                       Antibiotic   antb"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## simplify the df\n",
    "SEMMED_semtypes = srdef[srdef['Record Type (RT)'] == 'STY'].copy()\n",
    "SEMMED_semtypes = SEMMED_semtypes[['Unique Identifier (UI)',\n",
    "                                   'Full Name (STY/RL)', \n",
    "                                   'Abbreviation (ABR)']].copy()\n",
    "SEMMED_semtypes.columns = ['UI', 'FullName', \"Abbrev\"]\n",
    "SEMMED_semtypes.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_semtypes.shape\n",
    "SEMMED_semtypes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b70c514f",
   "metadata": {},
   "source": [
    "#### Add SEMMED semantic types that are in the data but not in SRDEF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49397246",
   "metadata": {},
   "source": [
    "Yao [noticed](https://github.com/biothings/pending.api/issues/30#issuecomment-903609946) that the data file uses \"old SEMMED semantic types\" from 2013AA, and these semantic types didn't exist in the latest SRDEF file (2020AA).\n",
    "\n",
    "In response, Andrew said to [keep](https://github.com/biothings/pending.api/issues/30#issuecomment-903879782) this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8c8c04ce",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:20.458468Z",
     "start_time": "2023-11-01T00:03:20.452265Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "these semantic types are in the data, but not in SRDEF:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'carb',\n",
       " 'eico',\n",
       " 'invt',\n",
       " 'lipd',\n",
       " 'nsba',\n",
       " 'opco',\n",
       " 'orgm',\n",
       " 'rich',\n",
       " 'strd'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "only_in_data = data_types - set(SEMMED_semtypes['Abbrev'])\n",
    "\n",
    "print(\"these semantic types are in the data, but not in SRDEF:\")\n",
    "only_in_data\n",
    "\n",
    "len(only_in_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533341a2",
   "metadata": {},
   "source": [
    "How I found info on these semantic types from [previous semantic versions](https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html):\n",
    "\n",
    "3 were in 2009AB file (sets with SG_2?)\n",
    "* alga\n",
    "* invt\n",
    "* rich\n",
    "\n",
    "7 were in 2014AB file (sets with SG_3?)\n",
    "* carb \n",
    "* eico \n",
    "* lipd \n",
    "* nsba \n",
    "* opco\n",
    "* orgm\n",
    "* strd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0400e8e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:24.012577Z",
     "start_time": "2023-11-01T00:03:24.007343Z"
    }
   },
   "outputs": [],
   "source": [
    "## build the missing info and add to the table\n",
    "missing_semantic_types = [{\"UI\": \"T003\", \"FullName\": \"Alga\", \"Abbrev\": \"alga\"},\n",
    "                          {\"UI\": \"T118\", \"FullName\": \"Carbohydrate\", \"Abbrev\": \"carb\"},\n",
    "                          {\"UI\": \"T111\", \"FullName\": \"Eicosanoid\", \"Abbrev\": \"eico\"},\n",
    "                          {\"UI\": \"T009\", \"FullName\": \"Invertebrate\", \"Abbrev\": \"invt\"},\n",
    "                          {\"UI\": \"T119\", \"FullName\": \"Lipid\", \"Abbrev\": \"lipd\"},\n",
    "                          {\"UI\": \"T124\", \"FullName\": \"Neuroreactive Substance or Biogenic Amine\", \"Abbrev\": \"nsba\"},\n",
    "                          {\"UI\": \"T115\", \"FullName\": \"Organophosphorus Compound\", \"Abbrev\": \"opco\"},\n",
    "                          {\"UI\": \"T001\", \"FullName\": \"Organism\", \"Abbrev\": \"orgm\"},\n",
    "                          {\"UI\": \"T006\", \"FullName\": \"Rickettsia or Chlamydia\", \"Abbrev\": \"rich\"},\n",
    "                          {\"UI\": \"T110\", \"FullName\": \"Steroid\", \"Abbrev\": \"strd\"}\n",
    "                         ]\n",
    "\n",
    "missing_semantic_types = pd.DataFrame.from_records(missing_semantic_types)\n",
    "\n",
    "## add it to the table\n",
    "SEMMED_semtypes = pd.concat([SEMMED_semtypes, missing_semantic_types])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "683a1f47",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:26.182766Z",
     "start_time": "2023-11-01T00:03:26.174292Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(136, 3)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "0    T003                             Alga   alga\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_semtypes.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_semtypes.shape\n",
    "SEMMED_semtypes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f01880",
   "metadata": {},
   "source": [
    "#### Get the biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a7a170a6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:32.602632Z",
     "start_time": "2023-11-01T00:03:31.496557Z"
    }
   },
   "outputs": [],
   "source": [
    "## as of 2023-08-14: using default which is 3.5.3 \n",
    "bmt_tool = Toolkit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4cc1fad9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:33.349055Z",
     "start_time": "2023-11-01T00:03:33.325528Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Small Molecule'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## double-checking: this corresponds to bacs (Biologically Active Substance)\n",
    "bmt_tool.get_element_by_mapping('STY:T123').title()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2c14436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:34.381544Z",
     "start_time": "2023-11-01T00:03:33.944283Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting biolink-mapping, in the format needed to create operations\n",
    "SEMMED_semtypes['BiolinkMapping'] = [bmt_tool.get_element_by_mapping('STY:'+i)\n",
    "                                     for i in SEMMED_semtypes['UI']]\n",
    "## put these biolink-model categories in the correct format for Translator/TRAPI: PascalCase\n",
    "SEMMED_semtypes['BiolinkMapping'] = [i.title().replace(\" \", \"\") if isinstance(i, str)\n",
    "                                     else i for i in SEMMED_semtypes['BiolinkMapping']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51111407",
   "metadata": {},
   "source": [
    "#### Clean up to only have terms in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7fb08392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:38.801800Z",
     "start_time": "2023-11-01T00:03:38.797553Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "note that these semantic types are in SRDEF but not in the data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cnce', 'crbs', 'enty', 'grpa', 'lang', 'phob'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"note that these semantic types are in SRDEF but not in the data\")\n",
    "set(SEMMED_semtypes['Abbrev']) - data_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "22daf133",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:39.536838Z",
     "start_time": "2023-11-01T00:03:39.533499Z"
    }
   },
   "outputs": [],
   "source": [
    "## get the subset of semantic network terms that are actually in the data\n",
    "SEMMED_semtypes = SEMMED_semtypes[SEMMED_semtypes['Abbrev'].isin(data_types)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "044843d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:40.314858Z",
     "start_time": "2023-11-01T00:03:40.307716Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>T054</td>\n",
       "      <td>Social Behavior</td>\n",
       "      <td>socb</td>\n",
       "      <td>Behavior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>T184</td>\n",
       "      <td>Sign or Symptom</td>\n",
       "      <td>sosy</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>T082</td>\n",
       "      <td>Spatial Concept</td>\n",
       "      <td>spco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T110</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>strd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T024</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>tisu</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "46   T054                      Social Behavior   socb   \n",
       "113  T184                      Sign or Symptom   sosy   \n",
       "73   T082                      Spatial Concept   spco   \n",
       "9    T110                              Steroid   strd   \n",
       "19   T024                               Tissue   tisu   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "46                   Behavior  \n",
       "113         PhenotypicFeature  \n",
       "73   InformationContentEntity  \n",
       "9               SmallMolecule  \n",
       "19   GrossAnatomicalStructure  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## was exploring the mapping by looking at slices of df\n",
    "SEMMED_semtypes[120:130]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d30960",
   "metadata": {},
   "source": [
    "#### review mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd9e849",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "We are still reviewing / changing mappings, since BTE's responses would probably differ a lot if this behavior was changed. \n",
    "\n",
    "We are only removing SEMMED semantic types that lack biolink-model mappings. Other curated SEMMED semantic type removals will be done in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "I'm going to keep the old notes + code (commented out), just in case. \n",
    "\n",
    "---\n",
    "\n",
    "Old notes:\n",
    "\n",
    "This is a place to **STOP** and review all the SEMMED semantic types and their mapping to biolink categories...to decide what we are interested in keeping. This involves some knowledge of what biolink categories are prioritized in Translator. \n",
    "\n",
    "One can use the definitions of the SEMMED semantic types (from the SRDEF file or [browsing the UMLS vocab online](https://uts.nlm.nih.gov/uts/umls/semantic-network/root)) and the definitions of biolink categories (look for a comment with 'THINGS' in the biolink-model yaml file)\n",
    "\n",
    "See the last section (section 6) for notes on decisions that were made..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eb7ea639",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:49.251495Z",
     "start_time": "2023-11-01T00:03:49.247556Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Polypeptide', 'Disease', 'Activity', 'Cohort', None,\n",
       "       'Invertebrate', 'OrganismalEntity', 'AnatomicalEntity', 'Drug',\n",
       "       'SmallMolecule', 'Bacterium', 'Behavior', 'Phenomenon', 'Device',\n",
       "       'GrossAnatomicalStructure', 'CellularComponent',\n",
       "       'PhysiologicalProcess', 'Cell', 'ChemicalEntity',\n",
       "       'InformationContentEntity', 'ClinicalAttribute', 'Procedure',\n",
       "       'BiologicalEntity', 'Protein', 'Event',\n",
       "       'DiseaseOrPhenotypicFeature', 'Fungus', 'Food',\n",
       "       'GeographicLocation', 'GenomicEntity', 'Agent', 'Human',\n",
       "       'PathologicalProcess', 'Publication', 'DiagnosticAid', 'Mammal',\n",
       "       'PhysicalEntity', 'MolecularActivity', 'MolecularEntity',\n",
       "       'NucleicAcidEntity', 'OrganismAttribute', 'Plant',\n",
       "       'PopulationOfIndividualOrganisms', 'PhenotypicFeature', 'Virus',\n",
       "       'Vertebrate'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## code used to review \n",
    "SEMMED_semtypes['BiolinkMapping'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fc902a02",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:50.263726Z",
     "start_time": "2023-11-01T00:03:50.256005Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>T194</td>\n",
       "      <td>Archaeon</td>\n",
       "      <td>arch</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>T022</td>\n",
       "      <td>Body System</td>\n",
       "      <td>bdsy</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T012</td>\n",
       "      <td>Bird</td>\n",
       "      <td>bird</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>T091</td>\n",
       "      <td>Biomedical Occupation or Discipline</td>\n",
       "      <td>bmod</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>T120</td>\n",
       "      <td>Chemical Viewed Functionally</td>\n",
       "      <td>chvf</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>T104</td>\n",
       "      <td>Chemical Viewed Structurally</td>\n",
       "      <td>chvs</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>T204</td>\n",
       "      <td>Eukaryote</td>\n",
       "      <td>euka</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>T021</td>\n",
       "      <td>Fully Formed Anatomical Structure</td>\n",
       "      <td>ffas</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T013</td>\n",
       "      <td>Fish</td>\n",
       "      <td>fish</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>T045</td>\n",
       "      <td>Genetic Function</td>\n",
       "      <td>genf</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T009</td>\n",
       "      <td>Invertebrate</td>\n",
       "      <td>invt</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>T090</td>\n",
       "      <td>Occupation or Discipline</td>\n",
       "      <td>ocdi</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>T001</td>\n",
       "      <td>Organism</td>\n",
       "      <td>orgm</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T014</td>\n",
       "      <td>Reptile</td>\n",
       "      <td>rept</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>T006</td>\n",
       "      <td>Rickettsia or Chlamydia</td>\n",
       "      <td>rich</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>T167</td>\n",
       "      <td>Substance</td>\n",
       "      <td>sbst</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev BiolinkMapping\n",
       "0    T003                                 Alga   alga           None\n",
       "118  T194                             Archaeon   arch           None\n",
       "17   T022                          Body System   bdsy           None\n",
       "7    T012                                 Bird   bird           None\n",
       "81   T091  Biomedical Occupation or Discipline   bmod           None\n",
       "98   T120         Chemical Viewed Functionally   chvf           None\n",
       "94   T104         Chemical Viewed Structurally   chvs           None\n",
       "125  T204                            Eukaryote   euka           None\n",
       "16   T021    Fully Formed Anatomical Structure   ffas           None\n",
       "8    T013                                 Fish   fish           None\n",
       "37   T045                     Genetic Function   genf           None\n",
       "3    T009                         Invertebrate   invt           None\n",
       "80   T090             Occupation or Discipline   ocdi           None\n",
       "7    T001                             Organism   orgm           None\n",
       "9    T014                              Reptile   rept           None\n",
       "8    T006              Rickettsia or Chlamydia   rich           None\n",
       "108  T167                            Substance   sbst           None"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'] == 'Vertebrate']\n",
    "SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c626a79a",
   "metadata": {},
   "source": [
    "Still re-mapping (not following a strict biolink-model mapping). May revisit later, but concerned about affecting responses to creative-mode queries / requiring template adjustments..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e641ae5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:53.969497Z",
     "start_time": "2023-11-01T00:03:53.964711Z"
    }
   },
   "outputs": [],
   "source": [
    "## re-mapping based on putting IDs into normalization service / our operation system...\n",
    "##   we have UMLS for Disease (mydisease, mychem), SmallMolecule (idisk)\n",
    "\n",
    "## leaving aapp as Polypeptide, enzy mapped to Protein\n",
    "## leaving Disease mappings as-is, but some seem like they could be PathologicalAnatomicalStructure \n",
    "##   (acab, anab, cgab) or PathologicalProcess (comd) instead\n",
    "## left clnd (Clinical Drug) as Drug since it really seemed like a drug (dosage)\n",
    "\n",
    "## currently mapped to Drug, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'antb'),'BiolinkMapping'] = 'SmallMolecule'  ## Antibiotic\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'phsu'),'BiolinkMapping'] = 'SmallMolecule'  ## Pharmacologic Substance\n",
    "\n",
    "## currently mapped to GenomicEntity, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'gngm'),'BiolinkMapping'] = 'Gene'\n",
    "\n",
    "## currently mapped to NucleicAcidEntity, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'nnon'),'BiolinkMapping'] = 'SmallMolecule'  ## Nucleic Acid, Nucleoside, or Nucleotide"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f46a30",
   "metadata": {},
   "source": [
    "We are removing metatriples with SEMMED semantic types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "882e7bde",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:57.364145Z",
     "start_time": "2023-11-01T00:03:57.361050Z"
    }
   },
   "outputs": [],
   "source": [
    "currently_unmapped_types = set(SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'].isna()].Abbrev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f4220655",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:03:58.070864Z",
     "start_time": "2023-11-01T00:03:58.067224Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich',\n",
       " 'sbst'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6eae498",
   "metadata": {},
   "source": [
    "Commenting out: we aren't removing SEMMED semantic types here anymore, because we will do so in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a3a9da81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-08-16T00:18:25.086278Z",
     "start_time": "2023-08-16T00:18:25.083033Z"
    }
   },
   "outputs": [],
   "source": [
    "# currently_unused_entities = [\n",
    "#     ## missing mappings\n",
    "#     \"arch\", \"bdsy\", \"bird\", \"bmod\", \"chvf\", \"chvs\", \"euka\", \"ffas\", \n",
    "#     \"fish\", \"genf\", \"ocdi\", \"rept\", \"sbst\", \n",
    "#     ## old terms here\n",
    "#     \"alga\", \"invt\", \"orgm\", \"rich\",\n",
    "#     ## mapped to Polypeptide\n",
    "#     \"amas\",     \n",
    "#     ## all that are mapped to Activity\n",
    "#     \"acty\", \"dora\", \"edac\", \"gora\", \"hlca\", \"mcha\", \"ocac\", \"resa\",  \n",
    "#     ## all that are mapped to Cohort\n",
    "#     \"aggp\", \"famg\", \"podg\", \"prog\",  \n",
    "#     ## all that are mapped to Invertebrate\n",
    "#     \"amph\",\n",
    "#     ## all that are mapped to OrganismalEntity\n",
    "#     \"anim\", \n",
    "#     ## all that are mapped to Anatomical Entity\n",
    "#     \"anst\", \"bdsu\", \"blor\", \"bsoj\", \n",
    "#     ## mapped to SmallMolecule\n",
    "#     \"elii\",\n",
    "#     ## mapped to Bacterium\n",
    "#     \"bact\",\n",
    "#     ## all that are mapped to Behavior\n",
    "#     \"bhvr\", \"inbe\", \"menp\", \"socb\",\n",
    "#     ## all that are mapped to Phenomenon\n",
    "#     \"biof\", \"eehu\", \"hcpp\", \"lbtr\", \"npop\", \"phpr\",\n",
    "#     ## all that are mapped to Device\n",
    "#     \"bodm\", \"drdd\", \"medd\", \"resd\",\n",
    "#     ## mapped to GrossAnatomicalStructure\n",
    "#     \"emst\", \"tisu\", \n",
    "#     ## mapped to ChemicalEntity\n",
    "#     \"chem\",\n",
    "#     ## all that are mapped to InformationContentEntity\n",
    "#     \"clas\", \"ftcn\", \"idcn\", \"qlco\", \"qnco\", \"rnlw\", \"spco\", \"tmco\",\n",
    "#     ## all that are mapped to ClinicalAttribute\n",
    "#     \"clna\",\n",
    "#     ## mapped to Procedure\n",
    "#     \"lbpr\", \"mbrt\", \n",
    "#     ## mapped to BiologicalEntity\n",
    "#     \"emod\",\n",
    "#     ## mapped to Protein\n",
    "#     \"rcpt\",\n",
    "#     ## all that are mapped to Event\n",
    "#     \"evnt\",\n",
    "#     ## all that are mapped to DiseaseOrPhenotypicFeature\n",
    "#     \"fndg\",\n",
    "#     ## all that are mapped to Fungus\n",
    "#     \"fngs\",\n",
    "#     ## all that are mapped to GeographicLocation\n",
    "#     \"geoa\",\n",
    "#     ## all that are mapped to Agent\n",
    "#     \"grup\", \"hcro\", \"orgt\", \"pros\", \"shro\",\n",
    "#     ## all that are mapped to Human\n",
    "#     \"humn\", \n",
    "#     ## all that are mapped to Publication\n",
    "#     \"inpr\",\n",
    "#     ## all that are mapped to DiagnosticAid\n",
    "#     \"irda\", \n",
    "#     ## all that are mapped to Mammal\n",
    "#     \"mamm\",  \n",
    "#     ## all that are mapped to PhysicalEntity\n",
    "#     \"mnob\",\n",
    "#     ## all that are mapped to MolecularEntity\n",
    "#     \"mosq\",\n",
    "#     ## mapped to NucleicAcidEntity\n",
    "#     \"nusq\",\n",
    "#     ## all that are mapped to OrganismAttribute\n",
    "#     \"orga\",\n",
    "#     ## all that are mapped to Plant\n",
    "#     \"plnt\", \n",
    "#     ## all that are mapped to PopulationOfIndividualOrganisms\n",
    "#     \"popg\",\n",
    "#     ## all that are mapped to Virus\n",
    "#     \"virs\",\n",
    "#     ## all that are mapped to Vertebrate\n",
    "#     \"vtbt\"\n",
    "# ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da7aa1e",
   "metadata": {},
   "source": [
    "#### prune data based on what SEMMED semantic types to remove"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17055733",
   "metadata": {},
   "source": [
    "We are removing SEMMED semantic types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8cf5b450",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:05.105709Z",
     "start_time": "2023-11-01T00:04:05.101157Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich',\n",
       " 'sbst'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types\n",
    "len(currently_unmapped_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "308358e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:07.817182Z",
     "start_time": "2023-11-01T00:04:07.808963Z"
    }
   },
   "outputs": [],
   "source": [
    "## prune this reference table of SEMMED semantic types\n",
    "SEMMED_semtypes = SEMMED_semtypes[ ~ SEMMED_semtypes['Abbrev'].isin(currently_unmapped_types)]\n",
    "\n",
    "## prune the metatriples table\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['SUBJECT_SEMTYPE'].isin(currently_unmapped_types)]\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['OBJECT_SEMTYPE'].isin(currently_unmapped_types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "67861c2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:08.579079Z",
     "start_time": "2023-11-01T00:04:08.569312Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "110"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 127, now 112: decreased by 15\n",
    "len(object_semtypes)  ## was 127, now 110: decreased by 17 (number of unmapped entities)\n",
    "len(predicates)       ## was 28, still 28 (expected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "647e4065",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:11.694103Z",
     "start_time": "2023-11-01T00:04:11.686120Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "77   T087                  Amino Acid Sequence   amas   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "77                Polypeptide  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  \n",
       "\n",
       "[113 rows x 4 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_semtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36f25379",
   "metadata": {},
   "source": [
    "### Custom predicate processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf6f43",
   "metadata": {},
   "source": [
    "#### getting mappings for predicates and qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "52496738",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:15.671193Z",
     "start_time": "2023-11-01T00:04:15.666776Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates = srdef[srdef['Record Type (RT)'] == 'RL'].copy()\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates[['Full Name (STY/RL)'\n",
    "                                      ]].copy()\n",
    "SEMMED_predicates.columns = ['SemmedPred']\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "635827c9",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "The SRDEF file might be missing predicates that are in the data, so check that using the code block below and see if you would like to add those back in..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b05e90f6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:20.359831Z",
     "start_time": "2023-11-01T00:04:20.354968Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['administered_to',\n",
       " 'augments',\n",
       " 'coexists_with',\n",
       " 'converts_to',\n",
       " 'inhibits',\n",
       " 'predisposes',\n",
       " 'same_as',\n",
       " 'stimulates']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## predicates currently in the filtered data\n",
    "\n",
    "missing_preds = list()\n",
    "\n",
    "for i in predicates:\n",
    "    if i.lower() not in set(SEMMED_predicates['SemmedPred']):\n",
    "        missing_preds.append(i.lower())\n",
    "\n",
    "missing_preds.sort()\n",
    "missing_preds\n",
    "\n",
    "# [i for i in predicates if i.lower() in SEMMED_predicates['SemmedName']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e40796c",
   "metadata": {},
   "source": [
    "Currently we add them back in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "6a3d217d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:24.474441Z",
     "start_time": "2023-11-01T00:04:24.463684Z"
    }
   },
   "outputs": [],
   "source": [
    "## manually add some predicates that are in the data, but not in SRDEF...\n",
    "\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['administered_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['augments']\n",
    "## notice that the data doesn't have a hyphen for coexists...even though SRDEF has hyphen for co-occurs_with\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['coexists_with']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['converts_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['inhibits']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['predisposes']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['same_as']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['stimulates']\n",
    "\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ec402da4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:25.280546Z",
     "start_time": "2023-11-01T00:04:25.271368Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>adjacent_to</td>\n",
       "      <td>SEMMEDDB:ADJACENT_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>analyzes</td>\n",
       "      <td>SEMMEDDB:ANALYZES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>assesses_effect_of</td>\n",
       "      <td>SEMMEDDB:ASSESSES_EFFECT_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>temporally_related_to</td>\n",
       "      <td>SEMMEDDB:TEMPORALLY_RELATED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>traverses</td>\n",
       "      <td>SEMMEDDB:TRAVERSES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>tributary_of</td>\n",
       "      <td>SEMMEDDB:TRIBUTARY_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SemmedPred                    Semmed_in_BM\n",
       "162            adjacent_to            SEMMEDDB:ADJACENT_TO\n",
       "54         administered_to        SEMMEDDB:ADMINISTERED_TO\n",
       "145                affects                SEMMEDDB:AFFECTS\n",
       "176               analyzes               SEMMEDDB:ANALYZES\n",
       "158     assesses_effect_of     SEMMEDDB:ASSESSES_EFFECT_OF\n",
       "..                     ...                             ...\n",
       "130  temporally_related_to  SEMMEDDB:TEMPORALLY_RELATED_TO\n",
       "166              traverses              SEMMEDDB:TRAVERSES\n",
       "148                 treats                 SEMMEDDB:TREATS\n",
       "178           tributary_of           SEMMEDDB:TRIBUTARY_OF\n",
       "149                   uses                   SEMMEDDB:USES\n",
       "\n",
       "[62 rows x 2 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## since biolink-model mappings doesn't use hyphens\n",
    "SEMMED_predicates['Semmed_in_BM']= ['SEMMEDDB:'+i \n",
    "                                   for i in SEMMED_predicates['SemmedPred'].str.replace(\"-\", \"\")]\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.upper()\n",
    "\n",
    "## Specifically change same_as to be lower-case for the mapping process\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedPred'] == 'same_as'),'Semmed_in_BM'] = 'SEMMEDDB:same_as'\n",
    "\n",
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "979e0da8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:28.181252Z",
     "start_time": "2023-11-01T00:04:27.964310Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting the mapping\n",
    "SEMMED_predicates['BMPred'] = [bmt_tool.get_element_by_mapping(i)\n",
    "                                       for i in SEMMED_predicates['Semmed_in_BM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5d7f7944",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:04:29.089524Z",
     "start_time": "2023-11-01T00:04:29.075674Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>measurement_of</td>\n",
       "      <td>SEMMEDDB:MEASUREMENT_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO            related to\n",
       "145           affects           SEMMEDDB:AFFECTS               affects\n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related to\n",
       "55           augments          SEMMEDDB:AUGMENTS                  None\n",
       "141            causes            SEMMEDDB:CAUSES                causes\n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists with\n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates\n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives into\n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses\n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts\n",
       "58           inhibits          SEMMEDDB:INHIBITS                  None\n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts with\n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location of\n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation of\n",
       "170    measurement_of    SEMMEDDB:MEASUREMENT_OF                  None\n",
       "156          measures          SEMMEDDB:MEASURES  quantifier qualifier\n",
       "171         method_of         SEMMEDDB:METHOD_OF                  None\n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs in\n",
       "127           part_of           SEMMEDDB:PART_OF               part of\n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes\n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes\n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents\n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs in\n",
       "138          produces          SEMMEDDB:PRODUCES              produces\n",
       "60            same_as           SEMMEDDB:same_as           close match\n",
       "61         stimulates        SEMMEDDB:STIMULATES                  None\n",
       "148            treats            SEMMEDDB:TREATS                treats\n",
       "149              uses              SEMMEDDB:USES             has input"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>measurement_of</td>\n",
       "      <td>SEMMEDDB:MEASUREMENT_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SemmedPred             Semmed_in_BM BMPred\n",
       "55         augments        SEMMEDDB:AUGMENTS   None\n",
       "58         inhibits        SEMMEDDB:INHIBITS   None\n",
       "170  measurement_of  SEMMEDDB:MEASUREMENT_OF   None\n",
       "171       method_of       SEMMEDDB:METHOD_OF   None\n",
       "61       stimulates      SEMMEDDB:STIMULATES   None"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remember there are 28 predicates in the data\n",
    "len(predicates)\n",
    "pred_compare = [i.lower() for i in predicates]\n",
    "\n",
    "SEMMED_predicates[SEMMED_predicates['SemmedPred'].isin(pred_compare)]\n",
    "\n",
    "SEMMED_predicates[(SEMMED_predicates['SemmedPred'].isin(pred_compare)) & \n",
    "                  (SEMMED_predicates['BMPred'].isna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c658e0a2",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Review the mappings in the table printed by the previous code chunk. \n",
    "\n",
    "The mapping for several SEMMEDDB predicates are missing (`BMPred` value is `None`) \n",
    "\n",
    "I'll add the mappings for some predicates below: and will be added below\n",
    "* augments\n",
    "* inhibits\n",
    "* stimulates\n",
    "\n",
    "I'll keep some as missing (mapping doesn't exist in biolink-model):\n",
    "* measurement_of \n",
    "* method_of\n",
    "\n",
    "Notes:\n",
    "* semmeddb `administered_to` is \"trying something out with a goal\", not that it actually does something...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)\n",
    "* semmeddb `diagnoses` is more of a \"this thing distinguishes / identifies this other thing\"...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T163)\n",
    "* semmeddb `process_of` does seem to map well to biolink-model's `occurs in` so...okay. As an example, see the use of `process_of` in the [API's data](https://biothings.ncats.io/semmeddb/query?q=predicate:PROCESS_OF%20AND%20object.semantic_type_abbreviation:cell)\n",
    "* keeping semmeddb `treats` is more general than biolink-model's `treats`. It's more of \"trying this out as a treatment\". see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5040417f",
   "metadata": {},
   "source": [
    "Starting with the biolink 3.1.1 update, we:\n",
    "* map semmeddb `augments` and `stimulates` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: increased\n",
    "* map semmeddb `inhibits` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: decreased\n",
    "\n",
    "And for the reverse of these operations with qualifiers, we'll want to get the inverted qualified_predicates and we'll use \"subject_\" rather than \"object_\" qualifiers (but the values will remain the same). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "77adf097",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:05:35.007229Z",
     "start_time": "2023-11-01T00:05:34.996880Z"
    }
   },
   "outputs": [],
   "source": [
    "## complex change: modify table to use qualifiers for semmeddb augments, stimulates, and inhibits\n",
    "\n",
    "## At the moment, only need the following qualifiers added, to map these predicates\n",
    "##   QualPred = qualified_predicate\n",
    "##   ObjAsp = object_aspect_qualifier\n",
    "##   ObjDirect = object_direction_qualifier\n",
    "##   Inv_QualPred = for reverse operations, inverse of qualified_predicate\n",
    "##   Inv_SubjAsp = for reverse operations, subject_aspect_qualifier\n",
    "##   Inv_SubjDirect = for reverse operations, subject_direction_qualifier\n",
    "\n",
    "\n",
    "## First, add the columns. Most values will stay blank (None)\n",
    "SEMMED_predicates['BMQualPred'] = None  \n",
    "SEMMED_predicates['BMObjAsp'] = None \n",
    "SEMMED_predicates['BMObjDirect'] = None\n",
    "SEMMED_predicates['BM_Inv_QualPred'] = None  \n",
    "SEMMED_predicates['BM_Inv_SubjAsp'] = None \n",
    "SEMMED_predicates['BM_Inv_SubjDirect'] = None\n",
    "\n",
    "## Next, mutate the rows for biolink predicates being remapping to predicates + qualifiers\n",
    "for index in SEMMED_predicates.index:\n",
    "    ## semmeddb augments and stimulates should change\n",
    "    if (SEMMED_predicates.loc[index,'SemmedPred'] == 'augments') or \\\n",
    "    (SEMMED_predicates.loc[index,'SemmedPred'] == 'stimulates'):\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'increased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'increased'\n",
    "    ## semmeddb inhibits should change\n",
    "    elif SEMMED_predicates.loc[index,'SemmedPred'] == 'inhibits':\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'decreased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'decreased'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933327b9",
   "metadata": {},
   "source": [
    "Clean up table to only include what we have biolink-model mappings for: this removes measurement_of and method_of predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "80e84505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:01.446841Z",
     "start_time": "2023-11-01T00:06:01.443348Z"
    }
   },
   "outputs": [],
   "source": [
    "## only keep the stuff we found biolink mappings for\n",
    "SEMMED_predicates = SEMMED_predicates[SEMMED_predicates['BMPred'].notna()].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd31635",
   "metadata": {},
   "source": [
    "Now to get the inverse predicates. With biolink-model-toolkit > 0.9.0 and biolink-model > 3.1.1, I've encountered issues retrieving inverse predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f9acdf7f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:03.418291Z",
     "start_time": "2023-11-01T00:06:03.415731Z"
    }
   },
   "outputs": [],
   "source": [
    "## Testing methods of getting inverse\n",
    "# print(bmt_tool.get_element('causes').inverse)  ## get None\n",
    "# print(bmt_tool.get_inverse('subclass of'))   ## get None\n",
    "# print(bmt_tool.get_inverse('subclass_of'))   ## get AttributeError: 'NoneType' object has no attribute 'name'\n",
    "# print(bmt_tool.get_inverse('quantifier_qualifier'))    ## also get AttributeError\n",
    "\n",
    "# for i in SEMMED_predicates['BMPred']:\n",
    "#     print(i)\n",
    "#     if bmt_tool.get_element(i).symmetric:\n",
    "#         print('is symmetrical')\n",
    "#     elif bmt_tool.get_inverse(i):\n",
    "#         print('inverse: ' + bmt_tool.get_inverse(i))\n",
    "#     else:\n",
    "#         print('missing inverse')\n",
    "#     print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "52cfbbc4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:05.503393Z",
     "start_time": "2023-11-01T00:06:05.497055Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['BM_Inv_Pred'] = [i if bmt_tool.get_element(i).symmetric\n",
    "                                      else bmt_tool.get_inverse(i) if bmt_tool.get_inverse(i)\n",
    "                                      else 'MISSING' \n",
    "                                    for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates.reindex(columns=[\n",
    "    'SemmedPred', 'Semmed_in_BM', 'BMPred', 'BMQualPred', 'BMObjAsp', 'BMObjDirect', \n",
    "    'BM_Inv_Pred', 'BM_Inv_QualPred', 'BM_Inv_SubjAsp', 'BM_Inv_SubjDirect'])\n",
    "\n",
    "## old code that doesn't work anymore, was previously used to get inverses for predicates and qualified predicates\n",
    "## get inverses for basic predicates + qualified_predicates to generate reverse operations\n",
    "# SEMMED_predicates['BM_Inv_Pred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "# SEMMED_predicates['BM_Inv_QualPred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if i and isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMQualPred']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9718fe3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:06.119770Z",
     "start_time": "2023-11-01T00:06:06.108296Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is exacerbated by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is diagnosed by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has predisposing factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is input of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM                BMPred  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO            related to   \n",
       "145           affects           SEMMEDDB:AFFECTS               affects   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH            related to   \n",
       "55           augments          SEMMEDDB:AUGMENTS               affects   \n",
       "141            causes            SEMMEDDB:CAUSES                causes   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH         coexists with   \n",
       "143       complicates       SEMMEDDB:COMPLICATES           exacerbates   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO          derives into   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES             diagnoses   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS              disrupts   \n",
       "58           inhibits          SEMMEDDB:INHIBITS               affects   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH        interacts with   \n",
       "172               isa               SEMMEDDB:ISA           subclass of   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF           location of   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF      manifestation of   \n",
       "156          measures          SEMMEDDB:MEASURES  quantifier qualifier   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN             occurs in   \n",
       "127           part_of           SEMMEDDB:PART_OF               part of   \n",
       "132          precedes          SEMMEDDB:PRECEDES              precedes   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES           predisposes   \n",
       "142          prevents          SEMMEDDB:PREVENTS              prevents   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF             occurs in   \n",
       "138          produces          SEMMEDDB:PRODUCES              produces   \n",
       "60            same_as           SEMMEDDB:same_as           close match   \n",
       "61         stimulates        SEMMEDDB:STIMULATES               affects   \n",
       "148            treats            SEMMEDDB:TREATS                treats   \n",
       "149              uses              SEMMEDDB:USES             has input   \n",
       "\n",
       "    BMQualPred               BMObjAsp BMObjDirect              BM_Inv_Pred  \\\n",
       "54        None                   None        None               related to   \n",
       "145       None                   None        None              affected by   \n",
       "160       None                   None        None               related to   \n",
       "55      causes  activity_or_abundance   increased              affected by   \n",
       "141       None                   None        None                caused by   \n",
       "56        None                   None        None            coexists with   \n",
       "143       None                   None        None        is exacerbated by   \n",
       "57        None                   None        None             derives from   \n",
       "157       None                   None        None          is diagnosed by   \n",
       "140       None                   None        None             disrupted by   \n",
       "58      causes  activity_or_abundance   decreased              affected by   \n",
       "136       None                   None        None           interacts with   \n",
       "172       None                   None        None                  MISSING   \n",
       "129       None                   None        None               located in   \n",
       "144       None                   None        None        has manifestation   \n",
       "156       None                   None        None                  MISSING   \n",
       "146       None                   None        None         contains process   \n",
       "127       None                   None        None                 has part   \n",
       "132       None                   None        None              preceded by   \n",
       "59        None                   None        None  has predisposing factor   \n",
       "142       None                   None        None             prevented by   \n",
       "134       None                   None        None         contains process   \n",
       "138       None                   None        None              produced by   \n",
       "60        None                   None        None              close match   \n",
       "61      causes  activity_or_abundance   increased              affected by   \n",
       "148       None                   None        None               treated by   \n",
       "149       None                   None        None              is input of   \n",
       "\n",
       "    BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54             None                   None              None  \n",
       "145            None                   None              None  \n",
       "160            None                   None              None  \n",
       "55        caused_by  activity_or_abundance         increased  \n",
       "141            None                   None              None  \n",
       "56             None                   None              None  \n",
       "143            None                   None              None  \n",
       "57             None                   None              None  \n",
       "157            None                   None              None  \n",
       "140            None                   None              None  \n",
       "58        caused_by  activity_or_abundance         decreased  \n",
       "136            None                   None              None  \n",
       "172            None                   None              None  \n",
       "129            None                   None              None  \n",
       "144            None                   None              None  \n",
       "156            None                   None              None  \n",
       "146            None                   None              None  \n",
       "127            None                   None              None  \n",
       "132            None                   None              None  \n",
       "59             None                   None              None  \n",
       "142            None                   None              None  \n",
       "134            None                   None              None  \n",
       "138            None                   None              None  \n",
       "60             None                   None              None  \n",
       "61        caused_by  activity_or_abundance         increased  \n",
       "148            None                   None              None  \n",
       "149            None                   None              None  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b121018d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:09.372961Z",
     "start_time": "2023-11-01T00:06:09.363918Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SemmedPred       Semmed_in_BM                BMPred BMQualPred BMObjAsp  \\\n",
       "172        isa       SEMMEDDB:ISA           subclass of       None     None   \n",
       "156   measures  SEMMEDDB:MEASURES  quantifier qualifier       None     None   \n",
       "\n",
       "    BMObjDirect BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "172        None     MISSING            None           None              None  \n",
       "156        None     MISSING            None           None              None  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd4046d",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Check if the predicate inverses look correct. \n",
    "\n",
    "For bmt >=1.1.0 and biolink-model >=3.5.2, we want to:\n",
    "* leave the inverse of `subclass of'` as `MISSING` because no inverse exists, so we'll want to remove it in the next step\n",
    "* leave the inverse of `quanitifer qualifier` as `MISSING` since it's actually not a predicate, so we'll want to remove it in the next step\n",
    "\n",
    "Old notes: \n",
    "* Maybe some aren't correct because they don't have an inverse but they are directional...If so, I suggest going back and changing the predicates to ones with inverses. \n",
    "* Some predicates aren't directional (their entry in the biolink-model yaml will have the property `symmetric == true`) so those are identical in either direction..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46765d17",
   "metadata": {},
   "source": [
    "Now we'll make sure all biolink-model-mapped columns are in snake_case, and clean up the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "85224c41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:14.573869Z",
     "start_time": "2023-11-01T00:06:14.569546Z"
    }
   },
   "outputs": [],
   "source": [
    "## FOR RESOURCES, these predicates must all be in snake_case\n",
    "for col in ['BMPred', 'BMQualPred', 'BM_Inv_Pred', 'BM_Inv_QualPred']:\n",
    "    SEMMED_predicates[col] = SEMMED_predicates[col].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "50027e5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:15.169135Z",
     "start_time": "2023-11-01T00:06:15.163453Z"
    }
   },
   "outputs": [],
   "source": [
    "## can mutate and rename the Semmed_in_BM column (not needed anymore) with the format needed\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['SemmedPred'].str.upper()\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.replace(\"-\", \"\")\n",
    "SEMMED_predicates.rename(columns = {'Semmed_in_BM':'Semmed_in_Data'}, inplace = True)\n",
    "\n",
    "## special handling keep lowercase for same_as\n",
    "SEMMED_predicates['Semmed_in_Data'] = [i.lower() if i == 'SAME_AS' else i\n",
    "                                       for i in SEMMED_predicates['Semmed_in_Data']]\n",
    "\n",
    "## Can remove the SemmedPred column: not needed anymore\n",
    "SEMMED_predicates.drop(columns='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "946dd392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:15.785132Z",
     "start_time": "2023-11-01T00:06:15.774279Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data                BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO            related_to       None                   None   \n",
       "145           AFFECTS               affects       None                   None   \n",
       "160   ASSOCIATED_WITH            related_to       None                   None   \n",
       "55           AUGMENTS               affects     causes  activity_or_abundance   \n",
       "141            CAUSES                causes       None                   None   \n",
       "56      COEXISTS_WITH         coexists_with       None                   None   \n",
       "143       COMPLICATES           exacerbates       None                   None   \n",
       "57        CONVERTS_TO          derives_into       None                   None   \n",
       "157         DIAGNOSES             diagnoses       None                   None   \n",
       "140          DISRUPTS              disrupts       None                   None   \n",
       "58           INHIBITS               affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH        interacts_with       None                   None   \n",
       "172               ISA           subclass_of       None                   None   \n",
       "129       LOCATION_OF           location_of       None                   None   \n",
       "144  MANIFESTATION_OF      manifestation_of       None                   None   \n",
       "156          MEASURES  quantifier_qualifier       None                   None   \n",
       "146         OCCURS_IN             occurs_in       None                   None   \n",
       "127           PART_OF               part_of       None                   None   \n",
       "132          PRECEDES              precedes       None                   None   \n",
       "59        PREDISPOSES           predisposes       None                   None   \n",
       "142          PREVENTS              prevents       None                   None   \n",
       "134        PROCESS_OF             occurs_in       None                   None   \n",
       "138          PRODUCES              produces       None                   None   \n",
       "60            same_as           close_match       None                   None   \n",
       "61         STIMULATES               affects     causes  activity_or_abundance   \n",
       "148            TREATS                treats       None                   None   \n",
       "149              USES             has_input       None                   None   \n",
       "\n",
       "    BMObjDirect              BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54         None               related_to            None   \n",
       "145        None              affected_by            None   \n",
       "160        None               related_to            None   \n",
       "55    increased              affected_by       caused_by   \n",
       "141        None                caused_by            None   \n",
       "56         None            coexists_with            None   \n",
       "143        None        is_exacerbated_by            None   \n",
       "57         None             derives_from            None   \n",
       "157        None          is_diagnosed_by            None   \n",
       "140        None             disrupted_by            None   \n",
       "58    decreased              affected_by       caused_by   \n",
       "136        None           interacts_with            None   \n",
       "172        None                  MISSING            None   \n",
       "129        None               located_in            None   \n",
       "144        None        has_manifestation            None   \n",
       "156        None                  MISSING            None   \n",
       "146        None         contains_process            None   \n",
       "127        None                 has_part            None   \n",
       "132        None              preceded_by            None   \n",
       "59         None  has_predisposing_factor            None   \n",
       "142        None             prevented_by            None   \n",
       "134        None         contains_process            None   \n",
       "138        None              produced_by            None   \n",
       "60         None              close_match            None   \n",
       "61    increased              affected_by       caused_by   \n",
       "148        None               treated_by            None   \n",
       "149        None              is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "172                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "156                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c616c4ff",
   "metadata": {},
   "source": [
    "#### custom-prune the data of some predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6f0a864",
   "metadata": {},
   "source": [
    "We are still doing this because some semmeddb predicates or their mappings are problematic. \n",
    "\n",
    "As of 2023-08-07: the Translator exclusions weren't on a predicate level. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "9079161f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:30.230522Z",
     "start_time": "2023-11-01T00:06:30.224218Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in data but not in mapping table\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'MEASUREMENT_OF', 'METHOD_OF'}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in mapping file but not in data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ISA'}"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in data but not in mapping table\")\n",
    "predicates - set(SEMMED_predicates['Semmed_in_Data'])\n",
    "\n",
    "print(\"in mapping file but not in data\")\n",
    "set(SEMMED_predicates['Semmed_in_Data']) - predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ca959b8c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:31.593504Z",
     "start_time": "2023-11-01T00:06:31.584658Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Semmed_in_Data                BMPred BMQualPred BMObjAsp BMObjDirect  \\\n",
       "172            ISA           subclass_of       None     None        None   \n",
       "156       MEASURES  quantifier_qualifier       None     None        None   \n",
       "\n",
       "    BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "172     MISSING            None           None              None  \n",
       "156     MISSING            None           None              None  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8b1e2b",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Remove predicates in the mapping table, but we don't want to use for associations:\n",
    "- no biolink-model mapping: METHOD_OF, MEASUREMENT_OF\n",
    "- no biolink-model inverse (`MISSING`) and other issues:\n",
    "  - ISA: already removed from data during Basic Filtering section 2.3 (predicates)\n",
    "  - MEASURES: mapping isn't to a predicate (quantifier_qualifier)\n",
    "\n",
    "Past: used to exclude USES (not helpful?), but now keeping because the predicate has been reviewed in the Translator-curated exclusions effort. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5e25a929",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:40.160883Z",
     "start_time": "2023-11-01T00:06:40.157105Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ISA', 'MEASUREMENT_OF', 'MEASURES', 'METHOD_OF'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_removals = {'MEASUREMENT_OF', 'METHOD_OF', 'ISA', 'MEASURES'}\n",
    "more_removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5a46158c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:41.147796Z",
     "start_time": "2023-11-01T00:06:41.135054Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_predisposing_factor</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>prevents</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data            BMPred BMQualPred               BMObjAsp  \\\n",
       "54    ADMINISTERED_TO        related_to       None                   None   \n",
       "145           AFFECTS           affects       None                   None   \n",
       "160   ASSOCIATED_WITH        related_to       None                   None   \n",
       "55           AUGMENTS           affects     causes  activity_or_abundance   \n",
       "141            CAUSES            causes       None                   None   \n",
       "56      COEXISTS_WITH     coexists_with       None                   None   \n",
       "143       COMPLICATES       exacerbates       None                   None   \n",
       "57        CONVERTS_TO      derives_into       None                   None   \n",
       "157         DIAGNOSES         diagnoses       None                   None   \n",
       "140          DISRUPTS          disrupts       None                   None   \n",
       "58           INHIBITS           affects     causes  activity_or_abundance   \n",
       "136    INTERACTS_WITH    interacts_with       None                   None   \n",
       "129       LOCATION_OF       location_of       None                   None   \n",
       "144  MANIFESTATION_OF  manifestation_of       None                   None   \n",
       "146         OCCURS_IN         occurs_in       None                   None   \n",
       "127           PART_OF           part_of       None                   None   \n",
       "132          PRECEDES          precedes       None                   None   \n",
       "59        PREDISPOSES       predisposes       None                   None   \n",
       "142          PREVENTS          prevents       None                   None   \n",
       "134        PROCESS_OF         occurs_in       None                   None   \n",
       "138          PRODUCES          produces       None                   None   \n",
       "60            same_as       close_match       None                   None   \n",
       "61         STIMULATES           affects     causes  activity_or_abundance   \n",
       "148            TREATS            treats       None                   None   \n",
       "149              USES         has_input       None                   None   \n",
       "\n",
       "    BMObjDirect              BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54         None               related_to            None   \n",
       "145        None              affected_by            None   \n",
       "160        None               related_to            None   \n",
       "55    increased              affected_by       caused_by   \n",
       "141        None                caused_by            None   \n",
       "56         None            coexists_with            None   \n",
       "143        None        is_exacerbated_by            None   \n",
       "57         None             derives_from            None   \n",
       "157        None          is_diagnosed_by            None   \n",
       "140        None             disrupted_by            None   \n",
       "58    decreased              affected_by       caused_by   \n",
       "136        None           interacts_with            None   \n",
       "129        None               located_in            None   \n",
       "144        None        has_manifestation            None   \n",
       "146        None         contains_process            None   \n",
       "127        None                 has_part            None   \n",
       "132        None              preceded_by            None   \n",
       "59         None  has_predisposing_factor            None   \n",
       "142        None             prevented_by            None   \n",
       "134        None         contains_process            None   \n",
       "138        None              produced_by            None   \n",
       "60         None              close_match            None   \n",
       "61    increased              affected_by       caused_by   \n",
       "148        None               treated_by            None   \n",
       "149        None              is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(25, 9)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remove this set from the SEMMED_predicates table\n",
    "SEMMED_predicates = SEMMED_predicates[~ SEMMED_predicates['Semmed_in_Data'].isin(more_removals)]\n",
    "SEMMED_predicates\n",
    "SEMMED_predicates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "64cacd52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:43.956353Z",
     "start_time": "2023-11-01T00:06:43.951766Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove this set from the metatriples table\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(more_removals)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "530e3631",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:44.562797Z",
     "start_time": "2023-11-01T00:06:44.553765Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "110"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 112, still 112 (expected)\n",
    "len(object_semtypes)  ## was 110, still 110 (expected)\n",
    "len(predicates)       ## was 28, now 25: decreased by 3 (expected: ISA was already removed earlier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "863543b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:46.403124Z",
     "start_time": "2023-11-01T00:06:46.393943Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15764, 7)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2344345</td>\n",
       "      <td>2147949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1183119</td>\n",
       "      <td>1133560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1159426</td>\n",
       "      <td>1011905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1088609</td>\n",
       "      <td>998529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>876862</td>\n",
       "      <td>804853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>846743</td>\n",
       "      <td>770736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>934027</td>\n",
       "      <td>714885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>778276</td>\n",
       "      <td>678610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>715266</td>\n",
       "      <td>642000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>747981</td>\n",
       "      <td>620129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            bpoc      PART_OF          umls           mamm   \n",
       "2           umls            fndg   PROCESS_OF          umls           humn   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "5           umls            bpoc  LOCATION_OF          umls           patf   \n",
       "6           umls            topp       TREATS          umls           dsyn   \n",
       "7           umls            topp       TREATS          umls           neop   \n",
       "8           umls            dsyn   PROCESS_OF          umls           mamm   \n",
       "9           umls            bpoc  LOCATION_OF          umls           dsyn   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2344345      2147949  \n",
       "1  1183119      1133560  \n",
       "2  1159426      1011905  \n",
       "3  1088609       998529  \n",
       "4   876862       804853  \n",
       "5   846743       770736  \n",
       "6   934027       714885  \n",
       "7   778276       678610  \n",
       "8   715266       642000  \n",
       "9   747981       620129  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at number of metatriples after this removal\n",
    "imported_metatriples.shape\n",
    "imported_metatriples.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426473de",
   "metadata": {},
   "source": [
    "### Use translator-curated exclusions to prune metatriple tables "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e732098e",
   "metadata": {},
   "source": [
    "I'll use Translator-curated exclusions to remove some SEMMED semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n",
    "\n",
    "Translator-curated exclusions work on the metatriple-level. \n",
    "\n",
    "Note: the `_t_code` columns sometimes have the value of `not_found`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37afbf97",
   "metadata": {},
   "source": [
    "#### Checking and adjusting exclusions table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5f404ea6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:50.839721Z",
     "start_time": "2023-11-01T00:06:50.829804Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "0                chem                  T103              NaN   \n",
       "1                 NaN                   NaN              NaN   \n",
       "2                chvs                  T104              NaN   \n",
       "3                 NaN                   NaN              NaN   \n",
       "4                chvf                  T120              NaN   \n",
       "\n",
       "  semmed_object_code semmed_object_t_code           exclusion_type  \n",
       "0                NaN                  NaN  semantic type exclusion  \n",
       "1               chem                 T103  semantic type exclusion  \n",
       "2                NaN                  NaN  semantic type exclusion  \n",
       "3               chvs                 T104  semantic type exclusion  \n",
       "4                NaN                  NaN  semantic type exclusion  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(1443, 6)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['semantic type exclusion', 'Domain exclusion', 'Range exclusion'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions.head()\n",
    "imported_exclusions.shape\n",
    "imported_exclusions.exclusion_type.unique()  ## number of unique kinds of exclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99185fc",
   "metadata": {},
   "source": [
    "First, I'm checking that the data fits the \"rules\" I expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "efd0d556",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:55.186295Z",
     "start_time": "2023-11-01T00:06:55.179388Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    ## all 'semantic type exclusions' don't consider the predicates, \n",
    "    ##   just specific entities/instances of the types\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_predicate, str):\n",
    "            print('semantic type exclusion problem: predicate there?')\n",
    "            print(row)\n",
    "    ## 'Domain exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        if (not isinstance(row.semmed_subject_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('domain exclusion problem: missing subject or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_object_code, str):\n",
    "            print('domain exclusion problem: object present')\n",
    "            print(row)\n",
    "    ## 'Range exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        if (not isinstance(row.semmed_object_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('range exclusion problem: missing object or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_subject_code, str):\n",
    "            print('range exclusion problem: subject present')\n",
    "            print(row)\n",
    "    else: \n",
    "        print('some other kind of exclusion?')\n",
    "        print(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a6bb679",
   "metadata": {},
   "source": [
    "Then, I noticed the same_as predicate was not in the expected case, so adjusting this..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "32c37c3b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:06:58.848387Z",
     "start_time": "2023-11-01T00:06:58.835669Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          SAME_AS   \n",
       "716                 chvs                  T104          SAME_AS   \n",
       "717                 resa                  T062          SAME_AS   \n",
       "718                 chem                  T103          SAME_AS   \n",
       "719                 qnco                  T081          SAME_AS   \n",
       "720                 ftcn                  T169          SAME_AS   \n",
       "721                 inpr                  T170          SAME_AS   \n",
       "722                 tmco                  T079          SAME_AS   \n",
       "1387                 NaN                   NaN          SAME_AS   \n",
       "1388                 NaN                   NaN          SAME_AS   \n",
       "1389                 NaN                   NaN          SAME_AS   \n",
       "1390                 NaN                   NaN          SAME_AS   \n",
       "1391                 NaN                   NaN          SAME_AS   \n",
       "1392                 NaN                   NaN          SAME_AS   \n",
       "1393                 NaN                   NaN          SAME_AS   \n",
       "1394                 NaN                   NaN          SAME_AS   \n",
       "1395                 NaN                   NaN          SAME_AS   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## viewing\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "89bda465",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:02.274322Z",
     "start_time": "2023-11-01T00:07:02.271096Z"
    }
   },
   "outputs": [],
   "source": [
    "## fixing\n",
    "imported_exclusions['semmed_predicate'] = ['same_as' if i == 'SAME_AS' else i \\\n",
    "                                           for i in imported_exclusions['semmed_predicate']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1bb3d990",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:03.428725Z",
     "start_time": "2023-11-01T00:07:03.415993Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          same_as   \n",
       "716                 chvs                  T104          same_as   \n",
       "717                 resa                  T062          same_as   \n",
       "718                 chem                  T103          same_as   \n",
       "719                 qnco                  T081          same_as   \n",
       "720                 ftcn                  T169          same_as   \n",
       "721                 inpr                  T170          same_as   \n",
       "722                 tmco                  T079          same_as   \n",
       "1387                 NaN                   NaN          same_as   \n",
       "1388                 NaN                   NaN          same_as   \n",
       "1389                 NaN                   NaN          same_as   \n",
       "1390                 NaN                   NaN          same_as   \n",
       "1391                 NaN                   NaN          same_as   \n",
       "1392                 NaN                   NaN          same_as   \n",
       "1393                 NaN                   NaN          same_as   \n",
       "1394                 NaN                   NaN          same_as   \n",
       "1395                 NaN                   NaN          same_as   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## checking\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37364f4a",
   "metadata": {},
   "source": [
    "#### Pruning using exclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e6ba83",
   "metadata": {},
   "source": [
    "Now I can iterate through the `imported_exclusions` table and remove metatriples that match the exclusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e5bb1ec1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:11.763156Z",
     "start_time": "2023-11-01T00:07:09.339272Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_subject_code, str):  ## exclusion is in the subject\n",
    "            imported_metatriples = imported_metatriples[ ~ (imported_metatriples['SUBJECT_SEMTYPE'] == row.semmed_subject_code)]\n",
    "        else:  ## exclusion should be in object\n",
    "            imported_metatriples = imported_metatriples[ ~ (imported_metatriples['OBJECT_SEMTYPE'] == row.semmed_object_code)]\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        imported_metatriples = imported_metatriples[~ ( (imported_metatriples['SUBJECT_SEMTYPE'] == row.semmed_subject_code ) & \n",
    "                                                        (imported_metatriples['PREDICATE'] == row.semmed_predicate ))]\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        imported_metatriples = imported_metatriples[~ ( (imported_metatriples['OBJECT_SEMTYPE'] == row.semmed_object_code ) & \n",
    "                                                        (imported_metatriples['PREDICATE'] == row.semmed_predicate ))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "bf7a6309",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:14.100476Z",
     "start_time": "2023-11-01T00:07:14.091417Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11971, 7)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1088609</td>\n",
       "      <td>998529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>876862</td>\n",
       "      <td>804853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>846743</td>\n",
       "      <td>770736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>934027</td>\n",
       "      <td>714885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>778276</td>\n",
       "      <td>678610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>747981</td>\n",
       "      <td>620129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>739589</td>\n",
       "      <td>618156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>635370</td>\n",
       "      <td>608193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>878129</td>\n",
       "      <td>601951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>umls</td>\n",
       "      <td>phsu</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>723059</td>\n",
       "      <td>502492</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "3            umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4            umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "5            umls            bpoc  LOCATION_OF          umls           patf   \n",
       "6            umls            topp       TREATS          umls           dsyn   \n",
       "7            umls            topp       TREATS          umls           neop   \n",
       "9            umls            bpoc  LOCATION_OF          umls           dsyn   \n",
       "10           umls            bpoc  LOCATION_OF          umls           fndg   \n",
       "11           umls            cell      PART_OF          umls           bpoc   \n",
       "12           umls            bpoc  LOCATION_OF          umls           aapp   \n",
       "13           umls            phsu       TREATS          umls           dsyn   \n",
       "\n",
       "      DOC_N  DOC_N_VALID  \n",
       "3   1088609       998529  \n",
       "4    876862       804853  \n",
       "5    846743       770736  \n",
       "6    934027       714885  \n",
       "7    778276       678610  \n",
       "9    747981       620129  \n",
       "10   739589       618156  \n",
       "11   635370       608193  \n",
       "12   878129       601951  \n",
       "13   723059       502492  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape\n",
    "imported_metatriples.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "83675d58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:21.143780Z",
     "start_time": "2023-11-01T00:07:21.133512Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4595</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>697</td>\n",
       "      <td>317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11085</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>acab</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11623</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>cgab</td>\n",
       "      <td>14</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11964</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13245</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>mobd</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13256</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18245</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>inpo</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18247</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>anab</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SUBJECT_PREFIX SUBJECT_SEMTYPE PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "4595            umls            vita    TREATS          umls           dsyn   \n",
       "11085           umls            vita    TREATS          umls           acab   \n",
       "11623           umls            vita    TREATS          umls           cgab   \n",
       "11964           umls            vita    TREATS          umls           neop   \n",
       "13245           umls            vita    TREATS          umls           mobd   \n",
       "13256           umls            vita    TREATS          umls           patf   \n",
       "18245           umls            vita    TREATS          umls           inpo   \n",
       "18247           umls            vita    TREATS          umls           anab   \n",
       "\n",
       "       DOC_N  DOC_N_VALID  \n",
       "4595     697          317  \n",
       "11085     27           14  \n",
       "11623     14           11  \n",
       "11964     51           10  \n",
       "13245     23            6  \n",
       "13256     18            6  \n",
       "18245      6            0  \n",
       "18247      2            0  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[(imported_metatriples['SUBJECT_SEMTYPE'] == 'vita') &\n",
    "                      (imported_metatriples['PREDICATE'] == 'TREATS')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e9660d",
   "metadata": {},
   "source": [
    "## Final Filter: API-doc-counts per metatriple"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936210c1",
   "metadata": {},
   "source": [
    "### Exploring the count data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305f3a3a",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "use code blocks below to explore how many API docs there are per metatriple and decide what to use for a cutoff\n",
    "\n",
    "---\n",
    "\n",
    "Current decision: > 20 matching API docs, where each doc has pmid_count > 3 (will pass that filter)\n",
    "\n",
    "Notes:\n",
    "* Previously: 4974 metatriples are used to make operations. The cutoff is > 100 matching predications (not API-records/documents).\n",
    "* There are 7605 metatriples that have >=1 matching doc with pmid_count > 3. This set still has a VERY left-skewed distribution. \n",
    "* Our previous rule-of-thumb was creating operations when the number of matching API-docs > 10. This corresponds to 5582 metatriples to make operations from (+608 or +12%). Is this too much?\n",
    "* 20 is comparable to what we had before -> 5054 metatriples (+80 or +2%)\n",
    "* Maybe 100 is too high? -> 3655 metatriples (-1319 or -27%)\n",
    "* Andrew: My gut says that the value of semmeddb is overwhelmingly in the most common metatriples, so at any of these thresholds, I doubt we'll be losing much.  Having said that, as long as being inclusive doesn't result in a performance hit, I'm fine with being inclusive.  So for the moment, let's just shoot for the same basic \"size\" and use a threshold of 20, at least until we have the time to investigate more...\n",
    "* My reply: I agree that covering the most common metatriples is most important. I guess I'm thinking of the June Translator Relay, where people seemed very hesitant about \"removing data\" - which is what we do when we don't make operations for the rarer metatriples. However, I'm not aware of any feedback that BTE is missing expected semmeddb data because of this kind of cutoff....so yep, I agree with you. Going to implement a threshold of > 20 for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b8f25836",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:28.244196Z",
     "start_time": "2023-11-01T00:07:28.233522Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.197100e+04\n",
       "mean     4.683085e+03\n",
       "std      3.345748e+04\n",
       "min      1.000000e+00\n",
       "25%      4.000000e+00\n",
       "50%      4.100000e+01\n",
       "75%      6.645000e+02\n",
       "max      1.088609e+06\n",
       "Name: DOC_N, dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "count     11971.000000\n",
       "mean       2921.605964\n",
       "std       26229.351518\n",
       "min           0.000000\n",
       "25%           0.000000\n",
       "50%           7.000000\n",
       "75%         209.000000\n",
       "max      998529.000000\n",
       "Name: DOC_N_VALID, dtype: float64"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking at quick distribution of API doc counts\n",
    "imported_metatriples['DOC_N'].describe()        ## this is total number of docs\n",
    "imported_metatriples['DOC_N_VALID'].describe()  ## this is the total number of docs w/ pmid_count > 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b710b48",
   "metadata": {},
   "source": [
    "So...This output shows that LOTS of metatriples have 0 docs that will pass the current pmid_count filter (> 3). So let's look at the metatriples with documents that will pass the pmid_count filter (`DOC_N_VALID` > 1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "3d0ed289",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:30.677057Z",
     "start_time": "2023-11-01T00:07:30.670084Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count      7605.000000\n",
       "mean       4598.888231\n",
       "std       32791.494191\n",
       "min           1.000000\n",
       "25%           9.000000\n",
       "50%          85.000000\n",
       "75%         784.000000\n",
       "max      998529.000000\n",
       "Name: DOC_N_VALID, dtype: float64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 0]['DOC_N_VALID'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581add5d",
   "metadata": {},
   "source": [
    "So it's still very left-skewed, and there's lots of operations. Let's look at the distribution of the low end of the range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ddd78804",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:33.779523Z",
     "start_time": "2023-11-01T00:07:33.114980Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: ylabel='Frequency'>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAGdCAYAAAD0e7I1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAApSklEQVR4nO3df3DU9Z3H8deaX4ZcsiWJZN0SMJ6xVhI9DS0noqBAHH6Iws0hokCFm8HyQ1KICOVuij2bIIwBbQ4UjwGUw9h2wPPOigShaSnjGQLIDzvCaQQiSVNtbpNASELyuT8cvuOSIMlmw24+eT5mvjPs9/vZ777fSc2++vl+vrsuY4wRAACApa4JdQEAAADdibADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALBaZKgLCAetra06c+aM4uPj5XK5Ql0OAADoAGOM6urq5PV6dc01l5+/IexIOnPmjFJTU0NdBgAACMDp06fVv3//yx4n7EiKj4+X9PUPKyEhIcTVAACAjqitrVVqaqrzPn45hB3JuXSVkJBA2AEAoIe50hIUFigDAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWC0y1AX0Jjcsecf59+crxoWwEgAAeg9mdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGohDTvLly+Xy+Xy2zwej3PcGKPly5fL6/UqNjZWI0aM0LFjx/zO0djYqPnz5ys5OVlxcXGaMGGCKioqrnYrAAAgTIV8ZmfQoEGqrKx0tiNHjjjHVq5cqYKCAhUWFqq0tFQej0ejR49WXV2dMyYnJ0fbt29XUVGR9u7dq/r6eo0fP14tLS2haAcAAISZyJAXEBnpN5tzkTFGa9as0bJlyzRp0iRJ0ubNm5WSkqKtW7dq9uzZ8vl82rBhg15//XWNGjVKkrRlyxalpqZq165deuCBB65qLwAAIPyEfGbnxIkT8nq9SktL05QpU/TZZ59JksrLy1VVVaXs7GxnbExMjIYPH659+/ZJksrKytTc3Ow3xuv1KiMjwxnTnsbGRtXW1vptAADATiENO0OGDNFrr72m9957T6+++qqqqqo0dOhQffXVV6qqqpIkpaSk+D0nJSXFOVZVVaXo6Gj17dv3smPak5+fL7fb7WypqalB7gwAAISLkIadMWPG6B/+4R+UmZmpUaNG6Z133pH09eWqi1wul99zjDFt9l3qSmOWLl0qn8/nbKdPn+5CFwAAIJyF/DLWN8XFxSkzM1MnTpxw1vFcOkNTXV3tzPZ4PB41NTWppqbmsmPaExMTo4SEBL8NAADYKazCTmNjo/70pz/p+uuvV1pamjwej4qLi53jTU1NKikp0dChQyVJWVlZioqK8htTWVmpo0ePOmMAAEDvFtK7sXJzc/Xggw9qwIABqq6u1nPPPafa2lrNmDFDLpdLOTk5ysvLU3p6utLT05WXl6c+ffpo6tSpkiS3261Zs2Zp0aJFSkpKUmJionJzc53LYgAAACENOxUVFXr00Uf15Zdf6rrrrtPf//3f64MPPtDAgQMlSYsXL1ZDQ4PmzJmjmpoaDRkyRDt37lR8fLxzjtWrVysyMlKTJ09WQ0ODRo4cqU2bNikiIiJUbQEAgDDiMsaYUBcRarW1tXK73fL5fN26fueGJe84//58xbhuex0AAHqDjr5/h9WaHQAAgGAj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYLWwCTv5+flyuVzKyclx9hljtHz5cnm9XsXGxmrEiBE6duyY3/MaGxs1f/58JScnKy4uThMmTFBFRcVVrh4AAISrsAg7paWlWr9+vW677Ta//StXrlRBQYEKCwtVWloqj8ej0aNHq66uzhmTk5Oj7du3q6ioSHv37lV9fb3Gjx+vlpaWq90GAAAIQyEPO/X19Xrsscf06quvqm/fvs5+Y4zWrFmjZcuWadKkScrIyNDmzZt17tw5bd26VZLk8/m0YcMGvfDCCxo1apTuuOMObdmyRUeOHNGuXbtC1RIAAAgjIQ87c+fO1bhx4zRq1Ci//eXl5aqqqlJ2drazLyYmRsOHD9e+ffskSWVlZWpubvYb4/V6lZGR4YxpT2Njo2pra/02AABgp8hQvnhRUZEOHDig0tLSNseqqqokSSkpKX77U1JSdPLkSWdMdHS034zQxTEXn9+e/Px8Pfvss10tHwAA9AAhm9k5ffq0FixYoC1btujaa6+97DiXy+X32BjTZt+lrjRm6dKl8vl8znb69OnOFQ8AAHqMkIWdsrIyVVdXKysrS5GRkYqMjFRJSYleeuklRUZGOjM6l87QVFdXO8c8Ho+amppUU1Nz2THtiYmJUUJCgt8GAADsFLKwM3LkSB05ckSHDh1ytsGDB+uxxx7ToUOHdOONN8rj8ai4uNh5TlNTk0pKSjR06FBJUlZWlqKiovzGVFZW6ujRo84YAADQu4VszU58fLwyMjL89sXFxSkpKcnZn5OTo7y8PKWnpys9PV15eXnq06ePpk6dKklyu92aNWuWFi1apKSkJCUmJio3N1eZmZltFjwDAIDeKaQLlK9k8eLFamho0Jw5c1RTU6MhQ4Zo586dio+Pd8asXr1akZGRmjx5shoaGjRy5Eht2rRJERERIawcAACEC5cxxoS6iFCrra2V2+2Wz+fr1vU7Nyx5x/n35yvGddvrAADQG3T0/Tvkn7MDAADQnQg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVAgo75eXlwa4DAACgWwQUdm666Sbdd9992rJli86fPx/smgAAAIImoLDz0Ucf6Y477tCiRYvk8Xg0e/Zsffjhh8GuDQAAoMsCCjsZGRkqKCjQF198oY0bN6qqqkrDhg3ToEGDVFBQoL/85S/BrhMAACAgXVqgHBkZqYkTJ+pXv/qVnn/+eX366afKzc1V//79NX36dFVWVgarTgAAgIB0Kezs379fc+bM0fXXX6+CggLl5ubq008/1e7du/XFF1/ooYceCladAAAAAYkM5EkFBQXauHGjPvnkE40dO1avvfaaxo4dq2uu+To7paWl6ZVXXtEtt9wS1GIBAAA6K6Cws27dOs2cOVNPPPGEPB5Pu2MGDBigDRs2dKk4AACArgoo7Jw4ceKKY6KjozVjxoxATg8AABA0Aa3Z2bhxo37961+32f/rX/9amzdv7nJRAAAAwRJQ2FmxYoWSk5Pb7O/Xr5/y8vK6XBQAAECwBBR2Tp48qbS0tDb7Bw4cqFOnTnW5KAAAgGAJKOz069dPhw8fbrP/o48+UlJSUpeLAgAACJaAws6UKVP01FNPac+ePWppaVFLS4t2796tBQsWaMqUKcGuEQAAIGAB3Y313HPP6eTJkxo5cqQiI78+RWtrq6ZPn86aHQAAEFYCCjvR0dF688039a//+q/66KOPFBsbq8zMTA0cODDY9QEAAHRJQGHnoptvvlk333xzsGoBAAAIuoDCTktLizZt2qT3339f1dXVam1t9Tu+e/fuoBQHAADQVQGFnQULFmjTpk0aN26cMjIy5HK5gl0XAABAUAQUdoqKivSrX/1KY8eODXY9AAAAQRXQrefR0dG66aabuvzi69at02233aaEhAQlJCTorrvu0rvvvuscN8Zo+fLl8nq9io2N1YgRI3Ts2DG/czQ2Nmr+/PlKTk5WXFycJkyYoIqKii7XBgAA7BBQ2Fm0aJFefPFFGWO69OL9+/fXihUrtH//fu3fv1/333+/HnroISfQrFy5UgUFBSosLFRpaak8Ho9Gjx6turo65xw5OTnavn27ioqKtHfvXtXX12v8+PFqaWnpUm0AAMAOLhNAYpk4caL27NmjxMREDRo0SFFRUX7Ht23bFnBBiYmJWrVqlWbOnCmv16ucnBw988wzkr6exUlJSdHzzz+v2bNny+fz6brrrtPrr7+uRx55RJJ05swZpaam6re//a0eeOCBDr1mbW2t3G63fD6fEhISAq79Sm5Y8o7z789XjOu21wEAoDfo6Pt3QDM73/nOdzRx4kQNHz5cycnJcrvdflsgWlpaVFRUpLNnz+quu+5SeXm5qqqqlJ2d7YyJiYnR8OHDtW/fPklSWVmZmpub/cZ4vV5lZGQ4YwAAQO8W0ALljRs3Bq2AI0eO6K677tL58+f1N3/zN9q+fbtuvfVWJ6ykpKT4jU9JSdHJkyclSVVVVYqOjlbfvn3bjKmqqrrsazY2NqqxsdF5XFtbG6x2AABAmAloZkeSLly4oF27dumVV15x1tCcOXNG9fX1nTrP9773PR06dEgffPCBfvzjH2vGjBn6+OOPneOX3tZujLnire5XGpOfn+83E5WamtqpmgEAQM8RUNg5efKkMjMz9dBDD2nu3Ln6y1/+IunrBcW5ubmdOtfFO7sGDx6s/Px83X777XrxxRfl8Xgkqc0MTXV1tTPb4/F41NTUpJqamsuOac/SpUvl8/mc7fTp052qGQAA9BwBhZ0FCxZo8ODBqqmpUWxsrLN/4sSJev/997tUkDFGjY2NSktLk8fjUXFxsXOsqalJJSUlGjp0qCQpKytLUVFRfmMqKyt19OhRZ0x7YmJinNvdL24AAMBOAa3Z2bt3r/74xz8qOjrab//AgQP1xRdfdPg8P/3pTzVmzBilpqaqrq5ORUVF+t3vfqcdO3bI5XIpJydHeXl5Sk9PV3p6uvLy8tSnTx9NnTpVkuR2uzVr1iwtWrRISUlJSkxMVG5urjIzMzVq1KhAWgMAAJYJKOy0tra2+zk2FRUVio+P7/B5/vznP2vatGmqrKyU2+3Wbbfdph07dmj06NGSpMWLF6uhoUFz5sxRTU2NhgwZop07d/q9xurVqxUZGanJkyeroaFBI0eO1KZNmxQRERFIawAAwDIBfc7OI488IrfbrfXr1ys+Pl6HDx/Wddddp4ceekgDBgwI6t1aVwOfswMAQM/T0ffvgGZ2Vq9erfvuu0+33nqrzp8/r6lTp+rEiRNKTk7WG2+8EXDRAAAAwRZQ2PF6vTp06JDeeOMNHThwQK2trZo1a5Yee+wxvwXLAAAAoRZQ2JGk2NhYzZw5UzNnzgxmPQAAAEEVUNh57bXXvvX49OnTAyoGAAAg2AIKOwsWLPB73NzcrHPnzik6Olp9+vQh7AAAgLAR0IcK1tTU+G319fX65JNPNGzYMBYoAwCAsBLwd2NdKj09XStWrGgz6wMAABBKQQs7khQREaEzZ84E85QAAABdEtCanbffftvvsTFGlZWVKiws1N133x2UwgAAAIIhoLDz8MMP+z12uVy67rrrdP/99+uFF14IRl29Fp+yDABAcAX83VgAAAA9QVDX7AAAAISbgGZ2Fi5c2OGxBQUFgbwEAABAUAQUdg4ePKgDBw7owoUL+t73vidJOn78uCIiInTnnXc641wuV3CqtBBrcwAAuDoCCjsPPvig4uPjtXnzZvXt21fS1x80+MQTT+iee+7RokWLglokAABAoAJas/PCCy8oPz/fCTqS1LdvXz333HPcjQUAAMJKQGGntrZWf/7zn9vsr66uVl1dXZeLAgAACJaAws7EiRP1xBNP6De/+Y0qKipUUVGh3/zmN5o1a5YmTZoU7BoBAAACFtCanZdfflm5ubl6/PHH1dzc/PWJIiM1a9YsrVq1KqgFAgAAdEVAYadPnz5au3atVq1apU8//VTGGN10002Ki4sLdn0AAABd0qUPFaysrFRlZaVuvvlmxcXFyRgTrLoAAACCIqCw89VXX2nkyJG6+eabNXbsWFVWVkqS/umf/onbzgEAQFgJKOz85Cc/UVRUlE6dOqU+ffo4+x955BHt2LEjaMUBAAB0VUBrdnbu3Kn33ntP/fv399ufnp6ukydPBqUwAACAYAhoZufs2bN+MzoXffnll4qJielyUQAAAMESUNi599579dprrzmPXS6XWltbtWrVKt13331BKw4AAKCrArqMtWrVKo0YMUL79+9XU1OTFi9erGPHjumvf/2r/vjHPwa7RgAAgIAFNLNz66236vDhw/rhD3+o0aNH6+zZs5o0aZIOHjyov/3bvw12jQAAAAHr9MxOc3OzsrOz9corr+jZZ5/tjpoAAACCptMzO1FRUTp69KhcLld31AMAABBUAV3Gmj59ujZs2BDsWgAAAIIuoAXKTU1N+vd//3cVFxdr8ODBbb4Tq6CgICjFAQAAdFWnws5nn32mG264QUePHtWdd94pSTp+/LjfGC5vAQCAcNKpsJOenq7Kykrt2bNH0tdfD/HSSy8pJSWlW4oDAADoqk6t2bn0W83fffddnT17NqgFAQAABFNAC5QvujT8AAAAhJtOhR2Xy9VmTQ5rdAAAQDjr1JodY4x+9KMfOV/2ef78eT355JNt7sbatm1b8CoEAADogk6FnRkzZvg9fvzxx4NaDAAAQLB1Kuxs3Lixu+oAAADoFl1aoAwAABDuCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArBbSsJOfn68f/OAHio+PV79+/fTwww/rk08+8RtjjNHy5cvl9XoVGxurESNG6NixY35jGhsbNX/+fCUnJysuLk4TJkxQRUXF1WwFAACEqZCGnZKSEs2dO1cffPCBiouLdeHCBWVnZ+vs2bPOmJUrV6qgoECFhYUqLS2Vx+PR6NGjVVdX54zJycnR9u3bVVRUpL1796q+vl7jx49XS0tLKNoCAABhJDKUL75jxw6/xxs3blS/fv1UVlame++9V8YYrVmzRsuWLdOkSZMkSZs3b1ZKSoq2bt2q2bNny+fzacOGDXr99dc1atQoSdKWLVuUmpqqXbt26YEHHrjqfQEAgPARVmt2fD6fJCkxMVGSVF5erqqqKmVnZztjYmJiNHz4cO3bt0+SVFZWpubmZr8xXq9XGRkZzphLNTY2qra21m8DAAB2CpuwY4zRwoULNWzYMGVkZEiSqqqqJEkpKSl+Y1NSUpxjVVVVio6OVt++fS875lL5+flyu93OlpqaGux2AABAmAibsDNv3jwdPnxYb7zxRptjLpfL77Exps2+S33bmKVLl8rn8znb6dOnAy8cAACEtbAIO/Pnz9fbb7+tPXv2qH///s5+j8cjSW1maKqrq53ZHo/Ho6amJtXU1Fx2zKViYmKUkJDgtwEAADuFNOwYYzRv3jxt27ZNu3fvVlpamt/xtLQ0eTweFRcXO/uamppUUlKioUOHSpKysrIUFRXlN6ayslJHjx51xgAAgN4rpHdjzZ07V1u3btV//ud/Kj4+3pnBcbvdio2NlcvlUk5OjvLy8pSenq709HTl5eWpT58+mjp1qjN21qxZWrRokZKSkpSYmKjc3FxlZmY6d2cBAIDeK6RhZ926dZKkESNG+O3fuHGjfvSjH0mSFi9erIaGBs2ZM0c1NTUaMmSIdu7cqfj4eGf86tWrFRkZqcmTJ6uhoUEjR47Upk2bFBERcbVaAQAAYcpljDGhLiLUamtr5Xa75fP5unX9zg1L3ml3/+crxnVqDAAA6Pj7d1gsUAYAAOguhB0AAGA1wg4AALAaYQcAAFgtpHdj4WuXW5QMAAC6jpkdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGC1yFAXgM67Yck7zr8/XzEuhJUAABD+mNkBAABWI+wAAACrcRmrh/jmpSsAANBxzOwAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKzG5+z0cJd+/g5fHwEAgD9mdgAAgNUIOwAAwGqEHQAAYLWQhp3f//73evDBB+X1euVyufTWW2/5HTfGaPny5fJ6vYqNjdWIESN07NgxvzGNjY2aP3++kpOTFRcXpwkTJqiiouIqdgEAAMJZSMPO2bNndfvtt6uwsLDd4ytXrlRBQYEKCwtVWloqj8ej0aNHq66uzhmTk5Oj7du3q6ioSHv37lV9fb3Gjx+vlpaWq9VGWLlhyTvOBgAAQnw31pgxYzRmzJh2jxljtGbNGi1btkyTJk2SJG3evFkpKSnaunWrZs+eLZ/Ppw0bNuj111/XqFGjJElbtmxRamqqdu3apQceeOCq9QIAAMJT2K7ZKS8vV1VVlbKzs519MTExGj58uPbt2ydJKisrU3Nzs98Yr9erjIwMZ0x7GhsbVVtb67cBAAA7hW3YqaqqkiSlpKT47U9JSXGOVVVVKTo6Wn379r3smPbk5+fL7XY7W2pqapCrBwAA4SJsw85FLpfL77Exps2+S11pzNKlS+Xz+Zzt9OnTQakVAACEn7ANOx6PR5LazNBUV1c7sz0ej0dNTU2qqam57Jj2xMTEKCEhwW8DAAB2Ctuwk5aWJo/Ho+LiYmdfU1OTSkpKNHToUElSVlaWoqKi/MZUVlbq6NGjzhgAANC7hfRurPr6ev3v//6v87i8vFyHDh1SYmKiBgwYoJycHOXl5Sk9PV3p6enKy8tTnz59NHXqVEmS2+3WrFmztGjRIiUlJSkxMVG5ubnKzMx07s4CAAC9W0jDzv79+3Xfffc5jxcuXChJmjFjhjZt2qTFixeroaFBc+bMUU1NjYYMGaKdO3cqPj7eec7q1asVGRmpyZMnq6GhQSNHjtSmTZsUERFx1fsBAADhx2WMMaEuItRqa2vldrvl8/m6df3O1f6gP74BHQBgs46+f4ftmh0AAIBgIOwAAACrEXYAAIDVQrpAGd3rm2uEWL8DAOitmNkBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKvxdRFw8PUSAAAbEXZ6CYIMAKC3Iuz0QgQfAEBvwpodAABgNcIOAACwGmEHAABYjbADAACsxgLlXu6bi5UBALARMzsAAMBqhB0AAGA1LmOhU/iMHgBAT0PYQcAIPgCAnoCwgytiETMAoCcj7CAomOUBAIQrwg7axWwOAMAW3I0FAACsxswOwhaXxgAAwcDMDgAAsBozOwiJrszaMOMDAOgMZnYAAIDVmNlBt7qaszDM+AAA2kPYwVXD7ewAgFAg7CDkmJEBAHQnwg7CSnfM/hCmAKB3I+wg6Lo7sAAA0BmEHUAspAYAmxF2gG5CqAGA8EDYgZUud9mrIwGEkAIAdiHsAB1ECAKAnomwgx6tKwuXe9ui58v1S3ADYDvCDhCAzgaHrlxWAwB0DWEHCEOsLWqLnwmAQBF2gDAXDm/yofqW+t52qRFA9yDsAGGip7yxX81g1R3n/GbNzAQBvQNhB4Cjs0GjIyGiK+fvKYIVpghfQPcg7ADfwtY35+5g088qnC8L4urj99XzEXaAHqQrMy/BOme46e76u/uW/XD48ttw+LoULjV2HD+TziPsAJbp6eGlI7oS4oL18wlVyPqmjlw67I7XvVwNnT1POAS9jpwnWK/bE0NKsD5mI9T9EnYAIIh6Q9jszbojTPHlw93PmrCzdu1arVq1SpWVlRo0aJDWrFmje+65J9RloZfhja7nCoeZmu7WHW963X3OrrxWR2YZwuGjEbpyCe9q/vyDNT4UrAg7b775pnJycrR27VrdfffdeuWVVzRmzBh9/PHHGjBgQKjLA4BuEW5rsnrCm16468rlV2ZwLs9ljDGhLqKrhgwZojvvvFPr1q1z9n3/+9/Xww8/rPz8/Cs+v7a2Vm63Wz6fTwkJCd1WJ38IAKB36Y51VV0RrHo6e57uCl8dff/u8TM7TU1NKisr05IlS/z2Z2dna9++fe0+p7GxUY2Njc5jn88n6esfWndqbTzXrecHAISXAT/5dahL8BOsejp7nu56f7143ivN2/T4sPPll1+qpaVFKSkpfvtTUlJUVVXV7nPy8/P17LPPttmfmpraLTUCANCbudd07/nr6urkdrsve7zHh52LXC6X32NjTJt9Fy1dulQLFy50Hre2tuqvf/2rkpKSLvucQNTW1io1NVWnT5/u1stj4aS39Uy/dqNfu9Fvz2eMUV1dnbxe77eO6/FhJzk5WREREW1mcaqrq9vM9lwUExOjmJgYv33f+c53uqtEJSQkWPM/rI7qbT3Tr93o127027N924zORddchTq6VXR0tLKyslRcXOy3v7i4WEOHDg1RVQAAIFz0+JkdSVq4cKGmTZumwYMH66677tL69et16tQpPfnkk6EuDQAAhJgVYeeRRx7RV199pZ///OeqrKxURkaGfvvb32rgwIEhrSsmJkY/+9nP2lwys1lv65l+7Ua/dqPf3sOKz9kBAAC4nB6/ZgcAAODbEHYAAIDVCDsAAMBqhB0AAGA1wk43Wrt2rdLS0nTttdcqKytLf/jDH0JdUlDk5+frBz/4geLj49WvXz89/PDD+uSTT/zGGGO0fPlyeb1excbGasSIETp27FiIKg6u/Px8uVwu5eTkOPts6/eLL77Q448/rqSkJPXp00d/93d/p7KyMue4Tf1euHBB//zP/6y0tDTFxsbqxhtv1M9//nO1trY6Y3pyv7///e/14IMPyuv1yuVy6a233vI73pHeGhsbNX/+fCUnJysuLk4TJkxQRUXFVeyi476t3+bmZj3zzDPKzMxUXFycvF6vpk+frjNnzvidoyf1K135d/xNs2fPlsvl0po1a/z297SeO4uw003efPNN5eTkaNmyZTp48KDuuecejRkzRqdOnQp1aV1WUlKiuXPn6oMPPlBxcbEuXLig7OxsnT171hmzcuVKFRQUqLCwUKWlpfJ4PBo9erTq6upCWHnXlZaWav369brtttv89tvUb01Nje6++25FRUXp3Xff1ccff6wXXnjB71PGber3+eef18svv6zCwkL96U9/0sqVK7Vq1Sr98pe/dMb05H7Pnj2r22+/XYWFhe0e70hvOTk52r59u4qKirR3717V19dr/PjxamlpuVptdNi39Xvu3DkdOHBA//Iv/6IDBw5o27ZtOn78uCZMmOA3rif1K135d3zRW2+9pf/5n/9p96sVelrPnWbQLX74wx+aJ5980m/fLbfcYpYsWRKiirpPdXW1kWRKSkqMMca0trYaj8djVqxY4Yw5f/68cbvd5uWXXw5VmV1WV1dn0tPTTXFxsRk+fLhZsGCBMca+fp955hkzbNiwyx63rd9x48aZmTNn+u2bNGmSefzxx40xdvUryWzfvt153JHe/u///s9ERUWZoqIiZ8wXX3xhrrnmGrNjx46rVnsgLu23PR9++KGRZE6ePGmM6dn9GnP5nisqKsx3v/tdc/ToUTNw4ECzevVq51hP77kjmNnpBk1NTSorK1N2drbf/uzsbO3bty9EVXUfn88nSUpMTJQklZeXq6qqyq//mJgYDR8+vEf3P3fuXI0bN06jRo3y229bv2+//bYGDx6sf/zHf1S/fv10xx136NVXX3WO29bvsGHD9P777+v48eOSpI8++kh79+7V2LFjJdnX7zd1pLeysjI1Nzf7jfF6vcrIyOjx/Utf//1yuVzOzKWN/ba2tmratGl6+umnNWjQoDbHbez5UlZ8gnK4+fLLL9XS0tLmi0hTUlLafGFpT2eM0cKFCzVs2DBlZGRIktNje/2fPHnyqtcYDEVFRTpw4IBKS0vbHLOt388++0zr1q3TwoUL9dOf/lQffvihnnrqKcXExGj69OnW9fvMM8/I5/PplltuUUREhFpaWvSLX/xCjz76qCT7fr/f1JHeqqqqFB0drb59+7YZ09P/np0/f15LlizR1KlTnS/GtLHf559/XpGRkXrqqafaPW5jz5ci7HQjl8vl99gY02ZfTzdv3jwdPnxYe/fubXPMlv5Pnz6tBQsWaOfOnbr22msvO86WfltbWzV48GDl5eVJku644w4dO3ZM69at0/Tp051xtvT75ptvasuWLdq6dasGDRqkQ4cOKScnR16vVzNmzHDG2dJvewLpraf339zcrClTpqi1tVVr16694vie2m9ZWZlefPFFHThwoNP199Se28NlrG6QnJysiIiINom4urq6zf+D6snmz5+vt99+W3v27FH//v2d/R6PR5Ks6b+srEzV1dXKyspSZGSkIiMjVVJSopdeekmRkZFOT7b0e/311+vWW2/12/f973/fWVxv2+/36aef1pIlSzRlyhRlZmZq2rRp+slPfqL8/HxJ9vX7TR3pzePxqKmpSTU1NZcd09M0Nzdr8uTJKi8vV3FxsTOrI9nX7x/+8AdVV1drwIABzt+vkydPatGiRbrhhhsk2ddzewg73SA6OlpZWVkqLi72219cXKyhQ4eGqKrgMcZo3rx52rZtm3bv3q20tDS/42lpafJ4PH79NzU1qaSkpEf2P3LkSB05ckSHDh1ytsGDB+uxxx7ToUOHdOONN1rV7913393mowSOHz/ufLGubb/fc+fO6Zpr/P8URkREOLee29bvN3Wkt6ysLEVFRfmNqays1NGjR3tk/xeDzokTJ7Rr1y4lJSX5Hbet32nTpunw4cN+f7+8Xq+efvppvffee5Ls67ldIVoYbb2ioiITFRVlNmzYYD7++GOTk5Nj4uLizOeffx7q0rrsxz/+sXG73eZ3v/udqaysdLZz5845Y1asWGHcbrfZtm2bOXLkiHn00UfN9ddfb2pra0NYefB8824sY+zq98MPPzSRkZHmF7/4hTlx4oT5j//4D9OnTx+zZcsWZ4xN/c6YMcN897vfNf/93/9tysvLzbZt20xycrJZvHixM6Yn91tXV2cOHjxoDh48aCSZgoICc/DgQefuo4709uSTT5r+/fubXbt2mQMHDpj777/f3H777ebChQuhauuyvq3f5uZmM2HCBNO/f39z6NAhv79fjY2Nzjl6Ur/GXPl3fKlL78Yypuf13FmEnW70b//2b2bgwIEmOjra3Hnnnc6t2T2dpHa3jRs3OmNaW1vNz372M+PxeExMTIy59957zZEjR0JXdJBdGnZs6/e//uu/TEZGhomJiTG33HKLWb9+vd9xm/qtra01CxYsMAMGDDDXXnutufHGG82yZcv83vx6cr979uxp97/XGTNmGGM61ltDQ4OZN2+eSUxMNLGxsWb8+PHm1KlTIejmyr6t3/Ly8sv+/dqzZ49zjp7UrzFX/h1fqr2w09N67iyXMcZcjRkkAACAUGDNDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABW+38uq8/mcSdXSAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 0]['DOC_N_VALID'].plot.hist(bins=150, range=[0,150])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "665de94a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:36.202474Z",
     "start_time": "2023-11-01T00:07:36.189172Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6139, 7)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5582, 7)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5282, 7)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5054, 7)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(3655, 7)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking at how many metatriples would be left at different cutoffs\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 5].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 10].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 15].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 20].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 100].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d81fd09",
   "metadata": {},
   "source": [
    "After discussion with Andrew Su, we decided to set a threshold of 20 for now. \n",
    "\n",
    "I wondered what was going on with the same_as metatriples, so I ran the code block below. Even after the cutoff, it looks like some of these same_as metatriples probably mean \"similar effectiveness\" which is kind of awkward with the mapped biolink-model predicate of \"close_match\". \n",
    "* orch (organic chemical) same as medd (medical device)? or topp (therapeutic procedure)? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "86920c33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:40.654039Z",
     "start_time": "2023-11-01T00:07:40.643208Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4015</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>2910</td>\n",
       "      <td>450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8039</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>190</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8596</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>phsu</td>\n",
       "      <td>350</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9380</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>medd</td>\n",
       "      <td>249</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9494</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>258</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9548</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>antb</td>\n",
       "      <td>177</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9636</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>226</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_PREFIX SUBJECT_SEMTYPE PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "4015           umls            orch   same_as          umls           orch   \n",
       "8039           umls            orch   same_as          umls           hops   \n",
       "8596           umls            orch   same_as          umls           phsu   \n",
       "9380           umls            orch   same_as          umls           medd   \n",
       "9494           umls            orch   same_as          umls           aapp   \n",
       "9548           umls            orch   same_as          umls           antb   \n",
       "9636           umls            orch   same_as          umls           topp   \n",
       "\n",
       "      DOC_N  DOC_N_VALID  \n",
       "4015   2910          450  \n",
       "8039    190           54  \n",
       "8596    350           42  \n",
       "9380    249           30  \n",
       "9494    258           28  \n",
       "9548    177           28  \n",
       "9636    226           27  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[(imported_metatriples['DOC_N_VALID'] > 20) &\n",
    "                     (imported_metatriples['PREDICATE'] == 'same_as') & \n",
    "                     (imported_metatriples['SUBJECT_SEMTYPE'] == 'orch')\n",
    "                    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d79a90",
   "metadata": {},
   "source": [
    "### Pruning for `DOC_N_VALID` > 20 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "2c859df9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:45.637142Z",
     "start_time": "2023-11-01T00:07:45.626942Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5054, 5)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE\n",
       "3           umls            bpoc  LOCATION_OF          umls           neop\n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp\n",
       "5           umls            bpoc  LOCATION_OF          umls           patf"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples = imported_metatriples[(imported_metatriples['DOC_N_VALID'] > 20)].copy()\n",
    "final_metatriples.drop(columns=['DOC_N', 'DOC_N_VALID'], inplace=True)\n",
    "final_metatriples.shape\n",
    "final_metatriples[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dc02b1b",
   "metadata": {},
   "source": [
    "## Constructing final table with biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "b53030eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:55.565607Z",
     "start_time": "2023-11-01T00:07:55.558661Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all SEMMED subject semantic types to biolink categories\n",
    "\n",
    "final_metatriples = final_metatriples.merge(SEMMED_semtypes[['Abbrev', 'BiolinkMapping']], \n",
    "                      how='left', left_on='SUBJECT_SEMTYPE', right_on='Abbrev')\n",
    "\n",
    "final_metatriples.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "final_metatriples.columns = [\n",
    "    'SubjectPrefix', \n",
    "    'OriginalSubject', \n",
    "    'OriginalPredicate',\n",
    "    'ObjectPrefix',\n",
    "    'OriginalObject',\n",
    "    'BiolinkSubject'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "70092440",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:07:56.757985Z",
     "start_time": "2023-11-01T00:07:56.748795Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5049</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5050</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>umls</td>\n",
       "      <td>nusq</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5051</th>\n",
       "      <td>umls</td>\n",
       "      <td>imft</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5052</th>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5053</th>\n",
       "      <td>umls</td>\n",
       "      <td>elii</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5054 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5049          umls            eico          DISRUPTS         umls   \n",
       "5050      ncbigene            gngm       CONVERTS_TO         umls   \n",
       "5051          umls            imft    INTERACTS_WITH         umls   \n",
       "5052          umls            aapp           same_as         umls   \n",
       "5053          umls            elii           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject  \n",
       "0              neop  GrossAnatomicalStructure  \n",
       "1              aapp          AnatomicalEntity  \n",
       "2              patf  GrossAnatomicalStructure  \n",
       "3              dsyn                 Procedure  \n",
       "4              neop                 Procedure  \n",
       "...             ...                       ...  \n",
       "5049           dsyn             SmallMolecule  \n",
       "5050           nusq                      Gene  \n",
       "5051           cell          BiologicalEntity  \n",
       "5052           topp               Polypeptide  \n",
       "5053           fndg             SmallMolecule  \n",
       "\n",
       "[5054 rows x 6 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "9e435399",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:24.752998Z",
     "start_time": "2023-11-01T00:08:24.746236Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all SEMMED object semantic types to biolink categories\n",
    "\n",
    "final_metatriples = final_metatriples.merge(SEMMED_semtypes[['Abbrev', 'BiolinkMapping']],\n",
    "                      how='left', left_on='OriginalObject', right_on='Abbrev')\n",
    "\n",
    "final_metatriples.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "final_metatriples.columns = [\n",
    "    'SubjectPrefix', \n",
    "    'OriginalSubject', \n",
    "    'OriginalPredicate',\n",
    "    'ObjectPrefix',\n",
    "    'OriginalObject',\n",
    "    'BiolinkSubject',\n",
    "    'BiolinkObject'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a0fe51f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:25.430042Z",
     "start_time": "2023-11-01T00:08:25.420756Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5049</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5050</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>umls</td>\n",
       "      <td>nusq</td>\n",
       "      <td>Gene</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5051</th>\n",
       "      <td>umls</td>\n",
       "      <td>imft</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>Cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5052</th>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5053</th>\n",
       "      <td>umls</td>\n",
       "      <td>elii</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5054 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5049          umls            eico          DISRUPTS         umls   \n",
       "5050      ncbigene            gngm       CONVERTS_TO         umls   \n",
       "5051          umls            imft    INTERACTS_WITH         umls   \n",
       "5052          umls            aapp           same_as         umls   \n",
       "5053          umls            elii           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject               BiolinkObject  \n",
       "0              neop  GrossAnatomicalStructure                     Disease  \n",
       "1              aapp          AnatomicalEntity                 Polypeptide  \n",
       "2              patf  GrossAnatomicalStructure         PathologicalProcess  \n",
       "3              dsyn                 Procedure                     Disease  \n",
       "4              neop                 Procedure                     Disease  \n",
       "...             ...                       ...                         ...  \n",
       "5049           dsyn             SmallMolecule                     Disease  \n",
       "5050           nusq                      Gene           NucleicAcidEntity  \n",
       "5051           cell          BiologicalEntity                        Cell  \n",
       "5052           topp               Polypeptide                   Procedure  \n",
       "5053           fndg             SmallMolecule  DiseaseOrPhenotypicFeature  \n",
       "\n",
       "[5054 rows x 7 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "3f42bd07",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:33.439230Z",
     "start_time": "2023-11-01T00:08:33.432212Z"
    }
   },
   "outputs": [],
   "source": [
    "## add all biolink predicate / qualifier / inverse stuff\n",
    "final_metatriples = final_metatriples.merge(\n",
    "                      SEMMED_predicates,\n",
    "                      how='left', left_on='OriginalPredicate', right_on='Semmed_in_Data')\n",
    "\n",
    "final_metatriples.drop(columns = 'Semmed_in_Data', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "6c651d23",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:34.151632Z",
     "start_time": "2023-11-01T00:08:34.141051Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix OriginalObject  \\\n",
       "0          umls            bpoc       LOCATION_OF         umls           neop   \n",
       "1          umls            bdsu       LOCATION_OF         umls           aapp   \n",
       "2          umls            bpoc       LOCATION_OF         umls           patf   \n",
       "\n",
       "             BiolinkSubject        BiolinkObject       BMPred BMQualPred  \\\n",
       "0  GrossAnatomicalStructure              Disease  location_of       None   \n",
       "1          AnatomicalEntity          Polypeptide  location_of       None   \n",
       "2  GrossAnatomicalStructure  PathologicalProcess  location_of       None   \n",
       "\n",
       "  BMObjAsp BMObjDirect BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp  \\\n",
       "0     None        None  located_in            None           None   \n",
       "1     None        None  located_in            None           None   \n",
       "2     None        None  located_in            None           None   \n",
       "\n",
       "  BM_Inv_SubjDirect  \n",
       "0              None  \n",
       "1              None  \n",
       "2              None  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "cc7864f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:40.454232Z",
     "start_time": "2023-11-01T00:08:40.431252Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## check to make sure everything is mapped to Biolink successfully\n",
    "## should get all empty dataframes\n",
    "\n",
    "final_metatriples[final_metatriples['BiolinkSubject'].isna()]\n",
    "final_metatriples[final_metatriples['BiolinkObject'].isna()]\n",
    "\n",
    "final_metatriples[final_metatriples['BMPred'].isna()]\n",
    "final_metatriples[final_metatriples['BM_Inv_Pred'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2c500a",
   "metadata": {},
   "source": [
    "### Adjust Biolink categories based on NCBIGene prefix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5173b760",
   "metadata": {},
   "source": [
    "There's two ID-namespaces for subject + object: umls + ncbigene. But biolink-model only recognizes NCBIGene as a valid namespace for specific categories, like Gene. \n",
    "\n",
    "So we're going to replace the biolink subject / object mapping with Gene whenever it has the NCBIGene prefix..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b8d843fb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:08:54.933357Z",
     "start_time": "2023-11-01T00:08:54.904888Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>umls</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>virs</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Virus</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4961</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4983</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>imft</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4994</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>umls</td>\n",
       "      <td>nusq</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5040</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>horm</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>anab</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>282 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "146       ncbigene            bacs           PART_OF         umls   \n",
       "149       ncbigene            aapp        STIMULATES         umls   \n",
       "151       ncbigene            aapp    INTERACTS_WITH         umls   \n",
       "152       ncbigene            aapp   ASSOCIATED_WITH         umls   \n",
       "190       ncbigene            aapp           PART_OF         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "4961      ncbigene            phsu    INTERACTS_WITH     ncbigene   \n",
       "4983      ncbigene            imft   ASSOCIATED_WITH         umls   \n",
       "4994      ncbigene            aapp          INHIBITS         umls   \n",
       "4995      ncbigene            bacs    INTERACTS_WITH     ncbigene   \n",
       "5040      ncbigene            horm            TREATS         umls   \n",
       "\n",
       "     OriginalObject    BiolinkSubject      BiolinkObject          BMPred  \\\n",
       "146            aapp     SmallMolecule        Polypeptide         part_of   \n",
       "149            gngm       Polypeptide               Gene         affects   \n",
       "151            aapp       Polypeptide        Polypeptide  interacts_with   \n",
       "152            neop       Polypeptide            Disease      related_to   \n",
       "190            virs       Polypeptide              Virus         part_of   \n",
       "...             ...               ...                ...             ...   \n",
       "4961           aapp     SmallMolecule        Polypeptide  interacts_with   \n",
       "4983           neop  BiologicalEntity            Disease      related_to   \n",
       "4994           nusq       Polypeptide  NucleicAcidEntity         affects   \n",
       "4995           aapp     SmallMolecule        Polypeptide  interacts_with   \n",
       "5040           anab     SmallMolecule            Disease          treats   \n",
       "\n",
       "     BMQualPred               BMObjAsp BMObjDirect     BM_Inv_Pred  \\\n",
       "146        None                   None        None        has_part   \n",
       "149      causes  activity_or_abundance   increased     affected_by   \n",
       "151        None                   None        None  interacts_with   \n",
       "152        None                   None        None      related_to   \n",
       "190        None                   None        None        has_part   \n",
       "...         ...                    ...         ...             ...   \n",
       "4961       None                   None        None  interacts_with   \n",
       "4983       None                   None        None      related_to   \n",
       "4994     causes  activity_or_abundance   decreased     affected_by   \n",
       "4995       None                   None        None  interacts_with   \n",
       "5040       None                   None        None      treated_by   \n",
       "\n",
       "     BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "146             None                   None              None  \n",
       "149        caused_by  activity_or_abundance         increased  \n",
       "151             None                   None              None  \n",
       "152             None                   None              None  \n",
       "190             None                   None              None  \n",
       "...              ...                    ...               ...  \n",
       "4961            None                   None              None  \n",
       "4983            None                   None              None  \n",
       "4994       caused_by  activity_or_abundance         decreased  \n",
       "4995            None                   None              None  \n",
       "5040            None                   None              None  \n",
       "\n",
       "[282 rows x 15 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>umls</td>\n",
       "      <td>bacs</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>umls</td>\n",
       "      <td>tisu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5009</th>\n",
       "      <td>umls</td>\n",
       "      <td>phsf</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>imft</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5022</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5024</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>USES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5031</th>\n",
       "      <td>umls</td>\n",
       "      <td>enzy</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Protein</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "31            umls            bdsu       LOCATION_OF     ncbigene   \n",
       "39            umls            bpoc       LOCATION_OF     ncbigene   \n",
       "40            umls            cell       LOCATION_OF     ncbigene   \n",
       "63            umls            bacs           PART_OF     ncbigene   \n",
       "101           umls            tisu       LOCATION_OF     ncbigene   \n",
       "...            ...             ...               ...          ...   \n",
       "4995      ncbigene            bacs    INTERACTS_WITH     ncbigene   \n",
       "5009          umls            phsf          PRODUCES     ncbigene   \n",
       "5022          umls            eico          INHIBITS     ncbigene   \n",
       "5024          umls            topp              USES     ncbigene   \n",
       "5031          umls            enzy    INTERACTS_WITH     ncbigene   \n",
       "\n",
       "     OriginalObject            BiolinkSubject     BiolinkObject  \\\n",
       "31             aapp          AnatomicalEntity       Polypeptide   \n",
       "39             aapp  GrossAnatomicalStructure       Polypeptide   \n",
       "40             aapp                      Cell       Polypeptide   \n",
       "63             aapp             SmallMolecule       Polypeptide   \n",
       "101            aapp  GrossAnatomicalStructure       Polypeptide   \n",
       "...             ...                       ...               ...   \n",
       "4995           aapp             SmallMolecule       Polypeptide   \n",
       "5009           imft      PhysiologicalProcess  BiologicalEntity   \n",
       "5022           aapp             SmallMolecule       Polypeptide   \n",
       "5024           phsu                 Procedure     SmallMolecule   \n",
       "5031           aapp                   Protein       Polypeptide   \n",
       "\n",
       "              BMPred BMQualPred               BMObjAsp BMObjDirect  \\\n",
       "31       location_of       None                   None        None   \n",
       "39       location_of       None                   None        None   \n",
       "40       location_of       None                   None        None   \n",
       "63           part_of       None                   None        None   \n",
       "101      location_of       None                   None        None   \n",
       "...              ...        ...                    ...         ...   \n",
       "4995  interacts_with       None                   None        None   \n",
       "5009        produces       None                   None        None   \n",
       "5022         affects     causes  activity_or_abundance   decreased   \n",
       "5024       has_input       None                   None        None   \n",
       "5031  interacts_with       None                   None        None   \n",
       "\n",
       "         BM_Inv_Pred BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "31        located_in            None                   None              None  \n",
       "39        located_in            None                   None              None  \n",
       "40        located_in            None                   None              None  \n",
       "63          has_part            None                   None              None  \n",
       "101       located_in            None                   None              None  \n",
       "...              ...             ...                    ...               ...  \n",
       "4995  interacts_with            None                   None              None  \n",
       "5009     produced_by            None                   None              None  \n",
       "5022     affected_by       caused_by  activity_or_abundance         decreased  \n",
       "5024     is_input_of            None                   None              None  \n",
       "5031  interacts_with            None                   None              None  \n",
       "\n",
       "[163 rows x 15 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking\n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'ncbigene') &\n",
    "                  (final_metatriples['BiolinkSubject'] != 'Gene')]\n",
    "final_metatriples[(final_metatriples['ObjectPrefix'] == 'ncbigene') &\n",
    "                  (final_metatriples['BiolinkObject'] != 'Gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "480daedc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:09:04.851867Z",
     "start_time": "2023-11-01T00:09:04.713832Z"
    }
   },
   "outputs": [],
   "source": [
    "## Next, mutate the metatriples with ncbigene prefix so their Biolink categories are Gene \n",
    "for index in final_metatriples.index:\n",
    "    if (final_metatriples.loc[index,'SubjectPrefix'] == 'ncbigene'):\n",
    "        final_metatriples.loc[index,'BiolinkSubject'] = 'Gene'\n",
    "    if final_metatriples.loc[index,'ObjectPrefix'] == 'ncbigene':\n",
    "        final_metatriples.loc[index,'BiolinkObject'] = 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "08e436d1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:09:05.792681Z",
     "start_time": "2023-11-01T00:09:05.778673Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## double-checking: should now get empty dataframes\n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'ncbigene') &\n",
    "                (final_metatriples['BiolinkSubject'] != 'Gene')]\n",
    "final_metatriples[(final_metatriples['ObjectPrefix'] == 'ncbigene') &\n",
    "                (final_metatriples['BiolinkObject'] != 'Gene')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870db906",
   "metadata": {},
   "source": [
    "## Generate operation yaml!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344f9ec6",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Review / test the code below carefully!\n",
    "\n",
    "* If needed, change the code within the functions below to change the x-bte annotations that are made...\n",
    "  * review the prefix-handling code! This is to handle ncbigene ID-namespace for some metatriples\n",
    "  * review the qualifier-generating code! Currently it's very simple because all operations with qualifiers have the same set: qualified_predicate, an aspect_qualifier, and a direction_qualifier...\n",
    "* the yaml created below refers to umls-subj, umls-obj, ncbigene-subj, ncbigene-obj...those are specified [here close to the bottom](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf76bfcf",
   "metadata": {},
   "source": [
    "### function for forward operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "222c6bd1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:46:52.751245Z",
     "start_time": "2023-11-01T00:46:52.744920Z"
    }
   },
   "outputs": [],
   "source": [
    "## will use variables set in first section's config\n",
    "## original direction: subject -> object\n",
    "\n",
    "def generate_forward_op(\n",
    "    subj_prefix, obj_prefix,\n",
    "    original_subj, original_pred, original_obj,\n",
    "    biolink_subj, biolink_pred, \n",
    "    biolink_qualified_pred, biolink_obj_asp, biolink_obj_direct,\n",
    "    biolink_obj):\n",
    "    \n",
    "    ## get biolink-model spelling for prefixes\n",
    "    if subj_prefix == 'umls':\n",
    "        biolink_subj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_subj_prefix = 'NCBIGene'\n",
    "        \n",
    "    if obj_prefix == 'umls':\n",
    "        biolink_obj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_obj_prefix = 'NCBIGene'\n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    if (subj_prefix == 'umls') and (obj_prefix == 'umls'):\n",
    "        normal_op_name = f'{original_subj}-{original_pred}-{original_obj}'\n",
    "    else:\n",
    "        normal_op_name = f'{original_subj},{subj_prefix}-{original_pred}-{original_obj},{obj_prefix}'\n",
    "    \n",
    "    temp = {\n",
    "        ## original direction: subject -> object\n",
    "        normal_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': biolink_subj_prefix,\n",
    "                        'semantic': biolink_subj  ## input is subject!\n",
    "                    }\n",
    "                ],\n",
    "                ## notice that subject.{subj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'requestBody': {\n",
    "                    'body': {\n",
    "                        'q': '{{ queryInputs }}',\n",
    "                        'scopes': f'subject.{subj_prefix}'\n",
    "                    }\n",
    "                },\n",
    "                ## notice that object.{obj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'parameters': {\n",
    "                    'fields': f'object.{obj_prefix},predication.pmid,'\n",
    "                               'predication.sentence,subject.name,object.name',\n",
    "                    'size': POST_size,\n",
    "                    'filter': f'predicate:{original_pred} AND '\n",
    "                              f'object.semantic_type_abbreviation:{original_obj} AND '\n",
    "                              f'_exists_:object.{obj_prefix} AND '\n",
    "                              f'pmid_count:>{pmid_count}'\n",
    "                },\n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': biolink_obj_prefix,\n",
    "                        'semantic': biolink_obj  ## output is object\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_pred,\n",
    "                'source': doublequote('infores:semmeddb'),\n",
    "                'response_mapping': {\n",
    "                    ## matches output as object\n",
    "                    ## so umls-obj and ncbigene-obj are the response-mappings generated\n",
    "                    \"$ref\": f'#/components/x-bte-response-mapping/{obj_prefix}-obj'  \n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has obj aspect and obj direction\n",
    "    if biolink_qualified_pred:\n",
    "        temp[normal_op_name][0]['qualifiers'] = {\n",
    "            'qualified_predicate': biolink_qualified_pred,\n",
    "            'object_aspect_qualifier': biolink_obj_asp,\n",
    "            'object_direction_qualifier': biolink_obj_direct\n",
    "        }\n",
    "        \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "83b2edf4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:46:55.391168Z",
     "start_time": "2023-11-01T00:46:55.386875Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bpoc-LOCATION_OF-neop': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'GrossAnatomicalStructure'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.umls'}},\n",
       "   'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:LOCATION_OF AND object.semantic_type_abbreviation:neop AND _exists_:object.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'predicate': 'location_of',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-obj'}}]}"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - umls\n",
    "generate_forward_op(\n",
    "    'umls', 'umls',\n",
    "    'bpoc', 'LOCATION_OF', 'neop',\n",
    "    'GrossAnatomicalStructure', 'location_of', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "723f107d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:46:57.903622Z",
     "start_time": "2023-11-01T00:46:57.899194Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm,ncbigene-ASSOCIATED_WITH-neop,umls': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.ncbigene'}},\n",
       "   'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:ASSOCIATED_WITH AND object.semantic_type_abbreviation:neop AND _exists_:object.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'predicate': 'related_to',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-obj'}}]}"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - umls\n",
    "generate_forward_op(\n",
    "    'ncbigene', 'umls',\n",
    "    'gngm', 'ASSOCIATED_WITH', 'neop',\n",
    "    'Gene', 'related_to', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "c55ecc04",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:47:00.060571Z",
     "start_time": "2023-11-01T00:47:00.056163Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp,umls-STIMULATES-gngm,ncbigene': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.umls'}},\n",
       "   'parameters': {'fields': 'object.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND object.semantic_type_abbreviation:gngm AND _exists_:object.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affects',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-obj'},\n",
       "   'qualifiers': {'qualified_predicate': 'causes',\n",
       "    'object_aspect_qualifier': 'activity_or_abundance',\n",
       "    'object_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - ncbigene\n",
    "generate_forward_op(\n",
    "    'umls', 'ncbigene',\n",
    "    'aapp', 'STIMULATES', 'gngm',\n",
    "    'Polypeptide', 'affects', \n",
    "    'causes', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "ba5c9ffa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:47:01.901419Z",
     "start_time": "2023-11-01T00:47:01.896839Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm,ncbigene-STIMULATES-gngm,ncbigene': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.ncbigene'}},\n",
       "   'parameters': {'fields': 'object.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND object.semantic_type_abbreviation:gngm AND _exists_:object.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affects',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-obj'},\n",
       "   'qualifiers': {'qualified_predicate': 'causes',\n",
       "    'object_aspect_qualifier': 'activity_or_abundance',\n",
       "    'object_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - ncbigene\n",
    "generate_forward_op(\n",
    "    'ncbigene', 'ncbigene',\n",
    "    'gngm', 'STIMULATES', 'gngm',\n",
    "    'Gene', 'affects', \n",
    "    'causes', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "8effdfd9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:47:03.462234Z",
     "start_time": "2023-11-01T00:47:03.444109Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>265</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>591</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>902</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1355</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1831</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2276</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2796</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3239</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3468</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>same_as</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3782</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>enzy</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3978</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4166</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4213</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>same_as</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4424</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4455</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4471</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4652</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4703</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>horm</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4722</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4771</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>horm</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4961</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4995</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "265       ncbigene            aapp        STIMULATES     ncbigene   \n",
       "336       ncbigene            gngm        STIMULATES     ncbigene   \n",
       "343       ncbigene            aapp    INTERACTS_WITH     ncbigene   \n",
       "473       ncbigene            gngm    INTERACTS_WITH     ncbigene   \n",
       "478       ncbigene            gngm    INTERACTS_WITH     ncbigene   \n",
       "489       ncbigene            gngm          INHIBITS     ncbigene   \n",
       "516       ncbigene            aapp    INTERACTS_WITH     ncbigene   \n",
       "566       ncbigene            aapp     COEXISTS_WITH     ncbigene   \n",
       "591       ncbigene            bacs           PART_OF     ncbigene   \n",
       "659       ncbigene            aapp          INHIBITS     ncbigene   \n",
       "719       ncbigene            gngm     COEXISTS_WITH     ncbigene   \n",
       "868       ncbigene            aapp          INHIBITS     ncbigene   \n",
       "902       ncbigene            gngm          PRODUCES     ncbigene   \n",
       "1174      ncbigene            gngm     COEXISTS_WITH     ncbigene   \n",
       "1355      ncbigene            aapp        STIMULATES     ncbigene   \n",
       "1563      ncbigene            gngm        STIMULATES     ncbigene   \n",
       "1831      ncbigene            gngm          INHIBITS     ncbigene   \n",
       "2276      ncbigene            aapp     COEXISTS_WITH     ncbigene   \n",
       "2796      ncbigene            aapp          PRODUCES     ncbigene   \n",
       "3239      ncbigene            aapp           PART_OF     ncbigene   \n",
       "3468      ncbigene            aapp           same_as     ncbigene   \n",
       "3782      ncbigene            aapp    INTERACTS_WITH     ncbigene   \n",
       "3978      ncbigene            aapp       CONVERTS_TO     ncbigene   \n",
       "4166      ncbigene            gngm          PRODUCES     ncbigene   \n",
       "4213      ncbigene            gngm           same_as     ncbigene   \n",
       "4424      ncbigene            phsu          INHIBITS     ncbigene   \n",
       "4455      ncbigene            phsu        STIMULATES     ncbigene   \n",
       "4471      ncbigene            bacs        STIMULATES     ncbigene   \n",
       "4652      ncbigene            gngm       CONVERTS_TO     ncbigene   \n",
       "4703      ncbigene            horm    INTERACTS_WITH     ncbigene   \n",
       "4722      ncbigene            gngm    INTERACTS_WITH     ncbigene   \n",
       "4771      ncbigene            horm        STIMULATES     ncbigene   \n",
       "4961      ncbigene            phsu    INTERACTS_WITH     ncbigene   \n",
       "4995      ncbigene            bacs    INTERACTS_WITH     ncbigene   \n",
       "\n",
       "     OriginalObject BiolinkSubject BiolinkObject          BMPred BMQualPred  \\\n",
       "265            gngm           Gene          Gene         affects     causes   \n",
       "336            gngm           Gene          Gene         affects     causes   \n",
       "343            aapp           Gene          Gene  interacts_with       None   \n",
       "473            aapp           Gene          Gene  interacts_with       None   \n",
       "478            gngm           Gene          Gene  interacts_with       None   \n",
       "489            gngm           Gene          Gene         affects     causes   \n",
       "516            gngm           Gene          Gene  interacts_with       None   \n",
       "566            gngm           Gene          Gene   coexists_with       None   \n",
       "591            aapp           Gene          Gene         part_of       None   \n",
       "659            aapp           Gene          Gene         affects     causes   \n",
       "719            gngm           Gene          Gene   coexists_with       None   \n",
       "868            gngm           Gene          Gene         affects     causes   \n",
       "902            aapp           Gene          Gene        produces       None   \n",
       "1174           aapp           Gene          Gene   coexists_with       None   \n",
       "1355           aapp           Gene          Gene         affects     causes   \n",
       "1563           aapp           Gene          Gene         affects     causes   \n",
       "1831           aapp           Gene          Gene         affects     causes   \n",
       "2276           aapp           Gene          Gene   coexists_with       None   \n",
       "2796           aapp           Gene          Gene        produces       None   \n",
       "3239           aapp           Gene          Gene         part_of       None   \n",
       "3468           aapp           Gene          Gene     close_match       None   \n",
       "3782           enzy           Gene          Gene  interacts_with       None   \n",
       "3978           gngm           Gene          Gene    derives_into       None   \n",
       "4166           gngm           Gene          Gene        produces       None   \n",
       "4213           gngm           Gene          Gene     close_match       None   \n",
       "4424           gngm           Gene          Gene         affects     causes   \n",
       "4455           gngm           Gene          Gene         affects     causes   \n",
       "4471           gngm           Gene          Gene         affects     causes   \n",
       "4652           gngm           Gene          Gene    derives_into       None   \n",
       "4703           aapp           Gene          Gene  interacts_with       None   \n",
       "4722           phsu           Gene          Gene  interacts_with       None   \n",
       "4771           gngm           Gene          Gene         affects     causes   \n",
       "4961           aapp           Gene          Gene  interacts_with       None   \n",
       "4995           aapp           Gene          Gene  interacts_with       None   \n",
       "\n",
       "                   BMObjAsp BMObjDirect     BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "265   activity_or_abundance   increased     affected_by       caused_by   \n",
       "336   activity_or_abundance   increased     affected_by       caused_by   \n",
       "343                    None        None  interacts_with            None   \n",
       "473                    None        None  interacts_with            None   \n",
       "478                    None        None  interacts_with            None   \n",
       "489   activity_or_abundance   decreased     affected_by       caused_by   \n",
       "516                    None        None  interacts_with            None   \n",
       "566                    None        None   coexists_with            None   \n",
       "591                    None        None        has_part            None   \n",
       "659   activity_or_abundance   decreased     affected_by       caused_by   \n",
       "719                    None        None   coexists_with            None   \n",
       "868   activity_or_abundance   decreased     affected_by       caused_by   \n",
       "902                    None        None     produced_by            None   \n",
       "1174                   None        None   coexists_with            None   \n",
       "1355  activity_or_abundance   increased     affected_by       caused_by   \n",
       "1563  activity_or_abundance   increased     affected_by       caused_by   \n",
       "1831  activity_or_abundance   decreased     affected_by       caused_by   \n",
       "2276                   None        None   coexists_with            None   \n",
       "2796                   None        None     produced_by            None   \n",
       "3239                   None        None        has_part            None   \n",
       "3468                   None        None     close_match            None   \n",
       "3782                   None        None  interacts_with            None   \n",
       "3978                   None        None    derives_from            None   \n",
       "4166                   None        None     produced_by            None   \n",
       "4213                   None        None     close_match            None   \n",
       "4424  activity_or_abundance   decreased     affected_by       caused_by   \n",
       "4455  activity_or_abundance   increased     affected_by       caused_by   \n",
       "4471  activity_or_abundance   increased     affected_by       caused_by   \n",
       "4652                   None        None    derives_from            None   \n",
       "4703                   None        None  interacts_with            None   \n",
       "4722                   None        None  interacts_with            None   \n",
       "4771  activity_or_abundance   increased     affected_by       caused_by   \n",
       "4961                   None        None  interacts_with            None   \n",
       "4995                   None        None  interacts_with            None   \n",
       "\n",
       "             BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "265   activity_or_abundance         increased  \n",
       "336   activity_or_abundance         increased  \n",
       "343                    None              None  \n",
       "473                    None              None  \n",
       "478                    None              None  \n",
       "489   activity_or_abundance         decreased  \n",
       "516                    None              None  \n",
       "566                    None              None  \n",
       "591                    None              None  \n",
       "659   activity_or_abundance         decreased  \n",
       "719                    None              None  \n",
       "868   activity_or_abundance         decreased  \n",
       "902                    None              None  \n",
       "1174                   None              None  \n",
       "1355  activity_or_abundance         increased  \n",
       "1563  activity_or_abundance         increased  \n",
       "1831  activity_or_abundance         decreased  \n",
       "2276                   None              None  \n",
       "2796                   None              None  \n",
       "3239                   None              None  \n",
       "3468                   None              None  \n",
       "3782                   None              None  \n",
       "3978                   None              None  \n",
       "4166                   None              None  \n",
       "4213                   None              None  \n",
       "4424  activity_or_abundance         decreased  \n",
       "4455  activity_or_abundance         increased  \n",
       "4471  activity_or_abundance         increased  \n",
       "4652                   None              None  \n",
       "4703                   None              None  \n",
       "4722                   None              None  \n",
       "4771  activity_or_abundance         increased  \n",
       "4961                   None              None  \n",
       "4995                   None              None  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get an example to check on \n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'ncbigene') &\n",
    "                  (final_metatriples['ObjectPrefix'] == 'ncbigene')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e0079fb",
   "metadata": {},
   "source": [
    "### function for reverse operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "912aa75d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:49:34.972379Z",
     "start_time": "2023-11-01T00:49:34.965784Z"
    }
   },
   "outputs": [],
   "source": [
    "## will use variables set in first section's config\n",
    "## reverse direction: object -> subject\n",
    "\n",
    "def generate_reverse_op(\n",
    "    subj_prefix, obj_prefix,\n",
    "    original_subj, original_pred, original_obj,\n",
    "    biolink_subj, \n",
    "    ## NOTICE THE INVERSES USED HERE\n",
    "    biolink_inverse_pred, \n",
    "    biolink_inverse_qualified_pred, biolink_inverse_subj_asp, biolink_inverse_subj_direct,\n",
    "    biolink_obj):\n",
    "\n",
    "    ## get biolink-model spelling for prefixes\n",
    "    if subj_prefix == 'umls':\n",
    "        biolink_subj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_subj_prefix = 'NCBIGene'\n",
    "        \n",
    "    if obj_prefix == 'umls':\n",
    "        biolink_obj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_obj_prefix = 'NCBIGene'\n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    if (subj_prefix == 'umls') and (obj_prefix == 'umls'):\n",
    "        rev_op_name = f'{original_subj}-{original_pred}-{original_obj}-rev'\n",
    "    else:\n",
    "        rev_op_name = f'{original_subj},{subj_prefix}-{original_pred}-{original_obj},{obj_prefix}-rev'\n",
    "\n",
    "    temp = {\n",
    "        ## reverse direction: object -> subject\n",
    "        rev_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': biolink_obj_prefix,\n",
    "                        'semantic': biolink_obj  ## input is object!\n",
    "                    }\n",
    "                ],\n",
    "                ## notice that object.{obj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'requestBody': {\n",
    "                    'body': {\n",
    "                        'q': '{{ queryInputs }}',\n",
    "                        'scopes': f'object.{obj_prefix}'\n",
    "                    }\n",
    "                },\n",
    "                ## notice that subject.{subj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'parameters': {\n",
    "                    'fields': f'subject.{subj_prefix},predication.pmid,'\n",
    "                               'predication.sentence,subject.name,object.name',\n",
    "                    'size': POST_size,\n",
    "                    'filter': f'predicate:{original_pred} AND '\n",
    "                              f'subject.semantic_type_abbreviation:{original_subj} AND '\n",
    "                              f'_exists_:subject.{subj_prefix} AND '\n",
    "                              f'pmid_count:>{pmid_count}'\n",
    "                },        \n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': biolink_subj_prefix,\n",
    "                        'semantic': biolink_subj  ## output is subject\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                'source': doublequote('infores:semmeddb'),\n",
    "                'response_mapping': {\n",
    "                    ## matches output as object\n",
    "                    ## so umls-obj and ncbigene-obj are the response-mappings generated\n",
    "                    \"$ref\": f'#/components/x-bte-response-mapping/{subj_prefix}-subj'\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has subj aspect and subj direction\n",
    "    if biolink_inverse_qualified_pred:\n",
    "        temp[rev_op_name][0]['qualifiers'] = {\n",
    "            'qualified_predicate': biolink_inverse_qualified_pred,  ## use inverse pred!\n",
    "            ## use subject qualifiers\n",
    "            'subject_aspect_qualifier': biolink_inverse_subj_asp,         \n",
    "            'subject_direction_qualifier': biolink_inverse_subj_direct   \n",
    "        }\n",
    "    \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "955b36a2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:49:36.479854Z",
     "start_time": "2023-11-01T00:49:36.463087Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>treated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5047</th>\n",
       "      <td>umls</td>\n",
       "      <td>irda</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>phsu</td>\n",
       "      <td>DiagnosticAid</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5049</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5051</th>\n",
       "      <td>umls</td>\n",
       "      <td>imft</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>BiologicalEntity</td>\n",
       "      <td>Cell</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5052</th>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5053</th>\n",
       "      <td>umls</td>\n",
       "      <td>elii</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>DiseaseOrPhenotypicFeature</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4322 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5047          umls            irda     COEXISTS_WITH         umls   \n",
       "5049          umls            eico          DISRUPTS         umls   \n",
       "5051          umls            imft    INTERACTS_WITH         umls   \n",
       "5052          umls            aapp           same_as         umls   \n",
       "5053          umls            elii           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject               BiolinkObject  \\\n",
       "0              neop  GrossAnatomicalStructure                     Disease   \n",
       "1              aapp          AnatomicalEntity                 Polypeptide   \n",
       "2              patf  GrossAnatomicalStructure         PathologicalProcess   \n",
       "3              dsyn                 Procedure                     Disease   \n",
       "4              neop                 Procedure                     Disease   \n",
       "...             ...                       ...                         ...   \n",
       "5047           phsu             DiagnosticAid               SmallMolecule   \n",
       "5049           dsyn             SmallMolecule                     Disease   \n",
       "5051           cell          BiologicalEntity                        Cell   \n",
       "5052           topp               Polypeptide                   Procedure   \n",
       "5053           fndg             SmallMolecule  DiseaseOrPhenotypicFeature   \n",
       "\n",
       "              BMPred BMQualPred BMObjAsp BMObjDirect     BM_Inv_Pred  \\\n",
       "0        location_of       None     None        None      located_in   \n",
       "1        location_of       None     None        None      located_in   \n",
       "2        location_of       None     None        None      located_in   \n",
       "3             treats       None     None        None      treated_by   \n",
       "4             treats       None     None        None      treated_by   \n",
       "...              ...        ...      ...         ...             ...   \n",
       "5047   coexists_with       None     None        None   coexists_with   \n",
       "5049        disrupts       None     None        None    disrupted_by   \n",
       "5051  interacts_with       None     None        None  interacts_with   \n",
       "5052     close_match       None     None        None     close_match   \n",
       "5053         affects       None     None        None     affected_by   \n",
       "\n",
       "     BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "0               None           None              None  \n",
       "1               None           None              None  \n",
       "2               None           None              None  \n",
       "3               None           None              None  \n",
       "4               None           None              None  \n",
       "...              ...            ...               ...  \n",
       "5047            None           None              None  \n",
       "5049            None           None              None  \n",
       "5051            None           None              None  \n",
       "5052            None           None              None  \n",
       "5053            None           None              None  \n",
       "\n",
       "[4322 rows x 15 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get an example to check on \n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'umls') &\n",
    "                  (final_metatriples['ObjectPrefix'] == 'umls')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "98c5830f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:49:39.033284Z",
     "start_time": "2023-11-01T00:49:39.028807Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bpoc-LOCATION_OF-neop-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.umls'}},\n",
       "   'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:LOCATION_OF AND subject.semantic_type_abbreviation:bpoc AND _exists_:subject.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'GrossAnatomicalStructure'}],\n",
       "   'predicate': 'located_in',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'}}]}"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - umls\n",
    "generate_reverse_op(\n",
    "    'umls', 'umls',\n",
    "    'bpoc', 'LOCATION_OF', 'neop',\n",
    "    'GrossAnatomicalStructure', 'located_in', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b251cb75",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:50:14.499220Z",
     "start_time": "2023-11-01T00:50:14.494744Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm,ncbigene-ASSOCIATED_WITH-neop,umls-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.umls'}},\n",
       "   'parameters': {'fields': 'subject.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:ASSOCIATED_WITH AND subject.semantic_type_abbreviation:gngm AND _exists_:subject.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'related_to',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-subj'}}]}"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - umls\n",
    "generate_reverse_op(\n",
    "    'ncbigene', 'umls',\n",
    "    'gngm', 'ASSOCIATED_WITH', 'neop',\n",
    "    'Gene', 'related_to', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "586ebf88",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:50:40.117067Z",
     "start_time": "2023-11-01T00:50:40.112864Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp,umls-STIMULATES-gngm,ncbigene-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.ncbigene'}},\n",
       "   'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND subject.semantic_type_abbreviation:aapp AND _exists_:subject.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
       "   'predicate': 'affected_by',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'},\n",
       "   'qualifiers': {'qualified_predicate': 'caused_by',\n",
       "    'subject_aspect_qualifier': 'activity_or_abundance',\n",
       "    'subject_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - ncbigene\n",
    "generate_reverse_op(\n",
    "    'umls', 'ncbigene',\n",
    "    'aapp', 'STIMULATES', 'gngm',\n",
    "    'Polypeptide', 'affected_by', \n",
    "    'caused_by', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "a105373c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:51:10.701738Z",
     "start_time": "2023-11-01T00:51:10.697374Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm,ncbigene-STIMULATES-gngm,ncbigene-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.ncbigene'}},\n",
       "   'parameters': {'fields': 'subject.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND subject.semantic_type_abbreviation:gngm AND _exists_:subject.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affected_by',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-subj'},\n",
       "   'qualifiers': {'qualified_predicate': 'caused_by',\n",
       "    'subject_aspect_qualifier': 'activity_or_abundance',\n",
       "    'subject_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - ncbigene\n",
    "generate_reverse_op(\n",
    "    'ncbigene', 'ncbigene',\n",
    "    'gngm', 'STIMULATES', 'gngm',\n",
    "    'Gene', 'affected_by', \n",
    "    'caused_by', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6f70b8c",
   "metadata": {},
   "source": [
    "### generating all operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "db9f9f81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:51:41.118242Z",
     "start_time": "2023-11-01T00:51:41.111198Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_all_operations(metatriple_df):\n",
    "    op_tracking = set()\n",
    "    \n",
    "    saved = dict()\n",
    "    ## iterate through rows of metatriple dataframe\n",
    "    for row in metatriple_df.itertuples(index = False):\n",
    "        \n",
    "        ## forward: only make operation if it's not going to be a dupe\n",
    "        ##  dupes happen because we don't set the SEMMED semantic type of the subject (input / starting point)\n",
    "        ##  and multiple subject SEMMED semantic types can match 1 Biolink-Subject-category\n",
    "        ##  variables to have in common are the prefixes, semmed predicate, semmed object\n",
    "        forward_op_record = f\"{row.BiolinkSubject},{row.SubjectPrefix}-{row.OriginalPredicate}-{row.OriginalObject},{row.ObjectPrefix}\"\n",
    "        \n",
    "        if forward_op_record not in op_tracking:\n",
    "            op_tracking.add(forward_op_record)\n",
    "            \n",
    "            saved.update(generate_forward_op(subj_prefix = row.SubjectPrefix,\n",
    "                                             obj_prefix = row.ObjectPrefix,\n",
    "                                             original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_pred = row.BMPred,\n",
    "                                             biolink_qualified_pred = row.BMQualPred,\n",
    "                                             biolink_obj_asp = row.BMObjAsp,\n",
    "                                             biolink_obj_direct = row.BMObjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                            ))        \n",
    "        \n",
    "        ## reverse: make operation if it's not going to be a dupe\n",
    "        ##  dupes happen because we don't set the SEMMED semantic type of the object (input / starting point)\n",
    "        ##  and multiple object SEMMED semantic types can match 1 Biolink-Object-category\n",
    "        ##  variables to have in common are the prefixes, semmed predicate, semmed subject\n",
    "        reverse_op_record = f\"rev-{row.BiolinkObject},{row.ObjectPrefix}-{row.OriginalPredicate}-{row.OriginalSubject},{row.SubjectPrefix}\"\n",
    "        \n",
    "        if reverse_op_record not in op_tracking:\n",
    "            op_tracking.add(reverse_op_record)\n",
    "            saved.update(generate_reverse_op(subj_prefix = row.SubjectPrefix,\n",
    "                                             obj_prefix = row.ObjectPrefix,\n",
    "                                             original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_inverse_pred = row.BM_Inv_Pred,\n",
    "                                             biolink_inverse_qualified_pred = row.BM_Inv_QualPred,\n",
    "                                             biolink_inverse_subj_asp = row.BM_Inv_SubjAsp,\n",
    "                                             biolink_inverse_subj_direct = row.BM_Inv_SubjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                             ))           \n",
    "            \n",
    "    final = {\"x-bte-kgs-operations\": saved}\n",
    "    return final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ec9a4",
   "metadata": {},
   "source": [
    "Get the file made!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "3856781d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:51:44.910116Z",
     "start_time": "2023-11-01T00:51:44.792841Z"
    }
   },
   "outputs": [],
   "source": [
    "kgs_operations = generate_all_operations(final_metatriples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "fdcd4202",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:51:46.527854Z",
     "start_time": "2023-11-01T00:51:46.524319Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4715"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "5495d772",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:52:39.431821Z",
     "start_time": "2023-11-01T00:52:39.427896Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orch,umls-INHIBITS-gngm,ncbigene-rev\n",
      "[{'supportBatch': True,\n",
      "  'useTemplating': True,\n",
      "  'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
      "  'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
      "                           'scopes': 'object.ncbigene'}},\n",
      "  'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
      "                 'size': 1000,\n",
      "                 'filter': 'predicate:INHIBITS AND '\n",
      "                           'subject.semantic_type_abbreviation:orch AND '\n",
      "                           '_exists_:subject.umls AND pmid_count:>3'},\n",
      "  'outputs': [{'id': 'UMLS', 'semantic': 'SmallMolecule'}],\n",
      "  'predicate': 'affected_by',\n",
      "  'source': 'infores:semmeddb',\n",
      "  'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'},\n",
      "  'qualifiers': {'qualified_predicate': 'caused_by',\n",
      "                 'subject_aspect_qualifier': 'activity_or_abundance',\n",
      "                 'subject_direction_qualifier': 'decreased'}}]\n"
     ]
    }
   ],
   "source": [
    "## double-check qualifier + rev operation is written properly?\n",
    "for i in kgs_operations['x-bte-kgs-operations'].keys():\n",
    "    if ('INHIBITS' in i) and ('rev' in i) and ('ncbigene' in i):\n",
    "        print(i)\n",
    "        pprint.pprint(kgs_operations['x-bte-kgs-operations'][i], sort_dicts = False)\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd057af2",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* it's cool how condensed the operations become, due to the way querying is done (keeping track of not creating duplicated operations using the sets in generate_all_operations function)\n",
    "* set where to download the yamls in the code chunks below\n",
    "  * operations_path\n",
    "  * list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "239dce9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:53:30.164944Z",
     "start_time": "2023-11-01T00:53:22.224706Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.boolean_representation = ['False', 'True']\n",
    "\n",
    "## this path was set in the config section at the beginning\n",
    "yaml.dump(kgs_operations, operations_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c946247a",
   "metadata": {},
   "source": [
    "Wait a sec! Need the operations list too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "27403608",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:53:32.886628Z",
     "start_time": "2023-11-01T00:53:32.883648Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_kgs_operations_list(operations_dict):\n",
    "    kgs_op_list = []\n",
    "    for key in operations_dict.keys():\n",
    "        kgs_op_list.append( {\"$ref\": f\"#/components/x-bte-kgs-operations/{key}\"} )\n",
    "    final2 = {\"x-bte-kgs-operations\": kgs_op_list}\n",
    "    return final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "4bec97ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:53:33.912604Z",
     "start_time": "2023-11-01T00:53:33.909180Z"
    }
   },
   "outputs": [],
   "source": [
    "operations_list = generate_kgs_operations_list(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "e64c34db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2023-11-01T00:53:35.319258Z",
     "start_time": "2023-11-01T00:53:34.753670Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.dump(operations_list, list_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421fbcdc",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* now the yaml segments downloaded have to be indented manually and inserted into the correct sections of the smartapi yaml...\n",
    "  * It's easier to do with an IDE like Visual Code studio where one can select large sections of text\n",
    "  * the amount of indent to do and where to put things is specified in the [yaml that acts as a template](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)\n",
    "  * the finished file is meant to be [here](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) so one could select-paste the sections directly there"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa95cb1",
   "metadata": {},
   "source": [
    "## defunct: Notes on previously removed SEMMED semantic types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb04a8e",
   "metadata": {},
   "source": [
    "* amas: (Amino Acid Sequence) looks like protein \"domains\". Examples: Nuclear Export Signals, DNA Binding Domain\n",
    "* acty (Activity) Examples: War, Retirement, Euthanasia, Lifting\n",
    "* dora (Daily or Recreational Activity) Examples: Physical activity, Light Exercise, Relaxation\n",
    "* edac (Educational Activity) Examples: Training, Medical Residencies\n",
    "* gora (Governmental or Regulatory Activity) Examples: Health Care Reform, Advisory Committees\n",
    "* hlca (Health Care Activity) Examples: follow-up, Diagnosis\n",
    "* mcha (Machine Activity) Examples: Refridgeration, Neural Network Simulation\n",
    "* ocac (Occupational Activity) Examples: Promotion, Work, Mining\n",
    "* resa (Research Activity) Examples: Clinical Trials, research study\n",
    "* aggp (Age Group) Examples: Infant, Child, Adult, Elderly\n",
    "* famg (Family Group) Examples: spouse, Sister, Foster Parent\n",
    "* podg (Patient or Disabled Group) Examples: Patients\n",
    "* prog (Professional or Occupational Group) Examples: Administrators, Employee, Author\n",
    "* algae (Algae)\n",
    "* invt (Invertebrate)\n",
    "* orgm (Organism)\n",
    "* rich (Rickettsia or Chlamydia)\n",
    "* amph (Amphibian) Examples: Toad, Bufo boreas, Anura\n",
    "* anim (Animal) Examples: Animals, Laboratory /  Control Animal\n",
    "* arch (Archaeon) Examples: Archaea, halophilic bacteria, Thermoplasma acidophilum\n",
    "* bact (Bacterium) Examples: Escherichia coli, Salmonella, Borrelia burgdorferi\n",
    "* bdsu (Body Substance) too general. Examples: Urine, Milk, Lymph, Urine specimen\n",
    "* bdsy (Body System) too general. Examples: hypothalamic-pituitary-adrenal axis, Neurosecretory Systems\n",
    "* bird (Bird) Examples: Geese, Passeriformes, Raptors\n",
    "* blor (Body Location or Region) too general. Examples: Hepatic, Lysosomal, Cytoplasmic\n",
    "* bmod (Biomedical Occupation or Discipline) Examples: Medicine, Dentistry, Midwifery\n",
    "* bsoj (Body Space or Junction) too general. Examples: Compartments, Synapses, Cistern\n",
    "* chvf (Chemical Viewed Functionally) too general. Examples: inhibitors, antagonists, Agent\n",
    "* chvs (Chemical Viewed Structurally) too general. Examples: particle, solid state, vapor\n",
    "* euka (Eukaryote) Examples: Wasps, Protozoan parasite\n",
    "* ffas (Fully Formed Anatomical Structure) Examples: Carcass\n",
    "* fish (Fish) Examples: Eels, Fishes, Electric Fish\n",
    "* fngs (Fungus) Examples: Saccharomyces cerevisiae, Alternaria brassicicola, fungus\n",
    "* humn (Human) Examples: Family, Patients, Males\n",
    "* irda (Indicator, Reagent, or Diagnostic Aid) Examples: Fluorescent Probes, Chelating Agents\n",
    "* mamm (Mammal) Examples: Rattus norvegicus, Felis catus, Mus\n",
    "* ocdi (Occupation or Discipline) Examples: Science, Politics\n",
    "* plnt (Plant) Examples: Chrysanthemum x morifolium, Pollen, Oryza sativa\n",
    "* rept (Reptile) Examples: Snakes, Turtles, Reptiles\n",
    "* sbst (Substance) too general. Examples: Materials, Plastics, Photons, Substance\n",
    "* virs (Virus) Examples: Herpesvirus 4, Human / GB virus C / Herpesviridae\n",
    "* vtbt (Vertebrate) Examples: Vertebrates / Poikilotherm, NOS\n",
    "* anst (Anatomical Structure) Examples: Entire fetus, Whole body, Cadaver\n",
    "* bhvr (Behavior) too general. Examples: Sexuality, Nest Building, Behavioral phenotype\n",
    "* inbe (Individual Behavior) too general. Examples: impulsivity, Habits, Performance\n",
    "* menp (Mental Process) too general. Examples: mind control, Learning, experience\n",
    "* socb (Social Behavior) too general. Examples: Communication, Gestures, Marriage\n",
    "* biof (Biologic Function) too general. Examples: dose-response relationship, Pharmacodynamics, Anabolism\n",
    "* eehu (Environmental Effect of Humans) too general. Examples: Sewage, Pollution, Smoke\n",
    "* hcpp (Human-caused Phenomenon or Process) too general. Now not in API. Examples: particle beam, Conferences, Victimization\n",
    "* lbtr (Laboratory or Test Result) too general. Examples: False Positive Reactions, Bone Density, Serum Calcium Level\n",
    "* npop (Natural Phenomenon or Process) too general. Examples: Floods, Fluorescence, Freezing\n",
    "* phpr (Phenomenon or Process) too general. Examples: Disasters, Acceleration, Feedback\n",
    "* bodm (Biomedical or Dental Material) too general. Examples: Pill, Gel, Talc, calcium phosphate\n",
    "* drdd (Drug Delivery Device) too general. Examples: Epipen, Skin Patch, Lilly cyanide antidote kit\n",
    "* medd (Medical Device) too general. Examples: Implants / Denture, Overlay / Silicone gel implant / Swab\n",
    "* resd (Research Device) too general. Examples: Study models, Slide\n",
    "* emst (Embryonic Structure) Examples: Chick Embryo, Blastocyst structure, Placenta\n",
    "* tisu (Tissue) Examples: Tissue specimen, Blood, Human tissue, Mucous Membrane\n",
    "* genf (Genetic Function) too general. Examples: Transcription, Genetic / Transcriptional Activation / Recombination, Genetic\n",
    "* chem (Chemical) too general. Examples: Chemicals, Acids, Ligands, Ozone\n",
    "* clas (Classification) too general. Examples: Research Diagnostic Criteria, Group C\n",
    "* ftcn (Functional Concept) too general. Examples: Techniques, Intravenous Route of Drug Administration\n",
    "* idcn (Idea or Concept) too general. Examples: Significant, subject, Data, Owner\n",
    "* qlco (Qualitative Concept) too general. Examples: Effect, Associated with, Advanced phase\n",
    "* qnco (Quantitative Concept) too general. Examples: Calibration, occurrence, degrees Celsius\n",
    "* rnlw (Regulation or Law) too general. Examples: Medicare, Medicaid, regulatory\n",
    "* spco (Spatial Concept) too general. Examples: Structure, Longitudinal, Asymmetry\n",
    "* tmco (Temporal Concept) too general. Examples: New, /period, 24 Hours\n",
    "* clna (Clinical Attribute) too general. Examples: response, Renin secretion, BAND PATTERN\n",
    "* lbpr (Laboratory Procedure) too general. Examples: Western Blot, Radioimmunoassay, Staining method\n",
    "* mbrt (Molecular Biology Research Technique) too general. Examples: Polymerase Chain Reaction / Blotting, Northern\n",
    "* elii (Element, Ion, or Isotope) too general. Examples: Atom, Aluminum, Superoxides\n",
    "* emod (Experimental Model of Disease) too general. Examples: Experimental Autoimmune Encephalomyelitis, Rodent Model\n",
    "* rcpt (Receptor) too general. Examples: Binding Sites / Receptors, Metabotropic Glutamate\n",
    "* evnt (Event) too general. Now not in API. Examples: Stressful Events\n",
    "* fndg (Finding) too general. Examples: spinal cord; lesion, Normal birth weight, Sedentary job\n",
    "* geoa (Geographic Area) too general. Examples: Country, Canada\n",
    "* grup (Group) too general. Examples: Human, Individual\n",
    "* hcro (Health Care Related Organization) too general. Examples: Hospitals, Health System\n",
    "* orgt (Organization) too general. Examples: United Nations, Organization administrative structures\n",
    "* pros (Professional Society) too general. Examples: Professional Organizations, American Nurses' Association\n",
    "* shro (Self-help or Relief Organization) too general. Examples: Social Welfare, Support Groups\n",
    "* inpr (Intellectual Product) too general. Examples: Methodology, Study models\n",
    "* mnob (Manufactured Object) too general. Examples: Glass, Manuals\n",
    "* mosq (Molecular Sequence) too general. Now not in API. Examples: Genetic Code\n",
    "* nusq (Nucleotide Sequence) too general. Examples: Base Sequence, DNA Sequence, 22q11\n",
    "* orga (Organism Attribute) too general. Examples: Ability, Body Composition\n",
    "* popg (Population Group) too general. Examples: Male population group, Woman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf7f82a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "796.5px",
    "left": "39px",
    "top": "109.633px",
    "width": "319px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
