{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "47e7307b",
   "metadata": {},
   "source": [
    "# Auto-generate x-bte annotations for BioThings SEMMEDDB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58ab5663",
   "metadata": {},
   "source": [
    "This notebook walks a developer through the process of taking the [BioThings SEMMEDDB](https://biothings.transltr.io/semmeddb) meta-triple + API-doc-count file [generated during the API deployment process](https://github.com/biothings/semmeddb/blob/main/pred_meta_triples.ipynb) and using this data to auto-generate the x-bte operations for [BTE](https://github.com/biothings/BioThings_Explorer_TRAPI). This is needed for BTE to query + process the responses from the BioThings SEMMEDDB API. \n",
    "\n",
    "---\n",
    "\n",
    "When one sees this:  \n",
    "# PAUSE\n",
    "\n",
    "read the accompanying text that will explain what the developer needs to do before running the code chunks below that text block.\n",
    "\n",
    "---\n",
    "\n",
    "The [yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_operations.yaml) [segments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/generated_list.yaml) generated by this notebook are added to [this file](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml) to make the [SmartAPI yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) for BioThings SEMMEDDB API, which is registered in the [SmartAPI Registry](https://smart-api.info/registry)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43745613",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712186ee",
   "metadata": {},
   "source": [
    "Requirements:\n",
    "* install [Biolink-model Toolkit (bmt)](https://github.com/biolink/biolink-model-toolkit/). I'm [installing as a user](https://biolink.github.io/biolink-model-toolkit/intro/intro.html#for-users) with pip (`pip install bmt`). Using release 1.4.5 at the moment.\n",
    "* install ruamel.yaml with [pip](https://pypi.org/project/ruamel.yaml/). Using 0.18.15 at the moment. It's more recent than the version from [conda-forge](https://anaconda.org/conda-forge/ruamel_yaml). The documentation is [here](https://yaml.readthedocs.io/en/latest/)\n",
    "\n",
    "Files to obtain before starting:\n",
    "* Get the meta-triple + API-doc-count file from the dev for BioThings SEMMEDDB. I've saved the file I'm using to the same directory as this notebook. This file was provided 2025-08-08 by Nichollette, and it includes counts of matching API documents/records for each metatriple. Generally, the total set of meta-triples should be fairly constant between data releases, but the doc counts may differ. \n",
    "* Get SEMMEDDB SRDEF file needed for interpreting SEMMED semantic types: I'm using 2023AA right now. \n",
    "  * download the UMLS semantic network files from [here](https://www.nlm.nih.gov/research/umls/knowledge_sources/semantic_network/index.html)\n",
    "  * information on the SRDEF file [here](https://www.ncbi.nlm.nih.gov/books/NBK9679/#ch05.sec5.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11497815",
   "metadata": {},
   "source": [
    "### Set config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5ee5aac",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:00:40.722071Z",
     "start_time": "2025-08-20T07:00:35.488498Z"
    }
   },
   "outputs": [],
   "source": [
    "## CX: allows multiple lines of code to print from one code block\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "InteractiveShell.ast_node_interactivity = \"all\"\n",
    "\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "## see above for install instructions\n",
    "from bmt import Toolkit\n",
    "import ruamel.yaml as ryml    ## use ruamel_yaml if using the conda-forge version\n",
    "\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "## used in trying things out\n",
    "# import re\n",
    "# import matplotlib.pyplot as plt   ## for plotting doc counts\n",
    "\n",
    "\n",
    "## set up ruamel.yaml\n",
    "yaml=ryml.YAML()\n",
    "folded = ryml.scalarstring.FoldedScalarString\n",
    "doublequote = ryml.scalarstring.DoubleQuotedScalarString"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0af532d5",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* Set variables used in x-bte annotation\n",
    "  * pmid_count: discussed in [this Translator issue](https://github.com/NCATSTranslator/Feedback/issues/100#issuecomment-1632806388)\n",
    "* Set the `metatriples_path` to your local metatriples + API-doc-counts file\n",
    "* Set the `srdef_location` to your local SRDEF file\n",
    "* Check the url to the [Translator-curated exclusions (tsv file)](https://github.com/biolink/biolink-model/blob/master/SEMMEDDB_exclude_list.tsv) to remove some SEMMEDDB semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "73408978",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:02:27.313294Z",
     "start_time": "2025-08-20T07:02:27.040814Z"
    }
   },
   "outputs": [],
   "source": [
    "## variables used for x-bte annotation\n",
    "## set size parameter for biothings POST query. This is 1000 unless this API has custom settings for this\n",
    "POST_size = 1000  \n",
    "## pmid_count: only allow numbers greater than this\n",
    "pmid_count = 3\n",
    "\n",
    "## metatriples + API-doc-counts file\n",
    "metatriples_path = pathlib.Path.home().joinpath(\n",
    "    \"Desktop\", \"translator-api-registry\", \"semmeddb\", \n",
    "    \"meta_triple_document_stat_20250410_165534.tsv\")\n",
    "\n",
    "imported_metatriples = pd.read_csv(metatriples_path, sep=\"\\t\")\n",
    "\n",
    "\n",
    "## SEMMED Semantic info file\n",
    "srdef_location = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"RawDataFiles\", \"SEMMEDDB\", \"2023AA\", \"SRDEF\")\n",
    "\n",
    "srdef = pd.read_csv(srdef_location, sep=\"|\", header=0, index_col=False,\n",
    "                    names=[\"Record Type (RT)\",\n",
    "                           \"Unique Identifier (UI)\",\n",
    "                           \"Full Name (STY/RL)\", \n",
    "                           \"Tree Number (STN/RTN)\", \n",
    "                           \"Definition (DEF)\",\n",
    "                           \"Examples (EX)\",\n",
    "                           \"Usage Note (UN)\",\n",
    "                           \"Non-Human Flag (NH)\",\n",
    "                           \"Abbreviation (ABR)\",\n",
    "                           \"Inverse Relation (RIN)\"])\n",
    "\n",
    "\n",
    "## import the Translator-curated exclusions tsv\n",
    "imported_exclusions = pd.read_csv(\n",
    "    'https://raw.githubusercontent.com/biolink/biolink-model/master/SEMMEDDB_exclude_list.tsv',\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "\n",
    "## where the output yamls go\n",
    "operations_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_operations.yaml\")\n",
    "\n",
    "list_path = pathlib.Path.home().joinpath(\n",
    "            \"Desktop\", \"translator-api-registry\", \"semmeddb\", \"generated_list.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8150ed6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:02:33.285894Z",
     "start_time": "2025-08-20T07:02:33.279676Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(34831, 7)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "57d2dc4a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:02:37.517716Z",
     "start_time": "2025-08-20T07:02:37.504992Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2530738</td>\n",
       "      <td>2321777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1324267</td>\n",
       "      <td>1160368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1204184</td>\n",
       "      <td>1153238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1139224</td>\n",
       "      <td>1045335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>923819</td>\n",
       "      <td>846909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            fndg   PROCESS_OF          umls           humn   \n",
       "2           umls            bpoc      PART_OF          umls           mamm   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2530738      2321777  \n",
       "1  1324267      1160368  \n",
       "2  1204184      1153238  \n",
       "3  1139224      1045335  \n",
       "4   923819       846909  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b68180",
   "metadata": {},
   "source": [
    "## Basic Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32b8cf0b",
   "metadata": {},
   "source": [
    "Notes:\n",
    "* This BioThings API's parser already filters out all API docs where novelty == 0, since those with novelty == 0 probably aren't very helpful / interesting to Translator. The entities with novelty == 0 are [listed](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/dbinfo.html) in the SEMMEDDB GENERIC_CONCEPT table files, which can be downloaded [here](https://lhncbc.nlm.nih.gov/ii/tools/SemRep_SemMedDB_SKR/SemMedDB_download.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "403fe7b4",
   "metadata": {},
   "source": [
    "First, I'm doing a quick look at what exists in this meta-triple file:\n",
    "* two ID-namespaces for subject + object: umls + ncbigene\n",
    "  * will want special handling for ncbigene: biolink-model only recognizes NCBIGene as a valid namespace for specific categories, like Gene \n",
    "* all semmeddb semantic types are 4-letter abbreviations / codes\n",
    "  * 132 unique values for subject\n",
    "  * 131 unique values for object\n",
    "* semmeddb predicates are a mix of different cases...\n",
    "  * 63 unique values\n",
    "\n",
    "Side-note: the stats from the [archived website](https://web.archive.org/web/20241127183822/https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say semmeddb's 2023AA vocab has 127 semantic types and 54 predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "181df62c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:03:06.739380Z",
     "start_time": "2025-08-20T07:03:06.716765Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['umls', 'ncbigene'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['umls', 'ncbigene'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['aapp', 'gngm', 'bacs', 'horm', 'phsu', 'enzy', 'imft', 'rcpt',\n",
       "       'nsba', 'nusq', 'orch', 'nnon', 'lipd', 'irda', 'hops', 'carb',\n",
       "       'bodm', 'antb', 'vita', 'strd'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['aapp', 'gngm', 'enzy', 'phsu', 'bacs', 'horm', 'irda', 'imft',\n",
       "       'nsba', 'nusq', 'rcpt', 'nnon', 'antb', 'orch', 'hops', 'bodm'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.SUBJECT_PREFIX.unique()\n",
    "imported_metatriples.OBJECT_PREFIX.unique()\n",
    "\n",
    "imported_metatriples[imported_metatriples['SUBJECT_PREFIX'] == 'ncbigene'].SUBJECT_SEMTYPE.unique()\n",
    "imported_metatriples[imported_metatriples['OBJECT_PREFIX'] == 'ncbigene'].OBJECT_SEMTYPE.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a52d0b4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:03:40.668194Z",
     "start_time": "2025-08-20T07:03:40.652899Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "132"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "131"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "63"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## 132\n",
    "len(object_semtypes)  ## 131\n",
    "len(predicates)       ## 63"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d0dd2ed7",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:06:53.514593Z",
     "start_time": "2025-08-20T07:06:53.507465Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'evnt',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.SUBJECT_SEMTYPE.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b27b5f43",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:07:04.951909Z",
     "start_time": "2025-08-20T07:07:04.943617Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp',\n",
       " 'acab',\n",
       " 'acty',\n",
       " 'aggp',\n",
       " 'alga',\n",
       " 'amas',\n",
       " 'amph',\n",
       " 'anab',\n",
       " 'anim',\n",
       " 'anst',\n",
       " 'antb',\n",
       " 'arch',\n",
       " 'bacs',\n",
       " 'bact',\n",
       " 'bdsu',\n",
       " 'bdsy',\n",
       " 'bhvr',\n",
       " 'biof',\n",
       " 'bird',\n",
       " 'blor',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'bpoc',\n",
       " 'bsoj',\n",
       " 'carb',\n",
       " 'celc',\n",
       " 'celf',\n",
       " 'cell',\n",
       " 'cgab',\n",
       " 'chem',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'clas',\n",
       " 'clna',\n",
       " 'clnd',\n",
       " 'comd',\n",
       " 'diap',\n",
       " 'dora',\n",
       " 'drdd',\n",
       " 'dsyn',\n",
       " 'edac',\n",
       " 'eehu',\n",
       " 'eico',\n",
       " 'elii',\n",
       " 'emod',\n",
       " 'emst',\n",
       " 'enzy',\n",
       " 'euka',\n",
       " 'famg',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'fndg',\n",
       " 'fngs',\n",
       " 'food',\n",
       " 'ftcn',\n",
       " 'genf',\n",
       " 'geoa',\n",
       " 'gngm',\n",
       " 'gora',\n",
       " 'grup',\n",
       " 'hcpp',\n",
       " 'hcro',\n",
       " 'hlca',\n",
       " 'hops',\n",
       " 'horm',\n",
       " 'humn',\n",
       " 'idcn',\n",
       " 'imft',\n",
       " 'inbe',\n",
       " 'inch',\n",
       " 'inpo',\n",
       " 'inpr',\n",
       " 'invt',\n",
       " 'irda',\n",
       " 'lang',\n",
       " 'lbpr',\n",
       " 'lbtr',\n",
       " 'lipd',\n",
       " 'mamm',\n",
       " 'mbrt',\n",
       " 'mcha',\n",
       " 'medd',\n",
       " 'menp',\n",
       " 'mnob',\n",
       " 'mobd',\n",
       " 'moft',\n",
       " 'mosq',\n",
       " 'neop',\n",
       " 'nnon',\n",
       " 'npop',\n",
       " 'nsba',\n",
       " 'nusq',\n",
       " 'ocac',\n",
       " 'ocdi',\n",
       " 'opco',\n",
       " 'orch',\n",
       " 'orga',\n",
       " 'orgf',\n",
       " 'orgm',\n",
       " 'orgt',\n",
       " 'ortf',\n",
       " 'patf',\n",
       " 'phob',\n",
       " 'phpr',\n",
       " 'phsf',\n",
       " 'phsu',\n",
       " 'plnt',\n",
       " 'podg',\n",
       " 'popg',\n",
       " 'prog',\n",
       " 'pros',\n",
       " 'qlco',\n",
       " 'qnco',\n",
       " 'rcpt',\n",
       " 'rept',\n",
       " 'resa',\n",
       " 'resd',\n",
       " 'rich',\n",
       " 'rnlw',\n",
       " 'sbst',\n",
       " 'shro',\n",
       " 'socb',\n",
       " 'sosy',\n",
       " 'spco',\n",
       " 'strd',\n",
       " 'tisu',\n",
       " 'tmco',\n",
       " 'topp',\n",
       " 'virs',\n",
       " 'vita',\n",
       " 'vtbt'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.OBJECT_SEMTYPE.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5765d868",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:07:10.604886Z",
     "start_time": "2025-08-20T07:07:10.597797Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ADMINISTERED_TO',\n",
       " 'AFFECTS',\n",
       " 'ASSOCIATED_WITH',\n",
       " 'AUGMENTS',\n",
       " 'CAUSES',\n",
       " 'COEXISTS_WITH',\n",
       " 'COMPLICATES',\n",
       " 'CONVERTS_TO',\n",
       " 'DIAGNOSES',\n",
       " 'DISRUPTS',\n",
       " 'INHIBITS',\n",
       " 'INTERACTS_WITH',\n",
       " 'ISA',\n",
       " 'LOCATION_OF',\n",
       " 'MANIFESTATION_OF',\n",
       " 'MEASUREMENT_OF',\n",
       " 'MEASURES',\n",
       " 'METHOD_OF',\n",
       " 'NEG_ADMINISTERED_TO',\n",
       " 'NEG_AFFECTS',\n",
       " 'NEG_ASSOCIATED_WITH',\n",
       " 'NEG_AUGMENTS',\n",
       " 'NEG_CAUSES',\n",
       " 'NEG_COEXISTS_WITH',\n",
       " 'NEG_COMPLICATES',\n",
       " 'NEG_CONVERTS_TO',\n",
       " 'NEG_DIAGNOSES',\n",
       " 'NEG_DISRUPTS',\n",
       " 'NEG_INHIBITS',\n",
       " 'NEG_INTERACTS_WITH',\n",
       " 'NEG_ISA',\n",
       " 'NEG_LOCATION_OF',\n",
       " 'NEG_MANIFESTATION_OF',\n",
       " 'NEG_MEASUREMENT_OF',\n",
       " 'NEG_MEASURES',\n",
       " 'NEG_METHOD_OF',\n",
       " 'NEG_OCCURS_IN',\n",
       " 'NEG_PART_OF',\n",
       " 'NEG_PRECEDES',\n",
       " 'NEG_PREDISPOSES',\n",
       " 'NEG_PREVENTS',\n",
       " 'NEG_PROCESS_OF',\n",
       " 'NEG_PRODUCES',\n",
       " 'NEG_STIMULATES',\n",
       " 'NEG_TREATS',\n",
       " 'NEG_USES',\n",
       " 'NEG_higher_than',\n",
       " 'NEG_lower_than',\n",
       " 'NEG_same_as',\n",
       " 'OCCURS_IN',\n",
       " 'PART_OF',\n",
       " 'PRECEDES',\n",
       " 'PREDISPOSES',\n",
       " 'PREVENTS',\n",
       " 'PROCESS_OF',\n",
       " 'PRODUCES',\n",
       " 'STIMULATES',\n",
       " 'TREATS',\n",
       " 'USES',\n",
       " 'compared_with',\n",
       " 'higher_than',\n",
       " 'lower_than',\n",
       " 'same_as'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(imported_metatriples.PREDICATE.unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff20381b",
   "metadata": {},
   "source": [
    "### Remove Predicates we don't want to make operations from"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "106d9b99",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* Decide what predicates you want to remove\n",
    "\n",
    "**Current logic**\n",
    "\n",
    "I'm using biolink-model master branch right now. \n",
    "\n",
    "With biolink-model >= 3.5.1, Translator has reviewed some predicates for domain-predicate and predicate-range combo exclusions - ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "We'll want to keep the predicates that have been reviewed and only do the custom removals below for predicates that:\n",
    "* haven't been reviewed yet\n",
    "* are problematic to use right now: \n",
    "    * Translator currently isn't great with negation, so removing those\n",
    "    * I didn't find the following relationships useful: \n",
    "      * ISA (subclass relationship)\n",
    "      * lower_than / higher_than (these two entities were compared and 1 was better (higher) or worse (lower) than the other)\n",
    "      * compared_with...\n",
    "    \n",
    "And we'll want to adjust the lists below if Translator reviews more predicates in the future. \n",
    "\n",
    "Old notes:\n",
    "* same_as (previously excluded): its meaning seems to be \"equivalent / it did just as well as some other thing\"\n",
    "* If you see `[\"VERB\", \"NOM\", \"PREP\"]`, these seem to be practice phrases that we wouldn't want to use. see the [article](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-486#Sec26)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dec1bcc5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:08:54.889673Z",
     "start_time": "2025-08-20T07:08:54.882188Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "negative_preds = []\n",
    "\n",
    "for i in imported_metatriples.PREDICATE.unique():\n",
    "    if i.startswith('NEG_'):\n",
    "        negative_preds.append(i)\n",
    "        \n",
    "len(negative_preds)\n",
    "        \n",
    "        \n",
    "dont_like = [\"ISA\", \"lower_than\", \"higher_than\", \"compared_with\"]\n",
    "\n",
    "## so 31 + 4 = 35 predicates are being removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0ab643b0",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:09:04.446849Z",
     "start_time": "2025-08-20T07:09:04.436360Z"
    }
   },
   "outputs": [],
   "source": [
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(negative_preds)]\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(dont_like)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f3d145",
   "metadata": {},
   "source": [
    "now look at the stats. \n",
    "\n",
    "(interesting since the stats from the [archived website](https://web.archive.org/web/20241127183822/https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html) say 127 semantic types and 54 predicates for 2023AA. 54 / 2 = 27 predicates without negation + 1 for same_as inclusion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6244458",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:09:08.491543Z",
     "start_time": "2025-08-20T07:09:08.480327Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 132, now 127: decreased by 5\n",
    "len(object_semtypes)  ## was 131, now 127: decreased by 4\n",
    "len(predicates)       ## was 63, now 28: decreased by 35 (was expected)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94e31d94",
   "metadata": {},
   "source": [
    "so....that's how many meta-triples we have to go forward with..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24036b97",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:10:14.339836Z",
     "start_time": "2025-08-20T07:10:14.335595Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18368, 7)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape\n",
    "## so that's still a lot...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "949ccb58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:10:17.007579Z",
     "start_time": "2025-08-20T07:10:16.997886Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2530738</td>\n",
       "      <td>2321777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1324267</td>\n",
       "      <td>1160368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1204184</td>\n",
       "      <td>1153238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1139224</td>\n",
       "      <td>1045335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>923819</td>\n",
       "      <td>846909</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            fndg   PROCESS_OF          umls           humn   \n",
       "2           umls            bpoc      PART_OF          umls           mamm   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2530738      2321777  \n",
       "1  1324267      1160368  \n",
       "2  1204184      1153238  \n",
       "3  1139224      1045335  \n",
       "4   923819       846909  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e998d9e8",
   "metadata": {},
   "source": [
    "## Mapping and Specific Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "330bba7a",
   "metadata": {},
   "source": [
    "First, I'm going to map SEMMED semantic info to biolink-model - this is a legacy from when I used the full names + mapping to decide what SEMMED semantic types and predicates to keep for x-bte annotation with. Now, I'll do minimal adjustments.  \n",
    "\n",
    "Then I'll use Translator-curated exclusions to remove some SEMMED semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d470fa70",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:10:31.330420Z",
     "start_time": "2025-08-20T07:10:31.319507Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Record Type (RT)</th>\n",
       "      <th>Unique Identifier (UI)</th>\n",
       "      <th>Full Name (STY/RL)</th>\n",
       "      <th>Tree Number (STN/RTN)</th>\n",
       "      <th>Definition (DEF)</th>\n",
       "      <th>Examples (EX)</th>\n",
       "      <th>Usage Note (UN)</th>\n",
       "      <th>Non-Human Flag (NH)</th>\n",
       "      <th>Abbreviation (ABR)</th>\n",
       "      <th>Inverse Relation (RIN)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>STY</td>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>A1.4.1.2.1.7</td>\n",
       "      <td>Amino acids and chains of amino acids connecte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>When the concept is both an enzyme and a prote...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>aapp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Record Type (RT) Unique Identifier (UI)               Full Name (STY/RL)  \\\n",
       "97              STY                   T116  Amino Acid, Peptide, or Protein   \n",
       "\n",
       "   Tree Number (STN/RTN)                                   Definition (DEF)  \\\n",
       "97          A1.4.1.2.1.7  Amino acids and chains of amino acids connecte...   \n",
       "\n",
       "    Examples (EX)                                    Usage Note (UN)  \\\n",
       "97            NaN  When the concept is both an enzyme and a prote...   \n",
       "\n",
       "   Non-Human Flag (NH) Abbreviation (ABR) Inverse Relation (RIN)  \n",
       "97                 NaN               aapp                    NaN  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## quick look at srdef info\n",
    "srdef[srdef['Abbreviation (ABR)'] == 'aapp']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e245634",
   "metadata": {},
   "source": [
    "### Custom SEMMED semantic type processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "89c14528",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:10:37.059040Z",
     "start_time": "2025-08-20T07:10:37.046010Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(126, 3)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>T195</td>\n",
       "      <td>Antibiotic</td>\n",
       "      <td>antb</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst\n",
       "119  T195                       Antibiotic   antb"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## simplify the df\n",
    "SEMMED_semtypes = srdef[srdef['Record Type (RT)'] == 'STY'].copy()\n",
    "SEMMED_semtypes = SEMMED_semtypes[['Unique Identifier (UI)',\n",
    "                                   'Full Name (STY/RL)', \n",
    "                                   'Abbreviation (ABR)']].copy()\n",
    "SEMMED_semtypes.columns = ['UI', 'FullName', \"Abbrev\"]\n",
    "SEMMED_semtypes.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_semtypes.shape\n",
    "SEMMED_semtypes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b70c514f",
   "metadata": {},
   "source": [
    "#### Add SEMMED semantic types that are in the data but not in SRDEF"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49397246",
   "metadata": {},
   "source": [
    "Yao [noticed](https://github.com/biothings/pending.api/issues/30#issuecomment-903609946) that the data file uses \"old SEMMED semantic types\" from 2013AA, and these semantic types didn't exist in the latest SRDEF file (2023AA).\n",
    "\n",
    "In response, Andrew said to [keep](https://github.com/biothings/pending.api/issues/30#issuecomment-903879782) this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8c8c04ce",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:11:06.000817Z",
     "start_time": "2025-08-20T07:11:05.993989Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "these semantic types are in the data, but not in SRDEF:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'carb',\n",
       " 'eico',\n",
       " 'invt',\n",
       " 'lipd',\n",
       " 'nsba',\n",
       " 'opco',\n",
       " 'orgm',\n",
       " 'rich',\n",
       " 'strd'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get all semantic types in the data\n",
    "data_types = subject_semtypes.union(object_semtypes)\n",
    "only_in_data = data_types - set(SEMMED_semtypes['Abbrev'])\n",
    "\n",
    "print(\"these semantic types are in the data, but not in SRDEF:\")\n",
    "only_in_data\n",
    "\n",
    "len(only_in_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "533341a2",
   "metadata": {},
   "source": [
    "How I found info on these semantic types from [previous semantic versions](http://web.archive.org/web/20241127183822/https://lhncbc.nlm.nih.gov/semanticnetwork/SemanticNetworkArchive.html):\n",
    "\n",
    "3 were in 2009AB file (sets with SG_2?)\n",
    "* alga\n",
    "* invt\n",
    "* rich\n",
    "\n",
    "7 were in 2014AB file (sets with SG_3?)\n",
    "* carb \n",
    "* eico \n",
    "* lipd \n",
    "* nsba \n",
    "* opco\n",
    "* orgm\n",
    "* strd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0400e8e3",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:11:42.107748Z",
     "start_time": "2025-08-20T07:11:42.101281Z"
    }
   },
   "outputs": [],
   "source": [
    "## build the missing info and add to the table\n",
    "missing_semantic_types = [{\"UI\": \"T003\", \"FullName\": \"Alga\", \"Abbrev\": \"alga\"},\n",
    "                          {\"UI\": \"T118\", \"FullName\": \"Carbohydrate\", \"Abbrev\": \"carb\"},\n",
    "                          {\"UI\": \"T111\", \"FullName\": \"Eicosanoid\", \"Abbrev\": \"eico\"},\n",
    "                          {\"UI\": \"T009\", \"FullName\": \"Invertebrate\", \"Abbrev\": \"invt\"},\n",
    "                          {\"UI\": \"T119\", \"FullName\": \"Lipid\", \"Abbrev\": \"lipd\"},\n",
    "                          {\"UI\": \"T124\", \"FullName\": \"Neuroreactive Substance or Biogenic Amine\", \"Abbrev\": \"nsba\"},\n",
    "                          {\"UI\": \"T115\", \"FullName\": \"Organophosphorus Compound\", \"Abbrev\": \"opco\"},\n",
    "                          {\"UI\": \"T001\", \"FullName\": \"Organism\", \"Abbrev\": \"orgm\"},\n",
    "                          {\"UI\": \"T006\", \"FullName\": \"Rickettsia or Chlamydia\", \"Abbrev\": \"rich\"},\n",
    "                          {\"UI\": \"T110\", \"FullName\": \"Steroid\", \"Abbrev\": \"strd\"}\n",
    "                         ]\n",
    "\n",
    "missing_semantic_types = pd.DataFrame.from_records(missing_semantic_types)\n",
    "\n",
    "## add it to the table\n",
    "SEMMED_semtypes = pd.concat([SEMMED_semtypes, missing_semantic_types])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "683a1f47",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:11:44.466036Z",
     "start_time": "2025-08-20T07:11:44.455683Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(136, 3)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T003</td>\n",
       "      <td>Alga</td>\n",
       "      <td>alga</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>T011</td>\n",
       "      <td>Amphibian</td>\n",
       "      <td>amph</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>T190</td>\n",
       "      <td>Anatomical Abnormality</td>\n",
       "      <td>anab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T008</td>\n",
       "      <td>Animal</td>\n",
       "      <td>anim</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>T017</td>\n",
       "      <td>Anatomical Structure</td>\n",
       "      <td>anst</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                         FullName Abbrev\n",
       "97   T116  Amino Acid, Peptide, or Protein   aapp\n",
       "15   T020             Acquired Abnormality   acab\n",
       "44   T052                         Activity   acty\n",
       "90   T100                        Age Group   aggp\n",
       "0    T003                             Alga   alga\n",
       "77   T087              Amino Acid Sequence   amas\n",
       "6    T011                        Amphibian   amph\n",
       "115  T190           Anatomical Abnormality   anab\n",
       "4    T008                           Animal   anim\n",
       "12   T017             Anatomical Structure   anst"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_semtypes.sort_values(by='Abbrev', inplace=True)\n",
    "\n",
    "## quick view\n",
    "SEMMED_semtypes.shape\n",
    "SEMMED_semtypes[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8f01880",
   "metadata": {},
   "source": [
    "#### Get the biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a7a170a6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:12:57.129870Z",
     "start_time": "2025-08-20T07:12:53.951746Z"
    }
   },
   "outputs": [],
   "source": [
    "## set to use biolink-model current master branch\n",
    "bmt_tool = Toolkit('https://raw.githubusercontent.com/biolink/biolink-model/refs/heads/master/biolink-model.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4cc1fad9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:02.091823Z",
     "start_time": "2025-08-20T07:13:02.057400Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Chemical Entity'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## double-checking: this corresponds to bacs (Biologically Active Substance)\n",
    "bmt_tool.get_element_by_mapping('STY:T123').title()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2c14436",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:05.291613Z",
     "start_time": "2025-08-20T07:13:04.413693Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting biolink-mapping, in the format needed to create operations\n",
    "SEMMED_semtypes['BiolinkMapping'] = [bmt_tool.get_element_by_mapping('STY:'+i)\n",
    "                                     for i in SEMMED_semtypes['UI']]\n",
    "## put these biolink-model categories in the correct format for Translator/TRAPI: PascalCase\n",
    "SEMMED_semtypes['BiolinkMapping'] = [i.title().replace(\" \", \"\") if isinstance(i, str)\n",
    "                                     else i for i in SEMMED_semtypes['BiolinkMapping']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51111407",
   "metadata": {},
   "source": [
    "#### Clean up to only have terms in the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7fb08392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:10.707113Z",
     "start_time": "2025-08-20T07:13:10.702564Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "note that these semantic types are in SRDEF but not in the data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'cnce', 'crbs', 'enty', 'grpa', 'lang', 'phob'}"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"note that these semantic types are in SRDEF but not in the data\")\n",
    "set(SEMMED_semtypes['Abbrev']) - data_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "22daf133",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:12.761774Z",
     "start_time": "2025-08-20T07:13:12.758158Z"
    }
   },
   "outputs": [],
   "source": [
    "## get the subset of semantic network terms that are actually in the data\n",
    "SEMMED_semtypes = SEMMED_semtypes[SEMMED_semtypes['Abbrev'].isin(data_types)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "044843d6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:13:13.823414Z",
     "start_time": "2025-08-20T07:13:13.813681Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(130, 4)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>T054</td>\n",
       "      <td>Social Behavior</td>\n",
       "      <td>socb</td>\n",
       "      <td>Behavior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>T184</td>\n",
       "      <td>Sign or Symptom</td>\n",
       "      <td>sosy</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>T082</td>\n",
       "      <td>Spatial Concept</td>\n",
       "      <td>spco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T110</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>strd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T024</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>tisu</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "46   T054                      Social Behavior   socb   \n",
       "113  T184                      Sign or Symptom   sosy   \n",
       "73   T082                      Spatial Concept   spco   \n",
       "9    T110                              Steroid   strd   \n",
       "19   T024                               Tissue   tisu   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "46                   Behavior  \n",
       "113         PhenotypicFeature  \n",
       "73   InformationContentEntity  \n",
       "9               SmallMolecule  \n",
       "19   GrossAnatomicalStructure  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## was exploring the mapping by looking at slices of df\n",
    "SEMMED_semtypes.shape\n",
    "SEMMED_semtypes[120:131]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89d30960",
   "metadata": {},
   "source": [
    "#### review mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fd9e849",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "We are still reviewing / changing mappings, since BTE's responses would probably differ a lot if this behavior was changed. \n",
    "\n",
    "We are only removing SEMMED semantic types that lack biolink-model mappings. Other curated SEMMED semantic type removals will be done in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264).\n",
    "\n",
    "I'm going to keep the old notes + code (commented out), just in case. \n",
    "\n",
    "---\n",
    "\n",
    "Old notes:\n",
    "\n",
    "This is a place to **STOP** and review all the SEMMED semantic types and their mapping to biolink categories...to decide what we are interested in keeping. This involves some knowledge of what biolink categories are prioritized in Translator. \n",
    "\n",
    "One can use the definitions of the SEMMED semantic types (from the SRDEF file or [browsing the UMLS vocab online](https://uts.nlm.nih.gov/uts/umls/semantic-network/root)) and the definitions of biolink categories (look for a comment with 'THINGS' in the biolink-model yaml file)\n",
    "\n",
    "See the last section (section 6) for notes on decisions that were made..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "eb7ea639",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:14:09.336932Z",
     "start_time": "2025-08-20T07:14:09.332450Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Polypeptide', 'Disease', 'Activity', 'Cohort', None,\n",
       "       'Invertebrate', 'PhenotypicFeature', 'OrganismalEntity',\n",
       "       'AnatomicalEntity', 'Drug', 'ChemicalEntity', 'Bacterium',\n",
       "       'Behavior', 'Phenomenon', 'GrossAnatomicalStructure',\n",
       "       'SmallMolecule', 'CellularComponent', 'PhysiologicalProcess',\n",
       "       'Cell', 'InformationContentEntity', 'ClinicalAttribute',\n",
       "       'Procedure', 'Device', 'BiologicalEntity', 'Protein', 'Event',\n",
       "       'DiseaseOrPhenotypicFeature', 'Fungus', 'GeographicLocation',\n",
       "       'GenomicEntity', 'Agent', 'Human', 'PathologicalProcess',\n",
       "       'Publication', 'DiagnosticAid', 'Mammal', 'PhysicalEntity',\n",
       "       'MolecularActivity', 'MolecularEntity', 'NucleicAcidEntity',\n",
       "       'OrganismAttribute', 'Plant', 'PopulationOfIndividualOrganisms',\n",
       "       'Virus', 'Vertebrate'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## code used to review \n",
    "SEMMED_semtypes['BiolinkMapping'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3fe82c9f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:00.439758Z",
     "start_time": "2025-08-20T07:24:00.431751Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>T054</td>\n",
       "      <td>Social Behavior</td>\n",
       "      <td>socb</td>\n",
       "      <td>Behavior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>T184</td>\n",
       "      <td>Sign or Symptom</td>\n",
       "      <td>sosy</td>\n",
       "      <td>PhenotypicFeature</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>T082</td>\n",
       "      <td>Spatial Concept</td>\n",
       "      <td>spco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>T110</td>\n",
       "      <td>Steroid</td>\n",
       "      <td>strd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>T024</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>tisu</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "46   T054                      Social Behavior   socb   \n",
       "113  T184                      Sign or Symptom   sosy   \n",
       "73   T082                      Spatial Concept   spco   \n",
       "9    T110                              Steroid   strd   \n",
       "19   T024                               Tissue   tisu   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "46                   Behavior  \n",
       "113         PhenotypicFeature  \n",
       "73   InformationContentEntity  \n",
       "9               SmallMolecule  \n",
       "19   GrossAnatomicalStructure  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'] == 'ChemicalEntity']\n",
    "# SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'].isna()]\n",
    "SEMMED_semtypes[120:140]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c626a79a",
   "metadata": {},
   "source": [
    "Still re-mapping (not following a strict biolink-model mapping). May revisit later, but concerned about affecting responses to creative-mode queries / requiring template adjustments..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "e641ae5c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:15.681817Z",
     "start_time": "2025-08-20T07:24:15.673960Z"
    }
   },
   "outputs": [],
   "source": [
    "## re-mapping based on putting IDs into normalization service / our operation system...\n",
    "##   we have UMLS for Disease (mydisease, mychem), SmallMolecule (idisk)\n",
    "\n",
    "## leaving aapp as Polypeptide, enzy mapped to Protein\n",
    "## leaving Disease mappings as-is, but some seem like they could be PathologicalAnatomicalStructure \n",
    "##   (acab, anab, cgab) or PathologicalProcess (comd) instead\n",
    "## left clnd (Clinical Drug) as Drug since it really seemed like a drug (dosage)\n",
    "\n",
    "## current biolink value is Invertebrate, which looks like an error. re-mapping\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'amph'),'BiolinkMapping'] = 'Vertebrate'   ## Amphibian\n",
    "\n",
    "## currently mapped to Drug, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'antb'),'BiolinkMapping'] = 'SmallMolecule'  ## Antibiotic\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'phsu'),'BiolinkMapping'] = 'SmallMolecule'  ## Pharmacologic Substance\n",
    "\n",
    "## currently mapped to GenomicEntity, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'gngm'),'BiolinkMapping'] = 'Gene'\n",
    "\n",
    "## currently mapped to NucleicAcidEntity, but re-mapping here\n",
    "SEMMED_semtypes.loc[\n",
    "    (SEMMED_semtypes['Abbrev'] == 'nnon'),'BiolinkMapping'] = 'SmallMolecule'  ## Nucleic Acid, Nucleoside, or Nucleotide"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f46a30",
   "metadata": {},
   "source": [
    "We are removing metatriples with SEMMED semantic types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "882e7bde",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:19.687857Z",
     "start_time": "2025-08-20T07:24:19.684432Z"
    }
   },
   "outputs": [],
   "source": [
    "currently_unmapped_types = set(SEMMED_semtypes[SEMMED_semtypes['BiolinkMapping'].isna()].Abbrev)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f4220655",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:20.424348Z",
     "start_time": "2025-08-20T07:24:20.420210Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'food',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6eae498",
   "metadata": {},
   "source": [
    "Commenting out: we aren't removing SEMMED semantic types here anymore, because we will do so in a later section using the Translator exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a3a9da81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-08-20T07:46:45.716566Z",
     "start_time": "2024-08-20T07:46:45.712274Z"
    }
   },
   "outputs": [],
   "source": [
    "# currently_unused_entities = [\n",
    "#     ## missing mappings\n",
    "#     \"arch\", \"bdsy\", \"bird\", \"bmod\", \"chvf\", \"chvs\", \"euka\", \"ffas\", \n",
    "#     \"fish\", \"genf\", \"ocdi\", \"rept\", \"sbst\", \n",
    "#     ## old terms here\n",
    "#     \"alga\", \"invt\", \"orgm\", \"rich\",\n",
    "#     ## mapped to Polypeptide\n",
    "#     \"amas\",     \n",
    "#     ## all that are mapped to Activity\n",
    "#     \"acty\", \"dora\", \"edac\", \"gora\", \"hlca\", \"mcha\", \"ocac\", \"resa\",  \n",
    "#     ## all that are mapped to Cohort\n",
    "#     \"aggp\", \"famg\", \"podg\", \"prog\",  \n",
    "#     ## all that are mapped to Invertebrate\n",
    "#     \"amph\",\n",
    "#     ## all that are mapped to OrganismalEntity\n",
    "#     \"anim\", \n",
    "#     ## all that are mapped to Anatomical Entity\n",
    "#     \"anst\", \"bdsu\", \"blor\", \"bsoj\", \n",
    "#     ## mapped to SmallMolecule\n",
    "#     \"elii\",\n",
    "#     ## mapped to Bacterium\n",
    "#     \"bact\",\n",
    "#     ## all that are mapped to Behavior\n",
    "#     \"bhvr\", \"inbe\", \"menp\", \"socb\",\n",
    "#     ## all that are mapped to Phenomenon\n",
    "#     \"biof\", \"eehu\", \"hcpp\", \"lbtr\", \"npop\", \"phpr\",\n",
    "#     ## all that are mapped to Device\n",
    "#     \"bodm\", \"drdd\", \"medd\", \"resd\",\n",
    "#     ## mapped to GrossAnatomicalStructure\n",
    "#     \"emst\", \"tisu\", \n",
    "#     ## mapped to ChemicalEntity\n",
    "#     \"chem\",\n",
    "#     ## all that are mapped to InformationContentEntity\n",
    "#     \"clas\", \"ftcn\", \"idcn\", \"qlco\", \"qnco\", \"rnlw\", \"spco\", \"tmco\",\n",
    "#     ## all that are mapped to ClinicalAttribute\n",
    "#     \"clna\",\n",
    "#     ## mapped to Procedure\n",
    "#     \"lbpr\", \"mbrt\", \n",
    "#     ## mapped to BiologicalEntity\n",
    "#     \"emod\",\n",
    "#     ## mapped to Protein\n",
    "#     \"rcpt\",\n",
    "#     ## all that are mapped to Event\n",
    "#     \"evnt\",\n",
    "#     ## all that are mapped to DiseaseOrPhenotypicFeature\n",
    "#     \"fndg\",\n",
    "#     ## all that are mapped to Fungus\n",
    "#     \"fngs\",\n",
    "#     ## all that are mapped to GeographicLocation\n",
    "#     \"geoa\",\n",
    "#     ## all that are mapped to Agent\n",
    "#     \"grup\", \"hcro\", \"orgt\", \"pros\", \"shro\",\n",
    "#     ## all that are mapped to Human\n",
    "#     \"humn\", \n",
    "#     ## all that are mapped to Publication\n",
    "#     \"inpr\",\n",
    "#     ## all that are mapped to DiagnosticAid\n",
    "#     \"irda\", \n",
    "#     ## all that are mapped to Mammal\n",
    "#     \"mamm\",  \n",
    "#     ## all that are mapped to PhysicalEntity\n",
    "#     \"mnob\",\n",
    "#     ## all that are mapped to MolecularEntity\n",
    "#     \"mosq\",\n",
    "#     ## mapped to NucleicAcidEntity\n",
    "#     \"nusq\",\n",
    "#     ## all that are mapped to OrganismAttribute\n",
    "#     \"orga\",\n",
    "#     ## all that are mapped to Plant\n",
    "#     \"plnt\", \n",
    "#     ## all that are mapped to PopulationOfIndividualOrganisms\n",
    "#     \"popg\",\n",
    "#     ## all that are mapped to Virus\n",
    "#     \"virs\",\n",
    "#     ## all that are mapped to Vertebrate\n",
    "#     \"vtbt\"\n",
    "# ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da7aa1e",
   "metadata": {},
   "source": [
    "#### prune data based on what SEMMED semantic types to remove"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17055733",
   "metadata": {},
   "source": [
    "We are removing SEMMED semantic types that lack biolink-model mappings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8cf5b450",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:29.041381Z",
     "start_time": "2025-08-20T07:24:29.036025Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'alga',\n",
       " 'arch',\n",
       " 'bdsy',\n",
       " 'bird',\n",
       " 'bmod',\n",
       " 'bodm',\n",
       " 'chvf',\n",
       " 'chvs',\n",
       " 'euka',\n",
       " 'ffas',\n",
       " 'fish',\n",
       " 'food',\n",
       " 'genf',\n",
       " 'invt',\n",
       " 'ocdi',\n",
       " 'orgm',\n",
       " 'rept',\n",
       " 'rich'}"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "18"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "currently_unmapped_types\n",
    "len(currently_unmapped_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "308358e6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:34.735717Z",
     "start_time": "2025-08-20T07:24:34.726020Z"
    }
   },
   "outputs": [],
   "source": [
    "## prune this reference table of SEMMED semantic types\n",
    "SEMMED_semtypes = SEMMED_semtypes[ ~ SEMMED_semtypes['Abbrev'].isin(currently_unmapped_types)]\n",
    "\n",
    "## prune the metatriples table\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['SUBJECT_SEMTYPE'].isin(currently_unmapped_types)]\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['OBJECT_SEMTYPE'].isin(currently_unmapped_types)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "67861c2e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:24:39.710748Z",
     "start_time": "2025-08-20T07:24:39.700062Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "109"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## was 127, now 111: decreased by 16\n",
    "len(object_semtypes)  ## was 127, now 109: decreased by 18 (number of unmapped entities)\n",
    "len(predicates)       ## was 28, still 28 (expected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "647e4065",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:19.187639Z",
     "start_time": "2025-08-20T07:25:19.178148Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UI</th>\n",
       "      <th>FullName</th>\n",
       "      <th>Abbrev</th>\n",
       "      <th>BiolinkMapping</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>T116</td>\n",
       "      <td>Amino Acid, Peptide, or Protein</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>T020</td>\n",
       "      <td>Acquired Abnormality</td>\n",
       "      <td>acab</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>T052</td>\n",
       "      <td>Activity</td>\n",
       "      <td>acty</td>\n",
       "      <td>Activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>T100</td>\n",
       "      <td>Age Group</td>\n",
       "      <td>aggp</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>T087</td>\n",
       "      <td>Amino Acid Sequence</td>\n",
       "      <td>amas</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>T079</td>\n",
       "      <td>Temporal Concept</td>\n",
       "      <td>tmco</td>\n",
       "      <td>InformationContentEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>T061</td>\n",
       "      <td>Therapeutic or Preventive Procedure</td>\n",
       "      <td>topp</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T005</td>\n",
       "      <td>Virus</td>\n",
       "      <td>virs</td>\n",
       "      <td>Virus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>T127</td>\n",
       "      <td>Vitamin</td>\n",
       "      <td>vita</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>T010</td>\n",
       "      <td>Vertebrate</td>\n",
       "      <td>vtbt</td>\n",
       "      <td>Vertebrate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>112 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       UI                             FullName Abbrev  \\\n",
       "97   T116      Amino Acid, Peptide, or Protein   aapp   \n",
       "15   T020                 Acquired Abnormality   acab   \n",
       "44   T052                             Activity   acty   \n",
       "90   T100                            Age Group   aggp   \n",
       "77   T087                  Amino Acid Sequence   amas   \n",
       "..    ...                                  ...    ...   \n",
       "70   T079                     Temporal Concept   tmco   \n",
       "53   T061  Therapeutic or Preventive Procedure   topp   \n",
       "2    T005                                Virus   virs   \n",
       "104  T127                              Vitamin   vita   \n",
       "5    T010                           Vertebrate   vtbt   \n",
       "\n",
       "               BiolinkMapping  \n",
       "97                Polypeptide  \n",
       "15                    Disease  \n",
       "44                   Activity  \n",
       "90                     Cohort  \n",
       "77                Polypeptide  \n",
       "..                        ...  \n",
       "70   InformationContentEntity  \n",
       "53                  Procedure  \n",
       "2                       Virus  \n",
       "104             SmallMolecule  \n",
       "5                  Vertebrate  \n",
       "\n",
       "[112 rows x 4 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_semtypes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36f25379",
   "metadata": {},
   "source": [
    "### Custom predicate processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf6f43",
   "metadata": {},
   "source": [
    "#### getting mappings for predicates and qualifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "52496738",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:27.212834Z",
     "start_time": "2025-08-20T07:25:27.207663Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates = srdef[srdef['Record Type (RT)'] == 'RL'].copy()\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates[['Full Name (STY/RL)'\n",
    "                                      ]].copy()\n",
    "SEMMED_predicates.columns = ['SemmedPred']\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "635827c9",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "The SRDEF file might be missing predicates that are in the data, so check that using the code block below and see if you would like to add those back in..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b05e90f6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:31.432829Z",
     "start_time": "2025-08-20T07:25:31.427376Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['administered_to',\n",
       " 'augments',\n",
       " 'coexists_with',\n",
       " 'converts_to',\n",
       " 'inhibits',\n",
       " 'predisposes',\n",
       " 'same_as',\n",
       " 'stimulates']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## predicates currently in the filtered data\n",
    "\n",
    "missing_preds = list()\n",
    "\n",
    "for i in predicates:\n",
    "    if i.lower() not in set(SEMMED_predicates['SemmedPred']):\n",
    "        missing_preds.append(i.lower())\n",
    "\n",
    "missing_preds.sort()\n",
    "missing_preds\n",
    "\n",
    "# [i for i in predicates if i.lower() in SEMMED_predicates['SemmedName']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e40796c",
   "metadata": {},
   "source": [
    "Currently we add them back in "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6a3d217d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:42.608158Z",
     "start_time": "2025-08-20T07:25:42.595695Z"
    }
   },
   "outputs": [],
   "source": [
    "## manually add some predicates that are in the data, but not in SRDEF...\n",
    "\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['administered_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['augments']\n",
    "## notice that the data doesn't have a hyphen for coexists...even though SRDEF has hyphen for co-occurs_with\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['coexists_with']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['converts_to']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['inhibits']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['predisposes']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['same_as']\n",
    "SEMMED_predicates.loc[len(SEMMED_predicates.index)] = ['stimulates']\n",
    "\n",
    "SEMMED_predicates.sort_values(by='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "ec402da4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:48.268843Z",
     "start_time": "2025-08-20T07:25:48.257860Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>adjacent_to</td>\n",
       "      <td>SEMMEDDB:ADJACENT_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>analyzes</td>\n",
       "      <td>SEMMEDDB:ANALYZES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>assesses_effect_of</td>\n",
       "      <td>SEMMEDDB:ASSESSES_EFFECT_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>temporally_related_to</td>\n",
       "      <td>SEMMEDDB:TEMPORALLY_RELATED_TO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>traverses</td>\n",
       "      <td>SEMMEDDB:TRAVERSES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>tributary_of</td>\n",
       "      <td>SEMMEDDB:TRIBUTARY_OF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SemmedPred                    Semmed_in_BM\n",
       "162            adjacent_to            SEMMEDDB:ADJACENT_TO\n",
       "54         administered_to        SEMMEDDB:ADMINISTERED_TO\n",
       "145                affects                SEMMEDDB:AFFECTS\n",
       "176               analyzes               SEMMEDDB:ANALYZES\n",
       "158     assesses_effect_of     SEMMEDDB:ASSESSES_EFFECT_OF\n",
       "..                     ...                             ...\n",
       "130  temporally_related_to  SEMMEDDB:TEMPORALLY_RELATED_TO\n",
       "166              traverses              SEMMEDDB:TRAVERSES\n",
       "148                 treats                 SEMMEDDB:TREATS\n",
       "178           tributary_of           SEMMEDDB:TRIBUTARY_OF\n",
       "149                   uses                   SEMMEDDB:USES\n",
       "\n",
       "[62 rows x 2 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## since biolink-model mappings doesn't use hyphens\n",
    "SEMMED_predicates['Semmed_in_BM']= ['SEMMEDDB:'+i \n",
    "                                   for i in SEMMED_predicates['SemmedPred'].str.replace(\"-\", \"\")]\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.upper()\n",
    "\n",
    "## Specifically change same_as to be lower-case for the mapping process\n",
    "SEMMED_predicates.loc[\n",
    "    (SEMMED_predicates['SemmedPred'] == 'same_as'),'Semmed_in_BM'] = 'SEMMEDDB:same_as'\n",
    "\n",
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "979e0da8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:52.270957Z",
     "start_time": "2025-08-20T07:25:51.869018Z"
    }
   },
   "outputs": [],
   "source": [
    "## getting the mapping\n",
    "SEMMED_predicates['BMPred'] = [bmt_tool.get_element_by_mapping(i)\n",
    "                                       for i in SEMMED_predicates['Semmed_in_BM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "5d7f7944",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:25:54.396687Z",
     "start_time": "2025-08-20T07:25:54.379668Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>measurement_of</td>\n",
       "      <td>SEMMEDDB:MEASUREMENT_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes to condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>preventative for condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats or applied or studied to treat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO   \n",
       "145           affects           SEMMEDDB:AFFECTS   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH   \n",
       "55           augments          SEMMEDDB:AUGMENTS   \n",
       "141            causes            SEMMEDDB:CAUSES   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH   \n",
       "143       complicates       SEMMEDDB:COMPLICATES   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS   \n",
       "58           inhibits          SEMMEDDB:INHIBITS   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF   \n",
       "170    measurement_of    SEMMEDDB:MEASUREMENT_OF   \n",
       "156          measures          SEMMEDDB:MEASURES   \n",
       "171         method_of         SEMMEDDB:METHOD_OF   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN   \n",
       "127           part_of           SEMMEDDB:PART_OF   \n",
       "132          precedes          SEMMEDDB:PRECEDES   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES   \n",
       "142          prevents          SEMMEDDB:PREVENTS   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF   \n",
       "138          produces          SEMMEDDB:PRODUCES   \n",
       "60            same_as           SEMMEDDB:same_as   \n",
       "61         stimulates        SEMMEDDB:STIMULATES   \n",
       "148            treats            SEMMEDDB:TREATS   \n",
       "149              uses              SEMMEDDB:USES   \n",
       "\n",
       "                                    BMPred  \n",
       "54                              related to  \n",
       "145                                affects  \n",
       "160                             related to  \n",
       "55                                    None  \n",
       "141                                 causes  \n",
       "56                           coexists with  \n",
       "143                  exacerbates condition  \n",
       "57                            derives into  \n",
       "157                              diagnoses  \n",
       "140                               disrupts  \n",
       "58                                    None  \n",
       "136                         interacts with  \n",
       "129                            location of  \n",
       "144                       manifestation of  \n",
       "170                                   None  \n",
       "156                   quantifier qualifier  \n",
       "171                                   None  \n",
       "146                              occurs in  \n",
       "127                                part of  \n",
       "132                               precedes  \n",
       "59                predisposes to condition  \n",
       "142             preventative for condition  \n",
       "134                              occurs in  \n",
       "138                               produces  \n",
       "60                             close match  \n",
       "61                                    None  \n",
       "148  treats or applied or studied to treat  \n",
       "149                              has input  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>measurement_of</td>\n",
       "      <td>SEMMEDDB:MEASUREMENT_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>method_of</td>\n",
       "      <td>SEMMEDDB:METHOD_OF</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SemmedPred             Semmed_in_BM BMPred\n",
       "55         augments        SEMMEDDB:AUGMENTS   None\n",
       "58         inhibits        SEMMEDDB:INHIBITS   None\n",
       "170  measurement_of  SEMMEDDB:MEASUREMENT_OF   None\n",
       "171       method_of       SEMMEDDB:METHOD_OF   None\n",
       "61       stimulates      SEMMEDDB:STIMULATES   None"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remember there are 28 predicates in the data\n",
    "len(predicates)\n",
    "pred_compare = [i.lower() for i in predicates]\n",
    "\n",
    "SEMMED_predicates[SEMMED_predicates['SemmedPred'].isin(pred_compare)]\n",
    "\n",
    "SEMMED_predicates[(SEMMED_predicates['SemmedPred'].isin(pred_compare)) & \n",
    "                  (SEMMED_predicates['BMPred'].isna())]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c658e0a2",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Review the mappings in the table printed by the previous code chunk. \n",
    "\n",
    "The mapping for several SEMMEDDB predicates are missing (`BMPred` value is `None`) \n",
    "\n",
    "I'll add the mappings for some predicates below: and will be added below\n",
    "* augments\n",
    "* inhibits\n",
    "* stimulates\n",
    "\n",
    "I'll keep some as missing (mapping doesn't exist in biolink-model):\n",
    "* measurement_of \n",
    "* method_of\n",
    "\n",
    "Notes:\n",
    "* semmeddb `administered_to` is \"trying something out with a goal\", not that it actually does something...had a link to definition but not anymore \n",
    "* semmeddb `diagnoses` is more of a \"this thing distinguishes / identifies this other thing\"...see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T163)\n",
    "* semmeddb `process_of` does seem to map well to biolink-model's `occurs in` so...okay. As an example, see the use of `process_of` in the [API's data](https://biothings.ncats.io/semmeddb/query?q=predicate:PROCESS_OF%20AND%20object.semantic_type_abbreviation:cell)\n",
    "* semmeddb `treats` is more general than biolink-model's `treats`. It's more of \"trying this out as a treatment\". see [definition](https://uts.nlm.nih.gov/uts/umls/semantic-network/T154)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5040417f",
   "metadata": {},
   "source": [
    "Starting with the biolink 3.1.1 update, we:\n",
    "* map semmeddb `augments` and `stimulates` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: increased\n",
    "* map semmeddb `inhibits` to biolink's `affects` with qualifiers: \n",
    "  * qualified_predicate: causes\n",
    "  * object_aspect_qualifier: activity_or_abundance\n",
    "  * object_direction_qualifier: decreased\n",
    "\n",
    "And for the reverse of these operations with qualifiers, we'll want to get the inverted qualified_predicates and we'll use \"subject_\" rather than \"object_\" qualifiers (but the values will remain the same). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "77adf097",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:30:37.816333Z",
     "start_time": "2025-08-20T07:30:37.803786Z"
    }
   },
   "outputs": [],
   "source": [
    "## complex change: modify table to use qualifiers for semmeddb augments, stimulates, and inhibits\n",
    "\n",
    "## At the moment, only need the following qualifiers added, to map these predicates\n",
    "##   QualPred = qualified_predicate\n",
    "##   ObjAsp = object_aspect_qualifier\n",
    "##   ObjDirect = object_direction_qualifier\n",
    "##   Inv_QualPred = for reverse operations, inverse of qualified_predicate\n",
    "##   Inv_SubjAsp = for reverse operations, subject_aspect_qualifier\n",
    "##   Inv_SubjDirect = for reverse operations, subject_direction_qualifier\n",
    "\n",
    "\n",
    "## First, add the columns. Most values will stay blank (None)\n",
    "SEMMED_predicates['BMQualPred'] = None  \n",
    "SEMMED_predicates['BMObjAsp'] = None \n",
    "SEMMED_predicates['BMObjDirect'] = None\n",
    "SEMMED_predicates['BM_Inv_QualPred'] = None  \n",
    "SEMMED_predicates['BM_Inv_SubjAsp'] = None \n",
    "SEMMED_predicates['BM_Inv_SubjDirect'] = None\n",
    "\n",
    "## Next, mutate the rows for biolink predicates being remapping to predicates + qualifiers\n",
    "for index in SEMMED_predicates.index:\n",
    "    ## semmeddb augments and stimulates should change\n",
    "    if (SEMMED_predicates.loc[index,'SemmedPred'] == 'augments') or \\\n",
    "    (SEMMED_predicates.loc[index,'SemmedPred'] == 'stimulates'):\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'increased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'increased'\n",
    "    ## semmeddb inhibits should change\n",
    "    elif SEMMED_predicates.loc[index,'SemmedPred'] == 'inhibits':\n",
    "        SEMMED_predicates.loc[index,'BMPred'] = 'affects'\n",
    "        SEMMED_predicates.loc[index,'BMQualPred'] = 'causes'\n",
    "        SEMMED_predicates.loc[index,'BMObjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BMObjDirect'] = 'decreased'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_QualPred'] = 'caused_by'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjAsp'] = 'activity_or_abundance'\n",
    "        SEMMED_predicates.loc[index,'BM_Inv_SubjDirect'] = 'decreased'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "933327b9",
   "metadata": {},
   "source": [
    "Clean up table to only include what we have biolink-model mappings for: this removes measurement_of and method_of predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "80e84505",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:30:41.288814Z",
     "start_time": "2025-08-20T07:30:41.284388Z"
    }
   },
   "outputs": [],
   "source": [
    "## only keep the stuff we found biolink mappings for\n",
    "SEMMED_predicates = SEMMED_predicates[SEMMED_predicates['BMPred'].notna()].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd31635",
   "metadata": {},
   "source": [
    "Now to get the inverse predicates. With biolink-model-toolkit > 0.9.0 and biolink-model > 3.1.1, I've encountered issues retrieving inverse predicates. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f9acdf7f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-06-12T06:52:49.729544Z",
     "start_time": "2024-06-12T06:52:49.727052Z"
    }
   },
   "outputs": [],
   "source": [
    "## Testing methods of getting inverse\n",
    "# print(bmt_tool.get_element('causes').inverse)  ## get None\n",
    "# print(bmt_tool.get_inverse('subclass of'))   ## get None\n",
    "# print(bmt_tool.get_inverse('subclass_of'))   ## get AttributeError: 'NoneType' object has no attribute 'name'\n",
    "# print(bmt_tool.get_inverse('quantifier_qualifier'))    ## also get AttributeError\n",
    "\n",
    "# for i in SEMMED_predicates['BMPred']:\n",
    "#     print(i)\n",
    "#     if bmt_tool.get_element(i).symmetric:\n",
    "#         print('is symmetrical')\n",
    "#     elif bmt_tool.get_inverse(i):\n",
    "#         print('inverse: ' + bmt_tool.get_inverse(i))\n",
    "#     else:\n",
    "#         print('missing inverse')\n",
    "#     print('\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "52cfbbc4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:30:48.771965Z",
     "start_time": "2025-08-20T07:30:48.764202Z"
    }
   },
   "outputs": [],
   "source": [
    "SEMMED_predicates['BM_Inv_Pred'] = [i if bmt_tool.get_element(i).symmetric\n",
    "                                      else bmt_tool.get_inverse(i) if bmt_tool.get_inverse(i)\n",
    "                                      else 'MISSING' \n",
    "                                    for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "SEMMED_predicates = SEMMED_predicates.reindex(columns=[\n",
    "    'SemmedPred', 'Semmed_in_BM', 'BMPred', 'BMQualPred', 'BMObjAsp', 'BMObjDirect', \n",
    "    'BM_Inv_Pred', 'BM_Inv_QualPred', 'BM_Inv_SubjAsp', 'BM_Inv_SubjDirect'])\n",
    "\n",
    "## old code that doesn't work anymore, was previously used to get inverses for predicates and qualified predicates\n",
    "## get inverses for basic predicates + qualified_predicates to generate reverse operations\n",
    "# SEMMED_predicates['BM_Inv_Pred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMPred']]\n",
    "\n",
    "# SEMMED_predicates['BM_Inv_QualPred'] = [bmt_tool.get_element(i).inverse \n",
    "#                                        if i and isinstance(bmt_tool.get_element(i).inverse, str)\n",
    "#                                        else i \n",
    "#                                        for i in SEMMED_predicates['BMQualPred']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "9718fe3a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:30:50.744323Z",
     "start_time": "2025-08-20T07:30:50.730114Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>administered_to</td>\n",
       "      <td>SEMMEDDB:ADMINISTERED_TO</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>affects</td>\n",
       "      <td>SEMMEDDB:AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>associated_with</td>\n",
       "      <td>SEMMEDDB:ASSOCIATED_WITH</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>augments</td>\n",
       "      <td>SEMMEDDB:AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>causes</td>\n",
       "      <td>SEMMEDDB:CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>coexists_with</td>\n",
       "      <td>SEMMEDDB:COEXISTS_WITH</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>complicates</td>\n",
       "      <td>SEMMEDDB:COMPLICATES</td>\n",
       "      <td>exacerbates condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition exacerbated by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>converts_to</td>\n",
       "      <td>SEMMEDDB:CONVERTS_TO</td>\n",
       "      <td>derives into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>diagnoses</td>\n",
       "      <td>SEMMEDDB:DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is diagnosed by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>disrupts</td>\n",
       "      <td>SEMMEDDB:DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>inhibits</td>\n",
       "      <td>SEMMEDDB:INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>interacts_with</td>\n",
       "      <td>SEMMEDDB:INTERACTS_WITH</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>isa</td>\n",
       "      <td>SEMMEDDB:ISA</td>\n",
       "      <td>subclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>superclass of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>location_of</td>\n",
       "      <td>SEMMEDDB:LOCATION_OF</td>\n",
       "      <td>location of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>SEMMEDDB:MANIFESTATION_OF</td>\n",
       "      <td>manifestation of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>occurs_in</td>\n",
       "      <td>SEMMEDDB:OCCURS_IN</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>part_of</td>\n",
       "      <td>SEMMEDDB:PART_OF</td>\n",
       "      <td>part of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>precedes</td>\n",
       "      <td>SEMMEDDB:PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>predisposes</td>\n",
       "      <td>SEMMEDDB:PREDISPOSES</td>\n",
       "      <td>predisposes to condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition predisposed by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>prevents</td>\n",
       "      <td>SEMMEDDB:PREVENTS</td>\n",
       "      <td>preventative for condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition prevented by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>process_of</td>\n",
       "      <td>SEMMEDDB:PROCESS_OF</td>\n",
       "      <td>occurs in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>produces</td>\n",
       "      <td>SEMMEDDB:PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>SEMMEDDB:same_as</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>stimulates</td>\n",
       "      <td>SEMMEDDB:STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>treats</td>\n",
       "      <td>SEMMEDDB:TREATS</td>\n",
       "      <td>treats or applied or studied to treat</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subject of treatment application or study for ...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>uses</td>\n",
       "      <td>SEMMEDDB:USES</td>\n",
       "      <td>has input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is input of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SemmedPred               Semmed_in_BM  \\\n",
       "54    administered_to   SEMMEDDB:ADMINISTERED_TO   \n",
       "145           affects           SEMMEDDB:AFFECTS   \n",
       "160   associated_with   SEMMEDDB:ASSOCIATED_WITH   \n",
       "55           augments          SEMMEDDB:AUGMENTS   \n",
       "141            causes            SEMMEDDB:CAUSES   \n",
       "56      coexists_with     SEMMEDDB:COEXISTS_WITH   \n",
       "143       complicates       SEMMEDDB:COMPLICATES   \n",
       "57        converts_to       SEMMEDDB:CONVERTS_TO   \n",
       "157         diagnoses         SEMMEDDB:DIAGNOSES   \n",
       "140          disrupts          SEMMEDDB:DISRUPTS   \n",
       "58           inhibits          SEMMEDDB:INHIBITS   \n",
       "136    interacts_with    SEMMEDDB:INTERACTS_WITH   \n",
       "172               isa               SEMMEDDB:ISA   \n",
       "129       location_of       SEMMEDDB:LOCATION_OF   \n",
       "144  manifestation_of  SEMMEDDB:MANIFESTATION_OF   \n",
       "156          measures          SEMMEDDB:MEASURES   \n",
       "146         occurs_in         SEMMEDDB:OCCURS_IN   \n",
       "127           part_of           SEMMEDDB:PART_OF   \n",
       "132          precedes          SEMMEDDB:PRECEDES   \n",
       "59        predisposes       SEMMEDDB:PREDISPOSES   \n",
       "142          prevents          SEMMEDDB:PREVENTS   \n",
       "134        process_of        SEMMEDDB:PROCESS_OF   \n",
       "138          produces          SEMMEDDB:PRODUCES   \n",
       "60            same_as           SEMMEDDB:same_as   \n",
       "61         stimulates        SEMMEDDB:STIMULATES   \n",
       "148            treats            SEMMEDDB:TREATS   \n",
       "149              uses              SEMMEDDB:USES   \n",
       "\n",
       "                                    BMPred BMQualPred               BMObjAsp  \\\n",
       "54                              related to       None                   None   \n",
       "145                                affects       None                   None   \n",
       "160                             related to       None                   None   \n",
       "55                                 affects     causes  activity_or_abundance   \n",
       "141                                 causes       None                   None   \n",
       "56                           coexists with       None                   None   \n",
       "143                  exacerbates condition       None                   None   \n",
       "57                            derives into       None                   None   \n",
       "157                              diagnoses       None                   None   \n",
       "140                               disrupts       None                   None   \n",
       "58                                 affects     causes  activity_or_abundance   \n",
       "136                         interacts with       None                   None   \n",
       "172                            subclass of       None                   None   \n",
       "129                            location of       None                   None   \n",
       "144                       manifestation of       None                   None   \n",
       "156                   quantifier qualifier       None                   None   \n",
       "146                              occurs in       None                   None   \n",
       "127                                part of       None                   None   \n",
       "132                               precedes       None                   None   \n",
       "59                predisposes to condition       None                   None   \n",
       "142             preventative for condition       None                   None   \n",
       "134                              occurs in       None                   None   \n",
       "138                               produces       None                   None   \n",
       "60                             close match       None                   None   \n",
       "61                                 affects     causes  activity_or_abundance   \n",
       "148  treats or applied or studied to treat       None                   None   \n",
       "149                              has input       None                   None   \n",
       "\n",
       "    BMObjDirect                                        BM_Inv_Pred  \\\n",
       "54         None                                         related to   \n",
       "145        None                                        affected by   \n",
       "160        None                                         related to   \n",
       "55    increased                                        affected by   \n",
       "141        None                                          caused by   \n",
       "56         None                                      coexists with   \n",
       "143        None                           condition exacerbated by   \n",
       "57         None                                       derives from   \n",
       "157        None                                    is diagnosed by   \n",
       "140        None                                       disrupted by   \n",
       "58    decreased                                        affected by   \n",
       "136        None                                     interacts with   \n",
       "172        None                                      superclass of   \n",
       "129        None                                         located in   \n",
       "144        None                                  has manifestation   \n",
       "156        None                                            MISSING   \n",
       "146        None                                   contains process   \n",
       "127        None                                           has part   \n",
       "132        None                                        preceded by   \n",
       "59         None                           condition predisposed by   \n",
       "142        None                             condition prevented by   \n",
       "134        None                                   contains process   \n",
       "138        None                                        produced by   \n",
       "60         None                                        close match   \n",
       "61    increased                                        affected by   \n",
       "148        None  subject of treatment application or study for ...   \n",
       "149        None                                        is input of   \n",
       "\n",
       "    BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54             None                   None              None  \n",
       "145            None                   None              None  \n",
       "160            None                   None              None  \n",
       "55        caused_by  activity_or_abundance         increased  \n",
       "141            None                   None              None  \n",
       "56             None                   None              None  \n",
       "143            None                   None              None  \n",
       "57             None                   None              None  \n",
       "157            None                   None              None  \n",
       "140            None                   None              None  \n",
       "58        caused_by  activity_or_abundance         decreased  \n",
       "136            None                   None              None  \n",
       "172            None                   None              None  \n",
       "129            None                   None              None  \n",
       "144            None                   None              None  \n",
       "156            None                   None              None  \n",
       "146            None                   None              None  \n",
       "127            None                   None              None  \n",
       "132            None                   None              None  \n",
       "59             None                   None              None  \n",
       "142            None                   None              None  \n",
       "134            None                   None              None  \n",
       "138            None                   None              None  \n",
       "60             None                   None              None  \n",
       "61        caused_by  activity_or_abundance         increased  \n",
       "148            None                   None              None  \n",
       "149            None                   None              None  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "b121018d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:31:49.059556Z",
     "start_time": "2025-08-20T07:31:49.049152Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SemmedPred</th>\n",
       "      <th>Semmed_in_BM</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>measures</td>\n",
       "      <td>SEMMEDDB:MEASURES</td>\n",
       "      <td>quantifier qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SemmedPred       Semmed_in_BM                BMPred BMQualPred BMObjAsp  \\\n",
       "156   measures  SEMMEDDB:MEASURES  quantifier qualifier       None     None   \n",
       "\n",
       "    BMObjDirect BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "156        None     MISSING            None           None              None  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd4046d",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Check if the predicate inverses look correct. \n",
    "\n",
    "For bmt >=1.2.0 and biolink-model >=4.1.5, we want to:\n",
    "* leave the inverse of `quanitifer qualifier` as `MISSING` since it's actually not a predicate, so we'll want to remove it in the next step\n",
    "\n",
    "Old notes: \n",
    "* used to have `subclass of` removed in this step because it didn't have an inverse. now it does\n",
    "* Maybe some aren't correct because they don't have an inverse but they are directional...If so, I suggest going back and changing the predicates to ones with inverses. \n",
    "* Some predicates aren't directional (their entry in the biolink-model yaml will have the property `symmetric == true`) so those are identical in either direction..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46765d17",
   "metadata": {},
   "source": [
    "Now we'll make sure all biolink-model-mapped columns are in snake_case, and clean up the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "85224c41",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:09.793914Z",
     "start_time": "2025-08-20T07:32:09.788503Z"
    }
   },
   "outputs": [],
   "source": [
    "## FOR RESOURCES, these predicates must all be in snake_case\n",
    "for col in ['BMPred', 'BMQualPred', 'BM_Inv_Pred', 'BM_Inv_QualPred']:\n",
    "    SEMMED_predicates[col] = SEMMED_predicates[col].str.replace(\" \", \"_\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "50027e5e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:14.018105Z",
     "start_time": "2025-08-20T07:32:14.011499Z"
    }
   },
   "outputs": [],
   "source": [
    "## can mutate and rename the Semmed_in_BM column (not needed anymore) with the format needed\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['SemmedPred'].str.upper()\n",
    "SEMMED_predicates['Semmed_in_BM'] = SEMMED_predicates['Semmed_in_BM'].str.replace(\"-\", \"\")\n",
    "SEMMED_predicates.rename(columns = {'Semmed_in_BM':'Semmed_in_Data'}, inplace = True)\n",
    "\n",
    "## special handling keep lowercase for same_as\n",
    "SEMMED_predicates['Semmed_in_Data'] = [i.lower() if i == 'SAME_AS' else i\n",
    "                                       for i in SEMMED_predicates['Semmed_in_Data']]\n",
    "\n",
    "## Can remove the SemmedPred column: not needed anymore\n",
    "SEMMED_predicates.drop(columns='SemmedPred', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "946dd392",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:29.523509Z",
     "start_time": "2025-08-20T07:32:29.510207Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>ISA</td>\n",
       "      <td>subclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>superclass_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes_to_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_predisposed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>preventative_for_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats_or_applied_or_studied_to_treat</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subject_of_treatment_application_or_study_for_...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data                                 BMPred BMQualPred  \\\n",
       "54    ADMINISTERED_TO                             related_to       None   \n",
       "145           AFFECTS                                affects       None   \n",
       "160   ASSOCIATED_WITH                             related_to       None   \n",
       "55           AUGMENTS                                affects     causes   \n",
       "141            CAUSES                                 causes       None   \n",
       "56      COEXISTS_WITH                          coexists_with       None   \n",
       "143       COMPLICATES                  exacerbates_condition       None   \n",
       "57        CONVERTS_TO                           derives_into       None   \n",
       "157         DIAGNOSES                              diagnoses       None   \n",
       "140          DISRUPTS                               disrupts       None   \n",
       "58           INHIBITS                                affects     causes   \n",
       "136    INTERACTS_WITH                         interacts_with       None   \n",
       "172               ISA                            subclass_of       None   \n",
       "129       LOCATION_OF                            location_of       None   \n",
       "144  MANIFESTATION_OF                       manifestation_of       None   \n",
       "156          MEASURES                   quantifier_qualifier       None   \n",
       "146         OCCURS_IN                              occurs_in       None   \n",
       "127           PART_OF                                part_of       None   \n",
       "132          PRECEDES                               precedes       None   \n",
       "59        PREDISPOSES               predisposes_to_condition       None   \n",
       "142          PREVENTS             preventative_for_condition       None   \n",
       "134        PROCESS_OF                              occurs_in       None   \n",
       "138          PRODUCES                               produces       None   \n",
       "60            same_as                            close_match       None   \n",
       "61         STIMULATES                                affects     causes   \n",
       "148            TREATS  treats_or_applied_or_studied_to_treat       None   \n",
       "149              USES                              has_input       None   \n",
       "\n",
       "                  BMObjAsp BMObjDirect  \\\n",
       "54                    None        None   \n",
       "145                   None        None   \n",
       "160                   None        None   \n",
       "55   activity_or_abundance   increased   \n",
       "141                   None        None   \n",
       "56                    None        None   \n",
       "143                   None        None   \n",
       "57                    None        None   \n",
       "157                   None        None   \n",
       "140                   None        None   \n",
       "58   activity_or_abundance   decreased   \n",
       "136                   None        None   \n",
       "172                   None        None   \n",
       "129                   None        None   \n",
       "144                   None        None   \n",
       "156                   None        None   \n",
       "146                   None        None   \n",
       "127                   None        None   \n",
       "132                   None        None   \n",
       "59                    None        None   \n",
       "142                   None        None   \n",
       "134                   None        None   \n",
       "138                   None        None   \n",
       "60                    None        None   \n",
       "61   activity_or_abundance   increased   \n",
       "148                   None        None   \n",
       "149                   None        None   \n",
       "\n",
       "                                           BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54                                          related_to            None   \n",
       "145                                        affected_by            None   \n",
       "160                                         related_to            None   \n",
       "55                                         affected_by       caused_by   \n",
       "141                                          caused_by            None   \n",
       "56                                       coexists_with            None   \n",
       "143                           condition_exacerbated_by            None   \n",
       "57                                        derives_from            None   \n",
       "157                                    is_diagnosed_by            None   \n",
       "140                                       disrupted_by            None   \n",
       "58                                         affected_by       caused_by   \n",
       "136                                     interacts_with            None   \n",
       "172                                      superclass_of            None   \n",
       "129                                         located_in            None   \n",
       "144                                  has_manifestation            None   \n",
       "156                                            MISSING            None   \n",
       "146                                   contains_process            None   \n",
       "127                                           has_part            None   \n",
       "132                                        preceded_by            None   \n",
       "59                            condition_predisposed_by            None   \n",
       "142                             condition_prevented_by            None   \n",
       "134                                   contains_process            None   \n",
       "138                                        produced_by            None   \n",
       "60                                         close_match            None   \n",
       "61                                         affected_by       caused_by   \n",
       "148  subject_of_treatment_application_or_study_for_...            None   \n",
       "149                                        is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "172                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "156                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SEMMED_predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c616c4ff",
   "metadata": {},
   "source": [
    "#### custom-prune the data of some predicates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6f0a864",
   "metadata": {},
   "source": [
    "We are still doing this because some semmeddb predicates or their mappings are problematic. \n",
    "\n",
    "As of 2023-08-07: the Translator exclusions weren't on a predicate level. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "9079161f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:38.757518Z",
     "start_time": "2025-08-20T07:32:38.749676Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in data but not in mapping table\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'MEASUREMENT_OF', 'METHOD_OF'}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "in mapping file but not in data\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'ISA'}"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"in data but not in mapping table\")\n",
    "predicates - set(SEMMED_predicates['Semmed_in_Data'])\n",
    "\n",
    "print(\"in mapping file but not in data\")\n",
    "set(SEMMED_predicates['Semmed_in_Data']) - predicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ca959b8c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:40.355777Z",
     "start_time": "2025-08-20T07:32:40.345474Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>MEASURES</td>\n",
       "      <td>quantifier_qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>MISSING</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Semmed_in_Data                BMPred BMQualPred BMObjAsp BMObjDirect  \\\n",
       "156       MEASURES  quantifier_qualifier       None     None        None   \n",
       "\n",
       "    BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "156     MISSING            None           None              None  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## finding predicates where inverse is missing (issues retrieving from biolink-model-toolkit)\n",
    "SEMMED_predicates[SEMMED_predicates['BM_Inv_Pred'] == 'MISSING']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8b1e2b",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Remove predicates in the mapping table, but we don't want to use for associations:\n",
    "- no biolink-model mapping: METHOD_OF, MEASUREMENT_OF\n",
    "- no biolink-model inverse (`MISSING`) and mapping isn't to a predicate (quantifier_qualifier): MEASURES\n",
    "- already removed from data during Basic Filtering section 2.3 (predicates): ISA\n",
    "\n",
    "Past: used to exclude USES (not helpful?), but now keeping because the predicate has been reviewed in the Translator-curated exclusions effort. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5e25a929",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:47.675635Z",
     "start_time": "2025-08-20T07:32:47.671108Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ISA', 'MEASUREMENT_OF', 'MEASURES', 'METHOD_OF'}"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "more_removals = {'MEASUREMENT_OF', 'METHOD_OF', 'ISA', 'MEASURES'}\n",
    "more_removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "5a46158c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:48.557283Z",
     "start_time": "2025-08-20T07:32:48.541809Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Semmed_in_Data</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>ADMINISTERED_TO</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>CAUSES</td>\n",
       "      <td>causes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>COMPLICATES</td>\n",
       "      <td>exacerbates_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_exacerbated_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>CONVERTS_TO</td>\n",
       "      <td>derives_into</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>derives_from</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>DIAGNOSES</td>\n",
       "      <td>diagnoses</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_diagnosed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>MANIFESTATION_OF</td>\n",
       "      <td>manifestation_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_manifestation</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>OCCURS_IN</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PART_OF</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PRECEDES</td>\n",
       "      <td>precedes</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>preceded_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>PREDISPOSES</td>\n",
       "      <td>predisposes_to_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_predisposed_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PREVENTS</td>\n",
       "      <td>preventative_for_condition</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>condition_prevented_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>occurs_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>contains_process</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PRODUCES</td>\n",
       "      <td>produces</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>produced_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>same_as</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>close_match</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>TREATS</td>\n",
       "      <td>treats_or_applied_or_studied_to_treat</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subject_of_treatment_application_or_study_for_...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>USES</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Semmed_in_Data                                 BMPred BMQualPred  \\\n",
       "54    ADMINISTERED_TO                             related_to       None   \n",
       "145           AFFECTS                                affects       None   \n",
       "160   ASSOCIATED_WITH                             related_to       None   \n",
       "55           AUGMENTS                                affects     causes   \n",
       "141            CAUSES                                 causes       None   \n",
       "56      COEXISTS_WITH                          coexists_with       None   \n",
       "143       COMPLICATES                  exacerbates_condition       None   \n",
       "57        CONVERTS_TO                           derives_into       None   \n",
       "157         DIAGNOSES                              diagnoses       None   \n",
       "140          DISRUPTS                               disrupts       None   \n",
       "58           INHIBITS                                affects     causes   \n",
       "136    INTERACTS_WITH                         interacts_with       None   \n",
       "129       LOCATION_OF                            location_of       None   \n",
       "144  MANIFESTATION_OF                       manifestation_of       None   \n",
       "146         OCCURS_IN                              occurs_in       None   \n",
       "127           PART_OF                                part_of       None   \n",
       "132          PRECEDES                               precedes       None   \n",
       "59        PREDISPOSES               predisposes_to_condition       None   \n",
       "142          PREVENTS             preventative_for_condition       None   \n",
       "134        PROCESS_OF                              occurs_in       None   \n",
       "138          PRODUCES                               produces       None   \n",
       "60            same_as                            close_match       None   \n",
       "61         STIMULATES                                affects     causes   \n",
       "148            TREATS  treats_or_applied_or_studied_to_treat       None   \n",
       "149              USES                              has_input       None   \n",
       "\n",
       "                  BMObjAsp BMObjDirect  \\\n",
       "54                    None        None   \n",
       "145                   None        None   \n",
       "160                   None        None   \n",
       "55   activity_or_abundance   increased   \n",
       "141                   None        None   \n",
       "56                    None        None   \n",
       "143                   None        None   \n",
       "57                    None        None   \n",
       "157                   None        None   \n",
       "140                   None        None   \n",
       "58   activity_or_abundance   decreased   \n",
       "136                   None        None   \n",
       "129                   None        None   \n",
       "144                   None        None   \n",
       "146                   None        None   \n",
       "127                   None        None   \n",
       "132                   None        None   \n",
       "59                    None        None   \n",
       "142                   None        None   \n",
       "134                   None        None   \n",
       "138                   None        None   \n",
       "60                    None        None   \n",
       "61   activity_or_abundance   increased   \n",
       "148                   None        None   \n",
       "149                   None        None   \n",
       "\n",
       "                                           BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "54                                          related_to            None   \n",
       "145                                        affected_by            None   \n",
       "160                                         related_to            None   \n",
       "55                                         affected_by       caused_by   \n",
       "141                                          caused_by            None   \n",
       "56                                       coexists_with            None   \n",
       "143                           condition_exacerbated_by            None   \n",
       "57                                        derives_from            None   \n",
       "157                                    is_diagnosed_by            None   \n",
       "140                                       disrupted_by            None   \n",
       "58                                         affected_by       caused_by   \n",
       "136                                     interacts_with            None   \n",
       "129                                         located_in            None   \n",
       "144                                  has_manifestation            None   \n",
       "146                                   contains_process            None   \n",
       "127                                           has_part            None   \n",
       "132                                        preceded_by            None   \n",
       "59                            condition_predisposed_by            None   \n",
       "142                             condition_prevented_by            None   \n",
       "134                                   contains_process            None   \n",
       "138                                        produced_by            None   \n",
       "60                                         close_match            None   \n",
       "61                                         affected_by       caused_by   \n",
       "148  subject_of_treatment_application_or_study_for_...            None   \n",
       "149                                        is_input_of            None   \n",
       "\n",
       "            BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "54                    None              None  \n",
       "145                   None              None  \n",
       "160                   None              None  \n",
       "55   activity_or_abundance         increased  \n",
       "141                   None              None  \n",
       "56                    None              None  \n",
       "143                   None              None  \n",
       "57                    None              None  \n",
       "157                   None              None  \n",
       "140                   None              None  \n",
       "58   activity_or_abundance         decreased  \n",
       "136                   None              None  \n",
       "129                   None              None  \n",
       "144                   None              None  \n",
       "146                   None              None  \n",
       "127                   None              None  \n",
       "132                   None              None  \n",
       "59                    None              None  \n",
       "142                   None              None  \n",
       "134                   None              None  \n",
       "138                   None              None  \n",
       "60                    None              None  \n",
       "61   activity_or_abundance         increased  \n",
       "148                   None              None  \n",
       "149                   None              None  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(25, 9)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## remove this set from the SEMMED_predicates table\n",
    "SEMMED_predicates = SEMMED_predicates[~ SEMMED_predicates['Semmed_in_Data'].isin(more_removals)]\n",
    "SEMMED_predicates\n",
    "SEMMED_predicates.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "64cacd52",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:56.136798Z",
     "start_time": "2025-08-20T07:32:56.130865Z"
    }
   },
   "outputs": [],
   "source": [
    "## remove this set from the metatriples table\n",
    "imported_metatriples = imported_metatriples[ ~ imported_metatriples['PREDICATE'].isin(more_removals)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "530e3631",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:32:59.395228Z",
     "start_time": "2025-08-20T07:32:59.383503Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "111"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "109"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at the semantic types again after this removal\n",
    "subject_semtypes = set(imported_metatriples[\"SUBJECT_SEMTYPE\"].unique())\n",
    "object_semtypes = set(imported_metatriples[\"OBJECT_SEMTYPE\"].unique())\n",
    "predicates = set(imported_metatriples[\"PREDICATE\"].unique())\n",
    "\n",
    "len(subject_semtypes) ## same as before (expected)\n",
    "len(object_semtypes)  ## same as before (expected)\n",
    "len(predicates)       ## was 28, now 25: decreased by 3 (expected: ISA was already removed earlier)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "863543b6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:33:47.989552Z",
     "start_time": "2025-08-20T07:33:47.977551Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15311, 7)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>2530738</td>\n",
       "      <td>2321777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>humn</td>\n",
       "      <td>1324267</td>\n",
       "      <td>1160368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>1204184</td>\n",
       "      <td>1153238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1139224</td>\n",
       "      <td>1045335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>923819</td>\n",
       "      <td>846909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>895767</td>\n",
       "      <td>816152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>1016052</td>\n",
       "      <td>779870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>844184</td>\n",
       "      <td>735624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>PROCESS_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>mamm</td>\n",
       "      <td>759499</td>\n",
       "      <td>683358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>812383</td>\n",
       "      <td>678731</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "0           umls            dsyn   PROCESS_OF          umls           humn   \n",
       "1           umls            fndg   PROCESS_OF          umls           humn   \n",
       "2           umls            bpoc      PART_OF          umls           mamm   \n",
       "3           umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "5           umls            bpoc  LOCATION_OF          umls           patf   \n",
       "6           umls            topp       TREATS          umls           dsyn   \n",
       "7           umls            topp       TREATS          umls           neop   \n",
       "8           umls            dsyn   PROCESS_OF          umls           mamm   \n",
       "9           umls            bpoc  LOCATION_OF          umls           fndg   \n",
       "\n",
       "     DOC_N  DOC_N_VALID  \n",
       "0  2530738      2321777  \n",
       "1  1324267      1160368  \n",
       "2  1204184      1153238  \n",
       "3  1139224      1045335  \n",
       "4   923819       846909  \n",
       "5   895767       816152  \n",
       "6  1016052       779870  \n",
       "7   844184       735624  \n",
       "8   759499       683358  \n",
       "9   812383       678731  "
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## look at number of metatriples after this removal\n",
    "imported_metatriples.shape\n",
    "imported_metatriples.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "426473de",
   "metadata": {},
   "source": [
    "### Use translator-curated exclusions to prune metatriple tables "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e732098e",
   "metadata": {},
   "source": [
    "I'll use Translator-curated exclusions to remove some SEMMED semantic type, domain-predicate, and predicate-range exclusions. ref: [Translator google group, google sheet link](https://docs.google.com/spreadsheets/d/1c1gx0Jgm9rJUOXcQhBtZgvx50Cvz1-jh0DdGtg1zcd8/edit#gid=1801185264)\n",
    "\n",
    "Translator-curated exclusions work on the metatriple-level. \n",
    "\n",
    "Note: the `_t_code` columns sometimes have the value of `not_found`. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37afbf97",
   "metadata": {},
   "source": [
    "#### Checking and adjusting exclusions table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "5f404ea6",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:33:56.107752Z",
     "start_time": "2025-08-20T07:33:56.095475Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>semantic type exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "0                chem                  T103              NaN   \n",
       "1                 NaN                   NaN              NaN   \n",
       "2                chvs                  T104              NaN   \n",
       "3                 NaN                   NaN              NaN   \n",
       "4                chvf                  T120              NaN   \n",
       "\n",
       "  semmed_object_code semmed_object_t_code           exclusion_type  \n",
       "0                NaN                  NaN  semantic type exclusion  \n",
       "1               chem                 T103  semantic type exclusion  \n",
       "2                NaN                  NaN  semantic type exclusion  \n",
       "3               chvs                 T104  semantic type exclusion  \n",
       "4                NaN                  NaN  semantic type exclusion  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(1443, 6)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "array(['semantic type exclusion', 'Domain exclusion', 'Range exclusion'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_exclusions.head()\n",
    "imported_exclusions.shape\n",
    "imported_exclusions.exclusion_type.unique()  ## number of unique kinds of exclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99185fc",
   "metadata": {},
   "source": [
    "First, I'm checking that the data fits the \"rules\" I expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "efd0d556",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:00.053858Z",
     "start_time": "2025-08-20T07:34:00.045244Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    ## all 'semantic type exclusions' don't consider the predicates, \n",
    "    ##   just specific entities/instances of the types\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_predicate, str):\n",
    "            print('semantic type exclusion problem: predicate there?')\n",
    "            print(row)\n",
    "    ## 'Domain exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        if (not isinstance(row.semmed_subject_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('domain exclusion problem: missing subject or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_object_code, str):\n",
    "            print('domain exclusion problem: object present')\n",
    "            print(row)\n",
    "    ## 'Range exclusions' must have a domain/subject + predicate specified\n",
    "    ##   and no range/object specified\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        if (not isinstance(row.semmed_object_code, str)) or (not isinstance(row.semmed_predicate, str)):\n",
    "            print('range exclusion problem: missing object or predicate')\n",
    "            print(row)\n",
    "        elif isinstance(row.semmed_subject_code, str):\n",
    "            print('range exclusion problem: subject present')\n",
    "            print(row)\n",
    "    else: \n",
    "        print('some other kind of exclusion?')\n",
    "        print(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a6bb679",
   "metadata": {},
   "source": [
    "Then, I noticed the same_as predicate was not in the expected case, so adjusting this..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "32c37c3b",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:04.595836Z",
     "start_time": "2025-08-20T07:34:04.580078Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SAME_AS</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          SAME_AS   \n",
       "716                 chvs                  T104          SAME_AS   \n",
       "717                 resa                  T062          SAME_AS   \n",
       "718                 chem                  T103          SAME_AS   \n",
       "719                 qnco                  T081          SAME_AS   \n",
       "720                 ftcn                  T169          SAME_AS   \n",
       "721                 inpr                  T170          SAME_AS   \n",
       "722                 tmco                  T079          SAME_AS   \n",
       "1387                 NaN                   NaN          SAME_AS   \n",
       "1388                 NaN                   NaN          SAME_AS   \n",
       "1389                 NaN                   NaN          SAME_AS   \n",
       "1390                 NaN                   NaN          SAME_AS   \n",
       "1391                 NaN                   NaN          SAME_AS   \n",
       "1392                 NaN                   NaN          SAME_AS   \n",
       "1393                 NaN                   NaN          SAME_AS   \n",
       "1394                 NaN                   NaN          SAME_AS   \n",
       "1395                 NaN                   NaN          SAME_AS   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## viewing\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "89bda465",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:07.734285Z",
     "start_time": "2025-08-20T07:34:07.730733Z"
    }
   },
   "outputs": [],
   "source": [
    "## fixing\n",
    "imported_exclusions['semmed_predicate'] = ['same_as' if i == 'SAME_AS' else i \\\n",
    "                                           for i in imported_exclusions['semmed_predicate']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "1bb3d990",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:19.018672Z",
     "start_time": "2025-08-20T07:34:19.003696Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>715</th>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>717</th>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>718</th>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>ftcn</td>\n",
       "      <td>T169</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>inpr</td>\n",
       "      <td>T170</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>same_as</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Domain exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvf</td>\n",
       "      <td>T120</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>resa</td>\n",
       "      <td>T062</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1389</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chvs</td>\n",
       "      <td>T104</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>chem</td>\n",
       "      <td>T103</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1391</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>qnco</td>\n",
       "      <td>T081</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1392</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>mnob</td>\n",
       "      <td>T073</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1393</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>geoa</td>\n",
       "      <td>T083</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1394</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>prog</td>\n",
       "      <td>T097</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1395</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>same_as</td>\n",
       "      <td>tmco</td>\n",
       "      <td>T079</td>\n",
       "      <td>Range exclusion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     semmed_subject_code semmed_subject_t_code semmed_predicate  \\\n",
       "715                 chvf                  T120          same_as   \n",
       "716                 chvs                  T104          same_as   \n",
       "717                 resa                  T062          same_as   \n",
       "718                 chem                  T103          same_as   \n",
       "719                 qnco                  T081          same_as   \n",
       "720                 ftcn                  T169          same_as   \n",
       "721                 inpr                  T170          same_as   \n",
       "722                 tmco                  T079          same_as   \n",
       "1387                 NaN                   NaN          same_as   \n",
       "1388                 NaN                   NaN          same_as   \n",
       "1389                 NaN                   NaN          same_as   \n",
       "1390                 NaN                   NaN          same_as   \n",
       "1391                 NaN                   NaN          same_as   \n",
       "1392                 NaN                   NaN          same_as   \n",
       "1393                 NaN                   NaN          same_as   \n",
       "1394                 NaN                   NaN          same_as   \n",
       "1395                 NaN                   NaN          same_as   \n",
       "\n",
       "     semmed_object_code semmed_object_t_code    exclusion_type  \n",
       "715                 NaN                  NaN  Domain exclusion  \n",
       "716                 NaN                  NaN  Domain exclusion  \n",
       "717                 NaN                  NaN  Domain exclusion  \n",
       "718                 NaN                  NaN  Domain exclusion  \n",
       "719                 NaN                  NaN  Domain exclusion  \n",
       "720                 NaN                  NaN  Domain exclusion  \n",
       "721                 NaN                  NaN  Domain exclusion  \n",
       "722                 NaN                  NaN  Domain exclusion  \n",
       "1387               chvf                 T120   Range exclusion  \n",
       "1388               resa                 T062   Range exclusion  \n",
       "1389               chvs                 T104   Range exclusion  \n",
       "1390               chem                 T103   Range exclusion  \n",
       "1391               qnco                 T081   Range exclusion  \n",
       "1392               mnob                 T073   Range exclusion  \n",
       "1393               geoa                 T083   Range exclusion  \n",
       "1394               prog                 T097   Range exclusion  \n",
       "1395               tmco                 T079   Range exclusion  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>semmed_subject_code</th>\n",
       "      <th>semmed_subject_t_code</th>\n",
       "      <th>semmed_predicate</th>\n",
       "      <th>semmed_object_code</th>\n",
       "      <th>semmed_object_t_code</th>\n",
       "      <th>exclusion_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [semmed_subject_code, semmed_subject_t_code, semmed_predicate, semmed_object_code, semmed_object_t_code, exclusion_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## checking\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'same_as']\n",
    "imported_exclusions[imported_exclusions['semmed_predicate'] == 'SAME_AS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37364f4a",
   "metadata": {},
   "source": [
    "#### Pruning using exclusions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e6ba83",
   "metadata": {},
   "source": [
    "Now I can iterate through the `imported_exclusions` table and remove metatriples that match the exclusions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "e5bb1ec1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:27.015786Z",
     "start_time": "2025-08-20T07:34:23.271410Z"
    }
   },
   "outputs": [],
   "source": [
    "for row in imported_exclusions.itertuples(index=False):\n",
    "    if row.exclusion_type == 'semantic type exclusion':\n",
    "        if isinstance(row.semmed_subject_code, str):  ## exclusion is in the subject\n",
    "            imported_metatriples = imported_metatriples[ ~ (imported_metatriples['SUBJECT_SEMTYPE'] == row.semmed_subject_code)]\n",
    "        else:  ## exclusion should be in object\n",
    "            imported_metatriples = imported_metatriples[ ~ (imported_metatriples['OBJECT_SEMTYPE'] == row.semmed_object_code)]\n",
    "    elif row.exclusion_type == 'Domain exclusion':\n",
    "        imported_metatriples = imported_metatriples[~ ( (imported_metatriples['SUBJECT_SEMTYPE'] == row.semmed_subject_code ) & \n",
    "                                                        (imported_metatriples['PREDICATE'] == row.semmed_predicate ))]\n",
    "    elif row.exclusion_type == 'Range exclusion':\n",
    "        imported_metatriples = imported_metatriples[~ ( (imported_metatriples['OBJECT_SEMTYPE'] == row.semmed_object_code ) & \n",
    "                                                        (imported_metatriples['PREDICATE'] == row.semmed_predicate ))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "bf7a6309",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:28.129943Z",
     "start_time": "2025-08-20T07:34:28.119382Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11619, 7)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>1139224</td>\n",
       "      <td>1045335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>923819</td>\n",
       "      <td>846909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>895767</td>\n",
       "      <td>816152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>1016052</td>\n",
       "      <td>779870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>844184</td>\n",
       "      <td>735624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>fndg</td>\n",
       "      <td>812383</td>\n",
       "      <td>678731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>798443</td>\n",
       "      <td>664574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>668508</td>\n",
       "      <td>640356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>912167</td>\n",
       "      <td>623202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>umls</td>\n",
       "      <td>phsu</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>781003</td>\n",
       "      <td>544240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "3            umls            bpoc  LOCATION_OF          umls           neop   \n",
       "4            umls            bdsu  LOCATION_OF          umls           aapp   \n",
       "5            umls            bpoc  LOCATION_OF          umls           patf   \n",
       "6            umls            topp       TREATS          umls           dsyn   \n",
       "7            umls            topp       TREATS          umls           neop   \n",
       "9            umls            bpoc  LOCATION_OF          umls           fndg   \n",
       "10           umls            bpoc  LOCATION_OF          umls           dsyn   \n",
       "11           umls            cell      PART_OF          umls           bpoc   \n",
       "12           umls            bpoc  LOCATION_OF          umls           aapp   \n",
       "16           umls            phsu       TREATS          umls           dsyn   \n",
       "\n",
       "      DOC_N  DOC_N_VALID  \n",
       "3   1139224      1045335  \n",
       "4    923819       846909  \n",
       "5    895767       816152  \n",
       "6   1016052       779870  \n",
       "7    844184       735624  \n",
       "9    812383       678731  \n",
       "10   798443       664574  \n",
       "11   668508       640356  \n",
       "12   912167       623202  \n",
       "16   781003       544240  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples.shape\n",
    "imported_metatriples.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "83675d58",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:34:31.977821Z",
     "start_time": "2025-08-20T07:34:31.965435Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4699</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>716</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11337</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>acab</td>\n",
       "      <td>27</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11916</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>cgab</td>\n",
       "      <td>15</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12141</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>51</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13372</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>mobd</td>\n",
       "      <td>23</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13392</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19979</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>inpo</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19981</th>\n",
       "      <td>umls</td>\n",
       "      <td>vita</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>anab</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SUBJECT_PREFIX SUBJECT_SEMTYPE PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "4699            umls            vita    TREATS          umls           dsyn   \n",
       "11337           umls            vita    TREATS          umls           acab   \n",
       "11916           umls            vita    TREATS          umls           cgab   \n",
       "12141           umls            vita    TREATS          umls           neop   \n",
       "13372           umls            vita    TREATS          umls           mobd   \n",
       "13392           umls            vita    TREATS          umls           patf   \n",
       "19979           umls            vita    TREATS          umls           inpo   \n",
       "19981           umls            vita    TREATS          umls           anab   \n",
       "\n",
       "       DOC_N  DOC_N_VALID  \n",
       "4699     716          321  \n",
       "11337     27           14  \n",
       "11916     15           11  \n",
       "12141     51           10  \n",
       "13372     23            6  \n",
       "13392     18            6  \n",
       "19979      6            0  \n",
       "19981      2            0  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[(imported_metatriples['SUBJECT_SEMTYPE'] == 'vita') &\n",
    "                      (imported_metatriples['PREDICATE'] == 'TREATS')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1e9660d",
   "metadata": {},
   "source": [
    "## Final Filter: API-doc-counts per metatriple"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936210c1",
   "metadata": {},
   "source": [
    "### Exploring the count data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "305f3a3a",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "use code blocks below to explore how many API docs there are per metatriple and decide what to use for a cutoff\n",
    "\n",
    "---\n",
    "\n",
    "Current decision: > 20 matching API docs, where each doc has pmid_count > 3 (will pass that filter)\n",
    "\n",
    "Old Notes:\n",
    "* Previously: 4974 metatriples are used to make operations. The cutoff is > 100 matching predications (not API-records/documents).\n",
    "* There are 7605 metatriples that have >=1 matching doc with pmid_count > 3. This set still has a VERY left-skewed distribution. \n",
    "* Our previous rule-of-thumb was creating operations when the number of matching API-docs > 10. This corresponds to 5582 metatriples to make operations from (+608 or +12%). Is this too much?\n",
    "* 20 is comparable to what we had before -> 5054 metatriples (+80 or +2%)\n",
    "* Maybe 100 is too high? -> 3655 metatriples (-1319 or -27%)\n",
    "* Andrew: My gut says that the value of semmeddb is overwhelmingly in the most common metatriples, so at any of these thresholds, I doubt we'll be losing much.  Having said that, as long as being inclusive doesn't result in a performance hit, I'm fine with being inclusive.  So for the moment, let's just shoot for the same basic \"size\" and use a threshold of 20, at least until we have the time to investigate more...\n",
    "* My reply: I agree that covering the most common metatriples is most important. I guess I'm thinking of the June Translator Relay, where people seemed very hesitant about \"removing data\" - which is what we do when we don't make operations for the rarer metatriples. However, I'm not aware of any feedback that BTE is missing expected semmeddb data because of this kind of cutoff....so yep, I agree with you. Going to implement a threshold of > 20 for now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "b8f25836",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:02.239109Z",
     "start_time": "2025-08-20T07:35:02.225854Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    1.161900e+04\n",
       "mean     5.146300e+03\n",
       "std      3.612005e+04\n",
       "min      1.000000e+00\n",
       "25%      5.000000e+00\n",
       "50%      4.700000e+01\n",
       "75%      7.340000e+02\n",
       "max      1.139224e+06\n",
       "Name: DOC_N, dtype: float64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "count    1.161900e+04\n",
       "mean     3.214447e+03\n",
       "std      2.830455e+04\n",
       "min      0.000000e+00\n",
       "25%      0.000000e+00\n",
       "50%      8.000000e+00\n",
       "75%      2.420000e+02\n",
       "max      1.045335e+06\n",
       "Name: DOC_N_VALID, dtype: float64"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking at quick distribution of API doc counts\n",
    "imported_metatriples['DOC_N'].describe()        ## this is total number of docs\n",
    "imported_metatriples['DOC_N_VALID'].describe()  ## this is the total number of docs w/ pmid_count > 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b710b48",
   "metadata": {},
   "source": [
    "So...This output shows that LOTS of metatriples have 0 docs that will pass the current pmid_count filter (> 3). So let's look at the metatriples with documents that will pass the pmid_count filter (`DOC_N_VALID` > 1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "3d0ed289",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:18.190985Z",
     "start_time": "2025-08-20T07:35:18.182634Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    7.555000e+03\n",
       "mean     4.943568e+03\n",
       "std      3.498014e+04\n",
       "min      1.000000e+00\n",
       "25%      9.000000e+00\n",
       "50%      8.800000e+01\n",
       "75%      8.455000e+02\n",
       "max      1.045335e+06\n",
       "Name: DOC_N_VALID, dtype: float64"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 0]['DOC_N_VALID'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "581add5d",
   "metadata": {},
   "source": [
    "So it's still very left-skewed, and there's lots of operations. Let's look at the distribution of the low end of the range"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "ddd78804",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:35.412037Z",
     "start_time": "2025-08-20T07:35:33.148869Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: ylabel='Frequency'>"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjsAAAGdCAYAAAD0e7I1AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8pXeV/AAAACXBIWXMAAA9hAAAPYQGoP6dpAAApZklEQVR4nO3df3DU9Z3H8deaXyZpkiOJZNkSMJ6hVoKehpaKKFhIPH5IlZtDRIEKN4PHD0kJIpTeiD2bIAwBPQ4UjwGUw1jnwLNnRYJgLMd4YgD5YQc4jeFX0pw23QQIm5B87g+H77gkSLLZZDcfno+Z7wz7+X529/0Wmrz6+X72uy5jjBEAAIClrgt1AQAAAJ2JsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsFpkqAsIB83NzTpz5owSEhLkcrlCXQ4AAGgDY4zq6urk8Xh03XVXXr8h7Eg6c+aM0tPTQ10GAAAIwMmTJ9W7d+8rnifsSEpISJD0zX+sxMTEEFcDAADaora2Vunp6c7v8Ssh7EjOpavExETCDgAA3czVtqCwQRkAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAapGhLuBacuOCd5w/f7lkdAgrAQDg2sHKDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAVgtp2Fm8eLFcLpff4Xa7nfPGGC1evFgej0exsbEaNmyYjhw54vcaPp9Ps2fPVmpqquLj4zV27FidOnWqq1sBAABhKuQrO/3791dlZaVzHDp0yDm3dOlSFRUVadWqVdq7d6/cbrdycnJUV1fnzMnLy9PWrVtVXFys3bt36+zZsxozZoyamppC0Q4AAAgzkSEvIDLSbzXnEmOMVq5cqUWLFmncuHGSpI0bNyotLU2bN2/W9OnT5fV6tW7dOr322msaMWKEJGnTpk1KT0/Xjh07dP/993dpLwAAIPyEfGXn+PHj8ng8ysjI0IQJE/TFF19IksrLy1VVVaXc3FxnbkxMjIYOHao9e/ZIksrKytTY2Og3x+PxKCsry5nTGp/Pp9raWr8DAADYKaRhZ9CgQXr11Vf13nvv6ZVXXlFVVZUGDx6sr7/+WlVVVZKktLQ0v+ekpaU556qqqhQdHa0ePXpccU5rCgsLlZSU5Bzp6elB7gwAAISLkIadkSNH6u/+7u80YMAAjRgxQu+8846kby5XXeJyufyeY4xpMXa5q81ZuHChvF6vc5w8ebIDXQAAgHAW8stY3xYfH68BAwbo+PHjzj6ey1doqqurndUet9uthoYG1dTUXHFOa2JiYpSYmOh3AAAAO4VV2PH5fPrjH/+oXr16KSMjQ263WyUlJc75hoYGlZaWavDgwZKk7OxsRUVF+c2prKzU4cOHnTkAAODaFtJPY82bN08PPPCA+vTpo+rqaj333HOqra3VlClT5HK5lJeXp4KCAmVmZiozM1MFBQWKi4vTxIkTJUlJSUmaNm2a8vPzlZKSouTkZM2bN8+5LAYAABDSsHPq1Ck98sgj+uqrr3TDDTfoJz/5iT766CP17dtXkjR//nzV19drxowZqqmp0aBBg7R9+3YlJCQ4r7FixQpFRkZq/Pjxqq+v1/Dhw7VhwwZFRESEqi0AABBGXMYYE+oiQq22tlZJSUnyer2dun/nxgXvOH/+csnoTnsfAACuBW39/R1We3YAAACCjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVwibsFBYWyuVyKS8vzxkzxmjx4sXyeDyKjY3VsGHDdOTIEb/n+Xw+zZ49W6mpqYqPj9fYsWN16tSpLq4eAACEq7AIO3v37tXatWt12223+Y0vXbpURUVFWrVqlfbu3Su3262cnBzV1dU5c/Ly8rR161YVFxdr9+7dOnv2rMaMGaOmpqaubgMAAIShkIeds2fP6tFHH9Urr7yiHj16OOPGGK1cuVKLFi3SuHHjlJWVpY0bN+r8+fPavHmzJMnr9WrdunVavny5RowYoTvuuEObNm3SoUOHtGPHjlC1BAAAwkjIw87MmTM1evRojRgxwm+8vLxcVVVVys3NdcZiYmI0dOhQ7dmzR5JUVlamxsZGvzkej0dZWVnOnNb4fD7V1tb6HQAAwE6RoXzz4uJi7du3T3v37m1xrqqqSpKUlpbmN56WlqaKigpnTnR0tN+K0KU5l57fmsLCQj377LMdLR8AAHQDIVvZOXnypObMmaNNmzbp+uuvv+I8l8vl99gY02Lsclebs3DhQnm9Xuc4efJk+4oHAADdRsjCTllZmaqrq5Wdna3IyEhFRkaqtLRUL774oiIjI50VnctXaKqrq51zbrdbDQ0NqqmpueKc1sTExCgxMdHvAAAAdgpZ2Bk+fLgOHTqkAwcOOMfAgQP16KOP6sCBA7rpppvkdrtVUlLiPKehoUGlpaUaPHiwJCk7O1tRUVF+cyorK3X48GFnDgAAuLaFbM9OQkKCsrKy/Mbi4+OVkpLijOfl5amgoECZmZnKzMxUQUGB4uLiNHHiRElSUlKSpk2bpvz8fKWkpCg5OVnz5s3TgAEDWmx4BgAA16aQblC+mvnz56u+vl4zZsxQTU2NBg0apO3btyshIcGZs2LFCkVGRmr8+PGqr6/X8OHDtWHDBkVERISwcgAAEC5cxhgT6iJCrba2VklJSfJ6vZ26f+fGBe84f/5yyehOex8AAK4Fbf39HfL77AAAAHQmwg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUCCjvl5eXBrgMAAKBTBBR2br75Zt13333atGmTLly4EOyaAAAAgiagsPPpp5/qjjvuUH5+vtxut6ZPn66PP/442LUBAAB0WEBhJysrS0VFRTp9+rTWr1+vqqoqDRkyRP3791dRUZH+7//+L9h1AgAABKRDG5QjIyP10EMP6be//a2ef/55ff7555o3b5569+6tyZMnq7KyMlh1AgAABKRDYeeTTz7RjBkz1KtXLxUVFWnevHn6/PPPtXPnTp0+fVo/+9nPglUnAABAQCIDeVJRUZHWr1+vo0ePatSoUXr11Vc1atQoXXfdN9kpIyNDL7/8sm655ZagFgsAANBeAYWdNWvWaOrUqXr88cfldrtbndOnTx+tW7euQ8UBAAB0VEBh5/jx41edEx0drSlTpgTy8gAAAEET0J6d9evX680332wx/uabb2rjxo0dLgoAACBYAgo7S5YsUWpqaovxnj17qqCgoMNFAQAABEtAYaeiokIZGRktxvv27asTJ050uCgAAIBgCSjs9OzZUwcPHmwx/umnnyolJaXDRQEAAARLQGFnwoQJevLJJ7Vr1y41NTWpqalJO3fu1Jw5czRhwoRg1wgAABCwgD6N9dxzz6miokLDhw9XZOQ3L9Hc3KzJkyezZwcAAISVgMJOdHS03njjDf3zP/+zPv30U8XGxmrAgAHq27dvsOsDAADokIDCziX9+vVTv379glULAABA0AUUdpqamrRhwwa9//77qq6uVnNzs9/5nTt3BqU4AACAjgoo7MyZM0cbNmzQ6NGjlZWVJZfLFey6AAAAgiKgsFNcXKzf/va3GjVqVLDrAQAACKqAPnoeHR2tm2++Odi1AAAABF1AYSc/P18vvPCCjDEdevM1a9botttuU2JiohITE3XXXXfp3Xffdc4bY7R48WJ5PB7FxsZq2LBhOnLkiN9r+Hw+zZ49W6mpqYqPj9fYsWN16tSpDtUFAADsEdBlrN27d2vXrl1699131b9/f0VFRfmd37JlS5tep3fv3lqyZImzSrRx40b97Gc/0/79+9W/f38tXbpURUVF2rBhg/r166fnnntOOTk5Onr0qBISEiRJeXl5+t3vfqfi4mKlpKQoPz9fY8aMUVlZmSIiIgJpDwAAWMRlAlieefzxx7/z/Pr16wMuKDk5WcuWLdPUqVPl8XiUl5enp59+WtI3qzhpaWl6/vnnNX36dHm9Xt1www167bXX9PDDD0uSzpw5o/T0dP3+97/X/fff36b3rK2tVVJSkrxerxITEwOu/WpuXPCO8+cvl4zutPcBAOBa0Nbf3wGt7HQkzFxJU1OT3nzzTZ07d0533XWXysvLVVVVpdzcXGdOTEyMhg4dqj179mj69OkqKytTY2Oj3xyPx6OsrCzt2bOnzWEHAADYK+CbCl68eFEffPCBPv/8c02cOFEJCQk6c+aMEhMT9b3vfa/Nr3Po0CHdddddunDhgr73ve9p69atuvXWW7Vnzx5JUlpamt/8tLQ0VVRUSJKqqqoUHR2tHj16tJhTVVV1xff0+Xzy+XzO49ra2jbXCwAAupeAwk5FRYX+9m//VidOnJDP51NOTo4SEhK0dOlSXbhwQS+99FKbX+sHP/iBDhw4oL/85S/6j//4D02ZMkWlpaXO+cvv4WOMuep9fa42p7CwUM8++2ybawQAAN1XQJ/GmjNnjgYOHKiamhrFxsY64w899JDef//9dr3WpY+xDxw4UIWFhbr99tv1wgsvyO12S1KLFZrq6mpntcftdquhoUE1NTVXnNOahQsXyuv1OsfJkyfbVTMAAOg+Ago7u3fv1q9+9StFR0f7jfft21enT5/uUEHGGPl8PmVkZMjtdqukpMQ519DQoNLSUg0ePFiSlJ2draioKL85lZWVOnz4sDOnNTExMc7H3S8dAADATgFdxmpublZTU1OL8VOnTjkfCW+LX/7ylxo5cqTS09NVV1en4uJiffDBB9q2bZtcLpfy8vJUUFCgzMxMZWZmqqCgQHFxcZo4caIkKSkpSdOmTVN+fr5SUlKUnJysefPmacCAARoxYkQgrQEAAMsEFHZycnK0cuVKrV27VtI3+2rOnj2rZ555pl1fIfGnP/1JkyZNUmVlpZKSknTbbbdp27ZtysnJkSTNnz9f9fX1mjFjhmpqajRo0CBt377dL1CtWLFCkZGRGj9+vOrr6zV8+HBt2LCBe+wAAABJAd5n58yZM7rvvvsUERGh48ePa+DAgTp+/LhSU1P14YcfqmfPnp1Ra6fhPjsAAHQ/nXqfHY/HowMHDuj111/Xvn371NzcrGnTpunRRx/127AMAAAQagHfZyc2NlZTp07V1KlTg1kPAABAUAUUdl599dXvPD958uSAigEAAAi2gMLOnDlz/B43Njbq/Pnzio6OVlxcHGEHAACEjYDus1NTU+N3nD17VkePHtWQIUP0+uuvB7tGAACAgAUUdlqTmZmpJUuWtFj1AQAACKWghR1JioiI0JkzZ4L5kgAAAB0S0J6dt99+2++xMUaVlZVatWqV7r777qAUBgAAEAwBhZ0HH3zQ77HL5dINN9ygn/70p1q+fHkw6gIAAAiKgL8bCwAAoDsI6p4dAACAcBPQys7cuXPbPLeoqCiQtwAAAAiKgMLO/v37tW/fPl28eFE/+MEPJEnHjh1TRESE7rzzTmeey+UKTpUAAAABCijsPPDAA0pISNDGjRvVo0cPSd/caPDxxx/XPffco/z8/KAWCQAAEKiA9uwsX75chYWFTtCRpB49eui5557j01gAACCsBBR2amtr9ac//anFeHV1terq6jpcFAAAQLAEdBnroYce0uOPP67ly5frJz/5iSTpo48+0lNPPaVx48YFtcBrzY0L3nH+/OWS0SGsBAAAOwQUdl566SXNmzdPjz32mBobG795ochITZs2TcuWLQtqgQAAAB0RUNiJi4vT6tWrtWzZMn3++ecyxujmm29WfHx8sOsDAADokA7dVLCyslKVlZXq16+f4uPjZYwJVl0AAABBEVDY+frrrzV8+HD169dPo0aNUmVlpSTpH/7hH/jYOQAACCsBhZ1f/OIXioqK0okTJxQXF+eMP/zww9q2bVvQigMAAOiogPbsbN++Xe+995569+7tN56ZmamKioqgFAYAABAMAa3snDt3zm9F55KvvvpKMTExHS4KAAAgWAIKO/fee69effVV57HL5VJzc7OWLVum++67L2jF2ezGBe84BwAA6DwBXcZatmyZhg0bpk8++UQNDQ2aP3++jhw5oj//+c/67//+72DXCAAAELCAVnZuvfVWHTx4UD/+8Y+Vk5Ojc+fOady4cdq/f7/++q//Otg1AgAABKzdKzuNjY3Kzc3Vyy+/rGeffbYzagIAAAiadq/sREVF6fDhw3K5XJ1RDwAAQFAFdBlr8uTJWrduXbBrAQAACLqANig3NDTo3/7t31RSUqKBAwe2+E6soqKioBQHAADQUe0KO1988YVuvPFGHT58WHfeeack6dixY35zuLwFAADCSbvCTmZmpiorK7Vr1y5J33w9xIsvvqi0tLROKQ4AAKCj2rVn5/JvNX/33Xd17ty5oBYEAAAQTAFtUL7k8vADAAAQbtoVdlwuV4s9OezRAQAA4axde3aMMfr5z3/ufNnnhQsX9MQTT7T4NNaWLVuCVyEAAEAHtCvsTJkyxe/xY489FtRiAAAAgq1dYWf9+vWdVQcAAECn6NAGZQAAgHBH2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgtZCGncLCQv3oRz9SQkKCevbsqQcffFBHjx71m2OM0eLFi+XxeBQbG6thw4bpyJEjfnN8Pp9mz56t1NRUxcfHa+zYsTp16lRXtgIAAMJUSMNOaWmpZs6cqY8++kglJSW6ePGicnNzde7cOWfO0qVLVVRUpFWrVmnv3r1yu93KyclRXV2dMycvL09bt25VcXGxdu/erbNnz2rMmDFqamoKRVsAACCMRIbyzbdt2+b3eP369erZs6fKysp07733yhijlStXatGiRRo3bpwkaePGjUpLS9PmzZs1ffp0eb1erVu3Tq+99ppGjBghSdq0aZPS09O1Y8cO3X///V3eFwAACB9htWfH6/VKkpKTkyVJ5eXlqqqqUm5urjMnJiZGQ4cO1Z49eyRJZWVlamxs9Jvj8XiUlZXlzAEAANeukK7sfJsxRnPnztWQIUOUlZUlSaqqqpIkpaWl+c1NS0tTRUWFMyc6Olo9evRoMefS8y/n8/nk8/mcx7W1tUHrAwAAhJewWdmZNWuWDh48qNdff73FOZfL5ffYGNNi7HLfNaewsFBJSUnOkZ6eHnjhAAAgrIVF2Jk9e7befvtt7dq1S71793bG3W63JLVYoamurnZWe9xutxoaGlRTU3PFOZdbuHChvF6vc5w8eTKY7QAAgDAS0rBjjNGsWbO0ZcsW7dy5UxkZGX7nMzIy5Ha7VVJS4ow1NDSotLRUgwcPliRlZ2crKirKb05lZaUOHz7szLlcTEyMEhMT/Q4AAGCnkO7ZmTlzpjZv3qz//M//VEJCgrOCk5SUpNjYWLlcLuXl5amgoECZmZnKzMxUQUGB4uLiNHHiRGfutGnTlJ+fr5SUFCUnJ2vevHkaMGCA8+ksAABw7Qpp2FmzZo0kadiwYX7j69ev189//nNJ0vz581VfX68ZM2aopqZGgwYN0vbt25WQkODMX7FihSIjIzV+/HjV19dr+PDh2rBhgyIiIrqqFQAAEKZcxhgT6iJCrba2VklJSfJ6vZ16SevGBe+0Ov7lktGtzvn2OAAA8NfW399hsUEZAACgsxB2AACA1Qg7AADAamFzB2W0xP4dAAA6jpUdAABgNcIOAACwGmEHAABYjT07YeBK998BAAAdx8oOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKxG2AEAAFYj7AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABWI+wAAACrEXYAAIDVCDsAAMBqhB0AAGA1wg4AALAaYQcAAFiNsAMAAKwWGeoC0DY3Lnin1fEvl4zu4koAAOheWNkBAABWI+wAAACrEXYAAIDV2LPTzV2+l4c9PAAA+GNlBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1UIadj788EM98MAD8ng8crlceuutt/zOG2O0ePFieTwexcbGatiwYTpy5IjfHJ/Pp9mzZys1NVXx8fEaO3asTp061YVdAACAcBbSsHPu3DndfvvtWrVqVavnly5dqqKiIq1atUp79+6V2+1WTk6O6urqnDl5eXnaunWriouLtXv3bp09e1ZjxoxRU1NTV7UBAADCWEjvszNy5EiNHDmy1XPGGK1cuVKLFi3SuHHjJEkbN25UWlqaNm/erOnTp8vr9WrdunV67bXXNGLECEnSpk2blJ6erh07duj+++/vsl4AAEB4Cts9O+Xl5aqqqlJubq4zFhMTo6FDh2rPnj2SpLKyMjU2NvrN8Xg8ysrKcua0xufzqba21u8AAAB2CtuwU1VVJUlKS0vzG09LS3POVVVVKTo6Wj169LjinNYUFhYqKSnJOdLT04NcPQAACBdhG3Yucblcfo+NMS3GLne1OQsXLpTX63WOkydPBqVWAAAQfsL2u7Hcbrekb1ZvevXq5YxXV1c7qz1ut1sNDQ2qqanxW92prq7W4MGDr/jaMTExiomJ6aTKQ+vb35XF92QBABDGKzsZGRlyu90qKSlxxhoaGlRaWuoEmezsbEVFRfnNqays1OHDh78z7AAAgGtHSFd2zp49q//93/91HpeXl+vAgQNKTk5Wnz59lJeXp4KCAmVmZiozM1MFBQWKi4vTxIkTJUlJSUmaNm2a8vPzlZKSouTkZM2bN08DBgxwPp0FAACubSENO5988onuu+8+5/HcuXMlSVOmTNGGDRs0f/581dfXa8aMGaqpqdGgQYO0fft2JSQkOM9ZsWKFIiMjNX78eNXX12v48OHasGGDIiIiurwfAAAQflzGGBPqIkKttrZWSUlJ8nq9SkxM7LT3+fZ+mq7Anh0AgM3a+vs7bPfsAAAABANhBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAamH73VgILr4zCwBwrWJlBwAAWI2VHYt19R2bAQAIR6zsAAAAqxF2AACA1Qg7AADAaoQdAABgNcIOAACwGmEHAABYjbADAACsRtgBAABW46aC16Ar3WyQr5EAANiIlR0AAGA1wg4AALAaYQcAAFiNPTtwfHsvD/t3AAC2YGUHAABYjbADAACsxmUstKq9l7S4BAYACFeEHVwVQQYA0J1xGQsAAFiNsAMAAKzGZSy0y5W+agIAgHDFyg4AALAaKzsIOjY0AwDCCSs7AADAaoQdAABgNS5jISTacqmLy2EAgGAg7KBTdeROzAAABANhB91CR1aCWCECgGsbYQdd5kqrNu1dzWlv8AEAXNvYoAwAAKzGyg66NVZwAABXQ9jBNeta28tzrfULAJcQdgARBADAZuzZAQAAVmNlB7jMlfYBtXfFpy0fhe/I6wMA2oaVHQAAYDVWdnBN6cintzp7X084rPh0Ro/shwIQaoQdwDJd+b1jBJnA8N8N6FqEHaAbCYfVn87QlatmfJUIcO0h7AAB6MhXXHT2e7X3uW35Go+uvGwXbhu72/K+hCkgvBF2ALRLRwJUZ7xXOLxvd0coQ2cIp39XhB3AAuHwyzlUNYRD798WrIAWqr1U4bayhuAK1YptqBF2AFxVuP4AC4RNvYQbAhHCFWEHQJexKWh0di9tef2OrNp0pc64UeeVhMNtIdq7ehJOl3uk8KsnGKwJO6tXr9ayZctUWVmp/v37a+XKlbrnnntCXRaAbqI7BrH23magI6/flTp771Vn3EOqLXO68t5VwaqtvZc1w5UVYeeNN95QXl6eVq9erbvvvlsvv/yyRo4cqc8++0x9+vQJdXkAEBYrQfhGqFYuOmPFpyt76c7/xlzGGBPqIjpq0KBBuvPOO7VmzRpn7Ic//KEefPBBFRYWXvX5tbW1SkpKktfrVWJiYqfV2Z3/oQAAEKjOCmJt/f3d7Vd2GhoaVFZWpgULFviN5+bmas+ePa0+x+fzyefzOY+9Xq+kb/6jdaZm3/lOfX0AAMJRZ/1+vfS6V1u36fZh56uvvlJTU5PS0tL8xtPS0lRVVdXqcwoLC/Xss8+2GE9PT++UGgEAuJYlrezc16+rq1NSUtIVz3f7sHOJy+Xye2yMaTF2ycKFCzV37lzncXNzs/785z8rJSXlis8JRG1trdLT03Xy5MlOvTwWTq61nunXbvRrN/rt/owxqqurk8fj+c553T7spKamKiIiosUqTnV1dYvVnktiYmIUExPjN/ZXf/VXnVWiEhMTrfmH1VbXWs/0azf6tRv9dm/ftaJzyXVdUEenio6OVnZ2tkpKSvzGS0pKNHjw4BBVBQAAwkW3X9mRpLlz52rSpEkaOHCg7rrrLq1du1YnTpzQE088EerSAABAiFkRdh5++GF9/fXX+vWvf63KykplZWXp97//vfr27RvSumJiYvTMM8+0uGRms2utZ/q1G/3ajX6vHVbcZwcAAOBKuv2eHQAAgO9C2AEAAFYj7AAAAKsRdgAAgNUIO51o9erVysjI0PXXX6/s7Gz94Q9/CHVJQVFYWKgf/ehHSkhIUM+ePfXggw/q6NGjfnOMMVq8eLE8Ho9iY2M1bNgwHTlyJEQVB1dhYaFcLpfy8vKcMdv6PX36tB577DGlpKQoLi5Of/M3f6OysjLnvE39Xrx4Ub/61a+UkZGh2NhY3XTTTfr1r3+t5uZmZ0537vfDDz/UAw88II/HI5fLpbfeesvvfFt68/l8mj17tlJTUxUfH6+xY8fq1KlTXdhF231Xv42NjXr66ac1YMAAxcfHy+PxaPLkyTpz5ozfa3SnfqWr/x1/2/Tp0+VyubRy5Uq/8e7Wc3sRdjrJG2+8oby8PC1atEj79+/XPffco5EjR+rEiROhLq3DSktLNXPmTH300UcqKSnRxYsXlZubq3Pnzjlzli5dqqKiIq1atUp79+6V2+1WTk6O6urqQlh5x+3du1dr167Vbbfd5jduU781NTW6++67FRUVpXfffVefffaZli9f7neXcZv6ff755/XSSy9p1apV+uMf/6ilS5dq2bJl+pd/+RdnTnfu99y5c7r99tu1atWqVs+3pbe8vDxt3bpVxcXF2r17t86ePasxY8aoqampq9pos+/q9/z589q3b5/+6Z/+Sfv27dOWLVt07NgxjR071m9ed+pXuvrf8SVvvfWW/ud//qfVr1bobj23m0Gn+PGPf2yeeOIJv7FbbrnFLFiwIEQVdZ7q6mojyZSWlhpjjGlubjZut9ssWbLEmXPhwgWTlJRkXnrppVCV2WF1dXUmMzPTlJSUmKFDh5o5c+YYY+zr9+mnnzZDhgy54nnb+h09erSZOnWq39i4cePMY489Zoyxq19JZuvWrc7jtvT2l7/8xURFRZni4mJnzunTp811111ntm3b1mW1B+Lyflvz8ccfG0mmoqLCGNO9+zXmyj2fOnXKfP/73zeHDx82ffv2NStWrHDOdfee24KVnU7Q0NCgsrIy5ebm+o3n5uZqz549Iaqq83i9XklScnKyJKm8vFxVVVV+/cfExGjo0KHduv+ZM2dq9OjRGjFihN+4bf2+/fbbGjhwoP7+7/9ePXv21B133KFXXnnFOW9bv0OGDNH777+vY8eOSZI+/fRT7d69W6NGjZJkX7/f1pbeysrK1NjY6DfH4/EoKyur2/cvffPzy+VyOSuXNvbb3NysSZMm6amnnlL//v1bnLex58tZcQflcPPVV1+pqampxReRpqWltfjC0u7OGKO5c+dqyJAhysrKkiSnx9b6r6io6PIag6G4uFj79u3T3r17W5yzrd8vvvhCa9as0dy5c/XLX/5SH3/8sZ588knFxMRo8uTJ1vX79NNPy+v16pZbblFERISampr0m9/8Ro888ogk+/5+v60tvVVVVSk6Olo9evRoMae7/zy7cOGCFixYoIkTJzpfjGljv88//7wiIyP15JNPtnrexp4vR9jpRC6Xy++xMabFWHc3a9YsHTx4ULt3725xzpb+T548qTlz5mj79u26/vrrrzjPln6bm5s1cOBAFRQUSJLuuOMOHTlyRGvWrNHkyZOdebb0+8Ybb2jTpk3avHmz+vfvrwMHDigvL08ej0dTpkxx5tnSb2sC6a2799/Y2KgJEyaoublZq1evvur87tpvWVmZXnjhBe3bt6/d9XfXnlvDZaxOkJqaqoiIiBaJuLq6usX/g+rOZs+erbffflu7du1S7969nXG32y1J1vRfVlam6upqZWdnKzIyUpGRkSotLdWLL76oyMhIpydb+u3Vq5duvfVWv7Ef/vCHzuZ62/5+n3rqKS1YsEATJkzQgAEDNGnSJP3iF79QYWGhJPv6/ba29OZ2u9XQ0KCamporzuluGhsbNX78eJWXl6ukpMRZ1ZHs6/cPf/iDqqur1adPH+fnV0VFhfLz83XjjTdKsq/n1hB2OkF0dLSys7NVUlLiN15SUqLBgweHqKrgMcZo1qxZ2rJli3bu3KmMjAy/8xkZGXK73X79NzQ0qLS0tFv2P3z4cB06dEgHDhxwjoEDB+rRRx/VgQMHdNNNN1nV7913393iVgLHjh1zvljXtr/f8+fP67rr/H8URkREOB89t63fb2tLb9nZ2YqKivKbU1lZqcOHD3fL/i8FnePHj2vHjh1KSUnxO29bv5MmTdLBgwf9fn55PB499dRTeu+99yTZ13OrQrQx2nrFxcUmKirKrFu3znz22WcmLy/PxMfHmy+//DLUpXXYP/7jP5qkpCTzwQcfmMrKSuc4f/68M2fJkiUmKSnJbNmyxRw6dMg88sgjplevXqa2tjaElQfPtz+NZYxd/X788ccmMjLS/OY3vzHHjx83//7v/27i4uLMpk2bnDk29TtlyhTz/e9/3/zXf/2XKS8vN1u2bDGpqalm/vz5zpzu3G9dXZ3Zv3+/2b9/v5FkioqKzP79+51PH7WltyeeeML07t3b7Nixw+zbt8/89Kc/Nbfffru5ePFiqNq6ou/qt7Gx0YwdO9b07t3bHDhwwO/nl8/nc16jO/VrzNX/ji93+aexjOl+PbcXYacT/eu//qvp27eviY6ONnfeeafz0ezuTlKrx/r16505zc3N5plnnjFut9vExMSYe++91xw6dCh0RQfZ5WHHtn5/97vfmaysLBMTE2NuueUWs3btWr/zNvVbW1tr5syZY/r06WOuv/56c9NNN5lFixb5/fLrzv3u2rWr1f+9TpkyxRjTtt7q6+vNrFmzTHJysomNjTVjxowxJ06cCEE3V/dd/ZaXl1/x59euXbuc1+hO/Rpz9b/jy7UWdrpbz+3lMsaYrlhBAgAACAX27AAAAKsRdgAAgNUIOwAAwGqEHQAAYDXCDgAAsBphBwAAWI2wAwAArEbYAQAAViPsAAAAqxF2AACA1Qg7AADAaoQdAABgtf8Hp4qngL0cMR4AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 0]['DOC_N_VALID'].plot.hist(bins=150, range=[0,150])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "665de94a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:40.743795Z",
     "start_time": "2025-08-20T07:35:40.726638Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6122, 7)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5587, 7)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5279, 7)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(5049, 7)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/plain": [
       "(3668, 7)"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking at how many metatriples would be left at different cutoffs\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 5].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 10].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 15].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 20].shape\n",
    "imported_metatriples[imported_metatriples['DOC_N_VALID'] > 100].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d81fd09",
   "metadata": {},
   "source": [
    "After discussion with Andrew Su, we decided to set a threshold of 20 for now. \n",
    "\n",
    "I wondered what was going on with the same_as metatriples, so I ran the code block below. Even after the cutoff, it looks like some of these same_as metatriples probably mean \"similar effectiveness\" which is kind of awkward with the mapped biolink-model predicate of \"close_match\". \n",
    "* orch (organic chemical) same as medd (medical device)? or topp (therapeutic procedure)? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "86920c33",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:52.466032Z",
     "start_time": "2025-08-20T07:35:52.452773Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "      <th>DOC_N</th>\n",
       "      <th>DOC_N_VALID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4044</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>2971</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8185</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>198</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8651</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>phsu</td>\n",
       "      <td>356</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9511</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>medd</td>\n",
       "      <td>251</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9626</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>antb</td>\n",
       "      <td>181</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9631</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>264</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9730</th>\n",
       "      <td>umls</td>\n",
       "      <td>orch</td>\n",
       "      <td>same_as</td>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>236</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SUBJECT_PREFIX SUBJECT_SEMTYPE PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE  \\\n",
       "4044           umls            orch   same_as          umls           orch   \n",
       "8185           umls            orch   same_as          umls           hops   \n",
       "8651           umls            orch   same_as          umls           phsu   \n",
       "9511           umls            orch   same_as          umls           medd   \n",
       "9626           umls            orch   same_as          umls           antb   \n",
       "9631           umls            orch   same_as          umls           aapp   \n",
       "9730           umls            orch   same_as          umls           topp   \n",
       "\n",
       "      DOC_N  DOC_N_VALID  \n",
       "4044   2971          477  \n",
       "8185    198           54  \n",
       "8651    356           44  \n",
       "9511    251           30  \n",
       "9626    181           28  \n",
       "9631    264           28  \n",
       "9730    236           27  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imported_metatriples[(imported_metatriples['DOC_N_VALID'] > 20) &\n",
    "                     (imported_metatriples['PREDICATE'] == 'same_as') & \n",
    "                     (imported_metatriples['SUBJECT_SEMTYPE'] == 'orch')\n",
    "                    ]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7d79a90",
   "metadata": {},
   "source": [
    "### Pruning for `DOC_N_VALID` > 20 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2c859df9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:35:58.400995Z",
     "start_time": "2025-08-20T07:35:58.388150Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5049, 5)"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUBJECT_PREFIX</th>\n",
       "      <th>SUBJECT_SEMTYPE</th>\n",
       "      <th>PREDICATE</th>\n",
       "      <th>OBJECT_PREFIX</th>\n",
       "      <th>OBJECT_SEMTYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SUBJECT_PREFIX SUBJECT_SEMTYPE    PREDICATE OBJECT_PREFIX OBJECT_SEMTYPE\n",
       "3           umls            bpoc  LOCATION_OF          umls           neop\n",
       "4           umls            bdsu  LOCATION_OF          umls           aapp\n",
       "5           umls            bpoc  LOCATION_OF          umls           patf"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples = imported_metatriples[(imported_metatriples['DOC_N_VALID'] > 20)].copy()\n",
    "final_metatriples.drop(columns=['DOC_N', 'DOC_N_VALID'], inplace=True)\n",
    "final_metatriples.shape\n",
    "final_metatriples[0:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dc02b1b",
   "metadata": {},
   "source": [
    "## Constructing final table with biolink-mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "b53030eb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:04.401068Z",
     "start_time": "2025-08-20T07:36:04.392844Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all SEMMED subject semantic types to biolink categories\n",
    "\n",
    "final_metatriples = final_metatriples.merge(SEMMED_semtypes[['Abbrev', 'BiolinkMapping']], \n",
    "                      how='left', left_on='SUBJECT_SEMTYPE', right_on='Abbrev')\n",
    "\n",
    "final_metatriples.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "final_metatriples.columns = [\n",
    "    'SubjectPrefix', \n",
    "    'OriginalSubject', \n",
    "    'OriginalPredicate',\n",
    "    'ObjectPrefix',\n",
    "    'OriginalObject',\n",
    "    'BiolinkSubject'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "70092440",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:05.803490Z",
     "start_time": "2025-08-20T07:36:05.792733Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5044</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celc</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5045</th>\n",
       "      <td>umls</td>\n",
       "      <td>emst</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aggp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5046</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celf</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5047</th>\n",
       "      <td>umls</td>\n",
       "      <td>opco</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>tisu</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5048</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>comd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5049 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5044          umls            nsba          DISRUPTS         umls   \n",
       "5045          umls            emst           PART_OF         umls   \n",
       "5046      ncbigene            phsu           AFFECTS         umls   \n",
       "5047          umls            opco          AUGMENTS         umls   \n",
       "5048          umls            nsba           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject  \n",
       "0              neop  GrossAnatomicalStructure  \n",
       "1              aapp          AnatomicalEntity  \n",
       "2              patf  GrossAnatomicalStructure  \n",
       "3              dsyn                 Procedure  \n",
       "4              neop                 Procedure  \n",
       "...             ...                       ...  \n",
       "5044           celc             SmallMolecule  \n",
       "5045           aggp  GrossAnatomicalStructure  \n",
       "5046           celf             SmallMolecule  \n",
       "5047           tisu             SmallMolecule  \n",
       "5048           comd             SmallMolecule  \n",
       "\n",
       "[5049 rows x 6 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "9e435399",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:10.693851Z",
     "start_time": "2025-08-20T07:36:10.685525Z"
    }
   },
   "outputs": [],
   "source": [
    "## now have to map all SEMMED object semantic types to biolink categories\n",
    "\n",
    "final_metatriples = final_metatriples.merge(SEMMED_semtypes[['Abbrev', 'BiolinkMapping']],\n",
    "                      how='left', left_on='OriginalObject', right_on='Abbrev')\n",
    "\n",
    "final_metatriples.drop(columns = 'Abbrev', inplace=True)\n",
    "\n",
    "final_metatriples.columns = [\n",
    "    'SubjectPrefix', \n",
    "    'OriginalSubject', \n",
    "    'OriginalPredicate',\n",
    "    'ObjectPrefix',\n",
    "    'OriginalObject',\n",
    "    'BiolinkSubject',\n",
    "    'BiolinkObject'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "a0fe51f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:11.328857Z",
     "start_time": "2025-08-20T07:36:11.317803Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5044</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celc</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CellularComponent</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5045</th>\n",
       "      <td>umls</td>\n",
       "      <td>emst</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aggp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Cohort</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5046</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celf</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5047</th>\n",
       "      <td>umls</td>\n",
       "      <td>opco</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>tisu</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5048</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>comd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5049 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5044          umls            nsba          DISRUPTS         umls   \n",
       "5045          umls            emst           PART_OF         umls   \n",
       "5046      ncbigene            phsu           AFFECTS         umls   \n",
       "5047          umls            opco          AUGMENTS         umls   \n",
       "5048          umls            nsba           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject             BiolinkObject  \n",
       "0              neop  GrossAnatomicalStructure                   Disease  \n",
       "1              aapp          AnatomicalEntity               Polypeptide  \n",
       "2              patf  GrossAnatomicalStructure       PathologicalProcess  \n",
       "3              dsyn                 Procedure                   Disease  \n",
       "4              neop                 Procedure                   Disease  \n",
       "...             ...                       ...                       ...  \n",
       "5044           celc             SmallMolecule         CellularComponent  \n",
       "5045           aggp  GrossAnatomicalStructure                    Cohort  \n",
       "5046           celf             SmallMolecule      PhysiologicalProcess  \n",
       "5047           tisu             SmallMolecule  GrossAnatomicalStructure  \n",
       "5048           comd             SmallMolecule                   Disease  \n",
       "\n",
       "[5049 rows x 7 columns]"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "3f42bd07",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:15.118821Z",
     "start_time": "2025-08-20T07:36:15.110922Z"
    }
   },
   "outputs": [],
   "source": [
    "## add all biolink predicate / qualifier / inverse stuff\n",
    "final_metatriples = final_metatriples.merge(\n",
    "                      SEMMED_predicates,\n",
    "                      how='left', left_on='OriginalPredicate', right_on='Semmed_in_Data')\n",
    "\n",
    "final_metatriples.drop(columns = 'Semmed_in_Data', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6c651d23",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:15.897758Z",
     "start_time": "2025-08-20T07:36:15.883917Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix OriginalObject  \\\n",
       "0          umls            bpoc       LOCATION_OF         umls           neop   \n",
       "1          umls            bdsu       LOCATION_OF         umls           aapp   \n",
       "2          umls            bpoc       LOCATION_OF         umls           patf   \n",
       "\n",
       "             BiolinkSubject        BiolinkObject       BMPred BMQualPred  \\\n",
       "0  GrossAnatomicalStructure              Disease  location_of       None   \n",
       "1          AnatomicalEntity          Polypeptide  location_of       None   \n",
       "2  GrossAnatomicalStructure  PathologicalProcess  location_of       None   \n",
       "\n",
       "  BMObjAsp BMObjDirect BM_Inv_Pred BM_Inv_QualPred BM_Inv_SubjAsp  \\\n",
       "0     None        None  located_in            None           None   \n",
       "1     None        None  located_in            None           None   \n",
       "2     None        None  located_in            None           None   \n",
       "\n",
       "  BM_Inv_SubjDirect  \n",
       "0              None  \n",
       "1              None  \n",
       "2              None  "
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_metatriples[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "cc7864f5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:19.877646Z",
     "start_time": "2025-08-20T07:36:19.849184Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## check to make sure everything is mapped to Biolink successfully\n",
    "## should get all empty dataframes\n",
    "\n",
    "final_metatriples[final_metatriples['BiolinkSubject'].isna()]\n",
    "final_metatriples[final_metatriples['BiolinkObject'].isna()]\n",
    "\n",
    "final_metatriples[final_metatriples['BMPred'].isna()]\n",
    "final_metatriples[final_metatriples['BM_Inv_Pred'].isna()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de2c500a",
   "metadata": {},
   "source": [
    "### Adjust Biolink categories based on NCBIGene prefix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5173b760",
   "metadata": {},
   "source": [
    "There's two ID-namespaces for subject + object: umls + ncbigene. But biolink-model only recognizes NCBIGene as a valid namespace for specific categories, like Gene. \n",
    "\n",
    "So we're going to replace the biolink subject / object mapping with Gene whenever it has the NCBIGene prefix..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "b8d843fb",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:25.684382Z",
     "start_time": "2025-08-20T07:36:25.650399Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>umls</td>\n",
       "      <td>gngm</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ASSOCIATED_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Disease</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>related_to</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4993</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>umls</td>\n",
       "      <td>nusq</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>NucleicAcidEntity</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5005</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>umls</td>\n",
       "      <td>enzy</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>Protein</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5012</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>anst</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5038</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>orgf</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5046</th>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celf</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>PhysiologicalProcess</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>288 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "142       ncbigene            aapp        STIMULATES         umls   \n",
       "147       ncbigene            aapp   ASSOCIATED_WITH         umls   \n",
       "151       ncbigene            aapp    INTERACTS_WITH         umls   \n",
       "152       ncbigene            bacs           PART_OF         umls   \n",
       "160       ncbigene            aapp   ASSOCIATED_WITH         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "4993      ncbigene            aapp          INHIBITS         umls   \n",
       "5005      ncbigene            aapp    INTERACTS_WITH         umls   \n",
       "5012      ncbigene            aapp           AFFECTS         umls   \n",
       "5038      ncbigene            phsu          AUGMENTS         umls   \n",
       "5046      ncbigene            phsu           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject  BiolinkSubject         BiolinkObject          BMPred  \\\n",
       "142            gngm     Polypeptide                  Gene         affects   \n",
       "147            neop     Polypeptide               Disease      related_to   \n",
       "151            aapp     Polypeptide           Polypeptide  interacts_with   \n",
       "152            aapp  ChemicalEntity           Polypeptide         part_of   \n",
       "160            dsyn     Polypeptide               Disease      related_to   \n",
       "...             ...             ...                   ...             ...   \n",
       "4993           nusq     Polypeptide     NucleicAcidEntity         affects   \n",
       "5005           enzy     Polypeptide               Protein  interacts_with   \n",
       "5012           anst     Polypeptide      AnatomicalEntity         affects   \n",
       "5038           orgf   SmallMolecule  PhysiologicalProcess         affects   \n",
       "5046           celf   SmallMolecule  PhysiologicalProcess         affects   \n",
       "\n",
       "     BMQualPred               BMObjAsp BMObjDirect     BM_Inv_Pred  \\\n",
       "142      causes  activity_or_abundance   increased     affected_by   \n",
       "147        None                   None        None      related_to   \n",
       "151        None                   None        None  interacts_with   \n",
       "152        None                   None        None        has_part   \n",
       "160        None                   None        None      related_to   \n",
       "...         ...                    ...         ...             ...   \n",
       "4993     causes  activity_or_abundance   decreased     affected_by   \n",
       "5005       None                   None        None  interacts_with   \n",
       "5012       None                   None        None     affected_by   \n",
       "5038     causes  activity_or_abundance   increased     affected_by   \n",
       "5046       None                   None        None     affected_by   \n",
       "\n",
       "     BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "142        caused_by  activity_or_abundance         increased  \n",
       "147             None                   None              None  \n",
       "151             None                   None              None  \n",
       "152             None                   None              None  \n",
       "160             None                   None              None  \n",
       "...              ...                    ...               ...  \n",
       "4993       caused_by  activity_or_abundance         decreased  \n",
       "5005            None                   None              None  \n",
       "5012            None                   None              None  \n",
       "5038       caused_by  activity_or_abundance         increased  \n",
       "5046            None                   None              None  \n",
       "\n",
       "[288 rows x 15 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>umls</td>\n",
       "      <td>cell</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Cell</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>umls</td>\n",
       "      <td>bacs</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>umls</td>\n",
       "      <td>tisu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4978</th>\n",
       "      <td>umls</td>\n",
       "      <td>celc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>bacs</td>\n",
       "      <td>CellularComponent</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5013</th>\n",
       "      <td>umls</td>\n",
       "      <td>enzy</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>Protein</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5026</th>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5043</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>USES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>phsu</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>has_input</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>is_input_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>164 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "31            umls            bdsu       LOCATION_OF     ncbigene   \n",
       "39            umls            bpoc       LOCATION_OF     ncbigene   \n",
       "40            umls            cell       LOCATION_OF     ncbigene   \n",
       "58            umls            bacs           PART_OF     ncbigene   \n",
       "101           umls            tisu       LOCATION_OF     ncbigene   \n",
       "...            ...             ...               ...          ...   \n",
       "4977          umls            eico          INHIBITS     ncbigene   \n",
       "4978          umls            celc       LOCATION_OF     ncbigene   \n",
       "5013          umls            enzy    INTERACTS_WITH     ncbigene   \n",
       "5026          umls            aapp     COEXISTS_WITH     ncbigene   \n",
       "5043          umls            topp              USES     ncbigene   \n",
       "\n",
       "     OriginalObject            BiolinkSubject   BiolinkObject          BMPred  \\\n",
       "31             aapp          AnatomicalEntity     Polypeptide     location_of   \n",
       "39             aapp  GrossAnatomicalStructure     Polypeptide     location_of   \n",
       "40             aapp                      Cell     Polypeptide     location_of   \n",
       "58             aapp            ChemicalEntity     Polypeptide         part_of   \n",
       "101            aapp  GrossAnatomicalStructure     Polypeptide     location_of   \n",
       "...             ...                       ...             ...             ...   \n",
       "4977           aapp             SmallMolecule     Polypeptide         affects   \n",
       "4978           bacs         CellularComponent  ChemicalEntity     location_of   \n",
       "5013           aapp                   Protein     Polypeptide  interacts_with   \n",
       "5026           phsu               Polypeptide   SmallMolecule   coexists_with   \n",
       "5043           phsu                 Procedure   SmallMolecule       has_input   \n",
       "\n",
       "     BMQualPred               BMObjAsp BMObjDirect     BM_Inv_Pred  \\\n",
       "31         None                   None        None      located_in   \n",
       "39         None                   None        None      located_in   \n",
       "40         None                   None        None      located_in   \n",
       "58         None                   None        None        has_part   \n",
       "101        None                   None        None      located_in   \n",
       "...         ...                    ...         ...             ...   \n",
       "4977     causes  activity_or_abundance   decreased     affected_by   \n",
       "4978       None                   None        None      located_in   \n",
       "5013       None                   None        None  interacts_with   \n",
       "5026       None                   None        None   coexists_with   \n",
       "5043       None                   None        None     is_input_of   \n",
       "\n",
       "     BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "31              None                   None              None  \n",
       "39              None                   None              None  \n",
       "40              None                   None              None  \n",
       "58              None                   None              None  \n",
       "101             None                   None              None  \n",
       "...              ...                    ...               ...  \n",
       "4977       caused_by  activity_or_abundance         decreased  \n",
       "4978            None                   None              None  \n",
       "5013            None                   None              None  \n",
       "5026            None                   None              None  \n",
       "5043            None                   None              None  \n",
       "\n",
       "[164 rows x 15 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## looking\n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'ncbigene') &\n",
    "                  (final_metatriples['BiolinkSubject'] != 'Gene')]\n",
    "final_metatriples[(final_metatriples['ObjectPrefix'] == 'ncbigene') &\n",
    "                  (final_metatriples['BiolinkObject'] != 'Gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "480daedc",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:30.438918Z",
     "start_time": "2025-08-20T07:36:30.244787Z"
    }
   },
   "outputs": [],
   "source": [
    "## Next, mutate the metatriples with ncbigene prefix so their Biolink categories are Gene \n",
    "for index in final_metatriples.index:\n",
    "    if (final_metatriples.loc[index,'SubjectPrefix'] == 'ncbigene'):\n",
    "        final_metatriples.loc[index,'BiolinkSubject'] = 'Gene'\n",
    "    if final_metatriples.loc[index,'ObjectPrefix'] == 'ncbigene':\n",
    "        final_metatriples.loc[index,'BiolinkObject'] = 'Gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "08e436d1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:31.034515Z",
     "start_time": "2025-08-20T07:36:31.017393Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [SubjectPrefix, OriginalSubject, OriginalPredicate, ObjectPrefix, OriginalObject, BiolinkSubject, BiolinkObject, BMPred, BMQualPred, BMObjAsp, BMObjDirect, BM_Inv_Pred, BM_Inv_QualPred, BM_Inv_SubjAsp, BM_Inv_SubjDirect]\n",
       "Index: []"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## double-checking: should now get empty dataframes\n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'ncbigene') &\n",
    "                (final_metatriples['BiolinkSubject'] != 'Gene')]\n",
    "final_metatriples[(final_metatriples['ObjectPrefix'] == 'ncbigene') &\n",
    "                (final_metatriples['BiolinkObject'] != 'Gene')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870db906",
   "metadata": {},
   "source": [
    "## Generate operation yaml!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "344f9ec6",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "Review / test the code below carefully!\n",
    "\n",
    "* If needed, change the code within the functions below to change the x-bte annotations that are made...\n",
    "  * review the prefix-handling code! This is to handle ncbigene ID-namespace for some metatriples\n",
    "  * review the qualifier-generating code! Currently it's very simple because all operations with qualifiers have the same set: qualified_predicate, an aspect_qualifier, and a direction_qualifier...\n",
    "* the yaml created below refers to umls-subj, umls-obj, ncbigene-subj, ncbigene-obj...those are specified [here close to the bottom](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf76bfcf",
   "metadata": {},
   "source": [
    "### function for forward operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "222c6bd1",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:50.544521Z",
     "start_time": "2025-08-20T07:36:50.536611Z"
    }
   },
   "outputs": [],
   "source": [
    "## will use variables set in first section's config\n",
    "## original direction: subject -> object\n",
    "\n",
    "def generate_forward_op(\n",
    "    subj_prefix, obj_prefix,\n",
    "    original_subj, original_pred, original_obj,\n",
    "    biolink_subj, biolink_pred, \n",
    "    biolink_qualified_pred, biolink_obj_asp, biolink_obj_direct,\n",
    "    biolink_obj):\n",
    "    \n",
    "    ## get biolink-model spelling for prefixes\n",
    "    if subj_prefix == 'umls':\n",
    "        biolink_subj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_subj_prefix = 'NCBIGene'\n",
    "        \n",
    "    if obj_prefix == 'umls':\n",
    "        biolink_obj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_obj_prefix = 'NCBIGene'\n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    if (subj_prefix == 'umls') and (obj_prefix == 'umls'):\n",
    "        normal_op_name = f'{original_subj}-{original_pred}-{original_obj}'\n",
    "    else:\n",
    "        normal_op_name = f'{original_subj}_{subj_prefix}-{original_pred}-{original_obj}_{obj_prefix}'\n",
    "    \n",
    "    temp = {\n",
    "        ## original direction: subject -> object\n",
    "        normal_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': biolink_subj_prefix,\n",
    "                        'semantic': biolink_subj  ## input is subject!\n",
    "                    }\n",
    "                ],\n",
    "                ## notice that subject.{subj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'requestBody': {\n",
    "                    'body': {\n",
    "                        'q': '{{ queryInputs }}',\n",
    "                        'scopes': f'subject.{subj_prefix}'\n",
    "                    }\n",
    "                },\n",
    "                ## notice that object.{obj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'parameters': {\n",
    "                    'fields': f'object.{obj_prefix},predication.pmid,'\n",
    "                               'predication.sentence,subject.name,object.name',\n",
    "                    'size': POST_size,\n",
    "                    'filter': f'predicate:{original_pred} AND '\n",
    "                              f'object.semantic_type_abbreviation:{original_obj} AND '\n",
    "                              f'_exists_:object.{obj_prefix} AND '\n",
    "                              f'pmid_count:>{pmid_count}'\n",
    "                },\n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': biolink_obj_prefix,\n",
    "                        'semantic': biolink_obj  ## output is object\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_pred,\n",
    "                'source': doublequote('infores:semmeddb'),\n",
    "                'knowledge_level': 'not_provided',\n",
    "                'agent_type': 'text_mining_agent',\n",
    "                'response_mapping': {\n",
    "                    ## matches output as object\n",
    "                    ## so umls-obj and ncbigene-obj are the response-mappings generated\n",
    "                    \"$ref\": f'#/components/x-bte-response-mapping/{obj_prefix}-obj'  \n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has obj aspect and obj direction\n",
    "    if biolink_qualified_pred:\n",
    "        temp[normal_op_name][0]['qualifiers'] = {\n",
    "            'qualified_predicate': biolink_qualified_pred,\n",
    "            'object_aspect_qualifier': biolink_obj_asp,\n",
    "            'object_direction_qualifier': biolink_obj_direct\n",
    "        }\n",
    "        \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "83b2edf4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:52.169326Z",
     "start_time": "2025-08-20T07:36:52.164086Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bpoc-LOCATION_OF-neop': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'GrossAnatomicalStructure'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.umls'}},\n",
       "   'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:LOCATION_OF AND object.semantic_type_abbreviation:neop AND _exists_:object.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'predicate': 'location_of',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-obj'}}]}"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - umls\n",
    "generate_forward_op(\n",
    "    'umls', 'umls',\n",
    "    'bpoc', 'LOCATION_OF', 'neop',\n",
    "    'GrossAnatomicalStructure', 'location_of', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "723f107d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:55.899293Z",
     "start_time": "2025-08-20T07:36:55.894374Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm_ncbigene-ASSOCIATED_WITH-neop_umls': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.ncbigene'}},\n",
       "   'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:ASSOCIATED_WITH AND object.semantic_type_abbreviation:neop AND _exists_:object.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'predicate': 'related_to',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-obj'}}]}"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - umls\n",
    "generate_forward_op(\n",
    "    'ncbigene', 'umls',\n",
    "    'gngm', 'ASSOCIATED_WITH', 'neop',\n",
    "    'Gene', 'related_to', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "c55ecc04",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:36:58.402976Z",
     "start_time": "2025-08-20T07:36:58.397957Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp_umls-STIMULATES-gngm_ncbigene': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.umls'}},\n",
       "   'parameters': {'fields': 'object.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND object.semantic_type_abbreviation:gngm AND _exists_:object.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affects',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-obj'},\n",
       "   'qualifiers': {'qualified_predicate': 'causes',\n",
       "    'object_aspect_qualifier': 'activity_or_abundance',\n",
       "    'object_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - ncbigene\n",
    "generate_forward_op(\n",
    "    'umls', 'ncbigene',\n",
    "    'aapp', 'STIMULATES', 'gngm',\n",
    "    'Polypeptide', 'affects', \n",
    "    'causes', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ba5c9ffa",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:00.511586Z",
     "start_time": "2025-08-20T07:37:00.506193Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm_ncbigene-STIMULATES-gngm_ncbigene': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'subject.ncbigene'}},\n",
       "   'parameters': {'fields': 'object.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND object.semantic_type_abbreviation:gngm AND _exists_:object.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affects',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-obj'},\n",
       "   'qualifiers': {'qualified_predicate': 'causes',\n",
       "    'object_aspect_qualifier': 'activity_or_abundance',\n",
       "    'object_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - ncbigene\n",
    "generate_forward_op(\n",
    "    'ncbigene', 'ncbigene',\n",
    "    'gngm', 'STIMULATES', 'gngm',\n",
    "    'Gene', 'affects', \n",
    "    'causes', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "8effdfd9",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:02.594419Z",
     "start_time": "2025-08-20T07:37:02.578523Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1645</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2181</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2199</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>INTERACTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>interacts_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2753</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>gngm</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3007</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>COEXISTS_WITH</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>coexists_with</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3151</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>STIMULATES</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4504</th>\n",
       "      <td>umls</td>\n",
       "      <td>hops</td>\n",
       "      <td>INHIBITS</td>\n",
       "      <td>ncbigene</td>\n",
       "      <td>aapp</td>\n",
       "      <td>ChemicalEntity</td>\n",
       "      <td>Gene</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>decreased</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "971           umls            hops        STIMULATES     ncbigene   \n",
       "1645          umls            hops          INHIBITS     ncbigene   \n",
       "2181          umls            hops    INTERACTS_WITH     ncbigene   \n",
       "2199          umls            hops    INTERACTS_WITH     ncbigene   \n",
       "2753          umls            hops     COEXISTS_WITH     ncbigene   \n",
       "3007          umls            hops     COEXISTS_WITH     ncbigene   \n",
       "3151          umls            hops        STIMULATES     ncbigene   \n",
       "4504          umls            hops          INHIBITS     ncbigene   \n",
       "\n",
       "     OriginalObject  BiolinkSubject BiolinkObject          BMPred BMQualPred  \\\n",
       "971            gngm  ChemicalEntity          Gene         affects     causes   \n",
       "1645           gngm  ChemicalEntity          Gene         affects     causes   \n",
       "2181           aapp  ChemicalEntity          Gene  interacts_with       None   \n",
       "2199           gngm  ChemicalEntity          Gene  interacts_with       None   \n",
       "2753           gngm  ChemicalEntity          Gene   coexists_with       None   \n",
       "3007           aapp  ChemicalEntity          Gene   coexists_with       None   \n",
       "3151           aapp  ChemicalEntity          Gene         affects     causes   \n",
       "4504           aapp  ChemicalEntity          Gene         affects     causes   \n",
       "\n",
       "                   BMObjAsp BMObjDirect     BM_Inv_Pred BM_Inv_QualPred  \\\n",
       "971   activity_or_abundance   increased     affected_by       caused_by   \n",
       "1645  activity_or_abundance   decreased     affected_by       caused_by   \n",
       "2181                   None        None  interacts_with            None   \n",
       "2199                   None        None  interacts_with            None   \n",
       "2753                   None        None   coexists_with            None   \n",
       "3007                   None        None   coexists_with            None   \n",
       "3151  activity_or_abundance   increased     affected_by       caused_by   \n",
       "4504  activity_or_abundance   decreased     affected_by       caused_by   \n",
       "\n",
       "             BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "971   activity_or_abundance         increased  \n",
       "1645  activity_or_abundance         decreased  \n",
       "2181                   None              None  \n",
       "2199                   None              None  \n",
       "2753                   None              None  \n",
       "3007                   None              None  \n",
       "3151  activity_or_abundance         increased  \n",
       "4504  activity_or_abundance         decreased  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get an example to check on \n",
    "final_metatriples[(final_metatriples['OriginalSubject'] == 'hops') &\n",
    "                  (final_metatriples['ObjectPrefix'] == 'ncbigene')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e0079fb",
   "metadata": {},
   "source": [
    "### function for reverse operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "912aa75d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:05.258034Z",
     "start_time": "2025-08-20T07:37:05.250319Z"
    }
   },
   "outputs": [],
   "source": [
    "## will use variables set in first section's config\n",
    "## reverse direction: object -> subject\n",
    "\n",
    "def generate_reverse_op(\n",
    "    subj_prefix, obj_prefix,\n",
    "    original_subj, original_pred, original_obj,\n",
    "    biolink_subj, \n",
    "    ## NOTICE THE INVERSES USED HERE\n",
    "    biolink_inverse_pred, \n",
    "    biolink_inverse_qualified_pred, biolink_inverse_subj_asp, biolink_inverse_subj_direct,\n",
    "    biolink_obj):\n",
    "\n",
    "    ## get biolink-model spelling for prefixes\n",
    "    if subj_prefix == 'umls':\n",
    "        biolink_subj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_subj_prefix = 'NCBIGene'\n",
    "        \n",
    "    if obj_prefix == 'umls':\n",
    "        biolink_obj_prefix = 'UMLS'\n",
    "    else:\n",
    "        biolink_obj_prefix = 'NCBIGene'\n",
    "    \n",
    "    ## create the keys for the operation names\n",
    "    if (subj_prefix == 'umls') and (obj_prefix == 'umls'):\n",
    "        rev_op_name = f'{original_subj}-{original_pred}-{original_obj}-rev'\n",
    "    else:\n",
    "        rev_op_name = f'{original_subj}_{subj_prefix}-{original_pred}-{original_obj}_{obj_prefix}-rev'\n",
    "\n",
    "    temp = {\n",
    "        ## reverse direction: object -> subject\n",
    "        rev_op_name: [\n",
    "            {\n",
    "                'supportBatch': True,\n",
    "                'useTemplating': True,\n",
    "                'inputs': [\n",
    "                    {\n",
    "                        'id': biolink_obj_prefix,\n",
    "                        'semantic': biolink_obj  ## input is object!\n",
    "                    }\n",
    "                ],\n",
    "                ## notice that object.{obj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'requestBody': {\n",
    "                    'body': {\n",
    "                        'q': '{{ queryInputs }}',\n",
    "                        'scopes': f'object.{obj_prefix}'\n",
    "                    }\n",
    "                },\n",
    "                ## notice that subject.{subj_prefix} only works because \n",
    "                ##   the API fields are spelled 'umls' and 'ncbigene' too\n",
    "                ## if this changes, it'll need adjustment\n",
    "                'parameters': {\n",
    "                    'fields': f'subject.{subj_prefix},predication.pmid,'\n",
    "                               'predication.sentence,subject.name,object.name',\n",
    "                    'size': POST_size,\n",
    "                    'filter': f'predicate:{original_pred} AND '\n",
    "                              f'subject.semantic_type_abbreviation:{original_subj} AND '\n",
    "                              f'_exists_:subject.{subj_prefix} AND '\n",
    "                              f'pmid_count:>{pmid_count}'\n",
    "                },        \n",
    "                'outputs': [\n",
    "                    {\n",
    "                        'id': biolink_subj_prefix,\n",
    "                        'semantic': biolink_subj  ## output is subject\n",
    "                    }\n",
    "                ],\n",
    "                'predicate': biolink_inverse_pred,  ## use inverse pred!\n",
    "                'source': doublequote('infores:semmeddb'),\n",
    "                'knowledge_level': 'not_provided',\n",
    "                'agent_type': 'text_mining_agent',\n",
    "                'response_mapping': {\n",
    "                    ## matches output as object\n",
    "                    ## so umls-obj and ncbigene-obj are the response-mappings generated\n",
    "                    \"$ref\": f'#/components/x-bte-response-mapping/{subj_prefix}-subj'\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    ## if there is a qualified_predicate\n",
    "    ##   there's a set of qualifiers that also has subj aspect and subj direction\n",
    "    if biolink_inverse_qualified_pred:\n",
    "        temp[rev_op_name][0]['qualifiers'] = {\n",
    "            'qualified_predicate': biolink_inverse_qualified_pred,  ## use inverse pred!\n",
    "            ## use subject qualifiers\n",
    "            'subject_aspect_qualifier': biolink_inverse_subj_asp,         \n",
    "            'subject_direction_qualifier': biolink_inverse_subj_direct   \n",
    "        }\n",
    "    \n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "955b36a2",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:06.123300Z",
     "start_time": "2025-08-20T07:37:06.103120Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SubjectPrefix</th>\n",
       "      <th>OriginalSubject</th>\n",
       "      <th>OriginalPredicate</th>\n",
       "      <th>ObjectPrefix</th>\n",
       "      <th>OriginalObject</th>\n",
       "      <th>BiolinkSubject</th>\n",
       "      <th>BiolinkObject</th>\n",
       "      <th>BMPred</th>\n",
       "      <th>BMQualPred</th>\n",
       "      <th>BMObjAsp</th>\n",
       "      <th>BMObjDirect</th>\n",
       "      <th>BM_Inv_Pred</th>\n",
       "      <th>BM_Inv_QualPred</th>\n",
       "      <th>BM_Inv_SubjAsp</th>\n",
       "      <th>BM_Inv_SubjDirect</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>umls</td>\n",
       "      <td>bdsu</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aapp</td>\n",
       "      <td>AnatomicalEntity</td>\n",
       "      <td>Polypeptide</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umls</td>\n",
       "      <td>bpoc</td>\n",
       "      <td>LOCATION_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>patf</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>PathologicalProcess</td>\n",
       "      <td>location_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>located_in</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats_or_applied_or_studied_to_treat</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subject_of_treatment_application_or_study_for_...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>umls</td>\n",
       "      <td>topp</td>\n",
       "      <td>TREATS</td>\n",
       "      <td>umls</td>\n",
       "      <td>neop</td>\n",
       "      <td>Procedure</td>\n",
       "      <td>Disease</td>\n",
       "      <td>treats_or_applied_or_studied_to_treat</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>subject_of_treatment_application_or_study_for_...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5042</th>\n",
       "      <td>umls</td>\n",
       "      <td>eico</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>dsyn</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5044</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>DISRUPTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>celc</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>CellularComponent</td>\n",
       "      <td>disrupts</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>disrupted_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5045</th>\n",
       "      <td>umls</td>\n",
       "      <td>emst</td>\n",
       "      <td>PART_OF</td>\n",
       "      <td>umls</td>\n",
       "      <td>aggp</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>Cohort</td>\n",
       "      <td>part_of</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>has_part</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5047</th>\n",
       "      <td>umls</td>\n",
       "      <td>opco</td>\n",
       "      <td>AUGMENTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>tisu</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>GrossAnatomicalStructure</td>\n",
       "      <td>affects</td>\n",
       "      <td>causes</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>caused_by</td>\n",
       "      <td>activity_or_abundance</td>\n",
       "      <td>increased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5048</th>\n",
       "      <td>umls</td>\n",
       "      <td>nsba</td>\n",
       "      <td>AFFECTS</td>\n",
       "      <td>umls</td>\n",
       "      <td>comd</td>\n",
       "      <td>SmallMolecule</td>\n",
       "      <td>Disease</td>\n",
       "      <td>affects</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>affected_by</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4307 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     SubjectPrefix OriginalSubject OriginalPredicate ObjectPrefix  \\\n",
       "0             umls            bpoc       LOCATION_OF         umls   \n",
       "1             umls            bdsu       LOCATION_OF         umls   \n",
       "2             umls            bpoc       LOCATION_OF         umls   \n",
       "3             umls            topp            TREATS         umls   \n",
       "4             umls            topp            TREATS         umls   \n",
       "...            ...             ...               ...          ...   \n",
       "5042          umls            eico          DISRUPTS         umls   \n",
       "5044          umls            nsba          DISRUPTS         umls   \n",
       "5045          umls            emst           PART_OF         umls   \n",
       "5047          umls            opco          AUGMENTS         umls   \n",
       "5048          umls            nsba           AFFECTS         umls   \n",
       "\n",
       "     OriginalObject            BiolinkSubject             BiolinkObject  \\\n",
       "0              neop  GrossAnatomicalStructure                   Disease   \n",
       "1              aapp          AnatomicalEntity               Polypeptide   \n",
       "2              patf  GrossAnatomicalStructure       PathologicalProcess   \n",
       "3              dsyn                 Procedure                   Disease   \n",
       "4              neop                 Procedure                   Disease   \n",
       "...             ...                       ...                       ...   \n",
       "5042           dsyn             SmallMolecule                   Disease   \n",
       "5044           celc             SmallMolecule         CellularComponent   \n",
       "5045           aggp  GrossAnatomicalStructure                    Cohort   \n",
       "5047           tisu             SmallMolecule  GrossAnatomicalStructure   \n",
       "5048           comd             SmallMolecule                   Disease   \n",
       "\n",
       "                                     BMPred BMQualPred               BMObjAsp  \\\n",
       "0                               location_of       None                   None   \n",
       "1                               location_of       None                   None   \n",
       "2                               location_of       None                   None   \n",
       "3     treats_or_applied_or_studied_to_treat       None                   None   \n",
       "4     treats_or_applied_or_studied_to_treat       None                   None   \n",
       "...                                     ...        ...                    ...   \n",
       "5042                               disrupts       None                   None   \n",
       "5044                               disrupts       None                   None   \n",
       "5045                                part_of       None                   None   \n",
       "5047                                affects     causes  activity_or_abundance   \n",
       "5048                                affects       None                   None   \n",
       "\n",
       "     BMObjDirect                                        BM_Inv_Pred  \\\n",
       "0           None                                         located_in   \n",
       "1           None                                         located_in   \n",
       "2           None                                         located_in   \n",
       "3           None  subject_of_treatment_application_or_study_for_...   \n",
       "4           None  subject_of_treatment_application_or_study_for_...   \n",
       "...          ...                                                ...   \n",
       "5042        None                                       disrupted_by   \n",
       "5044        None                                       disrupted_by   \n",
       "5045        None                                           has_part   \n",
       "5047   increased                                        affected_by   \n",
       "5048        None                                        affected_by   \n",
       "\n",
       "     BM_Inv_QualPred         BM_Inv_SubjAsp BM_Inv_SubjDirect  \n",
       "0               None                   None              None  \n",
       "1               None                   None              None  \n",
       "2               None                   None              None  \n",
       "3               None                   None              None  \n",
       "4               None                   None              None  \n",
       "...              ...                    ...               ...  \n",
       "5042            None                   None              None  \n",
       "5044            None                   None              None  \n",
       "5045            None                   None              None  \n",
       "5047       caused_by  activity_or_abundance         increased  \n",
       "5048            None                   None              None  \n",
       "\n",
       "[4307 rows x 15 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## get an example to check on \n",
    "final_metatriples[(final_metatriples['SubjectPrefix'] == 'umls') &\n",
    "                  (final_metatriples['ObjectPrefix'] == 'umls')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "98c5830f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:08.463270Z",
     "start_time": "2025-08-20T07:37:08.458256Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bpoc-LOCATION_OF-neop-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.umls'}},\n",
       "   'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:LOCATION_OF AND subject.semantic_type_abbreviation:bpoc AND _exists_:subject.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'GrossAnatomicalStructure'}],\n",
       "   'predicate': 'located_in',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'}}]}"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - umls\n",
    "generate_reverse_op(\n",
    "    'umls', 'umls',\n",
    "    'bpoc', 'LOCATION_OF', 'neop',\n",
    "    'GrossAnatomicalStructure', 'located_in', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "b251cb75",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:09.354105Z",
     "start_time": "2025-08-20T07:37:09.348610Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm_ncbigene-ASSOCIATED_WITH-neop_umls-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.umls'}},\n",
       "   'parameters': {'fields': 'subject.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:ASSOCIATED_WITH AND subject.semantic_type_abbreviation:gngm AND _exists_:subject.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'related_to',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-subj'}}]}"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - umls\n",
    "generate_reverse_op(\n",
    "    'ncbigene', 'umls',\n",
    "    'gngm', 'ASSOCIATED_WITH', 'neop',\n",
    "    'Gene', 'related_to', \n",
    "     None, None, None,\n",
    "    'Disease')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "586ebf88",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:15.773792Z",
     "start_time": "2025-08-20T07:37:15.768187Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'aapp_umls-STIMULATES-gngm_ncbigene-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.ncbigene'}},\n",
       "   'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND subject.semantic_type_abbreviation:aapp AND _exists_:subject.umls AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'UMLS', 'semantic': 'Polypeptide'}],\n",
       "   'predicate': 'affected_by',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'},\n",
       "   'qualifiers': {'qualified_predicate': 'caused_by',\n",
       "    'subject_aspect_qualifier': 'activity_or_abundance',\n",
       "    'subject_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: umls - ncbigene\n",
    "generate_reverse_op(\n",
    "    'umls', 'ncbigene',\n",
    "    'aapp', 'STIMULATES', 'gngm',\n",
    "    'Polypeptide', 'affected_by', \n",
    "    'caused_by', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "a105373c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:18.218518Z",
     "start_time": "2025-08-20T07:37:18.213437Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gngm_ncbigene-STIMULATES-gngm_ncbigene-rev': [{'supportBatch': True,\n",
       "   'useTemplating': True,\n",
       "   'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
       "     'scopes': 'object.ncbigene'}},\n",
       "   'parameters': {'fields': 'subject.ncbigene,predication.pmid,predication.sentence,subject.name,object.name',\n",
       "    'size': 1000,\n",
       "    'filter': 'predicate:STIMULATES AND subject.semantic_type_abbreviation:gngm AND _exists_:subject.ncbigene AND pmid_count:>3'},\n",
       "   'outputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
       "   'predicate': 'affected_by',\n",
       "   'source': 'infores:semmeddb',\n",
       "   'knowledge_level': 'not_provided',\n",
       "   'agent_type': 'text_mining_agent',\n",
       "   'response_mapping': {'$ref': '#/components/x-bte-response-mapping/ncbigene-subj'},\n",
       "   'qualifiers': {'qualified_predicate': 'caused_by',\n",
       "    'subject_aspect_qualifier': 'activity_or_abundance',\n",
       "    'subject_direction_qualifier': 'increased'}}]}"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## testing: ncbigene - ncbigene\n",
    "generate_reverse_op(\n",
    "    'ncbigene', 'ncbigene',\n",
    "    'gngm', 'STIMULATES', 'gngm',\n",
    "    'Gene', 'affected_by', \n",
    "    'caused_by', 'activity_or_abundance', 'increased',\n",
    "    'Gene')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6f70b8c",
   "metadata": {},
   "source": [
    "### generating all operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "db9f9f81",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:21.372022Z",
     "start_time": "2025-08-20T07:37:21.365074Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_all_operations(metatriple_df):\n",
    "    op_tracking = set()\n",
    "    \n",
    "    saved = dict()\n",
    "    ## iterate through rows of metatriple dataframe\n",
    "    for row in metatriple_df.itertuples(index = False):\n",
    "        \n",
    "        ## forward: only make operation if it's not going to be a dupe\n",
    "        ##  dupes happen because we don't set the SEMMED semantic type of the subject (input / starting point)\n",
    "        ##  and multiple subject SEMMED semantic types can match 1 Biolink-Subject-category\n",
    "        ##  variables to have in common are the prefixes, semmed predicate, semmed object\n",
    "        forward_op_record = f\"{row.BiolinkSubject},{row.SubjectPrefix}-{row.OriginalPredicate}-{row.OriginalObject},{row.ObjectPrefix}\"\n",
    "        \n",
    "        if forward_op_record not in op_tracking:\n",
    "            op_tracking.add(forward_op_record)\n",
    "            \n",
    "            saved.update(generate_forward_op(subj_prefix = row.SubjectPrefix,\n",
    "                                             obj_prefix = row.ObjectPrefix,\n",
    "                                             original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_pred = row.BMPred,\n",
    "                                             biolink_qualified_pred = row.BMQualPred,\n",
    "                                             biolink_obj_asp = row.BMObjAsp,\n",
    "                                             biolink_obj_direct = row.BMObjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                            ))        \n",
    "        \n",
    "        ## reverse: make operation if it's not going to be a dupe\n",
    "        ##  dupes happen because we don't set the SEMMED semantic type of the object (input / starting point)\n",
    "        ##  and multiple object SEMMED semantic types can match 1 Biolink-Object-category\n",
    "        ##  variables to have in common are the prefixes, semmed predicate, semmed subject\n",
    "        reverse_op_record = f\"rev-{row.BiolinkObject},{row.ObjectPrefix}-{row.OriginalPredicate}-{row.OriginalSubject},{row.SubjectPrefix}\"\n",
    "        \n",
    "        if reverse_op_record not in op_tracking:\n",
    "            op_tracking.add(reverse_op_record)\n",
    "            saved.update(generate_reverse_op(subj_prefix = row.SubjectPrefix,\n",
    "                                             obj_prefix = row.ObjectPrefix,\n",
    "                                             original_subj = row.OriginalSubject,\n",
    "                                             original_pred = row.OriginalPredicate,\n",
    "                                             original_obj = row.OriginalObject,\n",
    "                                             biolink_subj = row.BiolinkSubject,\n",
    "                                             biolink_inverse_pred = row.BM_Inv_Pred,\n",
    "                                             biolink_inverse_qualified_pred = row.BM_Inv_QualPred,\n",
    "                                             biolink_inverse_subj_asp = row.BM_Inv_SubjAsp,\n",
    "                                             biolink_inverse_subj_direct = row.BM_Inv_SubjDirect,\n",
    "                                             biolink_obj = row.BiolinkObject\n",
    "                                             ))           \n",
    "            \n",
    "    final = {\"x-bte-kgs-operations\": saved}\n",
    "    return final"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ec9a4",
   "metadata": {},
   "source": [
    "Get the file made!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "3856781d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:25.778444Z",
     "start_time": "2025-08-20T07:37:25.642391Z"
    }
   },
   "outputs": [],
   "source": [
    "kgs_operations = generate_all_operations(final_metatriples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "fdcd4202",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:26.848560Z",
     "start_time": "2025-08-20T07:37:26.844185Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5075"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "5495d772",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:29.507522Z",
     "start_time": "2025-08-20T07:37:29.502876Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "orch_umls-INHIBITS-gngm_ncbigene-rev\n",
      "[{'supportBatch': True,\n",
      "  'useTemplating': True,\n",
      "  'inputs': [{'id': 'NCBIGene', 'semantic': 'Gene'}],\n",
      "  'requestBody': {'body': {'q': '{{ queryInputs }}',\n",
      "                           'scopes': 'object.ncbigene'}},\n",
      "  'parameters': {'fields': 'subject.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
      "                 'size': 1000,\n",
      "                 'filter': 'predicate:INHIBITS AND '\n",
      "                           'subject.semantic_type_abbreviation:orch AND '\n",
      "                           '_exists_:subject.umls AND pmid_count:>3'},\n",
      "  'outputs': [{'id': 'UMLS', 'semantic': 'SmallMolecule'}],\n",
      "  'predicate': 'affected_by',\n",
      "  'source': 'infores:semmeddb',\n",
      "  'knowledge_level': 'not_provided',\n",
      "  'agent_type': 'text_mining_agent',\n",
      "  'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-subj'},\n",
      "  'qualifiers': {'qualified_predicate': 'caused_by',\n",
      "                 'subject_aspect_qualifier': 'activity_or_abundance',\n",
      "                 'subject_direction_qualifier': 'decreased'}}]\n"
     ]
    }
   ],
   "source": [
    "## double-check qualifier + rev operation is written properly?\n",
    "for i in kgs_operations['x-bte-kgs-operations'].keys():\n",
    "    if ('INHIBITS' in i) and ('rev' in i) and ('ncbigene' in i):\n",
    "        print(i)\n",
    "        pprint.pprint(kgs_operations['x-bte-kgs-operations'][i], sort_dicts = False)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "84637c6f",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:37:32.369454Z",
     "start_time": "2025-08-20T07:37:32.364596Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hops-CAUSES-neop\n",
      "[{'supportBatch': True,\n",
      "  'useTemplating': True,\n",
      "  'inputs': [{'id': 'UMLS', 'semantic': 'ChemicalEntity'}],\n",
      "  'requestBody': {'body': {'q': '{{ queryInputs }}', 'scopes': 'subject.umls'}},\n",
      "  'parameters': {'fields': 'object.umls,predication.pmid,predication.sentence,subject.name,object.name',\n",
      "                 'size': 1000,\n",
      "                 'filter': 'predicate:CAUSES AND '\n",
      "                           'object.semantic_type_abbreviation:neop AND '\n",
      "                           '_exists_:object.umls AND pmid_count:>3'},\n",
      "  'outputs': [{'id': 'UMLS', 'semantic': 'Disease'}],\n",
      "  'predicate': 'causes',\n",
      "  'source': 'infores:semmeddb',\n",
      "  'knowledge_level': 'not_provided',\n",
      "  'agent_type': 'text_mining_agent',\n",
      "  'response_mapping': {'$ref': '#/components/x-bte-response-mapping/umls-obj'}}]\n"
     ]
    }
   ],
   "source": [
    "## double-check specific semantic type is written properly?\n",
    "for i in kgs_operations['x-bte-kgs-operations'].keys():\n",
    "    if ('hops' in i):\n",
    "        print(i)\n",
    "        pprint.pprint(kgs_operations['x-bte-kgs-operations'][i], sort_dicts = False)\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd057af2",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* it's cool how condensed the operations become, due to the way querying is done (keeping track of not creating duplicated operations using the sets in generate_all_operations function)\n",
    "* set where to download the yamls in the code chunks below\n",
    "  * operations_path\n",
    "  * list_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "239dce9d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T08:13:40.584384Z",
     "start_time": "2025-08-20T08:13:25.043569Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.boolean_representation = ['False', 'True']\n",
    "\n",
    "## this path was set in the config section at the beginning\n",
    "yaml.dump(kgs_operations, operations_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c946247a",
   "metadata": {},
   "source": [
    "Wait a sec! Need the operations list too!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "27403608",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:38:24.127140Z",
     "start_time": "2025-08-20T07:38:24.123637Z"
    }
   },
   "outputs": [],
   "source": [
    "def generate_kgs_operations_list(operations_dict):\n",
    "    kgs_op_list = []\n",
    "    for key in operations_dict.keys():\n",
    "        kgs_op_list.append( {\"$ref\": f\"#/components/x-bte-kgs-operations/{key}\"} )\n",
    "    final2 = {\"x-bte-kgs-operations\": kgs_op_list}\n",
    "    return final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "4bec97ed",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T07:38:25.027455Z",
     "start_time": "2025-08-20T07:38:25.023529Z"
    }
   },
   "outputs": [],
   "source": [
    "operations_list = generate_kgs_operations_list(kgs_operations['x-bte-kgs-operations'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "e64c34db",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-20T08:09:47.686040Z",
     "start_time": "2025-08-20T08:09:46.710715Z"
    }
   },
   "outputs": [],
   "source": [
    "yaml.dump(operations_list, list_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421fbcdc",
   "metadata": {},
   "source": [
    "# PAUSE\n",
    "\n",
    "* now the yaml segments downloaded have to be indented and manually inserted into the correct sections of the smartapi yaml...\n",
    "  * for these long files: \n",
    "    * use `sed` to add indent to the start of every line (ref: [Stack Overflow](https://unix.stackexchange.com/a/200618)): `sed -i -e 's/^/  /' generated_operations.yaml`. The number of spaces/text you want goes in-between the 2nd and 3rd `/`. Maybe the `-e` part of the command needs adjustment - generates another file with the old text and that -e in file name (ref: [Stack Overflow](https://unix.stackexchange.com/questions/715352/using-sed-with-i-option-creates-strangely-named-new-file-while-leaving-the))\n",
    "    * delete the old section, then paste the new section in VSCode. Less chance of freezing VS Code. \n",
    "  * the amount of indent to do and where to put things is specified in the [yaml that acts as a template](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/version_without_operations.yaml)\n",
    "  * the finished file is meant to be [here](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/semmeddb/smartapi.yaml) so one could select-paste the sections directly there"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa95cb1",
   "metadata": {},
   "source": [
    "## defunct: Notes on previously removed SEMMED semantic types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb04a8e",
   "metadata": {},
   "source": [
    "* amas: (Amino Acid Sequence) looks like protein \"domains\". Examples: Nuclear Export Signals, DNA Binding Domain\n",
    "* acty (Activity) Examples: War, Retirement, Euthanasia, Lifting\n",
    "* dora (Daily or Recreational Activity) Examples: Physical activity, Light Exercise, Relaxation\n",
    "* edac (Educational Activity) Examples: Training, Medical Residencies\n",
    "* gora (Governmental or Regulatory Activity) Examples: Health Care Reform, Advisory Committees\n",
    "* hlca (Health Care Activity) Examples: follow-up, Diagnosis\n",
    "* mcha (Machine Activity) Examples: Refridgeration, Neural Network Simulation\n",
    "* ocac (Occupational Activity) Examples: Promotion, Work, Mining\n",
    "* resa (Research Activity) Examples: Clinical Trials, research study\n",
    "* aggp (Age Group) Examples: Infant, Child, Adult, Elderly\n",
    "* famg (Family Group) Examples: spouse, Sister, Foster Parent\n",
    "* podg (Patient or Disabled Group) Examples: Patients\n",
    "* prog (Professional or Occupational Group) Examples: Administrators, Employee, Author\n",
    "* algae (Algae)\n",
    "* invt (Invertebrate)\n",
    "* orgm (Organism)\n",
    "* rich (Rickettsia or Chlamydia)\n",
    "* amph (Amphibian) Examples: Toad, Bufo boreas, Anura\n",
    "* anim (Animal) Examples: Animals, Laboratory /  Control Animal\n",
    "* arch (Archaeon) Examples: Archaea, halophilic bacteria, Thermoplasma acidophilum\n",
    "* bact (Bacterium) Examples: Escherichia coli, Salmonella, Borrelia burgdorferi\n",
    "* bdsu (Body Substance) too general. Examples: Urine, Milk, Lymph, Urine specimen\n",
    "* bdsy (Body System) too general. Examples: hypothalamic-pituitary-adrenal axis, Neurosecretory Systems\n",
    "* bird (Bird) Examples: Geese, Passeriformes, Raptors\n",
    "* blor (Body Location or Region) too general. Examples: Hepatic, Lysosomal, Cytoplasmic\n",
    "* bmod (Biomedical Occupation or Discipline) Examples: Medicine, Dentistry, Midwifery\n",
    "* bsoj (Body Space or Junction) too general. Examples: Compartments, Synapses, Cistern\n",
    "* chvf (Chemical Viewed Functionally) too general. Examples: inhibitors, antagonists, Agent\n",
    "* chvs (Chemical Viewed Structurally) too general. Examples: particle, solid state, vapor\n",
    "* euka (Eukaryote) Examples: Wasps, Protozoan parasite\n",
    "* ffas (Fully Formed Anatomical Structure) Examples: Carcass\n",
    "* fish (Fish) Examples: Eels, Fishes, Electric Fish\n",
    "* fngs (Fungus) Examples: Saccharomyces cerevisiae, Alternaria brassicicola, fungus\n",
    "* humn (Human) Examples: Family, Patients, Males\n",
    "* irda (Indicator, Reagent, or Diagnostic Aid) Examples: Fluorescent Probes, Chelating Agents\n",
    "* mamm (Mammal) Examples: Rattus norvegicus, Felis catus, Mus\n",
    "* ocdi (Occupation or Discipline) Examples: Science, Politics\n",
    "* plnt (Plant) Examples: Chrysanthemum x morifolium, Pollen, Oryza sativa\n",
    "* rept (Reptile) Examples: Snakes, Turtles, Reptiles\n",
    "* sbst (Substance) too general. Examples: Materials, Plastics, Photons, Substance\n",
    "* virs (Virus) Examples: Herpesvirus 4, Human / GB virus C / Herpesviridae\n",
    "* vtbt (Vertebrate) Examples: Vertebrates / Poikilotherm, NOS\n",
    "* anst (Anatomical Structure) Examples: Entire fetus, Whole body, Cadaver\n",
    "* bhvr (Behavior) too general. Examples: Sexuality, Nest Building, Behavioral phenotype\n",
    "* inbe (Individual Behavior) too general. Examples: impulsivity, Habits, Performance\n",
    "* menp (Mental Process) too general. Examples: mind control, Learning, experience\n",
    "* socb (Social Behavior) too general. Examples: Communication, Gestures, Marriage\n",
    "* biof (Biologic Function) too general. Examples: dose-response relationship, Pharmacodynamics, Anabolism\n",
    "* eehu (Environmental Effect of Humans) too general. Examples: Sewage, Pollution, Smoke\n",
    "* hcpp (Human-caused Phenomenon or Process) too general. Now not in API. Examples: particle beam, Conferences, Victimization\n",
    "* lbtr (Laboratory or Test Result) too general. Examples: False Positive Reactions, Bone Density, Serum Calcium Level\n",
    "* npop (Natural Phenomenon or Process) too general. Examples: Floods, Fluorescence, Freezing\n",
    "* phpr (Phenomenon or Process) too general. Examples: Disasters, Acceleration, Feedback\n",
    "* bodm (Biomedical or Dental Material) too general. Examples: Pill, Gel, Talc, calcium phosphate\n",
    "* drdd (Drug Delivery Device) too general. Examples: Epipen, Skin Patch, Lilly cyanide antidote kit\n",
    "* medd (Medical Device) too general. Examples: Implants / Denture, Overlay / Silicone gel implant / Swab\n",
    "* resd (Research Device) too general. Examples: Study models, Slide\n",
    "* emst (Embryonic Structure) Examples: Chick Embryo, Blastocyst structure, Placenta\n",
    "* tisu (Tissue) Examples: Tissue specimen, Blood, Human tissue, Mucous Membrane\n",
    "* genf (Genetic Function) too general. Examples: Transcription, Genetic / Transcriptional Activation / Recombination, Genetic\n",
    "* chem (Chemical) too general. Examples: Chemicals, Acids, Ligands, Ozone\n",
    "* clas (Classification) too general. Examples: Research Diagnostic Criteria, Group C\n",
    "* ftcn (Functional Concept) too general. Examples: Techniques, Intravenous Route of Drug Administration\n",
    "* idcn (Idea or Concept) too general. Examples: Significant, subject, Data, Owner\n",
    "* qlco (Qualitative Concept) too general. Examples: Effect, Associated with, Advanced phase\n",
    "* qnco (Quantitative Concept) too general. Examples: Calibration, occurrence, degrees Celsius\n",
    "* rnlw (Regulation or Law) too general. Examples: Medicare, Medicaid, regulatory\n",
    "* spco (Spatial Concept) too general. Examples: Structure, Longitudinal, Asymmetry\n",
    "* tmco (Temporal Concept) too general. Examples: New, /period, 24 Hours\n",
    "* clna (Clinical Attribute) too general. Examples: response, Renin secretion, BAND PATTERN\n",
    "* lbpr (Laboratory Procedure) too general. Examples: Western Blot, Radioimmunoassay, Staining method\n",
    "* mbrt (Molecular Biology Research Technique) too general. Examples: Polymerase Chain Reaction / Blotting, Northern\n",
    "* elii (Element, Ion, or Isotope) too general. Examples: Atom, Aluminum, Superoxides\n",
    "* emod (Experimental Model of Disease) too general. Examples: Experimental Autoimmune Encephalomyelitis, Rodent Model\n",
    "* rcpt (Receptor) too general. Examples: Binding Sites / Receptors, Metabotropic Glutamate\n",
    "* evnt (Event) too general. Now not in API. Examples: Stressful Events\n",
    "* fndg (Finding) too general. Examples: spinal cord; lesion, Normal birth weight, Sedentary job\n",
    "* geoa (Geographic Area) too general. Examples: Country, Canada\n",
    "* grup (Group) too general. Examples: Human, Individual\n",
    "* hcro (Health Care Related Organization) too general. Examples: Hospitals, Health System\n",
    "* orgt (Organization) too general. Examples: United Nations, Organization administrative structures\n",
    "* pros (Professional Society) too general. Examples: Professional Organizations, American Nurses' Association\n",
    "* shro (Self-help or Relief Organization) too general. Examples: Social Welfare, Support Groups\n",
    "* inpr (Intellectual Product) too general. Examples: Methodology, Study models\n",
    "* mnob (Manufactured Object) too general. Examples: Glass, Manuals\n",
    "* mosq (Molecular Sequence) too general. Now not in API. Examples: Genetic Code\n",
    "* nusq (Nucleotide Sequence) too general. Examples: Base Sequence, DNA Sequence, 22q11\n",
    "* orga (Organism Attribute) too general. Examples: Ability, Body Composition\n",
    "* popg (Population Group) too general. Examples: Male population group, Woman"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf7f82a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "796.5px",
    "left": "39px",
    "top": "109.633px",
    "width": "319px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
